{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "e586ae3d-a6c0-49b4-87c4-65a24e82ac04",
   "metadata": {},
   "source": [
    "# Compare DMS data to natural sequence data\n",
    "\n",
    "This notebook analyzes natural sequences for antibody escape predicted by DMS and shows the resulting neutralization validations. Additional analysis is performed by correlating natural sequence diversity with DMS data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ebdfd791-e188-4e4e-8aff-601433db2aee",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Imports\n",
    "import os\n",
    "import warnings\n",
    "import neutcurve\n",
    "import numpy as np\n",
    "import scipy as sp\n",
    "import pandas as pd\n",
    "import altair as alt\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "from Bio import SeqIO, AlignIO \n",
    "\n",
    "# Plotting colors\n",
    "tol_muted_adjusted = [\n",
    "    \"#000000\",\n",
    "    \"#CC6677\", \n",
    "    \"#1f78b4\", \n",
    "    \"#DDCC77\", \n",
    "    \"#117733\", \n",
    "    \"#882255\", \n",
    "    \"#88CCEE\",\n",
    "    \"#44AA99\", \n",
    "    \"#999933\", \n",
    "    \"#AA4499\", \n",
    "    \"#EE7733\",\n",
    "    \"#CC3311\",\n",
    "    \"#DDDDDD\",\n",
    "]\n",
    "\n",
    "# Seaborn style settings\n",
    "sns.set(rc={\"figure.dpi\":300, \"savefig.dpi\":300})\n",
    "sns.set_style(\"ticks\")\n",
    "sns.set_palette(tol_muted_adjusted)\n",
    "\n",
    "# Suppress warnings\n",
    "warnings.simplefilter(\"ignore\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "46eac608-f320-4cbe-bc08-d88609385d17",
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": ""
    },
    "tags": [
     "parameters"
    ]
   },
   "outputs": [],
   "source": [
    "# this cell is tagged as `parameters` for papermill parameterization\n",
    "filtered_escape_377H = None\n",
    "filtered_escape_89F = None\n",
    "filtered_escape_2510C = None\n",
    "filtered_escape_121F = None\n",
    "filtered_escape_256A = None\n",
    "filtered_escape_372D = None\n",
    "\n",
    "func_scores = None\n",
    "min_times_seen = None\n",
    "n_selections = None\n",
    "\n",
    "natural_sequence_variation = None\n",
    "natural_GPC_sequence_alignment = None\n",
    "\n",
    "fraction_infected_natural_isolates = None\n",
    "out_dir = None\n",
    "neuts_image_path = None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ca3d4132-3ddd-421b-8236-d79bbbc0578a",
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": ""
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "# # Uncomment for running interactive\n",
    "# filtered_escape_377H = \"../results/filtered_antibody_escape_CSVs/377H_filtered_mut_effect.csv\"\n",
    "# filtered_escape_89F = \"../results/filtered_antibody_escape_CSVs/89F_filtered_mut_effect.csv\"\n",
    "# filtered_escape_2510C = \"../results/filtered_antibody_escape_CSVs/2510C_filtered_mut_effect.csv\"\n",
    "# filtered_escape_121F = \"../results/filtered_antibody_escape_CSVs/121F_filtered_mut_effect.csv\"\n",
    "# filtered_escape_256A = \"../results/filtered_antibody_escape_CSVs/256A_filtered_mut_effect.csv\"\n",
    "# filtered_escape_372D = \"../results/filtered_antibody_escape_CSVs/372D_filtered_mut_effect.csv\"\n",
    "\n",
    "# func_scores = \"../results/func_effects/averages/293T_entry_func_effects.csv\"\n",
    "# min_times_seen = 2\n",
    "# n_selections = 8\n",
    "\n",
    "# natural_sequence_variation = \"../non-pipeline_analyses/LASV_phylogeny_analysis/Results/GPC_protein_variation.csv\"\n",
    "# natural_GPC_sequence_alignment = \"../non-pipeline_analyses/LASV_phylogeny_analysis/Results/LASV_GPC_protein_alignment.fasta\"\n",
    "\n",
    "# fraction_infected_natural_isolates = \"../data/validation_frac_infected_natural_isolates.csv\"\n",
    "# out_dir = \"../results/validation_plots/\"\n",
    "# neuts_image_path = \"../results/validation_plots/validation_neut_curves_natural_isolates.svg\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "95440d5a-e264-4630-82ef-984aba0eae5f",
   "metadata": {},
   "source": [
    "## Identify natural sequences that would potentially escape antibody neutralization\n",
    "\n",
    "To identify any natural isolates that could potentially escape any of the mapped antibodies, we want to identify high confidence escape mutations. First, we are going to filter for the top 5% escape mutants and then further filter this list to the mutations present in the top 5% of **summed escape** sites. Escape scores are clipped at the lower end of 0 to focus on escape mutations rather than sensitizing mutations. Any sequence with these mutations will be flagged as potential escape isolates."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7ddf6b27-d5d2-4f67-874d-d93a0ede9ed2",
   "metadata": {},
   "outputs": [],
   "source": [
    "def determine_escape(percentile_escape, sequence, escape_file, strain, print_results=False):\n",
    "    \"\"\"\n",
    "    Function that determines if a sequence contains any \n",
    "    escape mutations given a percentile cutoff.\n",
    "    \"\"\"\n",
    "\n",
    "    antibody_name = escape_file.split(\"/\")[-1].split(\"_\")[0]\n",
    "    \n",
    "    # Load data as dataframe\n",
    "    escape_df = pd.read_csv(escape_file)\n",
    "\n",
    "    # Clip lower scores to 0\n",
    "    escape_df[\"escape_median\"] = escape_df[\"escape_median\"].clip(lower=0)\n",
    "\n",
    "    # Get muts for top escape\n",
    "    cutoff = escape_df[\"escape_median\"].quantile(percentile_escape)\n",
    "    top_escape_muts = (\n",
    "        tuple(zip(\n",
    "            escape_df.loc[escape_df[\"escape_median\"] >= cutoff][\"mutation\"].tolist(),\n",
    "            escape_df.loc[escape_df[\"escape_median\"] >= cutoff][\"escape_median\"].tolist(),\n",
    "        ))\n",
    "    )\n",
    "\n",
    "    # Calculate summed escape sites and get sites for top escape\n",
    "    escape_df = (\n",
    "        escape_df\n",
    "        .groupby(\"site\")\n",
    "        .aggregate({\"escape_median\" : \"sum\"})\n",
    "        .reset_index()\n",
    "    )\n",
    "    cutoff = escape_df[\"escape_median\"].quantile(percentile_escape)\n",
    "    top_escape_sites = escape_df.loc[escape_df[\"escape_median\"] >= cutoff][\"site\"].tolist()\n",
    "\n",
    "    # Filter top escape muts bases on top escape sites\n",
    "    top_escape_muts = [x for x in top_escape_muts if int(x[0][1:-1]) in top_escape_sites]\n",
    "    \n",
    "    # Initialize escape flag\n",
    "    escape = 0\n",
    "    \n",
    "    # Iterate through list of escape mutations\n",
    "    for escape_mut in top_escape_muts:\n",
    "        site = int(escape_mut[0][1:-1])\n",
    "        if sequence[site-1] == escape_mut[0][-1]:\n",
    "            if print_results:\n",
    "                print(f\"{strain:<75} with \\t {escape_mut[0][0]}{site}{escape_mut[0][-1]} \\t DMS score: {escape_mut[1]}\")\n",
    "            escape += escape_mut[1]\n",
    "   \n",
    "    return escape\n",
    "\n",
    "# Load alignment and metadata info\n",
    "natural_seqs_df = pd.DataFrame(columns=[\"strain\", \"sequence\"])\n",
    "\n",
    "# Add alignment sequence to dataframe\n",
    "for curr_fasta in AlignIO.read(natural_GPC_sequence_alignment, \"fasta\"):\n",
    "    natural_seqs_df.loc[len(natural_seqs_df.index)] = [\n",
    "        str(curr_fasta.id),\n",
    "        str(curr_fasta.seq),\n",
    "    ] \n",
    "\n",
    "# Add antibody escape columns\n",
    "natural_seqs_df[\"377H_total_escape\"] = 0\n",
    "natural_seqs_df[\"89F_total_escape\"] = 0\n",
    "natural_seqs_df[\"2510C_total_escape\"] = 0\n",
    "natural_seqs_df[\"121F_total_escape\"] = 0\n",
    "natural_seqs_df[\"256A_total_escape\"] = 0\n",
    "natural_seqs_df[\"372D_total_escape\"] = 0\n",
    "\n",
    "# Antibody escape dataframes and percentile cutoffs to use\n",
    "antibody_files = [\n",
    "    (filtered_escape_2510C, 0.95),\n",
    "    (filtered_escape_121F, 0.95), \n",
    "    (filtered_escape_377H, 0.95),\n",
    "    (filtered_escape_256A, 0.95),\n",
    "    (filtered_escape_372D, 0.95),\n",
    "    (filtered_escape_89F, 0.95), \n",
    "]\n",
    "\n",
    "for antibody, percentile in antibody_files:\n",
    "    antibody_name = antibody.split(\"/\")[-1].split(\"_\")[0]\n",
    "    print(f\"{antibody_name} potentially escaped by:\")\n",
    "    natural_seqs_df[antibody_name + \"_total_escape\"] = (\n",
    "        natural_seqs_df.apply(lambda x: determine_escape(\n",
    "            percentile, \n",
    "            x[\"sequence\"], \n",
    "            antibody, \n",
    "            x[\"strain\"],\n",
    "            print_results=True,\n",
    "        ), axis=1)\n",
    "    )\n",
    "    print()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "18a72fb9-02b5-4b3d-8441-27af9fda8396",
   "metadata": {},
   "source": [
    "Four isolates were chosen from the above list based on the strongest escape mutations. We further calculate cumulative summed escape scores across all mutations present in each isolate even if the effects are small. However, this simple additive estimate probably does not correctly reflect the true impact of the combined mutations present in each isolate."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "84815ab3-f746-4323-aed7-91812f396eac",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Isolates chosen to validate for escape \n",
    "chosen_isolates = [\n",
    "    \"GA391_OL774861_reverse_complement_1977-XX-XX\",\n",
    "    \"LASV_H-sapiens-tc_NGA_2016_IRR_007_MK107922_2016-01-18\",\n",
    "    \"Lassa_virus_H-sapiens-wt_NGA_2018_ISTH_1024_MH157037_2018-02-14\",\n",
    "    \"LM395-SLE-2009_KM822115_2009-XX-XX\",\n",
    "]\n",
    "\n",
    "# Create subset of df for chosen isolates\n",
    "validation_isolates = (\n",
    "    natural_seqs_df.loc[natural_seqs_df[\"strain\"].isin(chosen_isolates)]\n",
    "    .reset_index(drop=True)\n",
    ")\n",
    "\n",
    "# Calculate total escape for all mutations for\n",
    "# each chosen isolate by adding all escape mutations\n",
    "for antibody, _ in antibody_files:\n",
    "    antibody_name = antibody.split(\"/\")[-1].split(\"_\")[0]\n",
    "    validation_isolates[antibody_name + \"_total_escape\"] = (\n",
    "        validation_isolates.apply(lambda x: determine_escape(\n",
    "            0, \n",
    "            x[\"sequence\"], \n",
    "            antibody, \n",
    "            x[\"strain\"],\n",
    "        ), axis=1)\n",
    "    )"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "59badf48-08ae-45c9-89e2-eb61f40af227",
   "metadata": {},
   "source": [
    "## Pseudovirus neutralization validation assays for chosen natural isolates and corresponding single mutants\n",
    "\n",
    "The following were chosen for validation:\n",
    "- 8.9F\n",
    "    - Natural isolate: Lassa_virus_H-sapiens-wt_NGA_2018_ISTH_1024_MH157037_2018-02-14\n",
    "        - Corresponding single mutant: K126N (DMS score: 0.7906)\n",
    "- 12.1F\n",
    "    - Natural isolate: LM395-SLE-2009_KM822115_2009-XX-XX\n",
    "        - Corresponding single mutant: N89D (DMS score: 2.515)\n",
    "- 25.10C\n",
    "    - Natural isolate: GA391_OL774861_reverse_complement_1977-XX-XX \n",
    "        - Corresponding single mutant: E228D (DMS score: 3.098)\n",
    "- 37.7H\n",
    "    - Natural isolate: GA391_OL774861_reverse_complement_1977-XX-XX \n",
    "        - Corresponding single mutant: H398K (DMS score: 2.949)\n",
    "    - Natural isolate: LASV_H-sapiens-tc_NGA_2016_IRR_007_MK107922_2016-01-18\n",
    "        - Corresponding single mutant: D401E (DMS score: 1.3)\n",
    "- 37.2D\n",
    "    - Natural isolate: GA391_OL774861_reverse_complement_1977-XX-XX\n",
    "        - Corresponding single mutant: H398K (DMS score: 1.078)\n",
    "\n",
    "*The D401E mutation present in the chosen natural isolates did not have high confidence in the DMS data for the 25.6A antibody selection so 25.6A will be omitted from validation neutralization assays. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "17c02dc5-16d8-4bfb-b6f4-fde34f6c64f8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read nuetralization data\n",
    "frac_infected_natural_isolates = pd.read_csv(fraction_infected_natural_isolates)\n",
    "\n",
    "# Rename viruses \n",
    "rename_dict = {\n",
    "    \"WT\" : \"unmutated\",\n",
    "}\n",
    "\n",
    "frac_infected_natural_isolates[\"virus\"] = (\n",
    "    frac_infected_natural_isolates[\"virus\"].replace(rename_dict)\n",
    ")\n",
    "\n",
    "# Markers\n",
    "markers = [\n",
    "    \"o\",\n",
    "    \"o\",\n",
    "    \"o\",\n",
    "    \"o\",\n",
    "    \"o\",\n",
    "    \"o\",\n",
    "    \"o\",\n",
    "    \"o\",\n",
    "    \"o\",\n",
    "    \"o\",\n",
    "]\n",
    "\n",
    "# Fit hill curves using neutcurve\n",
    "fits = neutcurve.curvefits.CurveFits(\n",
    "    data=frac_infected_natural_isolates,\n",
    "    fixbottom=0,\n",
    "    fixtop=1,\n",
    ")\n",
    "\n",
    "# IC values to calculate\n",
    "fitParams = fits.fitParams(ics=[50, 80, 90, 95, 97, 98, 99])\n",
    "\n",
    "fig, axes = fits.plotSera(\n",
    "    ncol=1,\n",
    "    markers=markers,\n",
    "    colors=tol_muted_adjusted,\n",
    "    attempt_shared_legend=False,\n",
    "    sharex=False,\n",
    "    sharey=False,\n",
    "    xlabel=\"\",\n",
    "    ylabel=\"\",\n",
    ")\n",
    "\n",
    "antibody_names = [\n",
    "    \"8.9F\",\n",
    "    \"12.1F\",\n",
    "    \"25.10C\",\n",
    "    \"37.7H\",\n",
    "    \"37.2D\",\n",
    "]\n",
    "\n",
    "for index in range(5):\n",
    "    axes[index,0].set_title(\n",
    "        antibody_names[index], \n",
    "        weight=\"bold\",\n",
    "        fontsize=8,\n",
    "    )\n",
    "    axes[index,0].set_xlabel(\n",
    "        \"concentration (\\u03BCg/mL)\", \n",
    "        fontsize=8,\n",
    "        # weight=\"bold\",\n",
    "    )\n",
    "    axes[index,0].set_ylabel(\n",
    "        \"fraction infectivity\", \n",
    "        fontsize=8,\n",
    "        # weight=\"bold\",\n",
    "    )\n",
    "    axes[index,0].set_ylim(-0.1, 1.3)\n",
    "    axes[index,0].set_yticks([0, 0.5, 1.0])\n",
    "    axes[index,0].set_yticklabels(labels=[0, 0.5, 1.0], fontsize=8)\n",
    "    axes[index,0].set_xlim(0.0005, 15)\n",
    "    axes[index,0].set_xticks([0.001, 0.01, 0.1, 1, 10])\n",
    "    axes[index,0].set_xticklabels(labels=[\"$10^{-3}$\", \"$10^{-2}$\", \"$10^{-1}$\", \"$10^0$\", \"$10^1$\"], fontsize=8)\n",
    "    plt.setp(axes[index,0].collections, alpha=0.8, linewidths=0.5, colors=\"black\") # for vertical error bar segment\n",
    "    plt.setp(axes[index,0].lines, alpha=0.8, markeredgewidth=0.5, markeredgecolor=\"black\", linewidth=1) # for the lines and markers\n",
    "    sns.move_legend(\n",
    "        axes[index,0], \n",
    "        loc=\"upper left\",\n",
    "        borderaxespad=0,\n",
    "        frameon=False,\n",
    "        bbox_to_anchor=(1, 1),\n",
    "        fontsize=8,\n",
    "        markerscale=1,\n",
    "        handletextpad=0.1,\n",
    "        title=\"amino acid\\nsubstitutions\",\n",
    "        title_fontproperties={\"weight\" : \"bold\", \"size\" : 8},\n",
    "        alignment=\"left\"\n",
    "    )\n",
    "\n",
    "    # Add edges to legend markers to match scatter plot\n",
    "    for ha in axes[index,0].legend_.legendHandles:\n",
    "        ha.set_markeredgecolor(\"black\")\n",
    "        ha.set_markeredgewidth(0.5)\n",
    "        ha.set_linewidth(0)\n",
    "        \n",
    "    # Change all spines\n",
    "    for axis in [\"top\", \"bottom\", \"left\", \"right\"]:\n",
    "        axes[index,0].spines[axis].set_linewidth(1)\n",
    "    axes[index,0].tick_params(axis=\"both\", length=4, width=1)\n",
    "\n",
    "width = 2.5\n",
    "height = 10\n",
    "fig.set_size_inches(width, height)\n",
    "\n",
    "# Make output dir if doesn't exist\n",
    "if not os.path.exists(out_dir):\n",
    "    os.mkdir(out_dir)\n",
    "\n",
    "# Save fig\n",
    "plt.savefig(neuts_image_path)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "80d7e3f1-41c7-444f-abc0-dc3320358d4a",
   "metadata": {},
   "source": [
    "## Compare natural diversity to functional scores\n",
    "\n",
    "Next, we are going to compare the DMS data to natural sequence diversity. In this case, natural sequence diversity is reflected by calculating effective amino acids* at each site based on all high quality available Lassa GPC sequences.\n",
    "\n",
    "*The formula for calculating effective amino acids is described in *Biophysical Models of Protein Evolution: Understanding the Patterns of Evolutionary Sequence Divergence*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "198dda5a-be8c-4c8e-89cb-3715f9a8cb66",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load data as dataframe\n",
    "functional_scores = pd.read_csv(func_scores)\n",
    "\n",
    "# Filter for minimum selections, times seen and no stop codons\n",
    "functional_scores = (\n",
    "    functional_scores.query(\n",
    "        \"n_selections >= @n_selections and times_seen >= @min_times_seen and mutant != '*'\"\n",
    "    )\n",
    "    .drop(columns=[\"mutant\", \"times_seen\"])\n",
    "    .groupby([\"site\", \"wildtype\"])\n",
    "    .aggregate({\n",
    "        \"effect\" : \"mean\"\n",
    "    })\n",
    "    .reset_index()\n",
    ")\n",
    "\n",
    "# Load data as dataframe\n",
    "natural_variation = pd.read_csv(natural_sequence_variation)\n",
    "\n",
    "# Drop individual amino acid counts\n",
    "natural_variation = natural_variation[[\"site\", \"entropy\", \"n_effective\"]]\n",
    "\n",
    "# Merge functional and natural dataframes\n",
    "merged_df = (\n",
    "    functional_scores.merge(\n",
    "        natural_variation,\n",
    "        how=\"left\",\n",
    "        on=[\"site\"],\n",
    "        validate=\"one_to_one\",\n",
    "    )\n",
    ")\n",
    "\n",
    "# Add escape to dataframe for each antibody\n",
    "for antibody_file,_ in antibody_files:\n",
    "\n",
    "    antibody_name = antibody_file.split(\"/\")[-1].split(\"_\")[0]\n",
    "\n",
    "    # Load data as dataframe\n",
    "    escape_df = pd.read_csv(antibody_file)\n",
    "\n",
    "    # Clip lower scores to 0\n",
    "    escape_df[\"escape_median\"] = escape_df[\"escape_median\"].clip(lower=0)\n",
    "\n",
    "    # Calculate summed escape sites and get sites for top escape\n",
    "    escape_df = (\n",
    "        escape_df\n",
    "        .groupby(\"site\")\n",
    "        .aggregate({\"escape_median\" : \"sum\"})\n",
    "        .reset_index()\n",
    "    )\n",
    "\n",
    "    # Rename escape column to include antibody name\n",
    "    escape_df = escape_df.rename(columns={\"escape_median\" : \"escape_\" + antibody_name})\n",
    "\n",
    "    # Merge dataframes\n",
    "    merged_df = (\n",
    "        merged_df.merge(\n",
    "            escape_df,\n",
    "            how=\"left\",\n",
    "            on=[\"site\"],\n",
    "            validate=\"one_to_one\",\n",
    "        )\n",
    "    )\n",
    "\n",
    "# Total summed escape per site for all antibodies\n",
    "merged_df[\"total_escape\"] = (\n",
    "    merged_df[\"escape_2510C\"]\n",
    "    + \n",
    "    merged_df[\"escape_121F\"]\n",
    "    +\n",
    "    merged_df[\"escape_377H\"]\n",
    "    +\n",
    "    merged_df[\"escape_256A\"]\n",
    "    +\n",
    "    merged_df[\"escape_372D\"]\n",
    "    +\n",
    "    merged_df[\"escape_89F\"]\n",
    ")\n",
    "\n",
    "# Sites of mutations for chosen validations\n",
    "validation_sites = [89, 126, 228, 398, 401]\n",
    "merged_df[\"site of validation\"] = (\n",
    "    merged_df[\"site\"].apply(lambda x: True if x in validation_sites else False)\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a08518c0-646c-4952-a5d8-3000872c7035",
   "metadata": {},
   "source": [
    "First, we look at the correlation of functional effects and antibody escape stratified by the different antibodies. As expected, the mutations that lead to escape tend to be more functionally tolerated. Furthermore, the mutations present in the natural isolates also tend to be functionally tolerated except for the notable exception of N89D which is quite deleterious. \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6c988965-6a04-48dd-b004-f1e4024624e1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Making two lists for values and colors \n",
    "dom = [True, False] \n",
    "rng = [\"#EE7733FF\", \"#00000026\"] \n",
    "\n",
    "subplots = []\n",
    "for index,antibody in enumerate(antibody_files):\n",
    "    antibody_name = antibody[0].split(\"/\")[-1].split(\"_\")[0]\n",
    "\n",
    "    curr_subplot = alt.Chart(merged_df, title=antibody_name).mark_point(\n",
    "        filled=True, \n",
    "        color=\"black\", \n",
    "        size=75\n",
    "    ).encode(\n",
    "        alt.X(\n",
    "            \"escape_\" + antibody_name,\n",
    "            axis=alt.Axis(\n",
    "                title=\"total site escape\", \n",
    "                domainWidth=1,\n",
    "                domainColor=\"black\",\n",
    "                tickColor=\"black\",\n",
    "            ),\n",
    "            scale=alt.Scale(domainMin=-1)\n",
    "        ),\n",
    "        alt.Y(\n",
    "            \"effect\",\n",
    "            axis=alt.Axis(\n",
    "                title=\"effect on cell entry\", \n",
    "                values=[-4,-3,-2,-1,0,1],\n",
    "                domainWidth=1,\n",
    "                domainColor=\"black\",\n",
    "                tickColor=\"black\",\n",
    "            ),\n",
    "            scale=alt.Scale(domain=[-4.5,1])\n",
    "        ),\n",
    "        tooltip=[\n",
    "            \"site\",\n",
    "            \"wildtype\",\n",
    "            \"effect\",\n",
    "            \"escape_\" + antibody_name,\n",
    "        ],\n",
    "        color=alt.Color(\n",
    "            \"site of validation\", \n",
    "            scale=alt.Scale(domain=dom, range=rng),\n",
    "        ), \n",
    "    ).properties(\n",
    "        width=150,\n",
    "        height=150,\n",
    "    )\n",
    "    \n",
    "    subplots.append(curr_subplot)\n",
    "\n",
    "func_vs_antibody = alt.hconcat(\n",
    "    subplots[0],\n",
    "    subplots[1],\n",
    "    subplots[2],\n",
    "    subplots[3],\n",
    "    subplots[4],\n",
    "    subplots[5],\n",
    "    spacing=5,\n",
    "    title=\"Correlations of functional effects and antibody escape\",\n",
    ").configure_axis(\n",
    "    grid=False,\n",
    "    labelFontSize=16,\n",
    "    titleFontSize=16,\n",
    "    labelFontWeight=\"normal\",\n",
    "    titleFontWeight=\"normal\",\n",
    ").configure_title(\n",
    "    fontSize=24,\n",
    ")\n",
    "\n",
    "func_vs_antibody"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c89fc6be-a8ff-4db6-b3f8-a5828d81ae03",
   "metadata": {},
   "source": [
    "Next, we look at the correlation of natural sequence diversity and antibody escape stratified by the different antibodies. The antibody 12.1F (and maybe 37.2D to a lesser extent) tends to be weakly escaped by mutations at sites with higher natural diversity. Of the chosen validations, site 126 tends to be most diverse but sites 228 and 398 also have increased diversity."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "754e7f17-e2bd-45b8-8548-75ff7643db59",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Making two lists for values and colors \n",
    "dom = [True, False] \n",
    "rng = [\"#EE7733FF\", \"#00000026\"] \n",
    "\n",
    "subplots = []\n",
    "for index,antibody in enumerate(antibody_files):\n",
    "    antibody_name = antibody[0].split(\"/\")[-1].split(\"_\")[0]\n",
    "\n",
    "    curr_subplot = alt.Chart(merged_df, title=antibody_name).mark_point(\n",
    "        filled=True, \n",
    "        color=\"black\", \n",
    "        size=75\n",
    "    ).encode(\n",
    "        alt.X(\n",
    "            \"escape_\" + antibody_name,\n",
    "            axis=alt.Axis(\n",
    "                title=\"total site escape\", \n",
    "                domainWidth=1,\n",
    "                domainColor=\"black\",\n",
    "                tickColor=\"black\",\n",
    "            ),\n",
    "            scale=alt.Scale(domainMin=-1)\n",
    "        ),\n",
    "        alt.Y(\n",
    "            \"n_effective\",\n",
    "            axis=alt.Axis(\n",
    "                title=[\"effective amino acids\", \"in natural sequences\"], \n",
    "                values=[1,2,3,4],\n",
    "                domainWidth=1,\n",
    "                domainColor=\"black\",\n",
    "                tickColor=\"black\",\n",
    "            ),\n",
    "            scale=alt.Scale(domain=[0.9,4])\n",
    "        ),\n",
    "        tooltip=[\n",
    "            \"site\",\n",
    "            \"wildtype\",\n",
    "            \"n_effective\",\n",
    "            \"escape_\" + antibody_name,\n",
    "        ],\n",
    "        color=alt.Color(\n",
    "            \"site of validation\", \n",
    "            scale=alt.Scale(domain=dom, range=rng),\n",
    "        ), \n",
    "    ).properties(\n",
    "        width=150,\n",
    "        height=150,\n",
    "    )\n",
    "    \n",
    "    subplots.append(curr_subplot)\n",
    "\n",
    "func_vs_antibody = alt.hconcat(\n",
    "    subplots[0],\n",
    "    subplots[1],\n",
    "    subplots[2],\n",
    "    subplots[3],\n",
    "    subplots[4],\n",
    "    subplots[5],\n",
    "    spacing=5,\n",
    "    title=\"Correlations of natural sequence diversity and antibody escape\",\n",
    ").configure_axis(\n",
    "    grid=False,\n",
    "    labelFontSize=16,\n",
    "    titleFontSize=16,\n",
    "    labelFontWeight=\"normal\",\n",
    "    titleFontWeight=\"normal\",\n",
    ").configure_title(\n",
    "    fontSize=24,\n",
    ")\n",
    "\n",
    "func_vs_antibody"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "375375cb-628b-4d7a-a6f2-c768ac82019d",
   "metadata": {},
   "source": [
    "Finally, we are going to look at summary correlations for antibody escape, functional effects on cell entry, and natural sequence diversity. The total antibody escape was calculated by summing escape at every site across all antibodies mapped."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1bbe4352-a10d-45db-9412-e2217985042e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Making two lists for values and colors \n",
    "dom = [True, False] \n",
    "rng = [\"#EE7733FF\", \"#00000026\"] \n",
    "\n",
    "# Calculate statistics\n",
    "r, p = sp.stats.pearsonr(\n",
    "    merged_df[[\"total_escape\",\"effect\"]].dropna()[\"total_escape\"],\n",
    "    merged_df[[\"total_escape\",\"effect\"]].dropna()[\"effect\"],\n",
    ")\n",
    "print(f\"r correlation of total escape and effect on cell entry: {r:.2f}\")\n",
    "\n",
    "effect_vs_antibody = alt.Chart(merged_df, title=f\"r = {r:.2f}\").mark_point(\n",
    "    filled=True, \n",
    "    color=\"black\", \n",
    "    size=75,\n",
    ").encode(\n",
    "    alt.X(\n",
    "        \"total_escape\",\n",
    "        axis=alt.Axis(\n",
    "            title=[\"total site escape across\", \"all mapped antibodies\"],  \n",
    "            domainWidth=1,\n",
    "            domainColor=\"black\",\n",
    "            tickColor=\"black\",\n",
    "        ),\n",
    "        scale=alt.Scale(domainMin=-1)\n",
    "    ),\n",
    "    alt.Y(\n",
    "        \"effect\",\n",
    "        axis=alt.Axis(\n",
    "            title=\"effect on cell entry\", \n",
    "            values=[-4,-3,-2,-1,0,1],\n",
    "            domainWidth=1,\n",
    "            domainColor=\"black\",\n",
    "            tickColor=\"black\",\n",
    "        ),\n",
    "        scale=alt.Scale(domain=[-4.5,1])\n",
    "    ),\n",
    "    tooltip=[\n",
    "        \"site\",\n",
    "        \"wildtype\",\n",
    "        \"effect\",\n",
    "        \"escape_2510C\",\n",
    "        \"escape_121F\",\n",
    "        \"escape_89F\",\n",
    "        \"escape_377H\",\n",
    "        \"escape_372D\",\n",
    "        \"escape_256A\",\n",
    "        \"total_escape\",\n",
    "    ],\n",
    "    color=alt.Color(\n",
    "        \"site of validation\", \n",
    "        scale=alt.Scale(domain=dom, range=rng),\n",
    "    ), \n",
    ").properties(\n",
    "    width=300,\n",
    "    height=300,\n",
    ")\n",
    "\n",
    "r, p = sp.stats.pearsonr(\n",
    "    merged_df[[\"total_escape\",\"n_effective\"]].dropna()[\"total_escape\"],\n",
    "    merged_df[[\"total_escape\",\"n_effective\"]].dropna()[\"n_effective\"],\n",
    ")\n",
    "print(f\"r correlation of total escape and natural sequence diversity: {r:.2f}\")\n",
    "\n",
    "natural_vs_antibody = alt.Chart(merged_df, title=f\"r = {r:.2f}\").mark_point(\n",
    "    filled=True, \n",
    "    color=\"black\", \n",
    "    size=75,\n",
    ").encode(\n",
    "    alt.Y(\n",
    "        \"total_escape\",\n",
    "        axis=alt.Axis(\n",
    "            title=[\"total site escape across\", \"all mapped antibodies\"], \n",
    "            domainWidth=1,\n",
    "            domainColor=\"black\",\n",
    "            tickColor=\"black\",\n",
    "        ),\n",
    "        scale=alt.Scale(domainMin=-1)\n",
    "    ),\n",
    "    alt.X(\n",
    "        \"n_effective\",\n",
    "        axis=alt.Axis(\n",
    "            title=[\"effective amino acids\", \"in natural sequences\"], \n",
    "            values=[1,2,3,4],\n",
    "            domainWidth=1,\n",
    "            domainColor=\"black\",\n",
    "            tickColor=\"black\",\n",
    "        ),\n",
    "        scale=alt.Scale(domain=[0.9,4])\n",
    "    ),\n",
    "    tooltip=[\n",
    "        \"site\",\n",
    "        \"wildtype\",\n",
    "        \"n_effective\",\n",
    "        \"escape_2510C\",\n",
    "        \"escape_121F\",\n",
    "        \"escape_89F\",\n",
    "        \"escape_377H\",\n",
    "        \"escape_372D\",\n",
    "        \"escape_256A\",\n",
    "        \"total_escape\",\n",
    "    ],\n",
    "    color=alt.Color(\n",
    "        \"site of validation\", \n",
    "        scale=alt.Scale(domain=dom, range=rng),\n",
    "    ), \n",
    ").properties(\n",
    "    width=300,\n",
    "    height=300,\n",
    ")\n",
    "\n",
    "r, p = sp.stats.pearsonr(\n",
    "    merged_df[[\"effect\",\"n_effective\"]].dropna()[\"effect\"],\n",
    "    merged_df[[\"effect\",\"n_effective\"]].dropna()[\"n_effective\"],\n",
    ")\n",
    "print(f\"r correlation of natural sequence diversity and effect on cell entry: {r:.2f}\")\n",
    "\n",
    "natural_vs_func = alt.Chart(merged_df, title=f\"r = {r:.2f}\").mark_point(\n",
    "    filled=True, \n",
    "    color=\"black\", \n",
    "    size=75,\n",
    ").encode(\n",
    "    alt.X(\n",
    "        \"n_effective\",\n",
    "        axis=alt.Axis(\n",
    "            title=[\"effective amino acids\", \"in natural sequences\"], \n",
    "            values=[1,2,3,4],\n",
    "            domainWidth=1,\n",
    "            domainColor=\"black\",\n",
    "            tickColor=\"black\",\n",
    "        ),\n",
    "        scale=alt.Scale(domain=[0.9,4])\n",
    "    ),\n",
    "    alt.Y(\n",
    "        \"effect\",\n",
    "        axis=alt.Axis(\n",
    "            title=\"effect on cell entry\", \n",
    "            values=[-4,-3,-2,-1,0,1],\n",
    "            domainWidth=1,\n",
    "            domainColor=\"black\",\n",
    "            tickColor=\"black\",\n",
    "        ),\n",
    "        scale=alt.Scale(domain=[-4.5,1])\n",
    "    ),\n",
    "    tooltip=[\n",
    "        \"site\",\n",
    "        \"wildtype\",\n",
    "        \"effect\",\n",
    "        \"n_effective\",\n",
    "        \"escape_2510C\",\n",
    "        \"escape_121F\",\n",
    "        \"escape_89F\",\n",
    "        \"escape_377H\",\n",
    "        \"escape_372D\",\n",
    "        \"escape_256A\",\n",
    "        \"total_escape\",\n",
    "    ],\n",
    "    color=alt.Color(\n",
    "        \"site of validation\", \n",
    "        scale=alt.Scale(domain=dom, range=rng),\n",
    "    ), \n",
    ").properties(\n",
    "    width=300,\n",
    "    height=300,\n",
    ")\n",
    "\n",
    "summary_corr_plot = alt.hconcat(\n",
    "    effect_vs_antibody,\n",
    "    natural_vs_antibody,\n",
    "    natural_vs_func,\n",
    "    spacing=5,\n",
    "    title=\"Correlations of summary statistics\",\n",
    ").configure_axis(\n",
    "    grid=False,\n",
    "    labelFontSize=16,\n",
    "    titleFontSize=16,\n",
    "    labelFontWeight=\"normal\",\n",
    "    titleFontWeight=\"normal\",\n",
    ").configure_title(\n",
    "    fontSize=24,\n",
    ")\n",
    "\n",
    "summary_corr_plot"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
