{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "e586ae3d-a6c0-49b4-87c4-65a24e82ac04",
   "metadata": {},
   "source": [
    "# Compare DMS data to natural sequence data\n",
    "\n",
    "This notebook analyzes natural sequences for antibody escape predicted by DMS and shows the resulting neutralization validations. Additional analysis is performed by correlating natural sequence diversity with DMS data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ebdfd791-e188-4e4e-8aff-601433db2aee",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Imports\n",
    "import os\n",
    "import warnings\n",
    "import dmslogo\n",
    "import neutcurve\n",
    "import numpy as np\n",
    "import scipy as sp\n",
    "import pandas as pd\n",
    "import altair as alt\n",
    "import seaborn as sns\n",
    "import matplotlib.colors\n",
    "import matplotlib.pyplot as plt\n",
    "from Bio import SeqIO, AlignIO \n",
    "\n",
    "# Plotting colors\n",
    "tol_muted_adjusted = [\n",
    "    \"#000000\",\n",
    "    \"#CC6677\", \n",
    "    \"#1f78b4\", \n",
    "    \"#DDCC77\", \n",
    "    \"#117733\", \n",
    "    \"#882255\", \n",
    "    \"#88CCEE\",\n",
    "    \"#44AA99\", \n",
    "    \"#999933\", \n",
    "    \"#AA4499\", \n",
    "    \"#EE7733\",\n",
    "    \"#CC3311\",\n",
    "    \"#DDDDDD\",\n",
    "]\n",
    "\n",
    "# Create color palette\n",
    "def color_gradient_hex(start, end, n):\n",
    "    \"\"\"Color function from polyclonal\"\"\"\n",
    "    cmap = matplotlib.colors.LinearSegmentedColormap.from_list(\n",
    "            name=\"_\", colors=[start, end], N=n\n",
    "        )\n",
    "    return [matplotlib.colors.rgb2hex(tup) for tup in cmap(list(range(0, n)))]\n",
    "\n",
    "# Seaborn style settings\n",
    "sns.set(rc={\"figure.dpi\":300, \"savefig.dpi\":300})\n",
    "sns.set_style(\"ticks\")\n",
    "sns.set_palette(tol_muted_adjusted)\n",
    "\n",
    "# Suppress warnings\n",
    "warnings.simplefilter(\"ignore\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "46eac608-f320-4cbe-bc08-d88609385d17",
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": ""
    },
    "tags": [
     "parameters"
    ]
   },
   "outputs": [],
   "source": [
    "# this cell is tagged as `parameters` for papermill parameterization\n",
    "filtered_escape_377H = None\n",
    "filtered_escape_89F = None\n",
    "filtered_escape_2510C = None\n",
    "filtered_escape_121F = None\n",
    "filtered_escape_256A = None\n",
    "filtered_escape_372D = None\n",
    "\n",
    "contacts_89F = None\n",
    "contacts_377H = None\n",
    "contacts_256A = None\n",
    "contacts_2510C = None\n",
    "contacts_121F = None\n",
    "contacts_372D = None\n",
    "\n",
    "func_scores = None\n",
    "min_times_seen = None\n",
    "n_selections = None\n",
    "\n",
    "natural_sequence_variation = None\n",
    "natural_GPC_sequence_alignment = None\n",
    "\n",
    "fraction_infected_natural_isolates = None\n",
    "out_dir = None\n",
    "neuts_image_path = None\n",
    "corr_image_path = None\n",
    "\n",
    "out_dir_escape = None\n",
    "escape_image_path = None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ca3d4132-3ddd-421b-8236-d79bbbc0578a",
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": ""
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "# # Uncomment for running interactive\n",
    "# filtered_escape_377H = \"../results/filtered_antibody_escape_CSVs/377H_filtered_mut_effect.csv\"\n",
    "# filtered_escape_89F = \"../results/filtered_antibody_escape_CSVs/89F_filtered_mut_effect.csv\"\n",
    "# filtered_escape_2510C = \"../results/filtered_antibody_escape_CSVs/2510C_filtered_mut_effect.csv\"\n",
    "# filtered_escape_121F = \"../results/filtered_antibody_escape_CSVs/121F_filtered_mut_effect.csv\"\n",
    "# filtered_escape_256A = \"../results/filtered_antibody_escape_CSVs/256A_filtered_mut_effect.csv\"\n",
    "# filtered_escape_372D = \"../results/filtered_antibody_escape_CSVs/372D_filtered_mut_effect.csv\"\n",
    "\n",
    "# contacts_89F = \"../data/antibody_contacts/antibody_contacts_89F.csv\"\n",
    "# contacts_377H = \"../data/antibody_contacts/antibody_contacts_377H.csv\"\n",
    "# contacts_256A = \"../data/antibody_contacts/antibody_contacts_256A.csv\"\n",
    "# contacts_2510C = \"../data/antibody_contacts/antibody_contacts_2510C.csv\"\n",
    "# contacts_121F = \"../data/antibody_contacts/antibody_contacts_121F.csv\"\n",
    "# contacts_372D = \"../data/antibody_contacts/antibody_contacts_372D.csv\"\n",
    "\n",
    "# func_scores = \"../results/func_effects/averages/293T_entry_func_effects.csv\"\n",
    "# min_times_seen = 2\n",
    "# n_selections = 8\n",
    "\n",
    "# natural_sequence_variation = \"../non-pipeline_analyses/LASV_phylogeny_analysis/Results/GPC_protein_variation.csv\"\n",
    "# natural_GPC_sequence_alignment = \"../non-pipeline_analyses/LASV_phylogeny_analysis/Results/LASV_GPC_protein_alignment.fasta\"\n",
    "\n",
    "# fraction_infected_natural_isolates = \"../data/validation_frac_infected_natural_isolates.csv\"\n",
    "# out_dir = \"../results/validation_plots/\"\n",
    "# neuts_image_path = \"../results/validation_plots/validation_neut_curves_natural_isolates.svg\"\n",
    "# corr_image_path = \"../results/validation_plots/natural_isolate_validation_correlation.svg\"\n",
    "\n",
    "# out_dir_escape = \"../results/antibody_escape_profiles/\"\n",
    "# escape_image_path = \"../results/antibody_escape_profiles/natural_isolate_escape_profiles.svg\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "95440d5a-e264-4630-82ef-984aba0eae5f",
   "metadata": {},
   "source": [
    "## Identify natural sequences that would potentially escape antibody neutralization\n",
    "\n",
    "To identify any natural isolates that could potentially escape any of the mapped antibodies, we want to identify high confidence escape mutations. First, we are going to filter for the top 5% escape mutants and then further filter this list to the mutations present in the top 5% of **summed escape** sites. Escape scores are clipped at the lower end of 0 to focus on escape mutations rather than sensitizing mutations. Any sequence with these mutations will be flagged as potential escape isolates."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7ddf6b27-d5d2-4f67-874d-d93a0ede9ed2",
   "metadata": {},
   "outputs": [],
   "source": [
    "def determine_escape(percentile_escape, sequence, escape_file, strain, print_results=False):\n",
    "    \"\"\"\n",
    "    Function that determines if a sequence contains any \n",
    "    escape mutations given a percentile cutoff.\n",
    "    \"\"\"\n",
    "\n",
    "    antibody_name = escape_file.split(\"/\")[-1].split(\"_\")[0]\n",
    "    \n",
    "    # Load data as dataframe\n",
    "    escape_df = pd.read_csv(escape_file)\n",
    "\n",
    "    # Clip lower scores to 0\n",
    "    escape_df[\"escape_median\"] = escape_df[\"escape_median\"].clip(lower=0)\n",
    "\n",
    "    # Get muts for top escape\n",
    "    cutoff = escape_df[\"escape_median\"].quantile(percentile_escape)\n",
    "    top_escape_muts = (\n",
    "        tuple(zip(\n",
    "            escape_df.loc[escape_df[\"escape_median\"] >= cutoff][\"mutation\"].tolist(),\n",
    "            escape_df.loc[escape_df[\"escape_median\"] >= cutoff][\"escape_median\"].tolist(),\n",
    "        ))\n",
    "    )\n",
    "\n",
    "    # Calculate summed escape sites and get sites for top escape\n",
    "    escape_df = (\n",
    "        escape_df\n",
    "        .groupby(\"site\")\n",
    "        .aggregate({\"escape_median\" : \"sum\"})\n",
    "        .reset_index()\n",
    "    )\n",
    "    cutoff = escape_df[\"escape_median\"].quantile(percentile_escape)\n",
    "    top_escape_sites = escape_df.loc[escape_df[\"escape_median\"] >= cutoff][\"site\"].tolist()\n",
    "\n",
    "    # Filter top escape muts bases on top escape sites\n",
    "    top_escape_muts = [x for x in top_escape_muts if int(x[0][1:-1]) in top_escape_sites]\n",
    "    \n",
    "    # Initialize escape flag\n",
    "    escape = 0\n",
    "    \n",
    "    # Iterate through list of escape mutations\n",
    "    for escape_mut in top_escape_muts:\n",
    "        site = int(escape_mut[0][1:-1])\n",
    "        if sequence[site-1] == escape_mut[0][-1]:\n",
    "            if print_results:\n",
    "                print(f\"{strain:<75} with \\t {escape_mut[0][0]}{site}{escape_mut[0][-1]} \\t DMS score: {escape_mut[1]}\")\n",
    "            escape += escape_mut[1]\n",
    "   \n",
    "    return escape\n",
    "\n",
    "# Load alignment and metadata info\n",
    "natural_seqs_df = pd.DataFrame(columns=[\"strain\", \"sequence\"])\n",
    "\n",
    "# Add alignment sequence to dataframe\n",
    "for curr_fasta in AlignIO.read(natural_GPC_sequence_alignment, \"fasta\"):\n",
    "    natural_seqs_df.loc[len(natural_seqs_df.index)] = [\n",
    "        str(curr_fasta.id),\n",
    "        str(curr_fasta.seq),\n",
    "    ] \n",
    "\n",
    "# Antibody escape dataframes and percentile cutoffs to use\n",
    "antibody_files = [\n",
    "    (filtered_escape_2510C, 0.95),\n",
    "    (filtered_escape_121F, 0.95), \n",
    "    (filtered_escape_377H, 0.95),\n",
    "    (filtered_escape_256A, 0.95),\n",
    "    (filtered_escape_372D, 0.95),\n",
    "    (filtered_escape_89F, 0.95), \n",
    "]\n",
    "\n",
    "for antibody, percentile in antibody_files:\n",
    "    antibody_name = antibody.split(\"/\")[-1].split(\"_\")[0]\n",
    "    print(f\"{antibody_name} potentially escaped by:\")\n",
    "    natural_seqs_df.apply(lambda x: determine_escape(\n",
    "        percentile, \n",
    "        x[\"sequence\"], \n",
    "        antibody, \n",
    "        x[\"strain\"],\n",
    "        print_results=True,\n",
    "    ), axis=1)\n",
    "    print()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "956ca791-89b9-44cb-b790-6db29ffadc95",
   "metadata": {},
   "source": [
    "The following were chosen for validation:\n",
    "- 8.9F\n",
    "    - Natural isolate: Lassa_virus_H-sapiens-wt_NGA_2018_ISTH_1024_MH157037_2018-02-14\n",
    "        - Corresponding single mutant: K126N (DMS score: 0.7906)\n",
    "- 12.1F\n",
    "    - Natural isolate: LM395-SLE-2009_KM822115_2009-XX-XX\n",
    "        - Corresponding single mutant: N89D (DMS score: 2.515)\n",
    "- 25.10C\n",
    "    - Natural isolate: GA391_OL774861_reverse_complement_1977-XX-XX \n",
    "        - Corresponding single mutant: E228D (DMS score: 3.098)\n",
    "- 37.7H\n",
    "    - Natural isolate: GA391_OL774861_reverse_complement_1977-XX-XX \n",
    "        - Corresponding single mutant: H398K (DMS score: 2.949)\n",
    "    - Natural isolate: LASV_H-sapiens-tc_NGA_2016_IRR_007_MK107922_2016-01-18\n",
    "        - Corresponding single mutant: D401E (DMS score: 1.3)\n",
    "- 37.2D\n",
    "    - Natural isolate: GA391_OL774861_reverse_complement_1977-XX-XX\n",
    "        - Corresponding single mutant: H398K (DMS score: 1.078)\n",
    "\n",
    "*The D401E mutation present in the chosen natural isolates did not have high confidence in the DMS data for the 25.6A antibody selection so 25.6A will be omitted from validation neutralization assays."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b2b93fa5-3bff-4fda-9729-5760a90424ff",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Isolates chosen to validate for escape \n",
    "chosen_isolates = [\n",
    "    \"Josiah_NC_004296_reverse_complement_2018-08-13\",\n",
    "    \"GA391_OL774861_reverse_complement_1977-XX-XX\",\n",
    "    \"LASV_H-sapiens-tc_NGA_2016_IRR_007_MK107922_2016-01-18\",\n",
    "    \"Lassa_virus_H-sapiens-wt_NGA_2018_ISTH_1024_MH157037_2018-02-14\",\n",
    "    \"LM395-SLE-2009_KM822115_2009-XX-XX\",\n",
    "]\n",
    "\n",
    "# Create subset of df for chosen isolates\n",
    "validation_isolates = (\n",
    "    natural_seqs_df.loc[natural_seqs_df[\"strain\"].isin(chosen_isolates)]\n",
    "    .reset_index(drop=True)\n",
    ")\n",
    "\n",
    "# Get Josiah sequence for comparison\n",
    "josiah_sequence = validation_isolates.loc[validation_isolates[\"strain\"] == \"Josiah_NC_004296_reverse_complement_2018-08-13\"].at[0,\"sequence\"]\n",
    "\n",
    "# Rename dictionary for virus names\n",
    "rename_dict = {\n",
    "    \"WT\" : \"unmutated Josiah\",\n",
    "    \"OL774861\" : \"GA391\",\n",
    "    \"KM822115\" : \"LM395\",\n",
    "    \"MK107922\" : \"IRR007\",\n",
    "    \"MH157037\" : \"ISTH1024\",\n",
    "    \"Josiah_NC_004296_reverse_complement_2018-08-13\" : \"unmutated Josiah\",\n",
    "    \"GA391_OL774861_reverse_complement_1977-XX-XX\" : \"GA391\",\n",
    "    \"LASV_H-sapiens-tc_NGA_2016_IRR_007_MK107922_2016-01-18\" : \"IRR007\",\n",
    "    \"Lassa_virus_H-sapiens-wt_NGA_2018_ISTH_1024_MH157037_2018-02-14\" : \"ISTH1024\",\n",
    "    \"LM395-SLE-2009_KM822115_2009-XX-XX\" : \"LM395\",\n",
    "    \"N89D\" : \"N89D Josiah\",\n",
    "    \"K126N\" : \"K126N Josiah\",\n",
    "    \"E228D\" : \"E228D Josiah\",\n",
    "    \"H398K\" : \"H398K Josiah\",\n",
    "    \"D401E\" : \"D401E Josiah\",\n",
    "}\n",
    "\n",
    "# Rename viruses and column name\n",
    "validation_isolates[\"strain\"] = (\n",
    "    validation_isolates[\"strain\"].replace(rename_dict)    \n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9cc12ace-7f42-40bc-9654-8797bfb6f18f",
   "metadata": {},
   "source": [
    "## Escape profiles for the sites that differ in the chosen isolates from the Josiah DMS strain"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "007859fe-c077-46ae-a1ba-913cc77d1e50",
   "metadata": {},
   "outputs": [],
   "source": [
    "def create_line_and_logoplots(\n",
    "    escape_file,  \n",
    "    contacts_file,\n",
    "    func_scores, \n",
    "    min_times_seen,  \n",
    "    n_selections, \n",
    "    line_plot,\n",
    "    logo_plot,\n",
    "    output_file = None,\n",
    "    sites = None,\n",
    "    name = None,\n",
    "):\n",
    "    \"\"\"\n",
    "    Function that filters and writes an antibody escape csv.\n",
    "    Also creates summed escape profiles and logoplots.\n",
    "    \"\"\"\n",
    "\n",
    "    antibody_name = escape_file.split(\"/\")[-1].split(\"_\")[0]\n",
    "\n",
    "    # Read data\n",
    "    escape_df = pd.read_csv(escape_file)\n",
    "    func_scores = pd.read_csv(func_scores)\n",
    "    contacts_df = pd.read_csv(contacts_file)\n",
    "    \n",
    "    # Create mutation column to match antibody df\n",
    "    func_scores[\"site\"] = func_scores[\"site\"].astype(str)\n",
    "    func_scores[\"mutation\"] = func_scores[\"wildtype\"] + func_scores[\"site\"] + func_scores[\"mutant\"]\n",
    "\n",
    "    # Clip lower scores to 0\n",
    "    escape_df[\"escape_median\"] = escape_df[\"escape_median\"].clip(lower=0)\n",
    "\n",
    "    # Summed escape to get top 20 escape sites to show\n",
    "    summed_df = (\n",
    "        escape_df\n",
    "        .groupby([\"site\", \"wildtype\"])\n",
    "        .aggregate({\n",
    "            \"escape_median\" : \"sum\",\n",
    "        })\n",
    "        .rename(columns={\"escape_median\" : \"site_escape\"})\n",
    "        .reset_index()\n",
    "    )\n",
    "    # Top escape sites for each antibody combined to show\n",
    "    if sites == None:\n",
    "        sites = sorted(summed_df.nlargest(18, \"site_escape\")[\"site\"].tolist())\n",
    "    escape_df[\"show_site\"] = escape_df.apply(lambda x: True if x[\"site\"] in sites else False, axis=1)\n",
    "    \n",
    "    # Shade contact sites in logo plot\n",
    "    shade_sites = list(contacts_df.loc[contacts_df[\"distance\"] == 4][\"position\"].unique())\n",
    "    \n",
    "    # **\n",
    "    # # Uncomment to show antibody contacts\n",
    "    # print(antibody_name)\n",
    "    # print(f\"Contact sites: {shade_sites}\")\n",
    "    # print()\n",
    "    # **\n",
    "    \n",
    "    escape_df[\"shade_site\"] = escape_df.apply(lambda x: \"#DDDDDD\" if x[\"site\"] in shade_sites else None, axis=1)\n",
    "    escape_df[\"shade_alpha\"] = 0.75\n",
    "    escape_df = (\n",
    "        escape_df.merge(\n",
    "            summed_df,\n",
    "            how=\"left\",\n",
    "            on=[\"site\", \"wildtype\"],\n",
    "            validate=\"many_to_one\",\n",
    "        )\n",
    "    )\n",
    "\n",
    "    # Drop extra columns\n",
    "    escape_df = (\n",
    "        escape_df.drop(\n",
    "            escape_df.columns.difference([\n",
    "                \"site\", \n",
    "                \"wildtype\",\n",
    "                \"mutant\", \n",
    "                \"escape_median\", \n",
    "                \"show_site\", \n",
    "                \"shade_site\",\n",
    "                \"shade_alpha\",\n",
    "                \"site_escape\", \n",
    "            ]), axis=1)\n",
    "    )\n",
    "\n",
    "    # Fill in missing sites\n",
    "    seen_list = [False]*491\n",
    "    for index in range(len(escape_df.index)):\n",
    "        site = escape_df.at[index, \"site\"] - 1 \n",
    "        seen_list[site] = True\n",
    "    for index, seen in enumerate(seen_list):\n",
    "        if seen == False:\n",
    "            site = index + 1\n",
    "            # Add missing sites\n",
    "            escape_df.loc[len(escape_df.index)] = [\n",
    "                site,\n",
    "                \"X\",\n",
    "                \"X\",\n",
    "                0,\n",
    "                False,\n",
    "                None,\n",
    "                None,\n",
    "                0,\n",
    "            ]\n",
    "\n",
    "    # Sort by site\n",
    "    escape_df = (\n",
    "        escape_df\n",
    "        .sort_values(by=\"site\")\n",
    "        .astype({\"mutant\" : \"str\"})\n",
    "        .reset_index(drop=True)\n",
    "    )\n",
    "\n",
    "    # Merge functional and escape dfs\n",
    "    func_scores[\"site\"] = func_scores[\"site\"].astype(\"int\")\n",
    "    func_scores = func_scores.loc[func_scores[\"mutant\"] != \"*\"] # remove stop codons\n",
    "    func_scores[\"effect\"] = func_scores[\"effect\"].clip(upper=0, lower=-2) # clip scores \n",
    "    escape_df = (\n",
    "        escape_df.merge(\n",
    "            func_scores,\n",
    "            how=\"left\",\n",
    "            on=[\"site\", \"wildtype\", \"mutant\"],\n",
    "            validate=\"one_to_one\",\n",
    "        )\n",
    "    )\n",
    "    escape_df[\"effect\"] = escape_df[\"effect\"].fillna(-2) # missing functional values are filled as -4 to make less visible\n",
    "    \n",
    "    # Add color column for logo plots\n",
    "    func_color_map = dmslogo.colorschemes.ValueToColorMap(\n",
    "        minvalue=func_scores[\"effect\"].min(),\n",
    "        maxvalue=func_scores[\"effect\"].max(),\n",
    "        cmap=matplotlib.colors.ListedColormap(color_gradient_hex(\"white\", \"#000000\", n=20))\n",
    "    )\n",
    "    escape_df = (\n",
    "        escape_df.assign(\n",
    "            color=lambda x: x[\"effect\"].map(func_color_map.val_to_color)\n",
    "        )\n",
    "    )\n",
    "\n",
    "    # Add wildtype to each site for logo plot\n",
    "    escape_df[\"wt_site\"] = escape_df[\"wildtype\"] + escape_df[\"site\"].map(str)\n",
    "\n",
    "    # Set ylim for each antibody\n",
    "    fixed_ymin = None\n",
    "    fixed_ymax = None\n",
    "    if antibody_name == \"2510C\":\n",
    "        fixed_ymin = -6.25\n",
    "        fixed_ymax = 56.25\n",
    "    elif antibody_name == \"121F\":\n",
    "        fixed_ymin = -2.5\n",
    "        fixed_ymax = 22.5\n",
    "    elif antibody_name == \"377H\":\n",
    "        fixed_ymin = -7.5 \n",
    "        fixed_ymax = 67.5\n",
    "    elif antibody_name == \"256A\":\n",
    "        fixed_ymin = -5 \n",
    "        fixed_ymax = 45\n",
    "    elif antibody_name == \"372D\":\n",
    "        fixed_ymin = -2.5 \n",
    "        fixed_ymax = 22.5\n",
    "    elif antibody_name == \"89F\":\n",
    "        fixed_ymin = -5 \n",
    "        fixed_ymax = 45\n",
    "    else:\n",
    "        print(\"Error! No ylims set!\")\n",
    "\n",
    "    # Plot escape profiles and logo plots\n",
    "    _, lineplot = dmslogo.draw_line(\n",
    "        escape_df,\n",
    "        x_col=\"site\",\n",
    "        height_col=\"site_escape\",\n",
    "        show_col=\"show_site\",\n",
    "        ax=line_plot,\n",
    "        show_color=\"#CC6677\",\n",
    "        linewidth=0.5,\n",
    "        fixed_ymin=fixed_ymin,\n",
    "        fixed_ymax=fixed_ymax,\n",
    "    )\n",
    "    lineplot.set(ylabel=None, xlabel=None)\n",
    "    lineplot.set_xlim(1,491)\n",
    "    xticks = [100, 200, 300, 400]\n",
    "    lineplot.set_xticks(xticks)\n",
    "    x_labels = [\n",
    "        \"100\", \n",
    "        \"200\",\n",
    "        \"300\",\n",
    "        \"400\",\n",
    "    ]\n",
    "    lineplot.set_xticklabels(labels=x_labels, rotation=90, horizontalalignment=\"center\", fontsize=6)\n",
    "    # Change all spines\n",
    "    for axis in [\"top\", \"bottom\", \"left\", \"right\"]:\n",
    "        lineplot.spines[axis].set_linewidth(1)\n",
    "    lineplot.tick_params(axis=\"both\", length=4, width=1, pad=1)\n",
    "    \n",
    "\n",
    "    _, logoplot = dmslogo.draw_logo(\n",
    "        escape_df.query(\"show_site == True\"),\n",
    "        x_col=\"site\",\n",
    "        letter_col=\"mutant\",\n",
    "        letter_height_col=\"escape_median\",\n",
    "        ax=logo_plot,\n",
    "        xtick_col=\"wt_site\",\n",
    "        color_col=\"color\",\n",
    "        shade_color_col=\"shade_site\",\n",
    "        shade_alpha_col=\"shade_alpha\",\n",
    "        draw_line_at_zero=\"never\",\n",
    "        fixed_ymin=fixed_ymin,\n",
    "        fixed_ymax=fixed_ymax,\n",
    "    )\n",
    "\n",
    "    logoplot.set(ylabel=None, xlabel=None)\n",
    "    x_labels = logoplot.get_xticklabels()\n",
    "    logoplot.set_xticklabels(labels=x_labels, rotation=90, horizontalalignment=\"center\", fontsize=6)\n",
    "    # Change all spines\n",
    "    for axis in [\"top\", \"bottom\", \"left\", \"right\"]:\n",
    "        logoplot.spines[axis].set_linewidth(1)\n",
    "    logoplot.tick_params(axis=\"both\", length=4, width=1, pad=1)\n",
    "\n",
    "\n",
    "    # Set antibody specific y axis ticks\n",
    "    if antibody_name == \"2510C\":\n",
    "        yticks = [0, 25, 50]\n",
    "        lineplot.set_yticks(yticks)\n",
    "        lineplot.set_yticklabels(labels=[\"0\", \"25\", \"50\"], fontsize=6)\n",
    "        logoplot.set_yticks(yticks)\n",
    "        logoplot.set_yticklabels(labels=[\"0\", \"25\", \"50\"], fontsize=6)\n",
    "        lineplot.set_title(\n",
    "            antibody_name[0:2] + \".\" + antibody_name[2:], \n",
    "            fontsize=8,\n",
    "            color=\"#44AA99\",\n",
    "        )\n",
    "        logoplot.set_title(\n",
    "            antibody_name[0:2] + \".\" + antibody_name[2:] + \" escape profile for sites different between Josiah and \" + name, \n",
    "            fontsize=8,\n",
    "            color=\"#44AA99\",\n",
    "        )\n",
    "    if antibody_name == \"121F\":\n",
    "        yticks = [0, 10, 20]\n",
    "        lineplot.set_yticks(yticks)\n",
    "        lineplot.set_yticklabels(labels=[\"0\", \"10\", \"20\"], fontsize=6)\n",
    "        logoplot.set_yticks(yticks)\n",
    "        logoplot.set_yticklabels(labels=[\"0\", \"10\", \"20\"], fontsize=6)\n",
    "        lineplot.set_title(\n",
    "            antibody_name[0:2] + \".\" + antibody_name[2:], \n",
    "            fontsize=8,\n",
    "            color=\"#999933\",\n",
    "        )\n",
    "        logoplot.set_title(\n",
    "            antibody_name[0:2] + \".\" + antibody_name[2:] + \" escape profile for sites different between Josiah and \" + name, \n",
    "            fontsize=8,\n",
    "            color=\"#999933\",\n",
    "        )\n",
    "    if antibody_name == \"377H\":\n",
    "        yticks = [0, 30, 60]\n",
    "        lineplot.set_yticks(yticks)\n",
    "        lineplot.set_yticklabels(labels=[\"0\", \"30\", \"60\"], fontsize=6)\n",
    "        logoplot.set_yticks(yticks)\n",
    "        logoplot.set_yticklabels(labels=[\"0\", \"30\", \"60\"], fontsize=6)\n",
    "        lineplot.set_title(\n",
    "            antibody_name[0:2] + \".\" + antibody_name[2:], \n",
    "            fontsize=8,\n",
    "            color=\"#AA4499\",\n",
    "        )\n",
    "        logoplot.set_title(\n",
    "            antibody_name[0:2] + \".\" + antibody_name[2:] + \" escape profile for sites different between Josiah and \" + name, \n",
    "            fontsize=8,\n",
    "            color=\"#AA4499\",\n",
    "        )\n",
    "    if antibody_name == \"256A\":\n",
    "        yticks = [0, 20, 40]\n",
    "        lineplot.set_yticks(yticks)\n",
    "        lineplot.set_yticklabels(labels=[\"0\", \"20\", \"40\"], fontsize=6)\n",
    "        logoplot.set_yticks(yticks)\n",
    "        logoplot.set_yticklabels(labels=[\"0\", \"20\", \"40\"], fontsize=6)\n",
    "        lineplot.set_title(\n",
    "            antibody_name[0:2] + \".\" + antibody_name[2:], \n",
    "            fontsize=8,\n",
    "            color=\"#AA4499\",\n",
    "        )\n",
    "        logoplot.set_title(\n",
    "            antibody_name[0:2] + \".\" + antibody_name[2:] + \" escape profile for sites different between Josiah and \" + name, \n",
    "            fontsize=8,\n",
    "            color=\"#AA4499\",\n",
    "        )\n",
    "    if antibody_name == \"372D\":\n",
    "        yticks = [0, 10, 20]\n",
    "        lineplot.set_yticks(yticks)\n",
    "        lineplot.set_yticklabels(labels=[\"0\", \"10\", \"20\"], fontsize=6)\n",
    "        logoplot.set_yticks(yticks)\n",
    "        logoplot.set_yticklabels(labels=[\"0\", \"10\", \"20\"], fontsize=6)\n",
    "        lineplot.set_title(\n",
    "            antibody_name[0:2] + \".\" + antibody_name[2:], \n",
    "            fontsize=8,\n",
    "            color=\"#AA4499\",\n",
    "        )\n",
    "        logoplot.set_title(\n",
    "            antibody_name[0:2] + \".\" + antibody_name[2:] + \" escape profile for sites different between Josiah and \" + name, \n",
    "            fontsize=8,\n",
    "            color=\"#AA4499\",\n",
    "        )\n",
    "    if antibody_name == \"89F\":\n",
    "        yticks = [0, 20, 40]\n",
    "        lineplot.set_yticks(yticks)\n",
    "        lineplot.set_yticklabels(labels=[\"0\", \"20\", \"40\"], fontsize=6)\n",
    "        logoplot.set_yticks(yticks)\n",
    "        logoplot.set_yticklabels(labels=[\"0\", \"20\", \"40\"], fontsize=6)\n",
    "        lineplot.set_title(\n",
    "            antibody_name[0] + \".\" + antibody_name[1:],\n",
    "            fontsize=8,\n",
    "            color=\"#117733\",\n",
    "        )\n",
    "        logoplot.set_title(\n",
    "            antibody_name[0] + \".\" + antibody_name[1:] + \" escape profile for sites different between Josiah and \" + name,\n",
    "            fontsize=8,\n",
    "            color=\"#117733\",\n",
    "        )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bdcbe358-00d1-41a4-afb0-95c5c721fc92",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_site_differences(seq1, seq2):\n",
    "    \"\"\"\n",
    "    Function that returns a list of all sites\n",
    "    that are different between two sequences.\n",
    "    \"\"\"\n",
    "    list_of_sites = []\n",
    "    site = 1\n",
    "    for s1, s2 in zip(seq1, seq2):\n",
    "        if s1 != s2 and s1 != \"-\" and s2 != \"-\":\n",
    "            list_of_sites.append(site)\n",
    "        site += 1\n",
    "    return list_of_sites\n",
    "\n",
    "# Get list of sequence differences compared to Josiah\n",
    "list_of_different_sites = []\n",
    "for index in validation_isolates.index:\n",
    "    curr_seq_name = validation_isolates.at[index, \"strain\"]\n",
    "    curr_seq = validation_isolates.at[index, \"sequence\"]\n",
    "    differences = get_site_differences(josiah_sequence, curr_seq)\n",
    "    list_of_different_sites.append((curr_seq_name, differences))\n",
    "\n",
    "# Antibody contact files and escape files\n",
    "contacts_files = [\n",
    "    contacts_2510C,\n",
    "    contacts_121F,\n",
    "    contacts_377H,\n",
    "    # contacts_256A,\n",
    "    contacts_377H, # duplicate 37.7H because validated with two strains\n",
    "    contacts_372D,\n",
    "    contacts_89F,\n",
    "]\n",
    "\n",
    "filtered_escape_files = [\n",
    "    filtered_escape_2510C,\n",
    "    filtered_escape_121F,\n",
    "    filtered_escape_377H,\n",
    "    # filtered_escape_256A,\n",
    "    filtered_escape_377H, # duplicate 37.7H because validated with two strains\n",
    "    filtered_escape_372D,\n",
    "    filtered_escape_89F,\n",
    "]\n",
    "\n",
    "# Set figure size and subplots\n",
    "fig, axes = plt.subplots(\n",
    "    6, \n",
    "    2, \n",
    "    gridspec_kw={\"width_ratios\":[1,8]},\n",
    "    figsize=(7, 6.25), \n",
    "    # sharex=\"col\"\n",
    ")\n",
    "# Adjust spacing of subplots\n",
    "fig.subplots_adjust(\n",
    "    bottom=0, \n",
    "    top=1, \n",
    "    wspace=0.1, \n",
    "    hspace=0.9,\n",
    ")\n",
    "\n",
    "# Iterate through list of antibody files\n",
    "for i in range(len(antibody_files)):\n",
    "    name = None\n",
    "    sites = None\n",
    "    if i == 0:\n",
    "        name = list_of_different_sites[1][0]\n",
    "        sites = list_of_different_sites[1][1]\n",
    "    elif i == 1:\n",
    "        name = list_of_different_sites[4][0]\n",
    "        sites = list_of_different_sites[4][1]\n",
    "    elif i == 2:\n",
    "        name = list_of_different_sites[1][0]\n",
    "        sites = list_of_different_sites[1][1]\n",
    "    elif i == 3:\n",
    "        name = list_of_different_sites[2][0]\n",
    "        sites = list_of_different_sites[2][1]\n",
    "    elif i == 4:\n",
    "        name = list_of_different_sites[1][0]\n",
    "        sites = list_of_different_sites[1][1]\n",
    "    elif i == 5:\n",
    "        name = list_of_different_sites[3][0]\n",
    "        sites = list_of_different_sites[3][1]\n",
    "    else:\n",
    "        print(\"Error! Index out of range!\")\n",
    "\n",
    "    create_line_and_logoplots(\n",
    "        filtered_escape_files[i], \n",
    "        contacts_files[i],\n",
    "        func_scores, \n",
    "        min_times_seen, \n",
    "        n_selections, \n",
    "        axes[i][0],\n",
    "        axes[i][1],\n",
    "        sites=sorted(sites),\n",
    "        name=name,\n",
    "    )\n",
    "\n",
    "# Common X and Y axis labels\n",
    "fig.text(0.5, -0.1, \"site\", ha=\"center\", fontsize=8)\n",
    "fig.text(-0.05, 0.5, \"site escape\", va=\"center\", rotation=\"vertical\", fontsize=8)\n",
    "\n",
    "# Make output dir if doesn't exist\n",
    "if not os.path.exists(out_dir_escape):\n",
    "    os.mkdir(out_dir_escape)\n",
    "\n",
    "# Save fig\n",
    "plt.savefig(escape_image_path)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "59badf48-08ae-45c9-89e2-eb61f40af227",
   "metadata": {},
   "source": [
    "## Pseudovirus neutralization validation assays for chosen natural isolates and corresponding single mutants\n",
    "\n",
    "The chosen isolates were validated with tradition pseudovirus neutralization assays in addition to the corresponding single mutation identified as the main one for escape."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "22c56711-4685-4950-8e5f-31cae7b00e7c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Rename column name\n",
    "validation_isolates = (\n",
    "    validation_isolates.rename(columns={\"strain\" : \"virus\"})\n",
    ")\n",
    "\n",
    "# Read nuetralization data\n",
    "frac_infected_natural_isolates = pd.read_csv(fraction_infected_natural_isolates)\n",
    "\n",
    "# Rename viruses\n",
    "frac_infected_natural_isolates[\"virus\"] = (\n",
    "    frac_infected_natural_isolates[\"virus\"].replace(rename_dict)\n",
    ")\n",
    "\n",
    "# Fit hill curves using neutcurve\n",
    "fits = neutcurve.curvefits.CurveFits(\n",
    "    data=frac_infected_natural_isolates,\n",
    "    fixbottom=0,\n",
    "    fixtop=1,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "17c02dc5-16d8-4bfb-b6f4-fde34f6c64f8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plot neut data\n",
    "fig, axes = fits.plotGrid(\n",
    "    {\n",
    "        (0, 0) : (\"25.10C\",\n",
    "                  [\n",
    "                      {\n",
    "                          \"serum\" : \"25.10C\", \n",
    "                          \"virus\" : \"unmutated Josiah\",\n",
    "                          \"replicate\" : \"average\",\n",
    "                          \"color\" : tol_muted_adjusted[0],\n",
    "                          \"marker\" : \"o\",\n",
    "                          \"label\" : \"unmutated Josiah\"\n",
    "                      },\n",
    "                      {\n",
    "                          \"serum\" : \"25.10C\", \n",
    "                          \"virus\" : \"GA391\",\n",
    "                          \"replicate\" : \"average\",\n",
    "                          \"color\" : tol_muted_adjusted[1],\n",
    "                          \"marker\" : \"o\",\n",
    "                          \"label\" : \"GA391\"\n",
    "                      },\n",
    "                      {\n",
    "                          \"serum\" : \"25.10C\", \n",
    "                          \"virus\" : \"E228D Josiah\",\n",
    "                          \"replicate\" : \"average\",\n",
    "                          \"color\" : tol_muted_adjusted[2],\n",
    "                          \"marker\" : \"o\",\n",
    "                          \"label\" : \"E228D Josiah\"\n",
    "                      },\n",
    "                  ]\n",
    "                 ),\n",
    "        (1, 0) : (\"12.1F\",\n",
    "                  [\n",
    "                      {\n",
    "                          \"serum\" : \"12.1F\", \n",
    "                          \"virus\" : \"unmutated Josiah\",\n",
    "                          \"replicate\" : \"average\",\n",
    "                          \"color\" : tol_muted_adjusted[0],\n",
    "                          \"marker\" : \"o\",\n",
    "                          \"label\" : \"unmutated Josiah\"\n",
    "                      },\n",
    "                      {\n",
    "                          \"serum\" : \"12.1F\", \n",
    "                          \"virus\" : \"LM395\",\n",
    "                          \"replicate\" : \"average\",\n",
    "                          \"color\" : tol_muted_adjusted[3],\n",
    "                          \"marker\" : \"o\",\n",
    "                          \"label\" : \"LM395\"\n",
    "                      },\n",
    "                      {\n",
    "                          \"serum\" : \"12.1F\", \n",
    "                          \"virus\" : \"N89D Josiah\",\n",
    "                          \"replicate\" : \"average\",\n",
    "                          \"color\" : tol_muted_adjusted[4],\n",
    "                          \"marker\" : \"o\",\n",
    "                          \"label\" : \"N89D Josiah\"\n",
    "                      },\n",
    "                  ]\n",
    "                 ),\n",
    "        (2, 0) : (\"37.7H\",\n",
    "                  [\n",
    "                      {\n",
    "                          \"serum\" : \"37.7H\", \n",
    "                          \"virus\" : \"unmutated Josiah\",\n",
    "                          \"replicate\" : \"average\",\n",
    "                          \"color\" : tol_muted_adjusted[0],\n",
    "                          \"marker\" : \"o\",\n",
    "                          \"label\" : \"unmutated Josiah\"\n",
    "                      },\n",
    "                      {\n",
    "                          \"serum\" : \"37.7H\", \n",
    "                          \"virus\" : \"GA391\",\n",
    "                          \"replicate\" : \"average\",\n",
    "                          \"color\" : tol_muted_adjusted[1],\n",
    "                          \"marker\" : \"o\",\n",
    "                          \"label\" : \"GA391\"\n",
    "                      },\n",
    "                      {\n",
    "                          \"serum\" : \"37.7H\", \n",
    "                          \"virus\" : \"H398K Josiah\",\n",
    "                          \"replicate\" : \"average\",\n",
    "                          \"color\" : tol_muted_adjusted[7],\n",
    "                          \"marker\" : \"o\",\n",
    "                          \"label\" : \"H398K Josiah\"\n",
    "                      },\n",
    "                  ]\n",
    "                 ),\n",
    "        (3, 0) : (\"37.7H\",\n",
    "                  [\n",
    "                      {\n",
    "                          \"serum\" : \"37.7H\", \n",
    "                          \"virus\" : \"unmutated Josiah\",\n",
    "                          \"replicate\" : \"average\",\n",
    "                          \"color\" : tol_muted_adjusted[0],\n",
    "                          \"marker\" : \"o\",\n",
    "                          \"label\" : \"unmutated Josiah\"\n",
    "                      },\n",
    "                      {\n",
    "                          \"serum\" : \"37.7H\", \n",
    "                          \"virus\" : \"IRR007\",\n",
    "                          \"replicate\" : \"average\",\n",
    "                          \"color\" : tol_muted_adjusted[6],\n",
    "                          \"marker\" : \"o\",\n",
    "                          \"label\" : \"IRR007\"\n",
    "                      },\n",
    "                      {\n",
    "                          \"serum\" : \"37.7H\", \n",
    "                          \"virus\" : \"D401E Josiah\",\n",
    "                          \"replicate\" : \"average\",\n",
    "                          \"color\" : tol_muted_adjusted[5],\n",
    "                          \"marker\" : \"o\",\n",
    "                          \"label\" : \"D401E Josiah\"\n",
    "                      },\n",
    "                  ]\n",
    "                 ),\n",
    "        (4, 0) : (\"37.2D\",\n",
    "                  [\n",
    "                      {\n",
    "                          \"serum\" : \"37.2D\", \n",
    "                          \"virus\" : \"unmutated Josiah\",\n",
    "                          \"replicate\" : \"average\",\n",
    "                          \"color\" : tol_muted_adjusted[0],\n",
    "                          \"marker\" : \"o\",\n",
    "                          \"label\" : \"unmutated Josiah\"\n",
    "                      },\n",
    "                      {\n",
    "                          \"serum\" : \"37.2D\", \n",
    "                          \"virus\" : \"GA391\",\n",
    "                          \"replicate\" : \"average\",\n",
    "                          \"color\" : tol_muted_adjusted[1],\n",
    "                          \"marker\" : \"o\",\n",
    "                          \"label\" : \"GA391\"\n",
    "                      },\n",
    "                      {\n",
    "                          \"serum\" : \"37.2D\", \n",
    "                          \"virus\" : \"H398K Josiah\",\n",
    "                          \"replicate\" : \"average\",\n",
    "                          \"color\" : tol_muted_adjusted[7],\n",
    "                          \"marker\" : \"o\",\n",
    "                          \"label\" : \"H398K Josiah\"\n",
    "                      },\n",
    "                  ]\n",
    "                 ),\n",
    "        (5, 0) : (\"8.9F\",\n",
    "                  [\n",
    "                      {\n",
    "                          \"serum\" : \"8.9F\", \n",
    "                          \"virus\" : \"unmutated Josiah\",\n",
    "                          \"replicate\" : \"average\",\n",
    "                          \"color\" : tol_muted_adjusted[0],\n",
    "                          \"marker\" : \"o\",\n",
    "                          \"label\" : \"unmutated Josiah\"\n",
    "                      },\n",
    "                      {\n",
    "                          \"serum\" : \"8.9F\", \n",
    "                          \"virus\" : \"ISTH1024\",\n",
    "                          \"replicate\" : \"average\",\n",
    "                          \"color\" : tol_muted_adjusted[8],\n",
    "                          \"marker\" : \"o\",\n",
    "                          \"label\" : \"ISTH1024\"\n",
    "                      },\n",
    "                      {\n",
    "                          \"serum\" : \"8.9F\", \n",
    "                          \"virus\" : \"K126N Josiah\",\n",
    "                          \"replicate\" : \"average\",\n",
    "                          \"color\" : tol_muted_adjusted[10],\n",
    "                          \"marker\" : \"o\",\n",
    "                          \"label\" : \"K126N Josiah\"\n",
    "                      },\n",
    "                  ]\n",
    "                 ),\n",
    "    },\n",
    "    sharex=True,\n",
    "    sharey=False,\n",
    "    xlabel=\"\",\n",
    "    ylabel=\"\",\n",
    "    attempt_shared_legend=False,\n",
    ")\n",
    "\n",
    "\n",
    "antibody_names = [\n",
    "    \"25.10C\",\n",
    "    \"12.1F\",\n",
    "    \"37.7H\",\n",
    "    \"37.7H\",\n",
    "    \"37.2D\",\n",
    "    \"8.9F\",\n",
    "]\n",
    "\n",
    "for index in range(6):\n",
    "    if antibody_names[index] == \"25.10C\":\n",
    "        axes[index,0].set_title(\n",
    "            antibody_names[index], \n",
    "            weight=\"bold\",\n",
    "            fontsize=8,\n",
    "            color=\"#44AA99\",\n",
    "        )\n",
    "    elif antibody_names[index] == \"12.1F\":\n",
    "        axes[index,0].set_title(\n",
    "            antibody_names[index], \n",
    "            weight=\"bold\",\n",
    "            fontsize=8,\n",
    "            color=\"#999933\",\n",
    "        )\n",
    "    elif antibody_names[index] == \"37.7H\" or antibody_names[index] == \"37.2D\" :\n",
    "        axes[index,0].set_title(\n",
    "            antibody_names[index], \n",
    "            weight=\"bold\",\n",
    "            fontsize=8,\n",
    "            color=\"#AA4499\",\n",
    "        )\n",
    "    elif antibody_names[index] == \"8.9F\":\n",
    "        axes[index,0].set_title(\n",
    "            antibody_names[index], \n",
    "            weight=\"bold\",\n",
    "            fontsize=8,\n",
    "            color=\"#117733\",\n",
    "        )\n",
    "    axes[index,0].set_ylim(-0.1, 1.3)\n",
    "    axes[index,0].set_yticks([0, 0.5, 1.0])\n",
    "    axes[index,0].set_yticklabels(labels=[0, 0.5, 1.0], fontsize=8)\n",
    "    axes[index,0].set_xlim(0.0005, 12.5)\n",
    "    axes[index,0].set_xticks([0.001, 0.01, 0.1, 1, 10,])\n",
    "    axes[index,0].set_xticklabels(labels=[\"$10^{-3}$\", \"$10^{-2}$\", \"$10^{-1}$\", \"$10^0$\", \"$10^1$\",], fontsize=8)\n",
    "    plt.setp(axes[index,0].collections, alpha=0.8, linewidths=0.5, colors=\"black\") # for vertical error bar segment\n",
    "    plt.setp(axes[index,0].lines, alpha=0.8, markeredgewidth=0.5, markeredgecolor=\"black\", linewidth=1) # for the lines and markers\n",
    "    sns.move_legend(\n",
    "        axes[index,0], \n",
    "        loc=\"upper left\",\n",
    "        borderaxespad=0,\n",
    "        frameon=False,\n",
    "        bbox_to_anchor=(1, 1),\n",
    "        fontsize=8,\n",
    "        markerscale=1,\n",
    "        handletextpad=0.1,\n",
    "        title=\"Lassa GPC\",\n",
    "        title_fontproperties={\"weight\" : \"bold\", \"size\" : 8},\n",
    "        alignment=\"left\"\n",
    "    )\n",
    "\n",
    "    # Add edges to legend markers to match scatter plot\n",
    "    for ha in axes[index,0].legend_.legendHandles:\n",
    "        ha.set_markeredgecolor(\"black\")\n",
    "        ha.set_markeredgewidth(0.5)\n",
    "        ha.set_linewidth(0)\n",
    "        \n",
    "    # Change all spines\n",
    "    for axis in [\"top\", \"bottom\", \"left\", \"right\"]:\n",
    "        axes[index,0].spines[axis].set_linewidth(1)\n",
    "    axes[index,0].tick_params(axis=\"both\", length=4, width=1)\n",
    "\n",
    "\n",
    "# Common X and Y axis labels\n",
    "fig.text(0.5, -0.1, \"concentration (\\u03BCg/mL)\", ha=\"center\", fontsize=8)\n",
    "fig.text(-0.05, 0.5, \"fraction infectivity\", va=\"center\", rotation=\"vertical\", fontsize=8)\n",
    "\n",
    "width = 2\n",
    "height = 7\n",
    "fig.set_size_inches(width, height)\n",
    "\n",
    "# Make output dir if doesn't exist\n",
    "if not os.path.exists(out_dir):\n",
    "    os.mkdir(out_dir)\n",
    "\n",
    "# Save fig\n",
    "plt.savefig(neuts_image_path)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "84a3dba5-c963-4fda-83b4-3f4c4a3bdfaa",
   "metadata": {},
   "source": [
    "We further calculate **cumulative summed** escape scores across all mutations present in each isolate even if the effects are small. However, this simple additive estimate probably does not correctly reflect the true impact of the combined mutations present in each isolate. These total DMS escape scores are then compared to log2 ic50 fold change values. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f4b0478a-243b-45f5-a063-f8305e2ed1f0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# IC values to calculate\n",
    "fitParams = fits.fitParams(ics=[50, 80, 99])[[\"serum\", \"virus\", \"ic50\", \"ic50_bound\"]]\n",
    "\n",
    "# Create dictionary for antibody ic50s\n",
    "wt_ic50s = {\n",
    "    \"8.9F\" : fitParams.query(\"serum == '8.9F' & virus == 'unmutated Josiah'\").reset_index(drop=True).at[0,\"ic50\"],\n",
    "    \"12.1F\" : fitParams.query(\"serum == '12.1F' & virus == 'unmutated Josiah'\").reset_index(drop=True).at[0,\"ic50\"],\n",
    "    \"25.10C\" : fitParams.query(\"serum == '25.10C' & virus == 'unmutated Josiah'\").reset_index(drop=True).at[0,\"ic50\"],\n",
    "    \"37.7H\" : fitParams.query(\"serum == '37.7H' & virus == 'unmutated Josiah'\").reset_index(drop=True).at[0,\"ic50\"],\n",
    "    \"37.2D\" : fitParams.query(\"serum == '37.2D' & virus == 'unmutated Josiah'\").reset_index(drop=True).at[0,\"ic50\"],\n",
    "}\n",
    "\n",
    "# Calculate log2 fold change of IC50 values\n",
    "fitParams[\"ic50 fold change\"] = (\n",
    "    fitParams.apply(lambda x: x[\"ic50\"]/wt_ic50s[x[\"serum\"]], axis=1)\n",
    ")\n",
    "fitParams[\"log2 ic50 fold change\"] = np.log2(fitParams[\"ic50 fold change\"])\n",
    "\n",
    "\n",
    "# Merge isolate sequence and neut dfs\n",
    "validation_df = (\n",
    "    fitParams.merge(\n",
    "        validation_isolates,\n",
    "        how=\"left\",\n",
    "        on=\"virus\",\n",
    "        validate=\"many_to_one\"\n",
    "    )\n",
    ")\n",
    "\n",
    "# Add sinlge mutant sequences\n",
    "validation_df.at[2, \"sequence\"] = (\n",
    "    josiah_sequence[:int(validation_df.at[2,\"virus\"].split()[0][1:-1])-1] + validation_df.at[2,\"virus\"].split()[0][-1] + josiah_sequence[int(validation_df.at[2,\"virus\"].split()[0][1:-1]):]\n",
    ")\n",
    "\n",
    "validation_df.at[5, \"sequence\"] = (\n",
    "    josiah_sequence[:int(validation_df.at[5,\"virus\"].split()[0][1:-1])-1] + validation_df.at[5,\"virus\"].split()[0][-1] + josiah_sequence[int(validation_df.at[5,\"virus\"].split()[0][1:-1]):]\n",
    ")\n",
    "\n",
    "validation_df.at[8, \"sequence\"] = (\n",
    "    josiah_sequence[:int(validation_df.at[8,\"virus\"].split()[0][1:-1])-1] + validation_df.at[8,\"virus\"].split()[0][-1] + josiah_sequence[int(validation_df.at[8,\"virus\"].split()[0][1:-1]):]\n",
    ")\n",
    "\n",
    "validation_df.at[10, \"sequence\"] = (\n",
    "    josiah_sequence[:int(validation_df.at[10,\"virus\"].split()[0][1:-1])-1] + validation_df.at[10,\"virus\"].split()[0][-1] + josiah_sequence[int(validation_df.at[10,\"virus\"].split()[0][1:-1]):]\n",
    ")\n",
    "\n",
    "validation_df.at[13, \"sequence\"] = (\n",
    "    josiah_sequence[:int(validation_df.at[13,\"virus\"].split()[0][1:-1])-1] + validation_df.at[13,\"virus\"].split()[0][-1] + josiah_sequence[int(validation_df.at[13,\"virus\"].split()[0][1:-1]):]\n",
    ")\n",
    "\n",
    "validation_df.at[16, \"sequence\"] = (\n",
    "    josiah_sequence[:int(validation_df.at[16,\"virus\"].split()[0][1:-1])-1] + validation_df.at[16,\"virus\"].split()[0][-1] + josiah_sequence[int(validation_df.at[16,\"virus\"].split()[0][1:-1]):]\n",
    ")\n",
    "\n",
    "# Antibody conversion dict\n",
    "antibody_file_dict = {\n",
    "    \"25.10C\" : filtered_escape_2510C,\n",
    "    \"12.1F\" : filtered_escape_121F,\n",
    "    \"37.7H\" : filtered_escape_377H,\n",
    "    \"25.6A\" : filtered_escape_256A,\n",
    "    \"37.2D\" : filtered_escape_372D,\n",
    "    \"8.9F\" : filtered_escape_89F,\n",
    "}\n",
    "\n",
    "# Calculate total escape for each sequence using an additive model\n",
    "validation_df[\"total_escape\"] = (\n",
    "    validation_df.apply(lambda x: determine_escape(\n",
    "        0, \n",
    "        x[\"sequence\"], \n",
    "        antibody_file_dict[x[\"serum\"]], \n",
    "        x[\"virus\"],\n",
    "    ), axis=1)\n",
    ")\n",
    "\n",
    "# Rename columns and reformat columns\n",
    "validation_df = (\n",
    "    validation_df.rename(columns={\n",
    "        \"ic50_bound\" : \"lower bound\",\n",
    "        \"serum\" : \"antibody\",\n",
    "        \"virus\" : \"Lassa GPC\"\n",
    "    })\n",
    ")\n",
    "validation_df[\"lower bound\"] = (\n",
    "    validation_df[\"lower bound\"].replace({\"interpolated\" : False, \"lower\" : True})\n",
    ")\n",
    "\n",
    "# Calculate correlation between predicted and measured\n",
    "r, p = sp.stats.pearsonr(\n",
    "    x=validation_df[\"log2 ic50 fold change\"], \n",
    "    y=validation_df[\"total_escape\"]\n",
    ")\n",
    "print(f\"R={r}\")\n",
    "print(f\"R^2={r**2}\")\n",
    "\n",
    "# Plot predicted vs measured\n",
    "fig, ax = plt.subplots(figsize=(2.25, 2.25))\n",
    "corr_chart = sns.scatterplot(\n",
    "    data=validation_df,\n",
    "    x=\"log2 ic50 fold change\", \n",
    "    y=\"total_escape\",\n",
    "    hue=\"Lassa GPC\",\n",
    "    palette={\n",
    "        \"unmutated Josiah\" : tol_muted_adjusted[0],\n",
    "        \"GA391\" : tol_muted_adjusted[1],\n",
    "        \"LM395\" : tol_muted_adjusted[3],\n",
    "        \"IRR007\" : tol_muted_adjusted[6],\n",
    "        \"ISTH1024\" : tol_muted_adjusted[8],\n",
    "        \"N89D Josiah\" : tol_muted_adjusted[4],\n",
    "        \"K126N Josiah\" : tol_muted_adjusted[10],\n",
    "        \"E228D Josiah\": tol_muted_adjusted[2],\n",
    "        \"H398K Josiah\" : tol_muted_adjusted[7],\n",
    "        \"D401E Josiah\" : tol_muted_adjusted[5],\n",
    "    },\n",
    "    style=\"antibody\",\n",
    "    markers=[\"o\", \"s\", \"D\", \"v\", \"^\", \"P\"],\n",
    "    alpha=0.8,\n",
    "    edgecolor=\"black\",\n",
    "    ax=ax,\n",
    "    s=40,\n",
    ")\n",
    "# Change all spines\n",
    "for axis in [\"top\", \"bottom\", \"left\", \"right\"]:\n",
    "    corr_chart.spines[axis].set_linewidth(1)\n",
    "sns.despine()\n",
    "corr_chart.tick_params(axis=\"both\", length=4, width=1)\n",
    "corr_chart.set_xlabel(\n",
    "    \"log2 fold change IC$_{50}$ measured\\nby pseuodvirus neutralization assay\", \n",
    "    # weight=\"bold\",\n",
    "    fontsize=8,\n",
    ")\n",
    "corr_chart.set_ylabel(\n",
    "    \"escape score predicted\\nby DMS (arbitrary units)\", \n",
    "    # weight=\"bold\",\n",
    "    fontsize=8,\n",
    ")\n",
    "corr_chart.set_xlim(-1, 12.5)\n",
    "corr_chart.set_ylim(-0.5, 5.5)\n",
    "corr_chart.set_xticks([0, 2, 4, 6, 8, 10, 12])\n",
    "corr_chart.set_xticklabels(corr_chart.get_xticks(), size=8)\n",
    "corr_chart.set_yticks([0, 1, 2, 3, 4, 5])\n",
    "corr_chart.set_yticklabels(corr_chart.get_yticks(), size=8)\n",
    "sns.move_legend(\n",
    "    corr_chart, \n",
    "    \"upper left\", \n",
    "    bbox_to_anchor=(1, 1),\n",
    "    fontsize=8,\n",
    "    markerscale=1,\n",
    "    handletextpad=0.1,\n",
    "    frameon=False,\n",
    "    borderaxespad=0.75,\n",
    ")\n",
    "corr_chart.get_legend().get_texts()[0].set_weight(\"bold\")\n",
    "corr_chart.get_legend().get_texts()[11].set_weight(\"bold\")\n",
    "corr_chart.text(\n",
    "    0, \n",
    "    5, \n",
    "    f\"r={r:.2f}\", \n",
    "    horizontalalignment=\"left\",  \n",
    "    weight=\"bold\",\n",
    "    fontsize=8,\n",
    ")\n",
    "\n",
    "# Add edges to legend markers to match scatter plot\n",
    "for ha in corr_chart.legend_.legendHandles:\n",
    "    ha.set_edgecolor(\"black\")\n",
    "    ha.set_linewidths(0.5)\n",
    "        \n",
    "# Set square ratio\n",
    "corr_chart.set_box_aspect(1)\n",
    "\n",
    "# Make output dir if doesn't exist\n",
    "if not os.path.exists(out_dir):\n",
    "    os.mkdir(out_dir)\n",
    "\n",
    "# Save fig\n",
    "plt.savefig(corr_image_path)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "80d7e3f1-41c7-444f-abc0-dc3320358d4a",
   "metadata": {},
   "source": [
    "## Correlation of antibody escape to functional scores\n",
    "\n",
    "Next, we are going to compare the DMS data to natural sequence diversity. In this case, natural sequence diversity is reflected by calculating effective amino acids at each site based on all high quality available Lassa GPC sequences. The formula for calculating effective amino acids is described in *Biophysical Models of Protein Evolution: Understanding the Patterns of Evolutionary Sequence Divergence*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "198dda5a-be8c-4c8e-89cb-3715f9a8cb66",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load data as dataframe\n",
    "functional_scores = pd.read_csv(func_scores)\n",
    "\n",
    "# Filter for minimum selections, times seen and no stop codons\n",
    "functional_scores = (\n",
    "    functional_scores.query(\n",
    "        \"n_selections >= @n_selections and times_seen >= @min_times_seen and mutant != '*'\"\n",
    "    )\n",
    "    .drop(columns=[\"mutant\", \"times_seen\"])\n",
    "    .groupby([\"site\", \"wildtype\"])\n",
    "    .aggregate({\n",
    "        \"effect\" : \"mean\"\n",
    "    })\n",
    "    .reset_index()\n",
    ")\n",
    "\n",
    "# Load data as dataframe\n",
    "natural_variation = pd.read_csv(natural_sequence_variation)\n",
    "\n",
    "# Drop individual amino acid counts\n",
    "natural_variation = natural_variation[[\"site\", \"entropy\", \"n_effective\"]]\n",
    "\n",
    "# Merge functional and natural dataframes\n",
    "merged_df = (\n",
    "    functional_scores.merge(\n",
    "        natural_variation,\n",
    "        how=\"left\",\n",
    "        on=[\"site\"],\n",
    "        validate=\"one_to_one\",\n",
    "    )\n",
    ")\n",
    "\n",
    "# Add escape to dataframe for each antibody\n",
    "for antibody_file,_ in antibody_files:\n",
    "\n",
    "    antibody_name = antibody_file.split(\"/\")[-1].split(\"_\")[0]\n",
    "\n",
    "    # Load data as dataframe\n",
    "    escape_df = pd.read_csv(antibody_file)\n",
    "\n",
    "    # Clip lower scores to 0\n",
    "    escape_df[\"escape_median\"] = escape_df[\"escape_median\"].clip(lower=0)\n",
    "\n",
    "    # Calculate summed escape sites and get sites for top escape\n",
    "    escape_df = (\n",
    "        escape_df\n",
    "        .groupby(\"site\")\n",
    "        .aggregate({\"escape_median\" : \"sum\"})\n",
    "        .reset_index()\n",
    "    )\n",
    "\n",
    "    # Rename escape column to include antibody name\n",
    "    escape_df = escape_df.rename(columns={\"escape_median\" : \"escape_\" + antibody_name})\n",
    "\n",
    "    # Merge dataframes\n",
    "    merged_df = (\n",
    "        merged_df.merge(\n",
    "            escape_df,\n",
    "            how=\"left\",\n",
    "            on=[\"site\"],\n",
    "            validate=\"one_to_one\",\n",
    "        )\n",
    "    )\n",
    "\n",
    "# Total summed escape per site for all antibodies\n",
    "merged_df[\"total_escape\"] = (\n",
    "    merged_df[[\n",
    "        \"escape_2510C\",\n",
    "        \"escape_121F\",\n",
    "        \"escape_377H\",\n",
    "        \"escape_256A\",\n",
    "        \"escape_372D\",\n",
    "        \"escape_89F\"\n",
    "    ]].median(axis=1)\n",
    ")\n",
    "\n",
    "# Sites of mutations for chosen validations\n",
    "validation_sites = [89, 126, 228, 398, 401]\n",
    "merged_df[\"site of validation\"] = (\n",
    "    merged_df[\"site\"].apply(lambda x: True if x in validation_sites else False)\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a08518c0-646c-4952-a5d8-3000872c7035",
   "metadata": {},
   "source": [
    "First, we look at the correlation of functional effects and antibody escape stratified by the different antibodies. As expected, the mutations that lead to escape tend to be more functionally tolerated. Furthermore, the mutations present in the natural isolates also tend to be functionally tolerated except for the notable exception of N89D which is quite deleterious. \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6c988965-6a04-48dd-b004-f1e4024624e1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Making two lists for values and colors \n",
    "dom = [True, False] \n",
    "rng = [\"#EE7733FF\", \"#00000026\"] \n",
    "\n",
    "subplots = []\n",
    "for index,antibody in enumerate(antibody_files):\n",
    "    antibody_name = antibody[0].split(\"/\")[-1].split(\"_\")[0]\n",
    "\n",
    "    curr_subplot = alt.Chart(merged_df, title=antibody_name).mark_point(\n",
    "        filled=True, \n",
    "        color=\"black\", \n",
    "        size=75\n",
    "    ).encode(\n",
    "        alt.X(\n",
    "            \"escape_\" + antibody_name,\n",
    "            axis=alt.Axis(\n",
    "                title=\"total site escape\", \n",
    "                domainWidth=1,\n",
    "                domainColor=\"black\",\n",
    "                tickColor=\"black\",\n",
    "            ),\n",
    "            scale=alt.Scale(domainMin=-1)\n",
    "        ),\n",
    "        alt.Y(\n",
    "            \"effect\",\n",
    "            axis=alt.Axis(\n",
    "                title=\"effect on cell entry\", \n",
    "                values=[-4,-3,-2,-1,0,1],\n",
    "                domainWidth=1,\n",
    "                domainColor=\"black\",\n",
    "                tickColor=\"black\",\n",
    "            ),\n",
    "            scale=alt.Scale(domain=[-4.5,1])\n",
    "        ),\n",
    "        tooltip=[\n",
    "            \"site\",\n",
    "            \"wildtype\",\n",
    "            \"effect\",\n",
    "            \"escape_\" + antibody_name,\n",
    "        ],\n",
    "        color=alt.Color(\n",
    "            \"site of validation\", \n",
    "            scale=alt.Scale(domain=dom, range=rng),\n",
    "        ), \n",
    "    ).properties(\n",
    "        width=150,\n",
    "        height=150,\n",
    "    )\n",
    "    \n",
    "    subplots.append(curr_subplot)\n",
    "\n",
    "func_vs_antibody = alt.hconcat(\n",
    "    subplots[0],\n",
    "    subplots[1],\n",
    "    subplots[2],\n",
    "    subplots[3],\n",
    "    subplots[4],\n",
    "    subplots[5],\n",
    "    spacing=5,\n",
    "    title=\"Correlations of functional effects and antibody escape\",\n",
    ").configure_axis(\n",
    "    grid=False,\n",
    "    labelFontSize=16,\n",
    "    titleFontSize=16,\n",
    "    labelFontWeight=\"normal\",\n",
    "    titleFontWeight=\"normal\",\n",
    ").configure_title(\n",
    "    fontSize=24,\n",
    ")\n",
    "\n",
    "func_vs_antibody"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c89fc6be-a8ff-4db6-b3f8-a5828d81ae03",
   "metadata": {},
   "source": [
    "## Correlation of antibody escape to natural sequence diversity\n",
    "\n",
    "Next, we look at the correlation of natural sequence diversity and antibody escape stratified by the different antibodies. The antibody 12.1F (and maybe 37.2D to a lesser extent) tends to be weakly escaped by mutations at sites with higher natural diversity. Of the chosen validations, site 126 tends to be most diverse but sites 228 and 398 also have increased diversity."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "754e7f17-e2bd-45b8-8548-75ff7643db59",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Making two lists for values and colors \n",
    "dom = [True, False] \n",
    "rng = [\"#EE7733FF\", \"#00000026\"] \n",
    "\n",
    "subplots = []\n",
    "for index,antibody in enumerate(antibody_files):\n",
    "    antibody_name = antibody[0].split(\"/\")[-1].split(\"_\")[0]\n",
    "\n",
    "    curr_subplot = alt.Chart(merged_df, title=antibody_name).mark_point(\n",
    "        filled=True, \n",
    "        color=\"black\", \n",
    "        size=75\n",
    "    ).encode(\n",
    "        alt.X(\n",
    "            \"escape_\" + antibody_name,\n",
    "            axis=alt.Axis(\n",
    "                title=\"total site escape\", \n",
    "                domainWidth=1,\n",
    "                domainColor=\"black\",\n",
    "                tickColor=\"black\",\n",
    "            ),\n",
    "            scale=alt.Scale(domainMin=-1)\n",
    "        ),\n",
    "        alt.Y(\n",
    "            \"n_effective\",\n",
    "            axis=alt.Axis(\n",
    "                title=[\"effective amino acids\", \"in natural sequences\"], \n",
    "                values=[1,2,3,4],\n",
    "                domainWidth=1,\n",
    "                domainColor=\"black\",\n",
    "                tickColor=\"black\",\n",
    "            ),\n",
    "            scale=alt.Scale(domain=[0.9,4])\n",
    "        ),\n",
    "        tooltip=[\n",
    "            \"site\",\n",
    "            \"wildtype\",\n",
    "            \"n_effective\",\n",
    "            \"escape_\" + antibody_name,\n",
    "        ],\n",
    "        color=alt.Color(\n",
    "            \"site of validation\", \n",
    "            scale=alt.Scale(domain=dom, range=rng),\n",
    "        ), \n",
    "    ).properties(\n",
    "        width=150,\n",
    "        height=150,\n",
    "    )\n",
    "    \n",
    "    subplots.append(curr_subplot)\n",
    "\n",
    "func_vs_antibody = alt.hconcat(\n",
    "    subplots[0],\n",
    "    subplots[1],\n",
    "    subplots[2],\n",
    "    subplots[3],\n",
    "    subplots[4],\n",
    "    subplots[5],\n",
    "    spacing=5,\n",
    "    title=\"Correlations of natural sequence diversity and antibody escape\",\n",
    ").configure_axis(\n",
    "    grid=False,\n",
    "    labelFontSize=16,\n",
    "    titleFontSize=16,\n",
    "    labelFontWeight=\"normal\",\n",
    "    titleFontWeight=\"normal\",\n",
    ").configure_title(\n",
    "    fontSize=24,\n",
    ")\n",
    "\n",
    "func_vs_antibody"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "375375cb-628b-4d7a-a6f2-c768ac82019d",
   "metadata": {},
   "source": [
    "## Correlation of summary metrics across all mapped antibodies\n",
    "\n",
    "Finally, we are going to look at summary correlations for antibody escape, functional effects on cell entry, and natural sequence diversity. The total antibody escape was calculated by summing escape at every site across all antibodies mapped."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1bbe4352-a10d-45db-9412-e2217985042e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Making two lists for values and colors \n",
    "dom = [True, False] \n",
    "rng = [\"#EE7733FF\", \"#00000026\"] \n",
    "\n",
    "# Calculate statistics\n",
    "r, p = sp.stats.pearsonr(\n",
    "    merged_df[[\"total_escape\",\"effect\"]].dropna()[\"total_escape\"],\n",
    "    merged_df[[\"total_escape\",\"effect\"]].dropna()[\"effect\"],\n",
    ")\n",
    "print(f\"r correlation of total escape and effect on cell entry: {r:.2f}\")\n",
    "\n",
    "effect_vs_antibody = alt.Chart(merged_df, title=f\"r = {r:.2f}\").mark_point(\n",
    "    filled=True, \n",
    "    color=\"black\", \n",
    "    size=75,\n",
    ").encode(\n",
    "    alt.X(\n",
    "        \"total_escape\",\n",
    "        axis=alt.Axis(\n",
    "            title=[\"median site escape across\", \"all mapped antibodies\"],  \n",
    "            values=[0,1,2,3,4,5],\n",
    "            domainWidth=1,\n",
    "            domainColor=\"black\",\n",
    "            tickColor=\"black\",\n",
    "        ),\n",
    "        scale=alt.Scale(domain=[-0.1, 5.1])\n",
    "    ),\n",
    "    alt.Y(\n",
    "        \"effect\",\n",
    "        axis=alt.Axis(\n",
    "            title=\"effect on cell entry\", \n",
    "            values=[-4,-3,-2,-1,0,1],\n",
    "            domainWidth=1,\n",
    "            domainColor=\"black\",\n",
    "            tickColor=\"black\",\n",
    "        ),\n",
    "        scale=alt.Scale(domain=[-4.5,1])\n",
    "    ),\n",
    "    tooltip=[\n",
    "        \"site\",\n",
    "        \"wildtype\",\n",
    "        \"effect\",\n",
    "        \"escape_2510C\",\n",
    "        \"escape_121F\",\n",
    "        \"escape_89F\",\n",
    "        \"escape_377H\",\n",
    "        \"escape_372D\",\n",
    "        \"escape_256A\",\n",
    "        \"total_escape\",\n",
    "    ],\n",
    "    color=alt.Color(\n",
    "        \"site of validation\", \n",
    "        scale=alt.Scale(domain=dom, range=rng),\n",
    "    ), \n",
    ").properties(\n",
    "    width=300,\n",
    "    height=300,\n",
    ")\n",
    "\n",
    "r, p = sp.stats.pearsonr(\n",
    "    merged_df[[\"total_escape\",\"n_effective\"]].dropna()[\"total_escape\"],\n",
    "    merged_df[[\"total_escape\",\"n_effective\"]].dropna()[\"n_effective\"],\n",
    ")\n",
    "print(f\"r correlation of total escape and natural sequence diversity: {r:.2f}\")\n",
    "\n",
    "natural_vs_antibody = alt.Chart(merged_df, title=f\"r = {r:.2f}\").mark_point(\n",
    "    filled=True, \n",
    "    color=\"black\", \n",
    "    size=75,\n",
    ").encode(\n",
    "    alt.Y(\n",
    "        \"total_escape\",\n",
    "        axis=alt.Axis(\n",
    "            title=[\"median site escape across\", \"all mapped antibodies\"], \n",
    "            values=[0,1,2,3,4,5],\n",
    "            domainWidth=1,\n",
    "            domainColor=\"black\",\n",
    "            tickColor=\"black\",\n",
    "        ),\n",
    "        scale=alt.Scale(domain=[-0.1, 5.1])\n",
    "    ),\n",
    "    alt.X(\n",
    "        \"n_effective\",\n",
    "        axis=alt.Axis(\n",
    "            title=[\"effective amino acids\", \"in natural sequences\"], \n",
    "            values=[1,2,3,4],\n",
    "            domainWidth=1,\n",
    "            domainColor=\"black\",\n",
    "            tickColor=\"black\",\n",
    "        ),\n",
    "        scale=alt.Scale(domain=[0.9,4])\n",
    "    ),\n",
    "    tooltip=[\n",
    "        \"site\",\n",
    "        \"wildtype\",\n",
    "        \"n_effective\",\n",
    "        \"escape_2510C\",\n",
    "        \"escape_121F\",\n",
    "        \"escape_89F\",\n",
    "        \"escape_377H\",\n",
    "        \"escape_372D\",\n",
    "        \"escape_256A\",\n",
    "        \"total_escape\",\n",
    "    ],\n",
    "    color=alt.Color(\n",
    "        \"site of validation\", \n",
    "        scale=alt.Scale(domain=dom, range=rng),\n",
    "    ), \n",
    ").properties(\n",
    "    width=300,\n",
    "    height=300,\n",
    ")\n",
    "\n",
    "r, p = sp.stats.pearsonr(\n",
    "    merged_df[[\"effect\",\"n_effective\"]].dropna()[\"effect\"],\n",
    "    merged_df[[\"effect\",\"n_effective\"]].dropna()[\"n_effective\"],\n",
    ")\n",
    "print(f\"r correlation of natural sequence diversity and effect on cell entry: {r:.2f}\")\n",
    "\n",
    "natural_vs_func = alt.Chart(merged_df, title=f\"r = {r:.2f}\").mark_point(\n",
    "    filled=True, \n",
    "    color=\"black\", \n",
    "    size=75,\n",
    ").encode(\n",
    "    alt.X(\n",
    "        \"n_effective\",\n",
    "        axis=alt.Axis(\n",
    "            title=[\"effective amino acids\", \"in natural sequences\"], \n",
    "            values=[1,2,3,4],\n",
    "            domainWidth=1,\n",
    "            domainColor=\"black\",\n",
    "            tickColor=\"black\",\n",
    "        ),\n",
    "        scale=alt.Scale(domain=[0.9,4])\n",
    "    ),\n",
    "    alt.Y(\n",
    "        \"effect\",\n",
    "        axis=alt.Axis(\n",
    "            title=\"effect on cell entry\", \n",
    "            values=[-4,-3,-2,-1,0,1],\n",
    "            domainWidth=1,\n",
    "            domainColor=\"black\",\n",
    "            tickColor=\"black\",\n",
    "        ),\n",
    "        scale=alt.Scale(domain=[-4.5,1])\n",
    "    ),\n",
    "    tooltip=[\n",
    "        \"site\",\n",
    "        \"wildtype\",\n",
    "        \"effect\",\n",
    "        \"n_effective\",\n",
    "        \"escape_2510C\",\n",
    "        \"escape_121F\",\n",
    "        \"escape_89F\",\n",
    "        \"escape_377H\",\n",
    "        \"escape_372D\",\n",
    "        \"escape_256A\",\n",
    "        \"total_escape\",\n",
    "    ],\n",
    "    color=alt.Color(\n",
    "        \"site of validation\", \n",
    "        scale=alt.Scale(domain=dom, range=rng),\n",
    "    ), \n",
    ").properties(\n",
    "    width=300,\n",
    "    height=300,\n",
    ")\n",
    "\n",
    "summary_corr_plot = alt.hconcat(\n",
    "    effect_vs_antibody,\n",
    "    natural_vs_antibody,\n",
    "    natural_vs_func,\n",
    "    spacing=5,\n",
    "    title=\"Correlations of summary statistics\",\n",
    ").configure_axis(\n",
    "    grid=False,\n",
    "    labelFontSize=16,\n",
    "    titleFontSize=16,\n",
    "    labelFontWeight=\"normal\",\n",
    "    titleFontWeight=\"normal\",\n",
    ").configure_title(\n",
    "    fontSize=24,\n",
    ").configure_legend(\n",
    "    labelFontSize=16,\n",
    "    titleFontSize=16,\n",
    ")\n",
    "\n",
    "summary_corr_plot"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
