{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "e586ae3d-a6c0-49b4-87c4-65a24e82ac04",
   "metadata": {},
   "source": [
    "# Compare DMS data to natural sequence data\n",
    "\n",
    "This notebook analyzes natural sequences for antibody escape predicted by DMS and shows the resulting neutralization validations. Additional analysis is performed by correlating natural sequence diversity with DMS data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ebdfd791-e188-4e4e-8aff-601433db2aee",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Imports\n",
    "import os\n",
    "import warnings\n",
    "import dmslogo\n",
    "import neutcurve\n",
    "import numpy as np\n",
    "import scipy as sp\n",
    "import pandas as pd\n",
    "import altair as alt\n",
    "import seaborn as sns\n",
    "import matplotlib.colors\n",
    "import matplotlib.pyplot as plt\n",
    "from Bio import SeqIO, AlignIO \n",
    "\n",
    "# Plotting colors\n",
    "tol_muted_adjusted = [\n",
    "    \"#000000\",\n",
    "    \"#CC6677\", \n",
    "    \"#1f78b4\", \n",
    "    \"#DDCC77\", \n",
    "    \"#117733\", \n",
    "    \"#882255\", \n",
    "    \"#88CCEE\",\n",
    "    \"#44AA99\", \n",
    "    \"#999933\", \n",
    "    \"#AA4499\", \n",
    "    \"#EE7733\",\n",
    "    \"#CC3311\",\n",
    "    \"#DDDDDD\",\n",
    "]\n",
    "\n",
    "# Create color palette\n",
    "def color_gradient_hex(start, end, n):\n",
    "    \"\"\"Color function from polyclonal\"\"\"\n",
    "    cmap = matplotlib.colors.LinearSegmentedColormap.from_list(\n",
    "            name=\"_\", colors=[start, end], N=n\n",
    "        )\n",
    "    return [matplotlib.colors.rgb2hex(tup) for tup in cmap(list(range(0, n)))]\n",
    "\n",
    "# Seaborn style settings\n",
    "sns.set(rc={\"figure.dpi\":300, \"savefig.dpi\":300})\n",
    "sns.set_style(\"ticks\")\n",
    "sns.set_palette(tol_muted_adjusted)\n",
    "\n",
    "# Suppress warnings\n",
    "warnings.simplefilter(\"ignore\")\n",
    "\n",
    "# Allow more rows for Altair\n",
    "_ = alt.data_transformers.disable_max_rows()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "46eac608-f320-4cbe-bc08-d88609385d17",
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": ""
    },
    "tags": [
     "parameters"
    ]
   },
   "outputs": [],
   "source": [
    "# this cell is tagged as `parameters` for papermill parameterization\n",
    "filtered_escape_377H = None\n",
    "filtered_escape_89F = None\n",
    "filtered_escape_2510C = None\n",
    "filtered_escape_121F = None\n",
    "filtered_escape_256A = None\n",
    "filtered_escape_372D = None\n",
    "\n",
    "contacts_89F = None\n",
    "contacts_377H = None\n",
    "contacts_256A = None\n",
    "contacts_2510C = None\n",
    "contacts_121F = None\n",
    "contacts_372D = None\n",
    "\n",
    "func_scores = None\n",
    "min_times_seen = None\n",
    "n_selections = None\n",
    "\n",
    "natural_sequence_variation = None\n",
    "natural_GPC_sequence_alignment = None\n",
    "\n",
    "fraction_infected_natural_isolates = None\n",
    "out_dir = None\n",
    "neuts_image_path = None\n",
    "corr_image_path = None\n",
    "\n",
    "out_dir_escape = None\n",
    "escape_top10_image_path = None\n",
    "escape_all_image_path = None\n",
    "\n",
    "out_dir_natural = None\n",
    "html_output = None\n",
    "natural_escape = None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ca3d4132-3ddd-421b-8236-d79bbbc0578a",
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": ""
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "# # Uncomment for running interactive\n",
    "# filtered_escape_377H = \"../results/filtered_antibody_escape_CSVs/377H_filtered_mut_effect.csv\"\n",
    "# filtered_escape_89F = \"../results/filtered_antibody_escape_CSVs/89F_filtered_mut_effect.csv\"\n",
    "# filtered_escape_2510C = \"../results/filtered_antibody_escape_CSVs/2510C_filtered_mut_effect.csv\"\n",
    "# filtered_escape_121F = \"../results/filtered_antibody_escape_CSVs/121F_filtered_mut_effect.csv\"\n",
    "# filtered_escape_256A = \"../results/filtered_antibody_escape_CSVs/256A_filtered_mut_effect.csv\"\n",
    "# filtered_escape_372D = \"../results/filtered_antibody_escape_CSVs/372D_filtered_mut_effect.csv\"\n",
    "\n",
    "# contacts_89F = \"../data/antibody_contacts/antibody_contacts_89F.csv\"\n",
    "# contacts_377H = \"../data/antibody_contacts/antibody_contacts_377H.csv\"\n",
    "# contacts_256A = \"../data/antibody_contacts/antibody_contacts_256A.csv\"\n",
    "# contacts_2510C = \"../data/antibody_contacts/antibody_contacts_2510C.csv\"\n",
    "# contacts_121F = \"../data/antibody_contacts/antibody_contacts_121F.csv\"\n",
    "# contacts_372D = \"../data/antibody_contacts/antibody_contacts_372D.csv\"\n",
    "\n",
    "# func_scores = \"../results/func_effects/averages/293T_entry_func_effects.csv\"\n",
    "# min_times_seen = 2\n",
    "# n_selections = 8\n",
    "\n",
    "# natural_sequence_variation = \"../non-pipeline_analyses/LASV_phylogeny_analysis/Results/GPC_protein_variation.csv\"\n",
    "# natural_GPC_sequence_alignment = \"../non-pipeline_analyses/LASV_phylogeny_analysis/Results/LASV_GPC_protein_alignment.fasta\"\n",
    "\n",
    "# fraction_infected_natural_isolates = \"../data/validation_frac_infected_natural_isolates.csv\"\n",
    "# out_dir = \"../results/validation_plots/\"\n",
    "# neuts_image_path = \"../results/validation_plots/validation_neut_curves_natural_isolates.svg\"\n",
    "# corr_image_path = \"../results/validation_plots/natural_isolate_validation_correlation.svg\"\n",
    "\n",
    "# out_dir_escape = \"../results/antibody_escape_profiles/\"\n",
    "# escape_top10_image_path = \"../results/antibody_escape_profiles/natural_isolate_top10_escape_profiles.svg\"\n",
    "# escape_all_image_path = \"../results/antibody_escape_profiles/natural_isolate_all_escape_profiles.svg\"\n",
    "\n",
    "# out_dir_natural = \"../results/natural_isolate_escape/\"\n",
    "# html_output = \"../results/natural_isolate_escape/natural_isolate_escape.html\"\n",
    "# natural_escape = \"../results/natural_isolate_escape/natural_isolate_escape.svg\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "95440d5a-e264-4630-82ef-984aba0eae5f",
   "metadata": {},
   "source": [
    "## Identify natural sequences that would potentially escape antibody neutralization\n",
    "\n",
    "To identify any natural isolates that could potentially escape any of the mapped antibodies, we want to identify high confidence escape mutations. We are going to use the following filters to identiy high confidence escape mutations: \n",
    "1. First, we are going to filter for the top 5% escape mutants for each antibody selection\n",
    "2. Second, we are going to further filter the list of mutations by keeping only mutations present in sites of  **strong escape** (e.g., summed escape sites 10 fold greater than the median of all summed sites)\n",
    "3. Finally, we are going to search all high quality sequences for any of the mutations identified as high confidence escape mutations.\n",
    "\n",
    "Other details of note are that escape scores are clipped at the lower end of 0 to focus on escape mutations rather than sensitizing mutations. Any sequence with these mutations will be flagged as potential escape isolates. This conservative method will filter many escape mutations but should result in high confidence mutations."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7ddf6b27-d5d2-4f67-874d-d93a0ede9ed2",
   "metadata": {},
   "outputs": [],
   "source": [
    "def determine_escape(percentile_escape, sequence, escape_file, strain, print_results=False, strong_escape_site=False):\n",
    "    \"\"\"\n",
    "    Function that determines if a sequence contains any \n",
    "    escape mutations given a percentile cutoff.\n",
    "    \"\"\"\n",
    "\n",
    "    antibody_name = escape_file.split(\"/\")[-1].split(\"_\")[0]\n",
    "    \n",
    "    # Load data as dataframe\n",
    "    escape_df = pd.read_csv(escape_file)\n",
    "\n",
    "    # Clip lower scores to 0\n",
    "    escape_df[\"escape_median\"] = escape_df[\"escape_median\"].clip(lower=0)\n",
    "\n",
    "    # Get muts for top escape\n",
    "    cutoff = escape_df[\"escape_median\"].quantile(percentile_escape)\n",
    "    top_escape_muts = (\n",
    "        tuple(zip(\n",
    "            escape_df.loc[escape_df[\"escape_median\"] >= cutoff][\"mutation\"].tolist(),\n",
    "            escape_df.loc[escape_df[\"escape_median\"] >= cutoff][\"escape_median\"].tolist(),\n",
    "        ))\n",
    "    )\n",
    "\n",
    "    # Calculate summed escape sites and get sites for top escape\n",
    "    # either by calculating 10-fold median or top percentile\n",
    "    escape_df = (\n",
    "        escape_df\n",
    "        .groupby(\"site\")\n",
    "        .aggregate({\"escape_median\" : \"sum\"})\n",
    "        .reset_index()\n",
    "    )\n",
    "    if strong_escape_site == True:\n",
    "        cutoff = escape_df[\"escape_median\"].median() * 10\n",
    "    else:\n",
    "        cutoff = escape_df[\"escape_median\"].quantile(percentile_escape)\n",
    "    top_escape_sites = escape_df.loc[escape_df[\"escape_median\"] >= cutoff][\"site\"].tolist()\n",
    "\n",
    "    # Filter top escape muts bases on top escape sites\n",
    "    top_escape_muts = [x for x in top_escape_muts if int(x[0][1:-1]) in top_escape_sites]\n",
    "    \n",
    "    # Initialize escape flag\n",
    "    escape = 0\n",
    "    \n",
    "    # Iterate through list of escape mutations\n",
    "    for escape_mut in top_escape_muts:\n",
    "        site = int(escape_mut[0][1:-1])\n",
    "        if sequence[site-1] == escape_mut[0][-1]:\n",
    "            if print_results:\n",
    "                print(f\"{strain:<75} with \\t {escape_mut[0][0]}{site}{escape_mut[0][-1]} \\t DMS score: {escape_mut[1]}\")\n",
    "            escape += escape_mut[1]\n",
    "   \n",
    "    return escape\n",
    "\n",
    "# Load alignment and metadata info\n",
    "natural_seqs_df = pd.DataFrame(columns=[\"strain\", \"sequence\"])\n",
    "\n",
    "# Add alignment sequence to dataframe\n",
    "for curr_fasta in AlignIO.read(natural_GPC_sequence_alignment, \"fasta\"):\n",
    "    natural_seqs_df.loc[len(natural_seqs_df.index)] = [\n",
    "        str(curr_fasta.id),\n",
    "        str(curr_fasta.seq),\n",
    "    ] \n",
    "\n",
    "# Antibody escape dataframes and percentile cutoffs to use\n",
    "antibody_files = [\n",
    "    (filtered_escape_2510C, 0.95),\n",
    "    (filtered_escape_121F, 0.95), \n",
    "    (filtered_escape_377H, 0.95),\n",
    "    (filtered_escape_256A, 0.95),\n",
    "    (filtered_escape_372D, 0.95),\n",
    "    (filtered_escape_89F, 0.95), \n",
    "]\n",
    "\n",
    "for antibody, percentile in antibody_files:\n",
    "    antibody_name = antibody.split(\"/\")[-1].split(\"_\")[0]\n",
    "    print(f\"{antibody_name} potentially escaped by:\")\n",
    "    natural_seqs_df.apply(lambda x: determine_escape(\n",
    "        percentile, \n",
    "        x[\"sequence\"], \n",
    "        antibody, \n",
    "        x[\"strain\"],\n",
    "        print_results=True,\n",
    "        strong_escape_site=True,\n",
    "    ), axis=1)\n",
    "    print()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "956ca791-89b9-44cb-b790-6db29ffadc95",
   "metadata": {},
   "source": [
    "The following were chosen for validation:\n",
    "- 8.9F\n",
    "    - Natural isolate: Lassa_virus_H-sapiens-wt_NGA_2018_ISTH_1024_MH157037_2018-02-14\n",
    "        - Corresponding single mutant: K126N (DMS score: 0.7906)\n",
    "- 12.1F\n",
    "    - Natural isolate: LM395-SLE-2009_KM822115_2009-XX-XX\n",
    "        - Corresponding single mutant: N89D (DMS score: 2.515)\n",
    "- 25.10C\n",
    "    - Natural isolate: GA391_OL774861_reverse_complement_1977-XX-XX \n",
    "        - Corresponding single mutant: E228D (DMS score: 3.098)\n",
    "- 37.7H\n",
    "    - Natural isolate: GA391_OL774861_reverse_complement_1977-XX-XX \n",
    "        - Corresponding single mutant: H398K (DMS score: 2.949)\n",
    "    - Natural isolate: LASV_H-sapiens-tc_NGA_2016_IRR_007_MK107922_2016-01-18\n",
    "        - Corresponding single mutant: D401E (DMS score: 1.3)\n",
    "- 37.2D\n",
    "    - Natural isolate: GA391_OL774861_reverse_complement_1977-XX-XX\n",
    "        - Corresponding single mutant: H398K (DMS score: 1.078)\n",
    "\n",
    "*The D401E mutation present in the chosen natural isolates did not have high confidence in the DMS data for the 25.6A antibody selection so 25.6A will be omitted from validation neutralization assays. In addition, it has been noted that antibodies 37.7H and 25.6A are very similar so both antibodies would be expected to have similar results."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b2b93fa5-3bff-4fda-9729-5760a90424ff",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Isolates chosen to validate for escape \n",
    "chosen_isolates = [\n",
    "    \"Josiah_NC_004296_reverse_complement_2018-08-13\",\n",
    "    \"GA391_OL774861_reverse_complement_1977-XX-XX\",\n",
    "    \"LASV_H-sapiens-tc_NGA_2016_IRR_007_MK107922_2016-01-18\",\n",
    "    \"Lassa_virus_H-sapiens-wt_NGA_2018_ISTH_1024_MH157037_2018-02-14\",\n",
    "    \"LM395-SLE-2009_KM822115_2009-XX-XX\",\n",
    "]\n",
    "\n",
    "# Create subset of df for chosen isolates\n",
    "validation_isolates = (\n",
    "    natural_seqs_df.loc[natural_seqs_df[\"strain\"].isin(chosen_isolates)]\n",
    "    .reset_index(drop=True)\n",
    ")\n",
    "\n",
    "# Get Josiah sequence for comparison\n",
    "josiah_sequence = validation_isolates.loc[validation_isolates[\"strain\"] == \"Josiah_NC_004296_reverse_complement_2018-08-13\"].at[0,\"sequence\"]\n",
    "\n",
    "# Rename dictionary for virus names\n",
    "rename_dict = {\n",
    "    \"WT\" : \"unmutated Josiah strain\",\n",
    "    \"OL774861\" : \"GA391 strain\",\n",
    "    \"KM822115\" : \"LM395 strain\",\n",
    "    \"MK107922\" : \"IRR007 strain\",\n",
    "    \"MH157037\" : \"ISTH1024 strain\",\n",
    "    \"Josiah_NC_004296_reverse_complement_2018-08-13\" : \"unmutated Josiah strain\",\n",
    "    \"GA391_OL774861_reverse_complement_1977-XX-XX\" : \"GA391 strain\",\n",
    "    \"LASV_H-sapiens-tc_NGA_2016_IRR_007_MK107922_2016-01-18\" : \"IRR007 strain\",\n",
    "    \"Lassa_virus_H-sapiens-wt_NGA_2018_ISTH_1024_MH157037_2018-02-14\" : \"ISTH1024 strain\",\n",
    "    \"LM395-SLE-2009_KM822115_2009-XX-XX\" : \"LM395 strain\",\n",
    "    \"N89D\" : \"N89D Josiah strain\",\n",
    "    \"K126N\" : \"K126N Josiah strain\",\n",
    "    \"E228D\" : \"E228D Josiah strain\",\n",
    "    \"H398K\" : \"H398K Josiah strain\",\n",
    "    \"D401E\" : \"D401E Josiah strain\",\n",
    "}\n",
    "\n",
    "# Rename viruses and column name\n",
    "validation_isolates[\"strain\"] = (\n",
    "    validation_isolates[\"strain\"].replace(rename_dict)    \n",
    ")\n",
    "\n",
    "# Get counts of variants for each validation site\n",
    "site_228_count = 0\n",
    "site_89_count = 0\n",
    "site_398_count = 0\n",
    "site_401_count = 0\n",
    "site_126_count = 0\n",
    "site_228D_count = 0\n",
    "site_89D_count = 0\n",
    "site_398K_count = 0\n",
    "site_401E_count = 0\n",
    "site_126N_count = 0\n",
    "all_position_variants = [0] * 491\n",
    "for seq in natural_seqs_df[\"sequence\"].tolist():\n",
    "    if seq[227] != \"E\":\n",
    "        site_228_count += 1\n",
    "        if seq[227] == \"D\":\n",
    "            site_228D_count += 1\n",
    "    if seq[88] != \"N\":\n",
    "        site_89_count += 1\n",
    "        if seq[88] == \"D\":\n",
    "            site_89D_count += 1\n",
    "    if seq[397] != \"H\":\n",
    "        site_398_count += 1\n",
    "        if seq[397] == \"K\":\n",
    "            site_398K_count += 1\n",
    "    if seq[400] != \"D\":\n",
    "        site_401_count += 1\n",
    "        if seq[400] == \"E\":\n",
    "            site_401E_count += 1\n",
    "    if seq[125] != \"K\":\n",
    "        site_126_count += 1\n",
    "        if seq[125] == \"N\":\n",
    "            site_126N_count += 1\n",
    "    for i in range(491):\n",
    "        if seq[i] != josiah_sequence[i] and seq[i] != \"-\":\n",
    "            all_position_variants[i] += 1\n",
    "    \n",
    "# Print counts of each variant for chosen validations\n",
    "print(f\"Total number of sequences analyzed: {len(natural_seqs_df['sequence'].tolist())}\")\n",
    "print(f\"Total number of non-Josiah strain variants at position 228: {site_228_count} (E228D variants: {site_228D_count})\")\n",
    "print(f\"Total number of non-Josiah strain variants at position 89: {site_89_count} (N89D variants: {site_89D_count})\")\n",
    "print(f\"Total number of non-Josiah strain variants at position 398: {site_398_count} (H398K variants: {site_398K_count})\")\n",
    "print(f\"Total number of non-Josiah strain variants at position 401: {site_401_count} (D401E variants: {site_401E_count})\")\n",
    "print(f\"Total number of non-Josiah strain variants at position 126: {site_126_count} (K126N variants: {site_126N_count})\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9cc12ace-7f42-40bc-9654-8797bfb6f18f",
   "metadata": {},
   "source": [
    "## Escape profiles for the sites that differ in the chosen isolates from the Josiah DMS strain\n",
    "\n",
    "Create escape logo plots for all sites that differ between isolate and Josiah strain as well as top 10 total escape sites that differ between isolate and Josiah."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "007859fe-c077-46ae-a1ba-913cc77d1e50",
   "metadata": {},
   "outputs": [],
   "source": [
    "def create_line_and_logoplots(\n",
    "    escape_file,  \n",
    "    contacts_file,\n",
    "    func_scores, \n",
    "    min_times_seen,  \n",
    "    n_selections, \n",
    "    logo_plot,\n",
    "    output_file = None,\n",
    "    sites = None,\n",
    "    name = None,\n",
    "    amino_acids_to_color = None,\n",
    "    only_top_10_sites=False,\n",
    "):\n",
    "    \"\"\"\n",
    "    Function that filters and writes an antibody escape csv.\n",
    "    Also creates summed escape profiles and logoplots.\n",
    "    \"\"\"\n",
    "\n",
    "    antibody_name = escape_file.split(\"/\")[-1].split(\"_\")[0]\n",
    "\n",
    "    # Read data\n",
    "    escape_df = pd.read_csv(escape_file)\n",
    "    func_scores = pd.read_csv(func_scores)\n",
    "    contacts_df = pd.read_csv(contacts_file)\n",
    "    \n",
    "    # Create mutation column to match antibody df\n",
    "    func_scores[\"site\"] = func_scores[\"site\"].astype(str)\n",
    "    func_scores[\"mutation\"] = func_scores[\"wildtype\"] + func_scores[\"site\"] + func_scores[\"mutant\"]\n",
    "\n",
    "    # Clip lower scores to 0\n",
    "    escape_df[\"escape_median\"] = escape_df[\"escape_median\"].clip(lower=0)\n",
    "\n",
    "    # Summed escape to get top escape sites to show\n",
    "    summed_df = (\n",
    "        escape_df\n",
    "        .groupby([\"site\", \"wildtype\"])\n",
    "        .aggregate({\n",
    "            \"escape_median\" : \"sum\",\n",
    "        })\n",
    "        .rename(columns={\"escape_median\" : \"site_escape\"})\n",
    "        .reset_index()\n",
    "    )\n",
    "    \n",
    "    # Top escape sites for each antibody combined to show\n",
    "    if sites == None:\n",
    "        sites = sorted(summed_df.nlargest(18, \"site_escape\")[\"site\"].tolist())\n",
    "    # Filter for top 10 sites if true\n",
    "    if only_top_10_sites == True:\n",
    "        summed_score_dict = dict(zip(summed_df[\"site\"].tolist(), summed_df[\"site_escape\"]))\n",
    "        filtered_sites = (\n",
    "            [x[0] for x in sorted([(x, summed_score_dict[x]) for x in sites], key = lambda x: x[1], reverse=True)[0:10]]\n",
    "        )\n",
    "        escape_df[\"show_site\"] = escape_df.apply(lambda x: True if (x[\"site\"] in sites and x[\"site\"] in filtered_sites) else False, axis=1)\n",
    "    else:\n",
    "        escape_df[\"show_site\"] = escape_df.apply(lambda x: True if x[\"site\"] in sites else False, axis=1)\n",
    "    \n",
    "    # Shade contact sites in logo plot\n",
    "    shade_sites = list(contacts_df.loc[contacts_df[\"distance\"] == 4][\"position\"].unique())\n",
    "    \n",
    "    # **\n",
    "    # # Uncomment to show antibody contacts\n",
    "    # print(antibody_name)\n",
    "    # print(f\"Contact sites: {shade_sites}\")\n",
    "    # print()\n",
    "    # **\n",
    "    \n",
    "    escape_df[\"shade_site\"] = escape_df.apply(lambda x: \"#DDDDDD\" if x[\"site\"] in shade_sites else None, axis=1)\n",
    "    escape_df[\"shade_alpha\"] = 0.75\n",
    "    escape_df = (\n",
    "        escape_df.merge(\n",
    "            summed_df,\n",
    "            how=\"left\",\n",
    "            on=[\"site\", \"wildtype\"],\n",
    "            validate=\"many_to_one\",\n",
    "        )\n",
    "    )\n",
    "\n",
    "    # Drop extra columns\n",
    "    escape_df = (\n",
    "        escape_df.drop(\n",
    "            escape_df.columns.difference([\n",
    "                \"site\", \n",
    "                \"wildtype\",\n",
    "                \"mutant\", \n",
    "                \"escape_median\", \n",
    "                \"show_site\", \n",
    "                \"shade_site\",\n",
    "                \"shade_alpha\",\n",
    "                \"site_escape\", \n",
    "            ]), axis=1)\n",
    "    )\n",
    "\n",
    "    # Fill in missing sites\n",
    "    seen_list = [False]*491\n",
    "    for index in range(len(escape_df.index)):\n",
    "        site = escape_df.at[index, \"site\"] - 1 \n",
    "        seen_list[site] = True\n",
    "    for index, seen in enumerate(seen_list):\n",
    "        if seen == False:\n",
    "            site = index + 1\n",
    "            # Add missing sites\n",
    "            escape_df.loc[len(escape_df.index)] = [\n",
    "                site,\n",
    "                \"X\",\n",
    "                \"X\",\n",
    "                0,\n",
    "                False,\n",
    "                None,\n",
    "                None,\n",
    "                0,\n",
    "            ]\n",
    "\n",
    "    # Sort by site\n",
    "    escape_df = (\n",
    "        escape_df\n",
    "        .sort_values(by=\"site\")\n",
    "        .astype({\"mutant\" : \"str\"})\n",
    "        .reset_index(drop=True)\n",
    "    )\n",
    "\n",
    "    # Merge functional and escape dfs\n",
    "    func_scores[\"site\"] = func_scores[\"site\"].astype(\"int\")\n",
    "    func_scores = func_scores.loc[func_scores[\"mutant\"] != \"*\"] # remove stop codons\n",
    "    func_scores[\"effect\"] = func_scores[\"effect\"].clip(upper=0, lower=-2) # clip scores \n",
    "    escape_df = (\n",
    "        escape_df.merge(\n",
    "            func_scores,\n",
    "            how=\"left\",\n",
    "            on=[\"site\", \"wildtype\", \"mutant\"],\n",
    "            validate=\"one_to_one\",\n",
    "        )\n",
    "    )\n",
    "    escape_df[\"effect\"] = escape_df[\"effect\"].fillna(-2) # missing functional values are filled as -4 to make less visible\n",
    "    \n",
    "    # Color specified mutants or color by functional effect\n",
    "    if amino_acids_to_color != None:\n",
    "        \n",
    "        escape_df[\"color\"] = (\n",
    "            escape_df.apply(\n",
    "                lambda x: \"#EE7733\" if (x[\"site\"], x[\"mutant\"]) in amino_acids_to_color else \"#000000\", axis=1\n",
    "            )  \n",
    "        )\n",
    "    else:\n",
    "        # Add color column for logo plots\n",
    "        func_color_map = dmslogo.colorschemes.ValueToColorMap(\n",
    "            minvalue=func_scores[\"effect\"].min(),\n",
    "            maxvalue=func_scores[\"effect\"].max(),\n",
    "            cmap=matplotlib.colors.ListedColormap(color_gradient_hex(\"white\", \"#000000\", n=20))\n",
    "        )\n",
    "        escape_df = (\n",
    "            escape_df.assign(\n",
    "                color=lambda x: x[\"effect\"].map(func_color_map.val_to_color)\n",
    "            )\n",
    "        )\n",
    "\n",
    "    # Add wildtype to each site for logo plot\n",
    "    escape_df[\"wt_site\"] = escape_df[\"wildtype\"] + escape_df[\"site\"].map(str)\n",
    "\n",
    "    # Set ylim for each antibody\n",
    "    fixed_ymin = None\n",
    "    fixed_ymax = None\n",
    "    if antibody_name == \"2510C\":\n",
    "        fixed_ymin = -6.25\n",
    "        fixed_ymax = 56.25\n",
    "    elif antibody_name == \"121F\":\n",
    "        fixed_ymin = -1.25\n",
    "        fixed_ymax = 11.25\n",
    "    elif antibody_name == \"377H\" and len(sites) == 33:\n",
    "        fixed_ymin = -1.25\n",
    "        fixed_ymax = 11.25\n",
    "    elif antibody_name == \"377H\" and len(sites) == 31:\n",
    "        fixed_ymin = -3.75 \n",
    "        fixed_ymax = 33.75\n",
    "    elif antibody_name == \"256A\":\n",
    "        fixed_ymin = -5 \n",
    "        fixed_ymax = 45\n",
    "    elif antibody_name == \"372D\":\n",
    "        fixed_ymin = -0.625\n",
    "        fixed_ymax = 5.625\n",
    "    elif antibody_name == \"89F\":\n",
    "        fixed_ymin = -1.25\n",
    "        fixed_ymax = 11.25\n",
    "    else:\n",
    "        print(\"Error! No ylims set!\")\n",
    "    \n",
    "\n",
    "    _, logoplot = dmslogo.draw_logo(\n",
    "        escape_df.query(\"show_site == True\"),\n",
    "        x_col=\"site\",\n",
    "        letter_col=\"mutant\",\n",
    "        letter_height_col=\"escape_median\",\n",
    "        ax=logo_plot,\n",
    "        xtick_col=\"wt_site\",\n",
    "        color_col=\"color\",\n",
    "        shade_color_col=\"shade_site\",\n",
    "        shade_alpha_col=\"shade_alpha\",\n",
    "        draw_line_at_zero=\"never\",\n",
    "        fixed_ymin=fixed_ymin,\n",
    "        fixed_ymax=fixed_ymax,\n",
    "    )\n",
    "\n",
    "    logoplot.set(ylabel=None, xlabel=None)\n",
    "    x_labels = logoplot.get_xticklabels()\n",
    "    logoplot.set_xticklabels(labels=x_labels, rotation=90, horizontalalignment=\"center\", fontsize=6)\n",
    "    # Change all spines\n",
    "    for axis in [\"top\", \"bottom\", \"left\", \"right\"]:\n",
    "        logoplot.spines[axis].set_linewidth(1)\n",
    "    logoplot.tick_params(axis=\"both\", length=2, width=1, pad=1)\n",
    "\n",
    "\n",
    "    # Set antibody specific y axis ticks\n",
    "    if antibody_name == \"2510C\":\n",
    "        yticks = [0, 25, 50]\n",
    "        logoplot.set_yticks(yticks)\n",
    "        logoplot.set_yticklabels(labels=[\"0\", \"25\", \"50\"], fontsize=6)\n",
    "        logoplot.set_title(\n",
    "            antibody_name[0:2] + \".\" + antibody_name[2:] + \" escape for sites different between Josiah and \" + name + \"s\", \n",
    "            fontsize=8,\n",
    "            color=\"#44AA99\",\n",
    "        )\n",
    "    if antibody_name == \"121F\":\n",
    "        yticks = [0, 5, 10]\n",
    "        logoplot.set_yticks(yticks)\n",
    "        logoplot.set_yticklabels(labels=[\"0\", \"5\", \"10\"], fontsize=6)\n",
    "        logoplot.set_title(\n",
    "            antibody_name[0:2] + \".\" + antibody_name[2:] + \" escape for sites different between Josiah and \" + name + \"s\", \n",
    "            fontsize=8,\n",
    "            color=\"#999933\",\n",
    "        )\n",
    "    if antibody_name == \"377H\" and len(sites) == 33:\n",
    "        yticks = [0, 5, 10]\n",
    "        logoplot.set_yticks(yticks)\n",
    "        logoplot.set_yticklabels(labels=[\"0\", \"5\", \"10\"], fontsize=6)\n",
    "        logoplot.set_title(\n",
    "            antibody_name[0:2] + \".\" + antibody_name[2:] + \" escape for sites different between Josiah and \" + name + \"s\", \n",
    "            fontsize=8,\n",
    "            color=\"#AA4499\",\n",
    "        )\n",
    "    if antibody_name == \"377H\" and len(sites) == 31:\n",
    "        yticks = [0, 15, 30]\n",
    "        logoplot.set_yticks(yticks)\n",
    "        logoplot.set_yticklabels(labels=[\"0\", \"15\", \"30\"], fontsize=6)\n",
    "        logoplot.set_title(\n",
    "            antibody_name[0:2] + \".\" + antibody_name[2:] + \" escape for sites different between Josiah and \" + name + \"s\", \n",
    "            fontsize=8,\n",
    "            color=\"#AA4499\",\n",
    "        )\n",
    "    if antibody_name == \"256A\":\n",
    "        yticks = [0, 20, 40]\n",
    "        logoplot.set_yticks(yticks)\n",
    "        logoplot.set_yticklabels(labels=[\"0\", \"20\", \"40\"], fontsize=6)\n",
    "        logoplot.set_title(\n",
    "            antibody_name[0:2] + \".\" + antibody_name[2:] + \" escape for sites different between Josiah and \" + name + \"s\", \n",
    "            fontsize=8,\n",
    "            color=\"#AA4499\",\n",
    "        )\n",
    "    if antibody_name == \"372D\":\n",
    "        yticks = [0, 2.5, 5]\n",
    "        logoplot.set_yticks(yticks)\n",
    "        logoplot.set_yticklabels(labels=[\"0\", \"2.5\", \"5\"], fontsize=6)\n",
    "        logoplot.set_title(\n",
    "            antibody_name[0:2] + \".\" + antibody_name[2:] + \" escape for sites different between Josiah and \" + name + \"s\", \n",
    "            fontsize=8,\n",
    "            color=\"#AA4499\",\n",
    "        )\n",
    "    if antibody_name == \"89F\":\n",
    "        yticks = [0, 5, 10]\n",
    "        logoplot.set_yticks(yticks)\n",
    "        logoplot.set_yticklabels(labels=[\"0\", \"5\", \"10\"], fontsize=6)\n",
    "        logoplot.set_title(\n",
    "            antibody_name[0] + \".\" + antibody_name[1:] + \" escape for sites different between Josiah and \" + name + \"s\",\n",
    "            fontsize=8,\n",
    "            color=\"#117733\",\n",
    "        )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bdcbe358-00d1-41a4-afb0-95c5c721fc92",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_site_differences(seq1, seq2):\n",
    "    \"\"\"\n",
    "    Function that returns a list of all sites\n",
    "    that are different between two sequences.\n",
    "    \"\"\"\n",
    "    list_of_sites = []\n",
    "    list_of_muts = []\n",
    "    site = 1\n",
    "    for s1, s2 in zip(seq1, seq2):\n",
    "        if s1 != s2 and s1 != \"-\" and s2 != \"-\":\n",
    "            list_of_sites.append(site)\n",
    "            list_of_muts.append(s2)\n",
    "        site += 1\n",
    "    return list_of_sites, tuple(zip(list_of_sites, list_of_muts))\n",
    "\n",
    "# Get list of sequence differences compared to Josiah\n",
    "list_of_different_sites = []\n",
    "for index in validation_isolates.index:\n",
    "    curr_seq_name = validation_isolates.at[index, \"strain\"]\n",
    "    curr_seq = validation_isolates.at[index, \"sequence\"]\n",
    "    site_differences, mut_differences = get_site_differences(josiah_sequence, curr_seq)\n",
    "    list_of_different_sites.append((curr_seq_name, site_differences, mut_differences))\n",
    "\n",
    "# Antibody contact files and escape files\n",
    "contacts_files = [\n",
    "    contacts_2510C,\n",
    "    contacts_121F,\n",
    "    contacts_377H,\n",
    "    # contacts_256A,\n",
    "    contacts_377H, # duplicate 37.7H because validated with two strains\n",
    "    contacts_372D,\n",
    "    contacts_89F,\n",
    "]\n",
    "\n",
    "filtered_escape_files = [\n",
    "    filtered_escape_2510C,\n",
    "    filtered_escape_121F,\n",
    "    filtered_escape_377H,\n",
    "    # filtered_escape_256A,\n",
    "    filtered_escape_377H, # duplicate 37.7H because validated with two strains\n",
    "    filtered_escape_372D,\n",
    "    filtered_escape_89F,\n",
    "]\n",
    "\n",
    "# Set figure size and subplots\n",
    "fig, axes = plt.subplots(\n",
    "    6, \n",
    "    1, \n",
    "    figsize=(3.5, 6), \n",
    "    # sharex=\"col\"\n",
    ")\n",
    "\n",
    "# Iterate through list of antibody files\n",
    "for i in range(len(antibody_files)):\n",
    "    name = None\n",
    "    sites = None\n",
    "    AA_to_color = None\n",
    "    if i == 0:\n",
    "        name = list_of_different_sites[1][0]\n",
    "        sites = list_of_different_sites[1][1]\n",
    "        AA_to_color = list_of_different_sites[1][2]\n",
    "    elif i == 1:\n",
    "        name = list_of_different_sites[4][0]\n",
    "        sites = list_of_different_sites[4][1]\n",
    "        AA_to_color = list_of_different_sites[4][2]\n",
    "    elif i == 2:\n",
    "        name = list_of_different_sites[1][0]\n",
    "        sites = list_of_different_sites[1][1]\n",
    "        AA_to_color = list_of_different_sites[1][2]\n",
    "    elif i == 3:\n",
    "        name = list_of_different_sites[2][0]\n",
    "        sites = list_of_different_sites[2][1]\n",
    "        AA_to_color = list_of_different_sites[2][2]\n",
    "    elif i == 4:\n",
    "        name = list_of_different_sites[1][0]\n",
    "        sites = list_of_different_sites[1][1]\n",
    "        AA_to_color = list_of_different_sites[1][2]\n",
    "    elif i == 5:\n",
    "        name = list_of_different_sites[3][0]\n",
    "        sites = list_of_different_sites[3][1]\n",
    "        AA_to_color = list_of_different_sites[3][2]\n",
    "    else:\n",
    "        print(\"Error! Index out of range!\")\n",
    "\n",
    "    create_line_and_logoplots(\n",
    "        filtered_escape_files[i], \n",
    "        contacts_files[i],\n",
    "        func_scores, \n",
    "        min_times_seen, \n",
    "        n_selections, \n",
    "        axes[i],\n",
    "        sites=sorted(sites),\n",
    "        name=name,\n",
    "        amino_acids_to_color=AA_to_color,\n",
    "        only_top_10_sites=True,\n",
    "    )\n",
    "\n",
    "fig.subplots_adjust(\n",
    "    left  = 0,  # the left side of the subplots of the figure\n",
    "    right = 1,    # the right side of the subplots of the figure\n",
    "    bottom = 0,   # the bottom of the subplots of the figure\n",
    "    top = 1,      # the top of the subplots of the figure\n",
    "    wspace = 0,   # the amount of width reserved for blank space between subplots\n",
    "    hspace = 0.8,   # the amount of height reserved for white space between subplots\n",
    ")\n",
    "\n",
    "# Common Y axis labels\n",
    "fig.text(-0.075, 0.5, \"site escape\", va=\"center\", rotation=\"vertical\", fontsize=8)\n",
    "fig.text(0.5, -0.07, \"site\", ha=\"center\", fontsize=8)\n",
    "\n",
    "# Make output dir if doesn't exist\n",
    "if not os.path.exists(out_dir_escape):\n",
    "    os.mkdir(out_dir_escape)\n",
    "\n",
    "# Save fig\n",
    "plt.savefig(escape_top10_image_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "228ff352-959a-42e6-9ab2-4eb2346e33f6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Set figure size and subplots\n",
    "fig, axes = plt.subplots(\n",
    "    6, \n",
    "    1, \n",
    "    figsize=(4.5, 6), \n",
    "    # sharex=\"col\"\n",
    ")\n",
    "\n",
    "# Iterate through list of antibody files\n",
    "for i in range(len(antibody_files)):\n",
    "    name = None\n",
    "    sites = None\n",
    "    AA_to_color = None\n",
    "    if i == 0:\n",
    "        name = list_of_different_sites[1][0]\n",
    "        sites = list_of_different_sites[1][1]\n",
    "        AA_to_color = list_of_different_sites[1][2]\n",
    "    elif i == 1:\n",
    "        name = list_of_different_sites[4][0]\n",
    "        sites = list_of_different_sites[4][1]\n",
    "        AA_to_color = list_of_different_sites[4][2]\n",
    "    elif i == 2:\n",
    "        name = list_of_different_sites[1][0]\n",
    "        sites = list_of_different_sites[1][1]\n",
    "        AA_to_color = list_of_different_sites[1][2]\n",
    "    elif i == 3:\n",
    "        name = list_of_different_sites[2][0]\n",
    "        sites = list_of_different_sites[2][1]\n",
    "        AA_to_color = list_of_different_sites[2][2]\n",
    "    elif i == 4:\n",
    "        name = list_of_different_sites[1][0]\n",
    "        sites = list_of_different_sites[1][1]\n",
    "        AA_to_color = list_of_different_sites[1][2]\n",
    "    elif i == 5:\n",
    "        name = list_of_different_sites[3][0]\n",
    "        sites = list_of_different_sites[3][1]\n",
    "        AA_to_color = list_of_different_sites[3][2]\n",
    "    else:\n",
    "        print(\"Error! Index out of range!\")\n",
    "\n",
    "    create_line_and_logoplots(\n",
    "        filtered_escape_files[i], \n",
    "        contacts_files[i],\n",
    "        func_scores, \n",
    "        min_times_seen, \n",
    "        n_selections, \n",
    "        axes[i],\n",
    "        sites=sorted(sites),\n",
    "        name=name,\n",
    "        amino_acids_to_color=AA_to_color,\n",
    "        only_top_10_sites=False,\n",
    "    )\n",
    "\n",
    "fig.subplots_adjust(\n",
    "    left  = 0,  # the left side of the subplots of the figure\n",
    "    right = 1,    # the right side of the subplots of the figure\n",
    "    bottom = 0,   # the bottom of the subplots of the figure\n",
    "    top = 1,      # the top of the subplots of the figure\n",
    "    wspace = 0,   # the amount of width reserved for blank space between subplots\n",
    "    hspace = 0.8,   # the amount of height reserved for white space between subplots\n",
    ")\n",
    "\n",
    "# Common Y axis labels\n",
    "fig.text(-0.075, 0.5, \"site escape\", va=\"center\", rotation=\"vertical\", fontsize=8)\n",
    "fig.text(0.5, -0.07, \"site\", ha=\"center\", fontsize=8)\n",
    "\n",
    "# Make output dir if doesn't exist\n",
    "if not os.path.exists(out_dir_escape):\n",
    "    os.mkdir(out_dir_escape)\n",
    "\n",
    "# Save fig\n",
    "plt.savefig(escape_all_image_path)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "59badf48-08ae-45c9-89e2-eb61f40af227",
   "metadata": {},
   "source": [
    "## Pseudovirus neutralization validation assays for chosen natural isolates and corresponding single mutants\n",
    "\n",
    "The chosen isolates were validated with tradition pseudovirus neutralization assays in addition to the corresponding single mutation identified as the main mutation for escape."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "22c56711-4685-4950-8e5f-31cae7b00e7c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Rename column name\n",
    "validation_isolates = (\n",
    "    validation_isolates.rename(columns={\"strain\" : \"virus\"})\n",
    ")\n",
    "\n",
    "# Read nuetralization data\n",
    "frac_infected_natural_isolates = pd.read_csv(fraction_infected_natural_isolates)\n",
    "\n",
    "# Rename viruses\n",
    "frac_infected_natural_isolates[\"virus\"] = (\n",
    "    frac_infected_natural_isolates[\"virus\"].replace(rename_dict)\n",
    ")\n",
    "\n",
    "# Fit hill curves using neutcurve\n",
    "fits = neutcurve.curvefits.CurveFits(\n",
    "    data=frac_infected_natural_isolates,\n",
    "    fixbottom=0,\n",
    "    fixtop=1,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "17c02dc5-16d8-4bfb-b6f4-fde34f6c64f8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plot neut data\n",
    "fig, axes = fits.plotGrid(\n",
    "    {\n",
    "        (0, 0) : (\"25.10C\",\n",
    "                  [\n",
    "                      {\n",
    "                          \"serum\" : \"25.10C\", \n",
    "                          \"virus\" : \"unmutated Josiah strain\",\n",
    "                          \"replicate\" : \"average\",\n",
    "                          \"color\" : tol_muted_adjusted[0],\n",
    "                          \"marker\" : \"o\",\n",
    "                          \"label\" : \"unmutated\\nJosiah\\nstrain\"\n",
    "                      },\n",
    "                      {\n",
    "                          \"serum\" : \"25.10C\", \n",
    "                          \"virus\" : \"GA391 strain\",\n",
    "                          \"replicate\" : \"average\",\n",
    "                          \"color\" : tol_muted_adjusted[12],\n",
    "                          \"marker\" : \"o\",\n",
    "                          \"label\" : \"GA391\\nstrain\"\n",
    "                      },\n",
    "                      {\n",
    "                          \"serum\" : \"25.10C\", \n",
    "                          \"virus\" : \"E228D Josiah strain\",\n",
    "                          \"replicate\" : \"average\",\n",
    "                          \"color\" : tol_muted_adjusted[10],\n",
    "                          \"marker\" : \"o\",\n",
    "                          \"label\" : \"E228D\\nJosiah\\nstrain\"\n",
    "                      },\n",
    "                  ]\n",
    "                 ),\n",
    "        (1, 0) : (\"12.1F\",\n",
    "                  [\n",
    "                      {\n",
    "                          \"serum\" : \"12.1F\", \n",
    "                          \"virus\" : \"unmutated Josiah strain\",\n",
    "                          \"replicate\" : \"average\",\n",
    "                          \"color\" : tol_muted_adjusted[0],\n",
    "                          \"marker\" : \"o\",\n",
    "                          \"label\" : \"unmutated\\nJosiah\\nstrain\"\n",
    "                      },\n",
    "                      {\n",
    "                          \"serum\" : \"12.1F\", \n",
    "                          \"virus\" : \"LM395 strain\",\n",
    "                          \"replicate\" : \"average\",\n",
    "                          \"color\" : tol_muted_adjusted[12],\n",
    "                          \"marker\" : \"o\",\n",
    "                          \"label\" : \"LM395\\nstrain\"\n",
    "                      },\n",
    "                      {\n",
    "                          \"serum\" : \"12.1F\", \n",
    "                          \"virus\" : \"N89D Josiah strain\",\n",
    "                          \"replicate\" : \"average\",\n",
    "                          \"color\" : tol_muted_adjusted[10],\n",
    "                          \"marker\" : \"o\",\n",
    "                          \"label\" : \"N89D\\nJosiah\\nstrain\"\n",
    "                      },\n",
    "                  ]\n",
    "                 ),\n",
    "        (2, 0) : (\"37.7H\",\n",
    "                  [\n",
    "                      {\n",
    "                          \"serum\" : \"37.7H\", \n",
    "                          \"virus\" : \"unmutated Josiah strain\",\n",
    "                          \"replicate\" : \"average\",\n",
    "                          \"color\" : tol_muted_adjusted[0],\n",
    "                          \"marker\" : \"o\",\n",
    "                          \"label\" : \"unmutated\\nJosiah\\nstrain\"\n",
    "                      },\n",
    "                      {\n",
    "                          \"serum\" : \"37.7H\", \n",
    "                          \"virus\" : \"GA391 strain\",\n",
    "                          \"replicate\" : \"average\",\n",
    "                          \"color\" : tol_muted_adjusted[12],\n",
    "                          \"marker\" : \"o\",\n",
    "                          \"label\" : \"GA391\\nstrain\"\n",
    "                      },\n",
    "                      {\n",
    "                          \"serum\" : \"37.7H\", \n",
    "                          \"virus\" : \"H398K Josiah strain\",\n",
    "                          \"replicate\" : \"average\",\n",
    "                          \"color\" : tol_muted_adjusted[10],\n",
    "                          \"marker\" : \"o\",\n",
    "                          \"label\" : \"H398K\\nJosiah\\nstrain\"\n",
    "                      },\n",
    "                  ]\n",
    "                 ),\n",
    "        (3, 0) : (\"37.7H\",\n",
    "                  [\n",
    "                      {\n",
    "                          \"serum\" : \"37.7H\", \n",
    "                          \"virus\" : \"unmutated Josiah strain\",\n",
    "                          \"replicate\" : \"average\",\n",
    "                          \"color\" : tol_muted_adjusted[0],\n",
    "                          \"marker\" : \"o\",\n",
    "                          \"label\" : \"unmutated\\nJosiah\\nstrain\"\n",
    "                      },\n",
    "                      {\n",
    "                          \"serum\" : \"37.7H\", \n",
    "                          \"virus\" : \"IRR007 strain\",\n",
    "                          \"replicate\" : \"average\",\n",
    "                          \"color\" : tol_muted_adjusted[12],\n",
    "                          \"marker\" : \"o\",\n",
    "                          \"label\" : \"IRR007\\nstrain\"\n",
    "                      },\n",
    "                      {\n",
    "                          \"serum\" : \"37.7H\", \n",
    "                          \"virus\" : \"D401E Josiah strain\",\n",
    "                          \"replicate\" : \"average\",\n",
    "                          \"color\" : tol_muted_adjusted[10],\n",
    "                          \"marker\" : \"o\",\n",
    "                          \"label\" : \"D401E\\nJosiah\\nstrain\"\n",
    "                      },\n",
    "                  ]\n",
    "                 ),\n",
    "        (4, 0) : (\"37.2D\",\n",
    "                  [\n",
    "                      {\n",
    "                          \"serum\" : \"37.2D\", \n",
    "                          \"virus\" : \"unmutated Josiah strain\",\n",
    "                          \"replicate\" : \"average\",\n",
    "                          \"color\" : tol_muted_adjusted[0],\n",
    "                          \"marker\" : \"o\",\n",
    "                          \"label\" : \"unmutated\\nJosiah\\nstrain\"\n",
    "                      },\n",
    "                      {\n",
    "                          \"serum\" : \"37.2D\", \n",
    "                          \"virus\" : \"GA391 strain\",\n",
    "                          \"replicate\" : \"average\",\n",
    "                          \"color\" : tol_muted_adjusted[12],\n",
    "                          \"marker\" : \"o\",\n",
    "                          \"label\" : \"GA391\\nstrain\"\n",
    "                      },\n",
    "                      {\n",
    "                          \"serum\" : \"37.2D\", \n",
    "                          \"virus\" : \"H398K Josiah strain\",\n",
    "                          \"replicate\" : \"average\",\n",
    "                          \"color\" : tol_muted_adjusted[10],\n",
    "                          \"marker\" : \"o\",\n",
    "                          \"label\" : \"H398K\\nJosiah\\nstrain\"\n",
    "                      },\n",
    "                  ]\n",
    "                 ),\n",
    "        (5, 0) : (\"8.9F\",\n",
    "                  [\n",
    "                      {\n",
    "                          \"serum\" : \"8.9F\", \n",
    "                          \"virus\" : \"unmutated Josiah strain\",\n",
    "                          \"replicate\" : \"average\",\n",
    "                          \"color\" : tol_muted_adjusted[0],\n",
    "                          \"marker\" : \"o\",\n",
    "                          \"label\" : \"unmutated\\nJosiah\\nstrain\"\n",
    "                      },\n",
    "                      {\n",
    "                          \"serum\" : \"8.9F\", \n",
    "                          \"virus\" : \"ISTH1024 strain\",\n",
    "                          \"replicate\" : \"average\",\n",
    "                          \"color\" : tol_muted_adjusted[12],\n",
    "                          \"marker\" : \"o\",\n",
    "                          \"label\" : \"ISTH1024\\nstrain\"\n",
    "                      },\n",
    "                      {\n",
    "                          \"serum\" : \"8.9F\", \n",
    "                          \"virus\" : \"K126N Josiah strain\",\n",
    "                          \"replicate\" : \"average\",\n",
    "                          \"color\" : tol_muted_adjusted[10],\n",
    "                          \"marker\" : \"o\",\n",
    "                          \"label\" : \"K126N\\nJosiah\\nstrain\"\n",
    "                      },\n",
    "                  ]\n",
    "                 ),\n",
    "    },\n",
    "    sharex=True,\n",
    "    sharey=False,\n",
    "    xlabel=\"\",\n",
    "    ylabel=\"\",\n",
    "    attempt_shared_legend=False,\n",
    "    despine=True,\n",
    ")\n",
    "\n",
    "\n",
    "antibody_names = [\n",
    "    \"25.10C\",\n",
    "    \"12.1F\",\n",
    "    \"37.7H\",\n",
    "    \"37.7H\",\n",
    "    \"37.2D\",\n",
    "    \"8.9F\",\n",
    "]\n",
    "\n",
    "for index in range(6):\n",
    "    if antibody_names[index] == \"25.10C\":\n",
    "        axes[index,0].set_title(\n",
    "            \"antibody \" + antibody_names[index], \n",
    "            # weight=\"bold\",\n",
    "            fontsize=8,\n",
    "            color=\"#44AA99\",\n",
    "        )\n",
    "    elif antibody_names[index] == \"12.1F\":\n",
    "        axes[index,0].set_title(\n",
    "            \"antibody \" + antibody_names[index], \n",
    "            # weight=\"bold\",\n",
    "            fontsize=8,\n",
    "            color=\"#999933\",\n",
    "        )\n",
    "    elif antibody_names[index] == \"37.7H\" or antibody_names[index] == \"37.2D\" :\n",
    "        axes[index,0].set_title(\n",
    "            \"antibody \" + antibody_names[index], \n",
    "            # weight=\"bold\",\n",
    "            fontsize=8,\n",
    "            color=\"#AA4499\",\n",
    "        )\n",
    "    elif antibody_names[index] == \"8.9F\":\n",
    "        axes[index,0].set_title(\n",
    "            \"antibody \" + antibody_names[index], \n",
    "            # weight=\"bold\",\n",
    "            fontsize=8,\n",
    "            color=\"#117733\",\n",
    "        )\n",
    "    axes[index,0].set_ylim(-0.1, 1.3)\n",
    "    axes[index,0].set_yticks([0, 0.5, 1.0])\n",
    "    axes[index,0].set_yticklabels(labels=[0, 0.5, 1.0], fontsize=6)\n",
    "    axes[index,0].set_xlim(0.0005, 12.5)\n",
    "    axes[index,0].set_xticks([0.001, 0.01, 0.1, 1, 10,])\n",
    "    axes[index,0].set_xticklabels(labels=[\"$10^{-3}$\", \"$10^{-2}$\", \"$10^{-1}$\", \"$10^0$\", \"$10^1$\",], fontsize=6)\n",
    "    plt.setp(axes[index,0].collections, alpha=0.8, linewidths=0.5, colors=\"black\") # for vertical error bar segment\n",
    "    plt.setp(axes[index,0].lines, alpha=0.8, markeredgewidth=0.5, markeredgecolor=\"black\", linewidth=1, markersize=4) # for the lines and markers\n",
    "    sns.move_legend(\n",
    "        axes[index,0], \n",
    "        loc=\"upper left\",\n",
    "        borderaxespad=0,\n",
    "        frameon=False,\n",
    "        bbox_to_anchor=(1, 1.3),\n",
    "        fontsize=6,\n",
    "        markerscale=1,\n",
    "        handletextpad=0.1,\n",
    "        title=\"Lassa GPC\",\n",
    "        title_fontproperties={\"weight\" : \"bold\", \"size\" : 6},\n",
    "        alignment=\"center\",\n",
    "    )\n",
    "\n",
    "    # Add edges to legend markers to match scatter plot\n",
    "    for ha in axes[index,0].legend_.legendHandles:\n",
    "        ha.set_markeredgecolor(\"black\")\n",
    "        ha.set_markeredgewidth(0.5)\n",
    "        ha.set_linewidth(0)\n",
    "        ha.set_markersize(5)\n",
    "        \n",
    "    # Change all spines\n",
    "    for axis in [\"top\", \"bottom\", \"left\", \"right\"]:\n",
    "        axes[index,0].spines[axis].set_linewidth(1)\n",
    "    axes[index,0].tick_params(axis=\"both\", length=2, width=1)\n",
    "\n",
    "# Adjust spacing of subplots\n",
    "fig.subplots_adjust(\n",
    "    left  = 0,  # the left side of the subplots of the figure\n",
    "    right = 1,    # the right side of the subplots of the figure\n",
    "    bottom = 0,   # the bottom of the subplots of the figure\n",
    "    top = 1,      # the top of the subplots of the figure\n",
    "    wspace = 0,   # the amount of width reserved for blank space between subplots\n",
    "    hspace = 0.8,   # the amount of height reserved for white space between subplots\n",
    ")\n",
    "\n",
    "# Common X and Y axis labels\n",
    "fig.text(-0.4, 0.5, \"fraction infectivity\", va=\"center\", rotation=\"vertical\", fontsize=8)\n",
    "fig.text(0.5, -0.05, \"concentration (\\u03BCg/mL)\", ha=\"center\", fontsize=8)\n",
    "\n",
    "width = 0.9\n",
    "height = 6\n",
    "fig.set_size_inches(width, height)\n",
    "\n",
    "# Make output dir if doesn't exist\n",
    "if not os.path.exists(out_dir):\n",
    "    os.mkdir(out_dir)\n",
    "\n",
    "# Save fig\n",
    "plt.savefig(neuts_image_path)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "84a3dba5-c963-4fda-83b4-3f4c4a3bdfaa",
   "metadata": {},
   "source": [
    "We further calculate **cumulative summed** escape scores across all mutations present in each isolate even if the effects are small. These total DMS escape scores for each isolate are then compared to log2 ic50 fold change values. The same comparisons are made for the corresponding single mutants. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f4b0478a-243b-45f5-a063-f8305e2ed1f0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# IC values to calculate\n",
    "fitParams = fits.fitParams(ics=[50, 80, 99])[[\"serum\", \"virus\", \"ic50\", \"ic50_bound\"]]\n",
    "\n",
    "# Create dictionary for antibody ic50s\n",
    "wt_ic50s = {\n",
    "    \"8.9F\" : fitParams.query(\"serum == '8.9F' & virus == 'unmutated Josiah strain'\").reset_index(drop=True).at[0,\"ic50\"],\n",
    "    \"12.1F\" : fitParams.query(\"serum == '12.1F' & virus == 'unmutated Josiah strain'\").reset_index(drop=True).at[0,\"ic50\"],\n",
    "    \"25.10C\" : fitParams.query(\"serum == '25.10C' & virus == 'unmutated Josiah strain'\").reset_index(drop=True).at[0,\"ic50\"],\n",
    "    \"37.7H\" : fitParams.query(\"serum == '37.7H' & virus == 'unmutated Josiah strain'\").reset_index(drop=True).at[0,\"ic50\"],\n",
    "    \"37.2D\" : fitParams.query(\"serum == '37.2D' & virus == 'unmutated Josiah strain'\").reset_index(drop=True).at[0,\"ic50\"],\n",
    "}\n",
    "\n",
    "# Calculate log2 fold change of IC50 values\n",
    "fitParams[\"ic50 fold change\"] = (\n",
    "    fitParams.apply(lambda x: x[\"ic50\"]/wt_ic50s[x[\"serum\"]], axis=1)\n",
    ")\n",
    "fitParams[\"log2 ic50 fold change\"] = np.log2(fitParams[\"ic50 fold change\"])\n",
    "\n",
    "\n",
    "# Merge isolate sequence and neut dfs\n",
    "validation_df = (\n",
    "    fitParams.merge(\n",
    "        validation_isolates,\n",
    "        how=\"left\",\n",
    "        on=\"virus\",\n",
    "        validate=\"many_to_one\"\n",
    "    )\n",
    ")\n",
    "\n",
    "# Add sinlge mutant sequences\n",
    "validation_df.at[2, \"sequence\"] = (\n",
    "    josiah_sequence[:int(validation_df.at[2,\"virus\"].split()[0][1:-1])-1] + validation_df.at[2,\"virus\"].split()[0][-1] + josiah_sequence[int(validation_df.at[2,\"virus\"].split()[0][1:-1]):]\n",
    ")\n",
    "\n",
    "validation_df.at[5, \"sequence\"] = (\n",
    "    josiah_sequence[:int(validation_df.at[5,\"virus\"].split()[0][1:-1])-1] + validation_df.at[5,\"virus\"].split()[0][-1] + josiah_sequence[int(validation_df.at[5,\"virus\"].split()[0][1:-1]):]\n",
    ")\n",
    "\n",
    "validation_df.at[8, \"sequence\"] = (\n",
    "    josiah_sequence[:int(validation_df.at[8,\"virus\"].split()[0][1:-1])-1] + validation_df.at[8,\"virus\"].split()[0][-1] + josiah_sequence[int(validation_df.at[8,\"virus\"].split()[0][1:-1]):]\n",
    ")\n",
    "\n",
    "validation_df.at[10, \"sequence\"] = (\n",
    "    josiah_sequence[:int(validation_df.at[10,\"virus\"].split()[0][1:-1])-1] + validation_df.at[10,\"virus\"].split()[0][-1] + josiah_sequence[int(validation_df.at[10,\"virus\"].split()[0][1:-1]):]\n",
    ")\n",
    "\n",
    "validation_df.at[13, \"sequence\"] = (\n",
    "    josiah_sequence[:int(validation_df.at[13,\"virus\"].split()[0][1:-1])-1] + validation_df.at[13,\"virus\"].split()[0][-1] + josiah_sequence[int(validation_df.at[13,\"virus\"].split()[0][1:-1]):]\n",
    ")\n",
    "\n",
    "validation_df.at[16, \"sequence\"] = (\n",
    "    josiah_sequence[:int(validation_df.at[16,\"virus\"].split()[0][1:-1])-1] + validation_df.at[16,\"virus\"].split()[0][-1] + josiah_sequence[int(validation_df.at[16,\"virus\"].split()[0][1:-1]):]\n",
    ")\n",
    "\n",
    "# Antibody conversion dict\n",
    "antibody_file_dict = {\n",
    "    \"25.10C\" : filtered_escape_2510C,\n",
    "    \"12.1F\" : filtered_escape_121F,\n",
    "    \"37.7H\" : filtered_escape_377H,\n",
    "    \"25.6A\" : filtered_escape_256A,\n",
    "    \"37.2D\" : filtered_escape_372D,\n",
    "    \"8.9F\" : filtered_escape_89F,\n",
    "}\n",
    "\n",
    "# Calculate total escape for each sequence using an additive model\n",
    "validation_df[\"total_escape\"] = (\n",
    "    validation_df.apply(lambda x: determine_escape(\n",
    "        0, \n",
    "        x[\"sequence\"], \n",
    "        antibody_file_dict[x[\"serum\"]], \n",
    "        x[\"virus\"],\n",
    "    ), axis=1)\n",
    ")\n",
    "\n",
    "# Rename columns and reformat columns\n",
    "validation_df = (\n",
    "    validation_df.rename(columns={\n",
    "        \"ic50_bound\" : \"lower bound\",\n",
    "        \"serum\" : \"antibody\",\n",
    "    })\n",
    ")\n",
    "validation_df[\"lower bound\"] = (\n",
    "    validation_df[\"lower bound\"].replace({\"interpolated\" : False, \"lower\" : True})\n",
    ")\n",
    "\n",
    "# Mark strains as natural isolates or single mutants\n",
    "validation_df[\"Lassa GPC\"] = (\n",
    "    validation_df.apply(lambda x: \"unmutated Josiah strain\" if x[\"virus\"] == \"unmutated Josiah strain\" else \n",
    "        (\"Josiah strain variant\" if x[\"virus\"] in [\n",
    "            \"N89D Josiah strain\",\n",
    "            \"K126N Josiah strain\",\n",
    "            \"E228D Josiah strain\",\n",
    "            \"H398K Josiah strain\",\n",
    "            \"D401E Josiah strain\",\n",
    "        ] else \"natural isolate strain\"), axis=1)\n",
    ")\n",
    "\n",
    "# Calculate correlation between predicted and measured\n",
    "r, p = sp.stats.pearsonr(\n",
    "    x=validation_df[\"log2 ic50 fold change\"], \n",
    "    y=validation_df[\"total_escape\"]\n",
    ")\n",
    "print(f\"R={r}\")\n",
    "print(f\"R^2={r**2}\")\n",
    "\n",
    "# Plot predicted vs measured\n",
    "fig, ax = plt.subplots(figsize=(2.25, 2.25))\n",
    "corr_chart = sns.scatterplot(\n",
    "    data=validation_df,\n",
    "    x=\"log2 ic50 fold change\", \n",
    "    y=\"total_escape\",\n",
    "    hue=\"Lassa GPC\",\n",
    "    palette={\n",
    "        \"unmutated Josiah strain\" : tol_muted_adjusted[0],\n",
    "        \"natural isolate strain\" : tol_muted_adjusted[12],\n",
    "        \"Josiah strain variant\" : tol_muted_adjusted[10],\n",
    "    },\n",
    "    style=\"antibody\",\n",
    "    markers=[\"o\", \"s\", \"D\", \"v\", \"^\", \"P\"],\n",
    "    alpha=0.8,\n",
    "    edgecolor=\"black\",\n",
    "    ax=ax,\n",
    "    s=40,\n",
    ")\n",
    "# Change all spines\n",
    "for axis in [\"top\", \"bottom\", \"left\", \"right\"]:\n",
    "    corr_chart.spines[axis].set_linewidth(1)\n",
    "sns.despine()\n",
    "corr_chart.tick_params(axis=\"both\", length=4, width=1)\n",
    "corr_chart.set_xlabel(\n",
    "    \"log2 fold change IC$_{50}$ measured\\nby pseuodvirus neutralization assay\", \n",
    "    # weight=\"bold\",\n",
    "    fontsize=8,\n",
    ")\n",
    "corr_chart.set_ylabel(\n",
    "    \"escape score predicted\\nby DMS (arbitrary units)\", \n",
    "    # weight=\"bold\",\n",
    "    fontsize=8,\n",
    ")\n",
    "corr_chart.set_xlim(-1, 12.5)\n",
    "corr_chart.set_ylim(-0.5, 5.5)\n",
    "corr_chart.set_xticks([0, 2, 4, 6, 8, 10, 12])\n",
    "corr_chart.set_xticklabels(corr_chart.get_xticks(), size=8)\n",
    "corr_chart.set_yticks([0, 1, 2, 3, 4, 5])\n",
    "corr_chart.set_yticklabels(corr_chart.get_yticks(), size=8)\n",
    "sns.move_legend(\n",
    "    corr_chart, \n",
    "    \"upper left\", \n",
    "    bbox_to_anchor=(1, 1),\n",
    "    fontsize=8,\n",
    "    markerscale=1,\n",
    "    handletextpad=0.1,\n",
    "    frameon=False,\n",
    "    borderaxespad=0.75,\n",
    ")\n",
    "corr_chart.get_legend().get_texts()[0].set_weight(\"bold\")\n",
    "corr_chart.get_legend().get_texts()[4].set_weight(\"bold\")\n",
    "corr_chart.text(\n",
    "    0, \n",
    "    5, \n",
    "    f\"r={r:.2f}\", \n",
    "    horizontalalignment=\"left\",  \n",
    "    weight=\"bold\",\n",
    "    fontsize=8,\n",
    ")\n",
    "\n",
    "# Add edges to legend markers to match scatter plot\n",
    "for ha in corr_chart.legend_.legendHandles:\n",
    "    ha.set_edgecolor(\"black\")\n",
    "    ha.set_linewidths(0.5)\n",
    "\n",
    "# Label points\n",
    "for i in range(0, validation_df.shape[0]):\n",
    "    x_pos = validation_df.at[i, \"log2 ic50 fold change\"]\n",
    "    y_pos = validation_df.at[i, \"total_escape\"]\n",
    "    name = validation_df.at[i, \"virus\"].split(\" \")[0]\n",
    "    if name == \"K126N\":\n",
    "        corr_chart.text(\n",
    "            x_pos+0.3, \n",
    "            y_pos-0.1, \n",
    "            f\"{name}\", \n",
    "            horizontalalignment=\"left\",  \n",
    "            # weight=\"bold\",\n",
    "            fontsize=6,\n",
    "        )\n",
    "    elif name == \"H398K\":\n",
    "        corr_chart.text(\n",
    "            x_pos+0.35, \n",
    "            y_pos+0.1, \n",
    "            f\"{name}\", \n",
    "            horizontalalignment=\"left\",  \n",
    "            # weight=\"bold\",\n",
    "            fontsize=6,\n",
    "        )\n",
    "    elif name == \"ISTH1024\":\n",
    "        corr_chart.text(\n",
    "            x_pos+0.35, \n",
    "            y_pos-0.25, \n",
    "            f\"{name}\", \n",
    "            horizontalalignment=\"left\",  \n",
    "            # weight=\"bold\",\n",
    "            fontsize=6,\n",
    "        )\n",
    "    elif name == \"N89D\":\n",
    "        corr_chart.text(\n",
    "            x_pos+0.35, \n",
    "            y_pos-0.1, \n",
    "            f\"{name}\", \n",
    "            horizontalalignment=\"left\",  \n",
    "            # weight=\"bold\",\n",
    "            fontsize=6,\n",
    "        )\n",
    "    elif name == \"IRR007\":\n",
    "        corr_chart.text(\n",
    "            x_pos+0.35, \n",
    "            y_pos+0.1, \n",
    "            f\"{name}\", \n",
    "            horizontalalignment=\"left\",  \n",
    "            # weight=\"bold\",\n",
    "            fontsize=6,\n",
    "        )\n",
    "    elif name == \"GA391\":\n",
    "        corr_chart.text(\n",
    "            x_pos+0.4, \n",
    "            y_pos+0.05, \n",
    "            f\"{name}\", \n",
    "            horizontalalignment=\"left\",  \n",
    "            # weight=\"bold\",\n",
    "            fontsize=6,\n",
    "        )\n",
    "    elif name == \"LM395\":\n",
    "        corr_chart.text(\n",
    "            x_pos+0.35, \n",
    "            y_pos+0.05, \n",
    "            f\"{name}\", \n",
    "            horizontalalignment=\"left\",  \n",
    "            # weight=\"bold\",\n",
    "            fontsize=6,\n",
    "        )\n",
    "    elif name == \"unmutated\":\n",
    "        corr_chart.text(\n",
    "            x_pos+0.4, \n",
    "            y_pos+0.05, \n",
    "            f\"{name} Josiah\", \n",
    "            horizontalalignment=\"left\",  \n",
    "            # weight=\"bold\",\n",
    "            fontsize=6,\n",
    "        )\n",
    "    else:\n",
    "        corr_chart.text(\n",
    "            x_pos, \n",
    "            y_pos+0.25, \n",
    "            f\"{name}\", \n",
    "            horizontalalignment=\"left\",  \n",
    "            # weight=\"bold\",\n",
    "            fontsize=6,\n",
    "        )\n",
    "        \n",
    "# Set square ratio\n",
    "corr_chart.set_box_aspect(1)\n",
    "\n",
    "# Make output dir if doesn't exist\n",
    "if not os.path.exists(out_dir):\n",
    "    os.mkdir(out_dir)\n",
    "\n",
    "# Save fig\n",
    "plt.savefig(corr_image_path)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0a643eba-b171-4341-b79c-8df9670e87ae",
   "metadata": {},
   "source": [
    "## Further antibody escape analysis\n",
    "\n",
    "First create a dataframe for amino acid level measurements and mutational frequencies with respect to the Josiah strain calculated from natural sequence alignments. In addition, create a site level dataframe with site level measurements, cumulative mutational frequencies, and effective number of amino acids calculated for each position. The formula for calculating effective amino acids is described in *Biophysical Models of Protein Evolution: Understanding the Patterns of Evolutionary Sequence Divergence*."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b8e7deba-42cc-4aa1-962e-c58818178562",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_natural_sequence_counts(site, amino_acid, natural_seqs_df):\n",
    "    \"\"\"\n",
    "    Function that counts occurences of an amino acid at a site\n",
    "    across a dataframe of sequences.\n",
    "    \"\"\"\n",
    "    count = 0\n",
    "    for seq in natural_seqs_df[\"sequence\"].tolist():\n",
    "        if seq[site-1] == amino_acid:\n",
    "            count += 1\n",
    "    return count\n",
    "            \n",
    "\n",
    "# Load data as dataframe\n",
    "functional_scores = pd.read_csv(func_scores)\n",
    "\n",
    "# Filter for minimum selections, times seen and no stop codons\n",
    "functional_scores = (\n",
    "    functional_scores.query(\n",
    "        \"n_selections >= @n_selections and times_seen >= @min_times_seen and mutant != '*'\"\n",
    "    )\n",
    "    .drop(columns=[\"times_seen\", \"effect_std\", \"n_selections\"])\n",
    "    .reset_index(drop=True)\n",
    ")\n",
    "\n",
    "# Create a wildtype df to fill in missing Josiah/wildtype measurements as 0\n",
    "josiah_df = pd.DataFrame(\n",
    "    zip(list(range(1,492)), josiah_sequence, josiah_sequence, [0]*491, [True]*491),\n",
    "    columns=[\"site\", \"wildtype\", \"mutant\", \"effect\", \"Josiah reference\"],\n",
    ")\n",
    "AA_level_df = (\n",
    "    pd.concat([functional_scores, josiah_df], ignore_index=True)\n",
    "    .sort_values(by=\"site\")\n",
    "    .reset_index(drop=True)\n",
    "    .fillna(False)\n",
    ")\n",
    "\n",
    "# Get natural sequence counts of each mutant and calculate mutation frequencies\n",
    "# compared to the Josiah reference\n",
    "AA_level_df[\"natural_counts\"] = (\n",
    "    AA_level_df.apply(lambda x: get_natural_sequence_counts(x[\"site\"], x[\"mutant\"], natural_seqs_df), axis=1)\n",
    ")\n",
    "number_sequences = AA_level_df[\"natural_counts\"].max() # 572 sequences\n",
    "AA_level_df[\"mutation_frequency\"] = (\n",
    "    AA_level_df.apply(lambda x: x[\"natural_counts\"]/number_sequences if x[\"Josiah reference\"] == False else None, axis=1)\n",
    ")\n",
    "\n",
    "\n",
    "# Add escape to dataframe for each antibody\n",
    "for antibody_file,_ in antibody_files:\n",
    "\n",
    "    antibody_name = antibody_file.split(\"/\")[-1].split(\"_\")[0]\n",
    "\n",
    "    # Load data as dataframe\n",
    "    escape_df = pd.read_csv(antibody_file)\n",
    "\n",
    "    # Clip lower scores to 0\n",
    "    escape_df[\"escape_median\"] = escape_df[\"escape_median\"].clip(lower=0)\n",
    "\n",
    "    # Rename escape column to include antibody name\n",
    "    escape_df = escape_df.rename(columns={\"escape_median\" : \"escape_\" + antibody_name})\n",
    "\n",
    "    # Merge dataframes\n",
    "    AA_level_df = (\n",
    "        AA_level_df.merge(\n",
    "            escape_df[[\"site\", \"wildtype\", \"mutant\", \"escape_\" + antibody_name]],\n",
    "            how=\"left\",\n",
    "            on=[\"site\", \"wildtype\", \"mutant\"],\n",
    "            validate=\"one_to_one\",\n",
    "        )\n",
    "    )\n",
    "\n",
    "# Fill in missing wildtype escape values\n",
    "AA_level_df[\"escape_2510C\"] = AA_level_df.apply(lambda x: 0 if x[\"wildtype\"] == x[\"mutant\"] else x[\"escape_2510C\"], axis=1)\n",
    "AA_level_df[\"escape_121F\"] = AA_level_df.apply(lambda x: 0 if x[\"wildtype\"] == x[\"mutant\"] else x[\"escape_121F\"], axis=1)\n",
    "AA_level_df[\"escape_377H\"] = AA_level_df.apply(lambda x: 0 if x[\"wildtype\"] == x[\"mutant\"] else x[\"escape_377H\"], axis=1)\n",
    "AA_level_df[\"escape_256A\"] = AA_level_df.apply(lambda x: 0 if x[\"wildtype\"] == x[\"mutant\"] else x[\"escape_256A\"], axis=1)\n",
    "AA_level_df[\"escape_372D\"] = AA_level_df.apply(lambda x: 0 if x[\"wildtype\"] == x[\"mutant\"] else x[\"escape_372D\"], axis=1)\n",
    "AA_level_df[\"escape_89F\"] = AA_level_df.apply(lambda x: 0 if x[\"wildtype\"] == x[\"mutant\"] else x[\"escape_89F\"], axis=1)\n",
    "\n",
    "# Calculate total escape\n",
    "AA_level_df[\"total_escape\"] = (\n",
    "    AA_level_df[[\n",
    "        \"escape_2510C\",\n",
    "        \"escape_121F\",\n",
    "        \"escape_377H\",\n",
    "        \"escape_256A\",\n",
    "        \"escape_372D\",\n",
    "        \"escape_89F\"\n",
    "    ]].sum(axis=1, skipna=False)\n",
    ")\n",
    "\n",
    "# Calculate epitope escape\n",
    "AA_level_df[\"GPC-A\"] = AA_level_df[\"escape_2510C\"]\n",
    "AA_level_df[\"GP1-A\"] = AA_level_df[\"escape_121F\"]\n",
    "AA_level_df[\"GPC-B\"] = (\n",
    "    AA_level_df[[\n",
    "        \"escape_377H\",\n",
    "        \"escape_256A\",\n",
    "        \"escape_372D\",\n",
    "    ]].mean(axis=1, skipna=False)\n",
    ")\n",
    "AA_level_df[\"GPC-C\"] = AA_level_df[\"escape_89F\"]\n",
    "\n",
    "# Mark mutations for chosen validations\n",
    "validation_muts = [\"89D\", \"126N\", \"228D\", \"398K\", \"401E\"]\n",
    "AA_level_df[\"site and mutant\"] = AA_level_df[\"site\"].astype(str) + AA_level_df[\"mutant\"]\n",
    "AA_level_df[\"validated\"] = (\n",
    "    AA_level_df[\"site and mutant\"].apply(lambda x: True if x in validation_muts else False)\n",
    ")\n",
    "\n",
    "# Create a site level dataframe\n",
    "site_level_df = (\n",
    "    AA_level_df.groupby([\"site\", \"wildtype\"])\n",
    "    .aggregate({\n",
    "        \"effect\" : \"mean\",\n",
    "        \"escape_2510C\" : \"sum\",\n",
    "        \"escape_121F\" : \"sum\",\n",
    "        \"escape_377H\" : \"sum\",\n",
    "        \"escape_256A\" : \"sum\",\n",
    "        \"escape_372D\" : \"sum\",\n",
    "        \"escape_89F\" : \"sum\",\n",
    "        \"total_escape\" : \"sum\",\n",
    "        \"natural_counts\" : \"sum\",\n",
    "        \"mutation_frequency\" : \"sum\",\n",
    "        \"GPC-A\" : \"sum\",\n",
    "        \"GP1-A\" : \"sum\",\n",
    "        \"GPC-B\" : \"sum\",\n",
    "        \"GPC-C\" : \"sum\",\n",
    "    })\n",
    "    .reset_index()\n",
    ")\n",
    "\n",
    "# Add site level escape to AA level df\n",
    "AA_level_df = (\n",
    "    AA_level_df.merge(\n",
    "        site_level_df[[\n",
    "            \"site\", \n",
    "            \"wildtype\", \n",
    "            \"escape_2510C\",\n",
    "            \"escape_121F\",\n",
    "            \"escape_377H\",\n",
    "            \"escape_256A\",\n",
    "            \"escape_372D\",\n",
    "            \"escape_89F\",\n",
    "            \"GPC-A\",\n",
    "            \"GP1-A\",\n",
    "            \"GPC-B\",\n",
    "            \"GPC-C\",\n",
    "            \"total_escape\",\n",
    "        ]].rename(columns={\n",
    "            \"escape_2510C\" : \"site_escape_2510C\",\n",
    "            \"escape_121F\" : \"site_escape_121F\",\n",
    "            \"escape_377H\" : \"site_escape_377H\",\n",
    "            \"escape_256A\" : \"site_escape_256A\",\n",
    "            \"escape_372D\" : \"site_escape_372D\",\n",
    "            \"escape_89F\" : \"site_escape_89F\",\n",
    "            \"total_escape\" : \"site_total_escape\",\n",
    "            \"GPC-A\" : \"site_GPC-A\",\n",
    "            \"GP1-A\" : \"site_GP1-A\",\n",
    "            \"GPC-B\" : \"site_GPC-B\",\n",
    "            \"GPC-C\" : \"site_GPC-C\",\n",
    "        }),\n",
    "        how=\"left\",\n",
    "        on=[\"site\", \"wildtype\"],\n",
    "        validate=\"many_to_one\",\n",
    "    )\n",
    ")\n",
    "\n",
    "# Load data as dataframe\n",
    "natural_variation = pd.read_csv(natural_sequence_variation)\n",
    "\n",
    "# Drop individual amino acid counts\n",
    "natural_variation = natural_variation[[\"site\", \"entropy\", \"n_effective\"]]\n",
    "\n",
    "# Merge functional and natural dataframes\n",
    "site_level_df = (\n",
    "    site_level_df.merge(\n",
    "        natural_variation,\n",
    "        how=\"left\",\n",
    "        on=[\"site\"],\n",
    "        validate=\"one_to_one\",\n",
    "    )\n",
    ")\n",
    "\n",
    "# Mark sites for chosen validations\n",
    "validation_sites = [89, 126, 228, 398, 401]\n",
    "site_level_df[\"validated\"] = (\n",
    "    site_level_df[\"site\"].apply(lambda x: True if x in validation_sites else False)\n",
    ")\n",
    "\n",
    "# Calculate log of mutational frequencies and replace inf with NaN\n",
    "AA_level_df[\"mutation_frequency\"] = np.log10(AA_level_df[\"mutation_frequency\"]).replace([np.inf, -np.inf], np.nan)\n",
    "site_level_df[\"mutation_frequency\"] = np.log10(site_level_df[\"mutation_frequency\"]).replace([np.inf, -np.inf], np.nan)\n",
    "\n",
    "# Mark mutations if in top percentile of summed escape sites\n",
    "for antibody in [\"GPC-A\", \"GP1-A\", \"GPC-B\", \"GPC-C\"]:\n",
    "\n",
    "    # Calculate cutoff \n",
    "    # cutoff = AA_level_df[\"site_\"+antibody].quantile(0.95)\n",
    "    cutoff = AA_level_df[\"site_\"+antibody].median() * 10\n",
    "\n",
    "    # Mark sites that are greater than or equal to the top 5% summed escape site\n",
    "    AA_level_df[\"top_site_escape_for_\"+antibody] = AA_level_df[\"site_\"+antibody].apply(lambda x: True if x >= cutoff else False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2077c3e9-8605-434f-b9dd-cb8b6d8a2b8d",
   "metadata": {},
   "source": [
    "## Comparisons of functional effects and antibody escape\n",
    "\n",
    "We are going to look at the correlation of functional effects and antibody escape for individual amino acid mutations as well as site averaged functional effects and site summed antibody escape."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c2a0e435-24ad-402d-afbc-b1729889c6c6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Making two lists for values and colors \n",
    "dom = [True, False] \n",
    "rng = [\"#EE7733FF\", \"#00000026\"] \n",
    "\n",
    "subplots = []\n",
    "for index,antibody in enumerate(antibody_files):\n",
    "    antibody_name = antibody[0].split(\"/\")[-1].split(\"_\")[0]\n",
    "\n",
    "    curr_subplot = alt.Chart(AA_level_df, title=antibody_name).mark_point(\n",
    "        filled=True, \n",
    "        color=\"black\", \n",
    "        size=75\n",
    "    ).encode(\n",
    "        alt.X(\n",
    "            \"escape_\" + antibody_name,\n",
    "            axis=alt.Axis(\n",
    "                title=\"escape\", \n",
    "                values=[0,1,2,3,4,5,6],\n",
    "                domainWidth=1,\n",
    "                domainColor=\"black\",\n",
    "                tickColor=\"black\",\n",
    "            ),\n",
    "            scale=alt.Scale(domain=[-0.1,6.5])\n",
    "        ),\n",
    "        alt.Y(\n",
    "            \"effect\",\n",
    "            axis=alt.Axis(\n",
    "                title=[\" \", \"effect on cell entry\"], \n",
    "                values=[-1,0,1],\n",
    "                domainWidth=1,\n",
    "                domainColor=\"black\",\n",
    "                tickColor=\"black\",\n",
    "            ),\n",
    "            scale=alt.Scale(domain=[-1.75,1.25])\n",
    "        ),\n",
    "        tooltip=[\n",
    "            \"site\",\n",
    "            \"wildtype\",\n",
    "            \"mutant\",\n",
    "            \"effect\",\n",
    "            \"escape_\" + antibody_name,\n",
    "        ],\n",
    "        color=alt.Color(\n",
    "            \"validated\", \n",
    "            scale=alt.Scale(domain=dom, range=rng),\n",
    "        ), \n",
    "    ).properties(\n",
    "        width=150,\n",
    "        height=150,\n",
    "    )\n",
    "    \n",
    "    subplots.append(curr_subplot)\n",
    "\n",
    "func_vs_antibody_AA = alt.hconcat(\n",
    "    subplots[0],\n",
    "    subplots[1],\n",
    "    subplots[2],\n",
    "    subplots[3],\n",
    "    subplots[4],\n",
    "    subplots[5],\n",
    "    spacing=5,\n",
    "    title=[\"Correlations of functional effects and antibody escape\", \"for individual amino-acid mutations\"],\n",
    ")\n",
    "\n",
    "subplots = []\n",
    "for index,antibody in enumerate(antibody_files):\n",
    "    antibody_name = antibody[0].split(\"/\")[-1].split(\"_\")[0]\n",
    "\n",
    "    curr_subplot = alt.Chart(site_level_df, title=antibody_name).mark_point(\n",
    "        filled=True, \n",
    "        color=\"black\", \n",
    "        size=75\n",
    "    ).encode(\n",
    "        alt.X(\n",
    "            \"escape_\" + antibody_name,\n",
    "            axis=alt.Axis(\n",
    "                title=\"total site escape\", \n",
    "                # values=[0,1,2,3,4,5,6],\n",
    "                domainWidth=1,\n",
    "                domainColor=\"black\",\n",
    "                tickColor=\"black\",\n",
    "            ),\n",
    "            # scale=alt.Scale(domain=[-0.1,6.5])\n",
    "        ),\n",
    "        alt.Y(\n",
    "            \"effect\",\n",
    "            axis=alt.Axis(\n",
    "                title=[\"average site\", \"effect on cell entry\"], \n",
    "                values=[-4,-3,-2,-1,0,1],\n",
    "                domainWidth=1,\n",
    "                domainColor=\"black\",\n",
    "                tickColor=\"black\",\n",
    "            ),\n",
    "            scale=alt.Scale(domain=[-4.1,1.1])\n",
    "        ),\n",
    "        tooltip=[\n",
    "            \"site\",\n",
    "            \"wildtype\",\n",
    "            \"effect\",\n",
    "            \"escape_\" + antibody_name,\n",
    "        ],\n",
    "        color=alt.Color(\n",
    "            \"validated\", \n",
    "            scale=alt.Scale(domain=dom, range=rng),\n",
    "        ), \n",
    "    ).properties(\n",
    "        width=150,\n",
    "        height=150,\n",
    "    )\n",
    "    \n",
    "    subplots.append(curr_subplot)\n",
    "\n",
    "func_vs_antibody_site = alt.hconcat(\n",
    "    subplots[0],\n",
    "    subplots[1],\n",
    "    subplots[2],\n",
    "    subplots[3],\n",
    "    subplots[4],\n",
    "    subplots[5],\n",
    "    spacing=5,\n",
    "    title=[\"Correlations of functional effects and antibody escape\", \"for summed site escape\"],\n",
    ")\n",
    "\n",
    "combined_plot = alt.vconcat(\n",
    "    func_vs_antibody_AA,\n",
    "    func_vs_antibody_site,\n",
    "    spacing=5,\n",
    ").configure_axis(\n",
    "    grid=False,\n",
    "    labelFontSize=16,\n",
    "    titleFontSize=16,\n",
    "    labelFontWeight=\"normal\",\n",
    "    titleFontWeight=\"normal\",\n",
    ").configure_title(\n",
    "    fontSize=24,\n",
    ").configure_legend(\n",
    "    labelFontSize=16,\n",
    "    titleFontSize=16,\n",
    ")\n",
    "\n",
    "combined_plot"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fc40b105-6763-4fe9-94f8-f17cfe5b9e96",
   "metadata": {},
   "source": [
    "Again look at correlation of amino acid functional effects and antibody escape but across all antibodies."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f3871fb7-27ab-457b-ae0d-a92d2d94eb27",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Making two lists for values and colors \n",
    "dom = [True, False] \n",
    "rng = [\"#EE7733FF\", \"#00000026\"] \n",
    "\n",
    "# Calculate statistics\n",
    "r, p = sp.stats.pearsonr(\n",
    "    AA_level_df[[\"total_escape\",\"effect\"]].dropna()[\"total_escape\"],\n",
    "    AA_level_df[[\"total_escape\",\"effect\"]].dropna()[\"effect\"],\n",
    ")\n",
    "print(f\"r correlation of total escape and effect on cell entry: {r:.2f}\")\n",
    "\n",
    "effect_vs_antibody_AA = alt.Chart(\n",
    "    AA_level_df, \n",
    "    title=\"for individual amino-acid mutations\",\n",
    ").mark_point(\n",
    "    filled=True, \n",
    "    color=\"black\", \n",
    "    size=75,\n",
    ").encode(\n",
    "    alt.X(\n",
    "        \"total_escape\",\n",
    "        axis=alt.Axis(\n",
    "            title=[\"escape across\", \"all mapped antibodies\"],  \n",
    "            # values=[0,1,2,3],\n",
    "            domainWidth=1,\n",
    "            domainColor=\"black\",\n",
    "            tickColor=\"black\",\n",
    "        ),\n",
    "        # scale=alt.Scale(domain=[-0.1, 3.1])\n",
    "    ),\n",
    "    alt.Y(\n",
    "        \"effect\",\n",
    "        axis=alt.Axis(\n",
    "            title=[\" \", \"effect on cell entry\"], \n",
    "            values=[-1,0,1],\n",
    "            domainWidth=1,\n",
    "            domainColor=\"black\",\n",
    "            tickColor=\"black\",\n",
    "        ),\n",
    "        scale=alt.Scale(domain=[-1.75,1.25])\n",
    "    ),\n",
    "    tooltip=[\n",
    "        \"site\",\n",
    "        \"wildtype\",\n",
    "        \"mutant\",\n",
    "        \"effect\",\n",
    "        \"escape_2510C\",\n",
    "        \"escape_121F\",\n",
    "        \"escape_89F\",\n",
    "        \"escape_377H\",\n",
    "        \"escape_372D\",\n",
    "        \"escape_256A\",\n",
    "        \"total_escape\",\n",
    "    ],\n",
    "    color=alt.Color(\n",
    "        \"validated\", \n",
    "        scale=alt.Scale(domain=dom, range=rng),\n",
    "    ), \n",
    ").properties(\n",
    "    width=300,\n",
    "    height=300,\n",
    ")\n",
    "\n",
    "\n",
    "# Calculate statistics\n",
    "r, p = sp.stats.pearsonr(\n",
    "    site_level_df[[\"total_escape\",\"effect\"]].dropna()[\"total_escape\"],\n",
    "    site_level_df[[\"total_escape\",\"effect\"]].dropna()[\"effect\"],\n",
    ")\n",
    "print(f\"r correlation of total escape and effect on cell entry: {r:.2f}\")\n",
    "\n",
    "effect_vs_antibody_site = alt.Chart(\n",
    "    site_level_df, \n",
    "    title=\"for summed site escape\",\n",
    ").mark_point(\n",
    "    filled=True, \n",
    "    color=\"black\", \n",
    "    size=75,\n",
    ").encode(\n",
    "    alt.X(\n",
    "        \"total_escape\",\n",
    "        axis=alt.Axis(\n",
    "            title=[\"total site escape across\", \"all mapped antibodies\"],  \n",
    "            # values=[0,1,2,3],\n",
    "            domainWidth=1,\n",
    "            domainColor=\"black\",\n",
    "            tickColor=\"black\",\n",
    "        ),\n",
    "        # scale=alt.Scale(domain=[-0.1, 3.1])\n",
    "    ),\n",
    "    alt.Y(\n",
    "        \"effect\",\n",
    "        axis=alt.Axis(\n",
    "            title=[\"average site\",\"effect on cell entry\"], \n",
    "            values=[-4,-3,-2,-1,0,1],\n",
    "            domainWidth=1,\n",
    "            domainColor=\"black\",\n",
    "            tickColor=\"black\",\n",
    "        ),\n",
    "        scale=alt.Scale(domain=[-4.1,1.1])\n",
    "    ),\n",
    "    tooltip=[\n",
    "        \"site\",\n",
    "        \"wildtype\",\n",
    "        \"effect\",\n",
    "        \"escape_2510C\",\n",
    "        \"escape_121F\",\n",
    "        \"escape_89F\",\n",
    "        \"escape_377H\",\n",
    "        \"escape_372D\",\n",
    "        \"escape_256A\",\n",
    "        \"total_escape\",\n",
    "    ],\n",
    "    color=alt.Color(\n",
    "        \"validated\", \n",
    "        scale=alt.Scale(domain=dom, range=rng),\n",
    "    ), \n",
    ").properties(\n",
    "    width=300,\n",
    "    height=300,\n",
    ")\n",
    "\n",
    "combined_plot = alt.hconcat(\n",
    "    effect_vs_antibody_AA,\n",
    "    effect_vs_antibody_site,\n",
    "    spacing=5,\n",
    "    title=\"Correlations of functional effects and antibody escape\"\n",
    ").configure_axis(\n",
    "    grid=False,\n",
    "    labelFontSize=16,\n",
    "    titleFontSize=16,\n",
    "    labelFontWeight=\"normal\",\n",
    "    titleFontWeight=\"normal\",\n",
    ").configure_title(\n",
    "    fontSize=24,\n",
    ").configure_legend(\n",
    "    labelFontSize=16,\n",
    "    titleFontSize=16,\n",
    ")\n",
    "\n",
    "combined_plot"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f9cf29d7-dfab-4bd4-9b57-ef8b0d348eea",
   "metadata": {},
   "source": [
    "## Compare DMS data to natural sequences\n",
    "\n",
    "Next, we are going to compare DMS antibody escape scores to natural sequence variation. Natural sequence variation is measured by two methods:\n",
    "1. Mutation frequency relative to Josiah strain for both single amino acid mutations as well as cumulative site mutation frequencies.\n",
    "2. Number of effective amino acids per site as described in *Biophysical Models of Protein Evolution: Understanding the Patterns of Evolutionary Sequence Divergence*.\n",
    "\n",
    "These diversity metrics will be compared to antibody escape as measured for individual amino acid mutations, site summed escape, and site summed escape across all mapped antibodies."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "793c19fb-335f-45bb-ae0b-5776908766a1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Making two lists for values and colors \n",
    "dom = [True, False] \n",
    "rng = [\"#EE7733FF\", \"#00000026\"] \n",
    "\n",
    "subplots = []\n",
    "for index,antibody in enumerate(antibody_files):\n",
    "    antibody_name = antibody[0].split(\"/\")[-1].split(\"_\")[0]\n",
    "\n",
    "    curr_subplot = alt.Chart(AA_level_df, title=antibody_name).mark_point(\n",
    "        filled=True, \n",
    "        color=\"black\", \n",
    "        size=75\n",
    "    ).encode(\n",
    "        alt.X(\n",
    "            \"mutation_frequency\",\n",
    "            axis=alt.Axis(\n",
    "                title=\"mutation frequency\", \n",
    "                values=[-3,-2,-1],\n",
    "                domainWidth=1,\n",
    "                domainColor=\"black\",\n",
    "                tickColor=\"black\",\n",
    "            ),\n",
    "            scale=alt.Scale(domain=[-3.1,0])\n",
    "        ),\n",
    "        alt.Y(\n",
    "            \"escape_\" + antibody_name,\n",
    "            axis=alt.Axis(\n",
    "                title=\"escape\", \n",
    "                values=[0,1,2,3],\n",
    "                domainWidth=1,\n",
    "                domainColor=\"black\",\n",
    "                tickColor=\"black\",\n",
    "            ),\n",
    "            scale=alt.Scale(domain=[0,3.1])\n",
    "        ),\n",
    "        tooltip=[\n",
    "            \"site\",\n",
    "            \"wildtype\",\n",
    "            \"mutant\",\n",
    "            \"mutation_frequency\",\n",
    "            \"effect\",\n",
    "            \"escape_\" + antibody_name,\n",
    "        ],\n",
    "        color=alt.Color(\n",
    "            \"validated\", \n",
    "            scale=alt.Scale(domain=dom, range=rng),\n",
    "        ), \n",
    "    ).properties(\n",
    "        width=150,\n",
    "        height=150,\n",
    "    )\n",
    "    \n",
    "    subplots.append(curr_subplot)\n",
    "\n",
    "natural_vs_antibody = alt.hconcat(\n",
    "    subplots[0],\n",
    "    subplots[1],\n",
    "    subplots[2],\n",
    "    subplots[3],\n",
    "    subplots[4],\n",
    "    subplots[5],\n",
    "    spacing=5,\n",
    "    title=[\"Correlations of mutational frequencies and antibody escape\", \"for individual amino-acid mutations\"],\n",
    ").configure_axis(\n",
    "    grid=False,\n",
    "    labelFontSize=16,\n",
    "    titleFontSize=16,\n",
    "    labelFontWeight=\"normal\",\n",
    "    titleFontWeight=\"normal\",\n",
    ").configure_title(\n",
    "    fontSize=24,\n",
    ").configure_legend(\n",
    "    labelFontSize=16,\n",
    "    titleFontSize=16,\n",
    ")\n",
    "\n",
    "natural_vs_antibody"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0587109d-936d-492a-bdb1-1b89870b9e8b",
   "metadata": {},
   "outputs": [],
   "source": [
    "r, p = sp.stats.pearsonr(\n",
    "    AA_level_df[[\"total_escape\",\"mutation_frequency\"]].replace([np.inf, -np.inf], np.nan).dropna()[\"total_escape\"],\n",
    "    AA_level_df[[\"total_escape\",\"mutation_frequency\"]].replace([np.inf, -np.inf], np.nan).dropna()[\"mutation_frequency\"],\n",
    ")\n",
    "print(f\"r correlation of total escape and natural sequence diversity: {r:.2f}\")\n",
    "\n",
    "natural_vs_antibody = alt.Chart(\n",
    "    AA_level_df,\n",
    "    title=[\"Correlations of mutational frequencies and antibody escape\", \"for individual amino-acid mutations\"]\n",
    ").mark_point(\n",
    "    filled=True, \n",
    "    color=\"black\", \n",
    "    size=75,\n",
    ").encode(\n",
    "    alt.Y(\n",
    "        \"total_escape\",\n",
    "        axis=alt.Axis(\n",
    "            title=[\"escape across\", \"all mapped antibodies\"], \n",
    "            # values=[0,1,2,3],\n",
    "            domainWidth=1,\n",
    "            domainColor=\"black\",\n",
    "            tickColor=\"black\",\n",
    "        ),\n",
    "        # scale=alt.Scale(domain=[-0.1, 3.1])\n",
    "    ),\n",
    "    alt.X(\n",
    "        \"mutation_frequency\",\n",
    "        axis=alt.Axis(\n",
    "            title=\"mutation frequency\", \n",
    "            values=[-3,-2,-1],\n",
    "            domainWidth=1,\n",
    "            domainColor=\"black\",\n",
    "            tickColor=\"black\",\n",
    "        ),\n",
    "        scale=alt.Scale(domain=[-3.1,0])\n",
    "    ),\n",
    "    tooltip=[\n",
    "        \"site\",\n",
    "        \"wildtype\",\n",
    "        \"mutant\",\n",
    "        \"mutation_frequency\",\n",
    "        \"escape_2510C\",\n",
    "        \"escape_121F\",\n",
    "        \"escape_89F\",\n",
    "        \"escape_377H\",\n",
    "        \"escape_372D\",\n",
    "        \"escape_256A\",\n",
    "        \"total_escape\",\n",
    "    ],\n",
    "    color=alt.Color(\n",
    "        \"validated\", \n",
    "        scale=alt.Scale(domain=dom, range=rng),\n",
    "    ), \n",
    ").properties(\n",
    "    width=300,\n",
    "    height=300,\n",
    ").configure_axis(\n",
    "    grid=False,\n",
    "    labelFontSize=16,\n",
    "    titleFontSize=16,\n",
    "    labelFontWeight=\"normal\",\n",
    "    titleFontWeight=\"normal\",\n",
    ").configure_title(\n",
    "    fontSize=24,\n",
    ").configure_legend(\n",
    "    labelFontSize=16,\n",
    "    titleFontSize=16,\n",
    ")\n",
    "\n",
    "natural_vs_antibody"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c9b1dede-5419-4b91-a43c-e6ac348a36cb",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Making two lists for values and colors \n",
    "dom = [True, False] \n",
    "rng = [\"#EE7733FF\", \"#00000026\"] \n",
    "\n",
    "subplots = []\n",
    "for index,antibody in enumerate(antibody_files):\n",
    "    antibody_name = antibody[0].split(\"/\")[-1].split(\"_\")[0]\n",
    "\n",
    "    curr_subplot = alt.Chart(site_level_df, title=antibody_name).mark_point(\n",
    "        filled=True, \n",
    "        color=\"black\", \n",
    "        size=75\n",
    "    ).encode(\n",
    "        alt.X(\n",
    "            \"escape_\" + antibody_name,\n",
    "            axis=alt.Axis(\n",
    "                title=\"escape\", \n",
    "                # values=[0,1,2,3,4,5,6],\n",
    "                domainWidth=1,\n",
    "                domainColor=\"black\",\n",
    "                tickColor=\"black\",\n",
    "            ),\n",
    "            # scale=alt.Scale(domain=[-0.1,6.5])\n",
    "        ),\n",
    "        alt.Y(\n",
    "            \"mutation_frequency\",\n",
    "            axis=alt.Axis(\n",
    "                title=[\"mutation frequencies\", \"in natural sequences\"], \n",
    "                values=[-3,-2,-1,0],\n",
    "                domainWidth=1,\n",
    "                domainColor=\"black\",\n",
    "                tickColor=\"black\",\n",
    "                # format=\"e\", \n",
    "            ),\n",
    "            scale=alt.Scale(domain=[-3.1,0])\n",
    "        ),\n",
    "        tooltip=[\n",
    "            \"site\",\n",
    "            \"wildtype\",\n",
    "            \"mutation_frequency\",\n",
    "            \"escape_\" + antibody_name,\n",
    "            \"effect\",\n",
    "        ],\n",
    "        color=alt.Color(\n",
    "            \"validated\", \n",
    "            scale=alt.Scale(domain=dom, range=rng),\n",
    "        ), \n",
    "    ).properties(\n",
    "        width=150,\n",
    "        height=150,\n",
    "    )\n",
    "    \n",
    "    subplots.append(curr_subplot)\n",
    "\n",
    "natural_vs_antibody_mut_freq = alt.hconcat(\n",
    "    subplots[0],\n",
    "    subplots[1],\n",
    "    subplots[2],\n",
    "    subplots[3],\n",
    "    subplots[4],\n",
    "    subplots[5],\n",
    "    spacing=5,\n",
    "    title=[\"Correlations of mutational frequencies and antibody escape\", \"for summed site escape\"],\n",
    ")\n",
    "\n",
    "subplots = []\n",
    "for index,antibody in enumerate(antibody_files):\n",
    "    antibody_name = antibody[0].split(\"/\")[-1].split(\"_\")[0]\n",
    "\n",
    "    curr_subplot = alt.Chart(site_level_df, title=antibody_name).mark_point(\n",
    "        filled=True, \n",
    "        color=\"black\", \n",
    "        size=75\n",
    "    ).encode(\n",
    "        alt.X(\n",
    "            \"escape_\" + antibody_name,\n",
    "            axis=alt.Axis(\n",
    "                title=\"total site escape\", \n",
    "                domainWidth=1,\n",
    "                domainColor=\"black\",\n",
    "                tickColor=\"black\",\n",
    "            ),\n",
    "            # scale=alt.Scale(domainMin=-1)\n",
    "        ),\n",
    "        alt.Y(\n",
    "            \"n_effective\",\n",
    "            axis=alt.Axis(\n",
    "                title=[\"effective amino acids\", \"in natural sequences\"], \n",
    "                values=[1,2,3,4],\n",
    "                domainWidth=1,\n",
    "                domainColor=\"black\",\n",
    "                tickColor=\"black\",\n",
    "            ),\n",
    "            scale=alt.Scale(domain=[0.9,4])\n",
    "        ),\n",
    "        tooltip=[\n",
    "            \"site\",\n",
    "            \"wildtype\",\n",
    "            \"n_effective\",\n",
    "            \"escape_\" + antibody_name,\n",
    "            \"effect\",\n",
    "        ],\n",
    "        color=alt.Color(\n",
    "            \"validated\", \n",
    "            scale=alt.Scale(domain=dom, range=rng),\n",
    "        ), \n",
    "    ).properties(\n",
    "        width=150,\n",
    "        height=150,\n",
    "    )\n",
    "    \n",
    "    subplots.append(curr_subplot)\n",
    "\n",
    "natural_vs_antibody_neff = alt.hconcat(\n",
    "    subplots[0],\n",
    "    subplots[1],\n",
    "    subplots[2],\n",
    "    subplots[3],\n",
    "    subplots[4],\n",
    "    subplots[5],\n",
    "    spacing=5,\n",
    "    title=[\"Correlations of effective amino acids and antibody escape\", \"for summed site escape\"]\n",
    ")\n",
    "\n",
    "combined_plot = alt.vconcat(\n",
    "    natural_vs_antibody_mut_freq,\n",
    "    natural_vs_antibody_neff,\n",
    "    spacing=5,\n",
    ").configure_axis(\n",
    "    grid=False,\n",
    "    labelFontSize=16,\n",
    "    titleFontSize=16,\n",
    "    labelFontWeight=\"normal\",\n",
    "    titleFontWeight=\"normal\",\n",
    ").configure_axis(\n",
    "    grid=False,\n",
    "    labelFontSize=16,\n",
    "    titleFontSize=16,\n",
    "    labelFontWeight=\"normal\",\n",
    "    titleFontWeight=\"normal\",\n",
    ").configure_title(\n",
    "    fontSize=24,\n",
    ").configure_legend(\n",
    "    labelFontSize=16,\n",
    "    titleFontSize=16,\n",
    ")\n",
    "\n",
    "combined_plot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c6cf537e-df4b-40e4-a7ab-6a4028780992",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Making two lists for values and colors \n",
    "dom = [True, False] \n",
    "rng = [\"#EE7733FF\", \"#00000026\"] \n",
    "\n",
    "# Calculate statistics\n",
    "r, p = sp.stats.pearsonr(\n",
    "    site_level_df[[\"total_escape\",\"mutation_frequency\"]].dropna()[\"total_escape\"],\n",
    "    site_level_df[[\"total_escape\",\"mutation_frequency\"]].dropna()[\"mutation_frequency\"],\n",
    ")\n",
    "print(f\"r correlation of total escape across all mapped antibodies and site mutational frequency: {r:.2f}\")\n",
    "\n",
    "natural_vs_antibody_mut_freq = alt.Chart(\n",
    "    site_level_df, \n",
    ").mark_point(\n",
    "    filled=True, \n",
    "    color=\"black\", \n",
    "    size=75,\n",
    ").encode(\n",
    "    alt.X(\n",
    "        \"total_escape\",\n",
    "        axis=alt.Axis(\n",
    "            title=[\"escape across\", \"all mapped antibodies\"],  \n",
    "            # values=[0,1,2,3],\n",
    "            domainWidth=1,\n",
    "            domainColor=\"black\",\n",
    "            tickColor=\"black\",\n",
    "        ),\n",
    "        # scale=alt.Scale(domain=[-0.1, 3.1])\n",
    "    ),\n",
    "    alt.Y(\n",
    "        \"mutation_frequency\",\n",
    "        axis=alt.Axis(\n",
    "            title=[\"mutation frequencies\", \"in natural sequences\"],\n",
    "            # values=[-2,-1,0,1],\n",
    "            domainWidth=1,\n",
    "            domainColor=\"black\",\n",
    "            tickColor=\"black\",\n",
    "        ),\n",
    "        # scale=alt.Scale(domain=[-2,1.5])\n",
    "    ),\n",
    "    tooltip=[\n",
    "        \"site\",\n",
    "        \"wildtype\",\n",
    "        \"mutation_frequency\",\n",
    "        \"escape_2510C\",\n",
    "        \"escape_121F\",\n",
    "        \"escape_89F\",\n",
    "        \"escape_377H\",\n",
    "        \"escape_372D\",\n",
    "        \"escape_256A\",\n",
    "        \"total_escape\",\n",
    "    ],\n",
    "    color=alt.Color(\n",
    "        \"validated\", \n",
    "        scale=alt.Scale(domain=dom, range=rng),\n",
    "    ), \n",
    ").properties(\n",
    "    width=300,\n",
    "    height=300,\n",
    ")\n",
    "\n",
    "\n",
    "# Calculate statistics\n",
    "r, p = sp.stats.pearsonr(\n",
    "    site_level_df[[\"total_escape\",\"n_effective\"]].dropna()[\"total_escape\"],\n",
    "    site_level_df[[\"total_escape\",\"n_effective\"]].dropna()[\"n_effective\"],\n",
    ")\n",
    "print(f\"r correlation of total escape across all mapped antibodies and site effective amino acids: {r:.2f}\")\n",
    "\n",
    "natural_vs_antibody_neff = alt.Chart(\n",
    "    site_level_df, \n",
    ").mark_point(\n",
    "    filled=True, \n",
    "    color=\"black\", \n",
    "    size=75,\n",
    ").encode(\n",
    "    alt.X(\n",
    "        \"total_escape\",\n",
    "        axis=alt.Axis(\n",
    "            title=[\"site escape across\", \"all mapped antibodies\"],  \n",
    "            # values=[0,1,2,3],\n",
    "            domainWidth=1,\n",
    "            domainColor=\"black\",\n",
    "            tickColor=\"black\",\n",
    "        ),\n",
    "        # scale=alt.Scale(domain=[-0.1, 3.1])\n",
    "    ),\n",
    "    alt.Y(\n",
    "        \"n_effective\",\n",
    "        axis=alt.Axis(\n",
    "            title=[\"effective amino acids\", \"in natural sequences\"], \n",
    "            values=[1,2,3,4],\n",
    "            domainWidth=1,\n",
    "            domainColor=\"black\",\n",
    "            tickColor=\"black\",\n",
    "        ),\n",
    "        scale=alt.Scale(domain=[1,4])\n",
    "    ),\n",
    "    tooltip=[\n",
    "        \"site\",\n",
    "        \"wildtype\",\n",
    "        \"n_effective\",\n",
    "        \"escape_2510C\",\n",
    "        \"escape_121F\",\n",
    "        \"escape_89F\",\n",
    "        \"escape_377H\",\n",
    "        \"escape_372D\",\n",
    "        \"escape_256A\",\n",
    "        \"total_escape\",\n",
    "    ],\n",
    "    color=alt.Color(\n",
    "        \"validated\", \n",
    "        scale=alt.Scale(domain=dom, range=rng),\n",
    "    ), \n",
    ").properties(\n",
    "    width=300,\n",
    "    height=300,\n",
    ")\n",
    "\n",
    "combined_plot = alt.hconcat(\n",
    "    natural_vs_antibody_mut_freq,\n",
    "    natural_vs_antibody_neff,\n",
    "    spacing=5,\n",
    "    title=\"Correlations of site natural sequence diversity and antibody escape\"\n",
    ").configure_axis(\n",
    "    grid=False,\n",
    "    labelFontSize=16,\n",
    "    titleFontSize=16,\n",
    "    labelFontWeight=\"normal\",\n",
    "    titleFontWeight=\"normal\",\n",
    ").configure_title(\n",
    "    fontSize=24,\n",
    ").configure_legend(\n",
    "    labelFontSize=16,\n",
    "    titleFontSize=16,\n",
    ")\n",
    "\n",
    "combined_plot"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "36951622-39eb-4b01-86e9-d3ee89af62dc",
   "metadata": {},
   "source": [
    "## Correlations of natural sequence diversity and antibody escape grouped by antibody epitope\n",
    "\n",
    "Finally, we are going to look at natural sequence diversity with respect to antibody escape when averaging across antibody epitopes. This only effects GPC-B antibodies because there are three of those antibodies but makes for more compact plots."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "33f3c9d7-d3f0-4cb8-be94-d67db6497486",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Making two lists for values and colors \n",
    "dom = [True, False] \n",
    "rng = [\"#EE7733FF\", \"#00000026\"] \n",
    "\n",
    "subplots = []\n",
    "for index,antibody in enumerate([\"GPC-A\", \"GP1-A\", \"GPC-B\", \"GPC-C\"]):\n",
    "\n",
    "    curr_subplot = alt.Chart(\n",
    "        site_level_df,\n",
    "        title=antibody,\n",
    "    ).mark_point(\n",
    "        filled=True, \n",
    "        color=\"black\", \n",
    "        size=75\n",
    "    ).encode(\n",
    "        alt.X(\n",
    "            antibody,\n",
    "            axis=alt.Axis(\n",
    "                title=\"escape\", \n",
    "                # values=[0,1,2,3,4,5,6],\n",
    "                domainWidth=1,\n",
    "                domainColor=\"black\",\n",
    "                tickColor=\"black\",\n",
    "            ),\n",
    "            # scale=alt.Scale(domain=[-0.1,6.5])\n",
    "        ),\n",
    "        alt.Y(\n",
    "            \"mutation_frequency\",\n",
    "            axis=alt.Axis(\n",
    "                title=[\"mutation frequencies\", \"in natural sequences\"], \n",
    "                # values=[0.001, 0.01, 0.1, 1],\n",
    "                domainWidth=1,\n",
    "                domainColor=\"black\",\n",
    "                tickColor=\"black\",\n",
    "                # format=\"e\", \n",
    "            ),\n",
    "            # scale=alt.Scale(domain=[0.001, 1], type=\"log\")\n",
    "        ),\n",
    "        tooltip=[\n",
    "            \"site\",\n",
    "            \"wildtype\",\n",
    "            \"mutation_frequency\",\n",
    "            \"effect\",\n",
    "        ],\n",
    "        color=alt.Color(\n",
    "            \"validated\", \n",
    "            scale=alt.Scale(domain=dom, range=rng),\n",
    "        ), \n",
    "    ).properties(\n",
    "        width=150,\n",
    "        height=150,\n",
    "    )\n",
    "    \n",
    "    subplots.append(curr_subplot)\n",
    "\n",
    "natural_vs_antibody_mut_freq = alt.hconcat(\n",
    "    subplots[0],\n",
    "    subplots[1],\n",
    "    subplots[2],\n",
    "    subplots[3],\n",
    "    spacing=5,\n",
    "    title=[\"Correlations of site mutational frequencies and antibody escape\", \"for summed site escape\"],\n",
    ")\n",
    "\n",
    "subplots = []\n",
    "for index,antibody in enumerate([\"GPC-A\", \"GP1-A\", \"GPC-B\", \"GPC-C\"]):\n",
    "\n",
    "    curr_subplot = alt.Chart(\n",
    "        site_level_df, \n",
    "        title=antibody,\n",
    "    ).mark_point(\n",
    "        filled=True, \n",
    "        color=\"black\", \n",
    "        size=75\n",
    "    ).encode(\n",
    "        alt.X(\n",
    "            antibody,\n",
    "            axis=alt.Axis(\n",
    "                title=\"total site escape\", \n",
    "                domainWidth=1,\n",
    "                domainColor=\"black\",\n",
    "                tickColor=\"black\",\n",
    "            ),\n",
    "            # scale=alt.Scale(domainMin=-1)\n",
    "        ),\n",
    "        alt.Y(\n",
    "            \"n_effective\",\n",
    "            axis=alt.Axis(\n",
    "                title=[\"effective amino acids\", \"in natural sequences\"], \n",
    "                values=[1,2,3,4],\n",
    "                domainWidth=1,\n",
    "                domainColor=\"black\",\n",
    "                tickColor=\"black\",\n",
    "            ),\n",
    "            scale=alt.Scale(domain=[1,4])\n",
    "        ),\n",
    "        tooltip=[\n",
    "            \"site\",\n",
    "            \"wildtype\",\n",
    "            \"n_effective\",\n",
    "            \"effect\",\n",
    "        ],\n",
    "        color=alt.Color(\n",
    "            \"validated\", \n",
    "            scale=alt.Scale(domain=dom, range=rng),\n",
    "        ), \n",
    "    ).properties(\n",
    "        width=150,\n",
    "        height=150,\n",
    "    )\n",
    "    \n",
    "    subplots.append(curr_subplot)\n",
    "\n",
    "natural_vs_antibody_neff = alt.hconcat(\n",
    "    subplots[0],\n",
    "    subplots[1],\n",
    "    subplots[2],\n",
    "    subplots[3],\n",
    "    spacing=5,\n",
    "    title=[\"Correlations of site effective amino acids and antibody escape\", \"for summed site escape\"],\n",
    ")\n",
    "\n",
    "combined_plot = alt.vconcat(\n",
    "    natural_vs_antibody_mut_freq,\n",
    "    natural_vs_antibody_neff,\n",
    "    spacing=5,\n",
    ").configure_axis(\n",
    "    grid=False,\n",
    "    labelFontSize=16,\n",
    "    titleFontSize=16,\n",
    "    labelFontWeight=\"normal\",\n",
    "    titleFontWeight=\"normal\",\n",
    ").configure_axis(\n",
    "    grid=False,\n",
    "    labelFontSize=16,\n",
    "    titleFontSize=16,\n",
    "    labelFontWeight=\"normal\",\n",
    "    titleFontWeight=\"normal\",\n",
    ").configure_title(\n",
    "    fontSize=24,\n",
    ").configure_legend(\n",
    "    labelFontSize=16,\n",
    "    titleFontSize=16,\n",
    ")\n",
    "\n",
    "combined_plot"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9f9671a9-a059-4e67-8ad6-d912e4fa2111",
   "metadata": {},
   "source": [
    "## Correlations of mutational frequency and escape\n",
    "\n",
    "The last plot shows the correlation of antibody escape for individual amino acids when averaged across all of the same epitope targetting antibodies compared to mutation frequency in natural sequences. In addition, any mutation that is part of a strong escape site (i.e., a summed site escape that is 10-fold greater than the median of all summed escape sites) are marked a different color. This plot most closely resembles the calculations conducted to identify natural isolates that would potentially escape the mapped antibodies. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "46971a05-a2e5-4cf9-abbf-8f7deb3fcafe",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Making two lists for values and colors \n",
    "dom = [True, False] \n",
    "rng = [\"#EE7733FF\", \"#00000026\"] \n",
    "\n",
    "subplots = []\n",
    "for index,antibody in enumerate([\"GPC-A\", \"GP1-A\", \"GPC-B\", \"GPC-C\"]):\n",
    "    \n",
    "    curr_subplot = alt.Chart(\n",
    "        # AA_level_df.loc[AA_level_df[\"site_\" + antibody] >= cutoff], \n",
    "        AA_level_df,\n",
    "        title=antibody\n",
    "    ).mark_point(\n",
    "        filled=True, \n",
    "        color=\"black\", \n",
    "        size=75\n",
    "    ).encode(\n",
    "        alt.X(\n",
    "            \"mutation_frequency\",\n",
    "            axis=alt.Axis(\n",
    "                title=\"mutation frequency\", \n",
    "                values=[-3,-2,-1],\n",
    "                domainWidth=1,\n",
    "                domainColor=\"black\",\n",
    "                tickColor=\"black\",\n",
    "            ),\n",
    "            scale=alt.Scale(domain=[-3.2,0])\n",
    "        ),\n",
    "        alt.Y(\n",
    "            antibody,\n",
    "            axis=alt.Axis(\n",
    "                title=\"escape\", \n",
    "                values=[0,1,2,3],\n",
    "                domainWidth=1,\n",
    "                domainColor=\"black\",\n",
    "                tickColor=\"black\",\n",
    "            ),\n",
    "            scale=alt.Scale(domain=[0,3.25])\n",
    "        ),\n",
    "        tooltip=[\n",
    "            \"site\",\n",
    "            \"wildtype\",\n",
    "            \"mutant\",\n",
    "            \"mutation_frequency\",\n",
    "            \"effect\",\n",
    "            antibody,\n",
    "            \"site_\" + antibody,\n",
    "        ],\n",
    "        color=alt.Color(\n",
    "            \"top_site_escape_for_\"+antibody, \n",
    "            scale=alt.Scale(domain=dom, range=rng),\n",
    "            title=[\"mutation present\", \"in site of strong\", \"escape\"],\n",
    "        ), \n",
    "    ).properties(\n",
    "        width=100,\n",
    "        height=150,\n",
    "    )\n",
    "    \n",
    "    subplots.append(curr_subplot)\n",
    "\n",
    "natural_vs_antibody = alt.hconcat(\n",
    "    subplots[0],\n",
    "    subplots[1],\n",
    "    subplots[2],\n",
    "    subplots[3],\n",
    "    spacing=5,\n",
    "    title=\"Correlations of amino-acid mutation frequencies and antibody escape\",\n",
    ").configure_axis(\n",
    "    grid=False,\n",
    "    labelFontSize=16,\n",
    "    titleFontSize=16,\n",
    "    labelFontWeight=\"normal\",\n",
    "    titleFontWeight=\"normal\",\n",
    ").configure_title(\n",
    "    fontSize=24,\n",
    ").configure_legend(\n",
    "    labelFontSize=16,\n",
    "    titleFontSize=16,\n",
    ")\n",
    "\n",
    "# Make output dir if doesn't exist\n",
    "if not os.path.exists(out_dir_natural):\n",
    "    os.mkdir(out_dir_natural)\n",
    "\n",
    "natural_vs_antibody.save(html_output)\n",
    "\n",
    "natural_vs_antibody"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c9b1b8f8-4da8-4cda-8ba8-424f6aace9ae",
   "metadata": {},
   "source": [
    "Recreate **same** plot below but formatted for a manuscript and in static form."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7ebea657-59ce-466f-b7de-b3cdc37ececd",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Set figure size and subplots\n",
    "fig, axes = plt.subplots(\n",
    "    1, \n",
    "    4, \n",
    "    figsize=(4.5, 1), \n",
    "    #sharey=True,\n",
    ")\n",
    "\n",
    "# Adjust spacing of subplots\n",
    "fig.subplots_adjust(\n",
    "    bottom=0, \n",
    "    top=1, \n",
    "    wspace=0.35, \n",
    "    hspace=0,\n",
    ")\n",
    "\n",
    "for index,antibody in enumerate([\"GPC-A\", \"GP1-A\", \"GPC-B\", \"GPC-C\"]):\n",
    "    chart = sns.scatterplot(\n",
    "        data=AA_level_df,\n",
    "        x=\"mutation_frequency\",\n",
    "        y=antibody,\n",
    "        hue=\"top_site_escape_for_\"+antibody,\n",
    "        edgecolor=None,\n",
    "        linewidth=0.5,\n",
    "        palette={False : \"#00000026\", True : \"#EE7733CC\"},\n",
    "        s=20,\n",
    "        ax=axes[index],\n",
    "    )\n",
    "    xticks = [-3, -2, -1]\n",
    "    chart.set_xticks(xticks)\n",
    "    x_labels=[\"$10^{-3}$\", \"$10^{-2}$\", \"$10^{-1}$\"]\n",
    "    chart.set_xticklabels(labels=x_labels, horizontalalignment=\"center\", fontsize=6)\n",
    "\n",
    "    chart.set_xlim(-3.1, 0.1)\n",
    "    chart.set(xlabel=None)\n",
    "\n",
    "\n",
    "    \n",
    "    if index == 0:\n",
    "        # Set only one y axis label\n",
    "        chart.set_ylabel(\"average escape\", fontsize=8)\n",
    "        chart.set_ylim(-0.2, 4.2)\n",
    "        yticks = [0, 1, 2, 3, 4]\n",
    "        chart.set_yticks(yticks)\n",
    "        chart.set_yticklabels(labels=[\"0\", \"1\", \"2\", \"3\", \"4\"], fontsize=6)\n",
    "        chart.set_title(\n",
    "            \"25.10C\", \n",
    "            fontsize=8,\n",
    "            color=\"#44AA99\",\n",
    "            horizontalalignment=\"center\",\n",
    "        )\n",
    "    if index == 1:\n",
    "        chart.set(ylabel=None)\n",
    "        chart.set_ylim(-0.15, 3.15)\n",
    "        yticks = [0, 1, 2, 3]\n",
    "        chart.set_yticks(yticks)\n",
    "        chart.set_yticklabels(labels=[\"0\", \"1\", \"2\", \"3\"], fontsize=6)\n",
    "        chart.set_title(\n",
    "            \"12.1F\", \n",
    "            fontsize=8,\n",
    "            color=\"#999933\",\n",
    "            horizontalalignment=\"center\",\n",
    "        )\n",
    "    if index == 2:\n",
    "        chart.set(ylabel=None)\n",
    "        chart.set_ylim(-0.1, 2.1)\n",
    "        yticks = [0, 1, 2]\n",
    "        chart.set_yticks(yticks)\n",
    "        chart.set_yticklabels(labels=[\"0\", \"1\", \"2\"], fontsize=6)\n",
    "        chart.set_title(\n",
    "            \"37.7H  25.6A  37.2D\", \n",
    "            fontsize=8,\n",
    "            color=\"#AA4499\",\n",
    "            horizontalalignment=\"center\",\n",
    "        )\n",
    "    if index == 3:\n",
    "        chart.set(ylabel=None)\n",
    "        chart.set_ylim(-0.1, 2.1)\n",
    "        yticks = [0, 1, 2]\n",
    "        chart.set_yticks(yticks)\n",
    "        chart.set_yticklabels(labels=[\"0\", \"1\", \"2\"], fontsize=6)\n",
    "        chart.set_title(\n",
    "            \"8.9F\", \n",
    "            fontsize=8,\n",
    "            color=\"#117733\",\n",
    "            horizontalalignment=\"center\",\n",
    "        )\n",
    "        # Make only one legend\n",
    "        sns.move_legend(\n",
    "            chart, \n",
    "            \"lower center\", \n",
    "            bbox_to_anchor=(-1.6, -0.5),\n",
    "            ncol=2,\n",
    "            fontsize=6,\n",
    "            markerscale=1,\n",
    "            handletextpad=0.1,\n",
    "            title=\"mutation in site of strong escape\",\n",
    "            title_fontproperties = {\n",
    "                \"size\" : 6, \n",
    "                # \"weight\" : \"bold\",\n",
    "            },\n",
    "            frameon=False,\n",
    "            borderaxespad=0.1,\n",
    "            reverse=True,\n",
    "            alignment=\"left\",\n",
    "        )\n",
    "        # Add edges to legend markers to match scatter plot\n",
    "        for ha in chart.legend_.legendHandles:\n",
    "            ha.set_edgecolor(None)\n",
    "            ha.set_linewidths(0.5)\n",
    "\n",
    "        def legend_title_left(leg):\n",
    "            \"\"\"\n",
    "            Function to move legend title\n",
    "            https://stackoverflow.com/questions/53316999/pyplot-legend-left-alignment-of-a-title\n",
    "            \"\"\"\n",
    "            c = leg.get_children()[0]\n",
    "            title = c.get_children()[0]\n",
    "            hpack = c.get_children()[1]\n",
    "            c._children = [hpack]\n",
    "            hpack._children = [title] + hpack.get_children()\n",
    "        legend_title_left(chart.legend_)\n",
    "        \n",
    "    else:\n",
    "        chart.get_legend().remove()\n",
    "\n",
    "    # Change all spines\n",
    "    for axis in [\"top\", \"bottom\", \"left\", \"right\"]:\n",
    "        chart.spines[axis].set_linewidth(1)\n",
    "    chart.tick_params(axis=\"both\", length=4, width=1)\n",
    "    chart.grid(False)\n",
    "    sns.despine()\n",
    "\n",
    "     # Label points on each scatter plot\n",
    "    for i in range(0, AA_level_df.shape[0]):\n",
    "        x_pos = AA_level_df.at[i, \"mutation_frequency\"]\n",
    "        y_pos = AA_level_df.at[i, antibody]\n",
    "        name = AA_level_df.at[i, \"wildtype\"] + str(AA_level_df.at[i, \"site\"]) + AA_level_df.at[i, \"mutant\"]\n",
    "        if antibody == \"GPC-A\":\n",
    "            if name == \"E228D\":\n",
    "                chart.text(\n",
    "                    x_pos+0.025,\n",
    "                    y_pos+0.15,\n",
    "                    f\"{name}\",\n",
    "                    fontsize=6,\n",
    "                    horizontalalignment=\"left\",\n",
    "                    color=\"#EE7733\",\n",
    "                )\n",
    "        if antibody == \"GP1-A\":\n",
    "            if name == \"N89D\":\n",
    "                chart.text(\n",
    "                    x_pos+0.025,\n",
    "                    y_pos+0.125,\n",
    "                    f\"{name}\",\n",
    "                    fontsize=6,\n",
    "                    horizontalalignment=\"left\",\n",
    "                    color=\"#EE7733\",\n",
    "                )\n",
    "        if antibody == \"GPC-B\":\n",
    "            if name == \"H398K\":\n",
    "                chart.text(\n",
    "                    x_pos+0.025,\n",
    "                    y_pos+0.08,\n",
    "                    f\"{name}\",\n",
    "                    fontsize=6,\n",
    "                    horizontalalignment=\"left\",\n",
    "                    color=\"#EE7733\",\n",
    "                )\n",
    "            if name == \"D401E\":\n",
    "                chart.text(\n",
    "                    x_pos+0.025,\n",
    "                    y_pos+0.08,\n",
    "                    f\"{name}\",\n",
    "                    fontsize=6,\n",
    "                    horizontalalignment=\"left\",\n",
    "                    color=\"#EE7733\",\n",
    "                )\n",
    "        if antibody == \"GPC-C\":\n",
    "            if name == \"K126N\":\n",
    "                chart.text(\n",
    "                    x_pos+0.025,\n",
    "                    y_pos+0.08,\n",
    "                    f\"{name}\",\n",
    "                    fontsize=6,\n",
    "                    horizontalalignment=\"left\",\n",
    "                    color=\"#EE7733\",\n",
    "                )\n",
    "# Common X axis labels\n",
    "fig.text(0.5, -0.3, \"frequency of amino-acid mutation\", ha=\"center\", rotation=\"horizontal\", fontsize=8)\n",
    "\n",
    "# Save fig\n",
    "fig.savefig(natural_escape)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
