{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "d14d13d8-e5f7-45b9-b8df-bdcf0ba65deb",
   "metadata": {},
   "source": [
    "# Analyze mutational accessibility for GPC to escape antibodies"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5e0ff593-084f-4e30-857e-a486362570c8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Imports\n",
    "import os\n",
    "import warnings\n",
    "import pandas as pd\n",
    "import altair as alt\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "from Bio import SeqIO, AlignIO \n",
    "\n",
    "# Plotting colors\n",
    "# re-arranged for plot\n",
    "tol_muted_adjusted = [\n",
    "    \"#AA4499\",\n",
    "    \"#88CCEE\",\n",
    "    \"#EE7733\",\n",
    "    \"#44AA99\",\n",
    "    \"#1f78b4\",\n",
    "    \"#CC6677\",\n",
    "    \"#117733\",\n",
    "    \"#999933\",\n",
    "    \"#DDCC77\",\n",
    "    \"#CC3311\",\n",
    "    \"#882255\",\n",
    "    \"#000000\",\n",
    "    \"#DDDDDD\",\n",
    "]\n",
    "\n",
    "# Seaborn style settings\n",
    "sns.set(rc={\n",
    "    \"figure.dpi\":300, \n",
    "    \"savefig.dpi\":300,\n",
    "    \"svg.fonttype\":\"none\",\n",
    "})\n",
    "sns.set_style(\"ticks\")\n",
    "sns.set_palette(tol_muted_adjusted)\n",
    "\n",
    "# Suppress warnings\n",
    "warnings.simplefilter(\"ignore\")\n",
    "\n",
    "# Allow more rows for Altair\n",
    "_ = alt.data_transformers.disable_max_rows()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c4ff435e-7305-4e9d-9145-a15ac5237f86",
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": ""
    },
    "tags": [
     "parameters"
    ]
   },
   "outputs": [],
   "source": [
    "# this cell is tagged as `parameters` for papermill parameterization\n",
    "filtered_escape_377H = None\n",
    "filtered_escape_89F = None\n",
    "filtered_escape_2510C = None\n",
    "filtered_escape_121F = None\n",
    "filtered_escape_256A = None\n",
    "filtered_escape_372D = None\n",
    "\n",
    "filtered_func_293T = None\n",
    "\n",
    "Josiah_sequence = None\n",
    "\n",
    "out_dir = None\n",
    "func_vs_escape = None\n",
    "func_vs_escape_svg = None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bb512831-a86e-4e28-abbb-38c0f53eb27e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Uncomment for running interactive\n",
    "# filtered_escape_377H = \"../results/filtered_antibody_escape_CSVs/377H_filtered_mut_effect.csv\"\n",
    "# filtered_escape_89F = \"../results/filtered_antibody_escape_CSVs/89F_filtered_mut_effect.csv\"\n",
    "# filtered_escape_2510C = \"../results/filtered_antibody_escape_CSVs/2510C_filtered_mut_effect.csv\"\n",
    "# filtered_escape_121F = \"../results/filtered_antibody_escape_CSVs/121F_filtered_mut_effect.csv\"\n",
    "# filtered_escape_256A = \"../results/filtered_antibody_escape_CSVs/256A_filtered_mut_effect.csv\"\n",
    "# filtered_escape_372D = \"../results/filtered_antibody_escape_CSVs/372D_filtered_mut_effect.csv\"\n",
    "\n",
    "# filtered_func_293T = \"../results/filtered_func_effect_CSVs/293T_filtered_func_effects.csv\"\n",
    "\n",
    "# Josiah_sequence = \"../data/Josiah_nucleotide_reference_NC_004296.fasta\"\n",
    "\n",
    "# out_dir = \"../results/antibody_escape_profiles/\"\n",
    "# func_vs_escape = \"../results/antibody_escape_profiles/antibody_escape_vs_func_effect_all_muts.html\"\n",
    "# func_vs_escape_svg = \"../results/antibody_escape_profiles/antibody_escape_vs_func_effect_all_muts.svg\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c3148886-f3a9-4743-aec7-203f324232d4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Dictionary of codon values and corresponding amino acids\n",
    "codon_chart = { \n",
    "    'ATA':'I', 'ATC':'I', 'ATT':'I', 'ATG':'M', \n",
    "    'ACA':'T', 'ACC':'T', 'ACG':'T', 'ACT':'T', \n",
    "    'AAC':'N', 'AAT':'N', 'AAA':'K', 'AAG':'K', \n",
    "    'AGC':'S', 'AGT':'S', 'AGA':'R', 'AGG':'R',                  \n",
    "    'CTA':'L', 'CTC':'L', 'CTG':'L', 'CTT':'L', \n",
    "    'CCA':'P', 'CCC':'P', 'CCG':'P', 'CCT':'P', \n",
    "    'CAC':'H', 'CAT':'H', 'CAA':'Q', 'CAG':'Q', \n",
    "    'CGA':'R', 'CGC':'R', 'CGG':'R', 'CGT':'R', \n",
    "    'GTA':'V', 'GTC':'V', 'GTG':'V', 'GTT':'V', \n",
    "    'GCA':'A', 'GCC':'A', 'GCG':'A', 'GCT':'A', \n",
    "    'GAC':'D', 'GAT':'D', 'GAA':'E', 'GAG':'E', \n",
    "    'GGA':'G', 'GGC':'G', 'GGG':'G', 'GGT':'G', \n",
    "    'TCA':'S', 'TCC':'S', 'TCG':'S', 'TCT':'S', \n",
    "    'TTC':'F', 'TTT':'F', 'TTA':'L', 'TTG':'L', \n",
    "    'TAC':'Y', 'TAT':'Y', 'TAA':'*', 'TAG':'*', \n",
    "    'TGC':'C', 'TGT':'C', 'TGA':'*', 'TGG':'W', \n",
    "} \n",
    "\n",
    "def check_if_AA_one_mutation_away(wt_codon, mutant_AA):\n",
    "    \"\"\"\n",
    "    This function check if codon is within \n",
    "    one mutation away from a mutant amino acid\n",
    "    \"\"\"\n",
    "    # Check if current codon is mutant AA\n",
    "    if codon_chart[wt_codon] == mutant_AA:\n",
    "        return True\n",
    "    else:\n",
    "        # Iterate through each position and check if each mutation\n",
    "        for i in range(3):\n",
    "            for new_base in [\"A\", \"T\", \"G\", \"C\"]:\n",
    "                mutated_codon = wt_codon[:i] + new_base + wt_codon[i + 1:]\n",
    "                if codon_chart[mutated_codon] == mutant_AA:\n",
    "                    return True\n",
    "    \n",
    "    # Return false if not mutant was found\n",
    "    return False\n",
    "\n",
    "escape = [\n",
    "    filtered_escape_2510C,\n",
    "    filtered_escape_121F,\n",
    "    filtered_escape_377H,\n",
    "    filtered_escape_256A,\n",
    "    filtered_escape_372D,\n",
    "    filtered_escape_89F,\n",
    "]\n",
    "\n",
    "merged_df = pd.read_csv(filtered_func_293T)\n",
    "\n",
    "# Add escape to dataframe for each antibody\n",
    "for antibody_file in escape:\n",
    "\n",
    "    antibody_name = antibody_file.split(\"/\")[-1].split(\"_\")[0]\n",
    "\n",
    "    # Load data as dataframe\n",
    "    escape_df = pd.read_csv(antibody_file)\n",
    "\n",
    "    # Filter escape df for low functional score mutations\n",
    "    escape_df = escape_df.query(\"poor_cell_entry == False\")\n",
    "\n",
    "    # Clip lower scores to 0\n",
    "    escape_df[\"escape_median\"] = escape_df[\"escape_median\"].clip(lower=0)\n",
    "\n",
    "    # Rename escape column to include antibody name\n",
    "    escape_df = escape_df.rename(columns={\"escape_median\" : \"escape_\" + antibody_name})\n",
    "\n",
    "    # Merge dataframes\n",
    "    merged_df  = (\n",
    "        merged_df.merge(\n",
    "            escape_df[[\"site\", \"wildtype\", \"mutant\", \"escape_\" + antibody_name]],\n",
    "            how=\"left\",\n",
    "            on=[\"site\", \"wildtype\", \"mutant\"],\n",
    "            validate=\"one_to_one\",\n",
    "        )\n",
    "    )\n",
    "\n",
    "    # Clip upper scores to 0\n",
    "    merged_df[\"effect\"] = merged_df[\"effect\"].clip(upper=0)\n",
    "\n",
    "# Read josiah sequence\n",
    "josiah_seq = SeqIO.read(Josiah_sequence, \"fasta\").seq\n",
    "\n",
    "# Mark each mutation as being one nucleotide accesible\n",
    "# from the josiah wt codon\n",
    "merged_df[\"single_nucleotide_accessible\"] = (\n",
    "    merged_df.apply(lambda x: check_if_AA_one_mutation_away(josiah_seq[((x[\"site\"]-1)*3):(((x[\"site\"]-1)*3)+3)], x[\"mutant\"]), axis=1)\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "61af83ba-9bf1-41b8-92c8-d03e49a56ab2",
   "metadata": {},
   "source": [
    "Compare antibody escape vs functional effects"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fb3209c4-9f90-4a37-a6c2-d2d6346b81ce",
   "metadata": {},
   "outputs": [],
   "source": [
    "subplots = []\n",
    "for antibody_file in escape:\n",
    "    \n",
    "    antibody_name = antibody_file.split(\"/\")[-1].split(\"_\")[0]\n",
    "\n",
    "    # Set lim for each antibody\n",
    "    fixed_min = None\n",
    "    fixed_max = None\n",
    "    if antibody_name == \"2510C\":\n",
    "        fixed_min = -0.4\n",
    "        fixed_max = 7\n",
    "    elif antibody_name == \"121F\":\n",
    "        fixed_min = -0.34\n",
    "        fixed_max = 6\n",
    "    elif antibody_name == \"377H\":\n",
    "        fixed_min = -0.43\n",
    "        fixed_max = 7.5\n",
    "    elif antibody_name == \"256A\":\n",
    "        fixed_min = -0.4\n",
    "        fixed_max = 7\n",
    "    elif antibody_name == \"372D\":\n",
    "        fixed_min = -0.2\n",
    "        fixed_max = 3.5\n",
    "    elif antibody_name == \"89F\":\n",
    "        fixed_min = -0.4\n",
    "        fixed_max = 7\n",
    "    else:\n",
    "        print(\"Error! No ylims set!\")\n",
    "\n",
    "    curr_subplot = alt.Chart(merged_df.query(\"single_nucleotide_accessible == True\"), title=antibody_name).mark_point(\n",
    "        filled=True, \n",
    "        size=75,\n",
    "        opacity=0.25,\n",
    "    ).encode(\n",
    "        alt.X(\n",
    "            \"escape_\"+antibody_name,\n",
    "            axis=alt.Axis(\n",
    "                title=\"escape\", \n",
    "                domainWidth=1,\n",
    "                domainColor=\"black\",\n",
    "                tickColor=\"black\",\n",
    "            ),\n",
    "            scale=alt.Scale(domain=[fixed_min, fixed_max])\n",
    "        ),\n",
    "        alt.Y(\n",
    "            \"effect\",\n",
    "            axis=alt.Axis(\n",
    "                title=\"effect on cell entry\", \n",
    "                labelExpr=(\n",
    "                    \"datum.label == 0 ? '≥0' : datum.label\"\n",
    "                ),\n",
    "                values=[-1.5,-1,-0.5,0],\n",
    "                domainWidth=1,\n",
    "                domainColor=\"black\",\n",
    "                tickColor=\"black\",\n",
    "            ),\n",
    "            scale=alt.Scale(domain=[-1.6,0.1])\n",
    "        ),\n",
    "        tooltip=[\n",
    "            \"site\",\n",
    "            \"wildtype\",\n",
    "            \"mutant\",\n",
    "            \"effect\",\n",
    "            \"escape_\" + antibody_name,\n",
    "            \"single_nucleotide_accessible\"\n",
    "        ],\n",
    "        color=alt.Color(\n",
    "            \"single_nucleotide_accessible:N\", \n",
    "            scale=alt.Scale(\n",
    "                domain=[True, False], \n",
    "                range=[\"#EE7733\", \"#000000\"]\n",
    "            ),\n",
    "            legend=alt.Legend(\n",
    "                title=[\"single nucleotide\", \"accessible mutation\"],\n",
    "            ),\n",
    "        ),\n",
    "    ).properties(\n",
    "        width=150,\n",
    "        height=150,\n",
    "    )\n",
    "    \n",
    "    subplots.append(curr_subplot)\n",
    "\n",
    "# Create row of plots\n",
    "row_1 = alt.hconcat(\n",
    "    subplots[0],\n",
    "    subplots[1],\n",
    "    subplots[2],\n",
    "    subplots[3],\n",
    "    subplots[4],\n",
    "    subplots[5],\n",
    "    spacing=5,\n",
    "    title=\"single nucleotide accessible mutations\",\n",
    ")\n",
    "\n",
    "subplots = []\n",
    "for antibody_file in escape:\n",
    "    \n",
    "    antibody_name = antibody_file.split(\"/\")[-1].split(\"_\")[0]\n",
    "\n",
    "    # Set lim for each antibody\n",
    "    fixed_min = None\n",
    "    fixed_max = None\n",
    "    if antibody_name == \"2510C\":\n",
    "        fixed_min = -0.4\n",
    "        fixed_max = 7\n",
    "    elif antibody_name == \"121F\":\n",
    "        fixed_min = -0.34\n",
    "        fixed_max = 6\n",
    "    elif antibody_name == \"377H\":\n",
    "        fixed_min = -0.43\n",
    "        fixed_max = 7.5\n",
    "    elif antibody_name == \"256A\":\n",
    "        fixed_min = -0.4\n",
    "        fixed_max = 7\n",
    "    elif antibody_name == \"372D\":\n",
    "        fixed_min = -0.2\n",
    "        fixed_max = 3.5\n",
    "    elif antibody_name == \"89F\":\n",
    "        fixed_min = -0.4\n",
    "        fixed_max = 7\n",
    "    else:\n",
    "        print(\"Error! No ylims set!\")\n",
    "\n",
    "    curr_subplot = alt.Chart(merged_df.query(\"single_nucleotide_accessible == False\"), title=antibody_name).mark_point(\n",
    "        filled=True, \n",
    "        size=75,\n",
    "        opacity=0.15,\n",
    "    ).encode(\n",
    "        alt.X(\n",
    "            \"escape_\"+antibody_name,\n",
    "            axis=alt.Axis(\n",
    "                title=\"escape\", \n",
    "                domainWidth=1,\n",
    "                domainColor=\"black\",\n",
    "                tickColor=\"black\",\n",
    "            ),\n",
    "            scale=alt.Scale(domain=[fixed_min, fixed_max])\n",
    "        ),\n",
    "        alt.Y(\n",
    "            \"effect\",\n",
    "            axis=alt.Axis(\n",
    "                title=\"effect on cell entry\",\n",
    "                labelExpr=(\n",
    "                    \"datum.label == 0 ? '≥0' : datum.label\"\n",
    "                ),\n",
    "                values=[-1.5,-1,-0.5,0],\n",
    "                domainWidth=1,\n",
    "                domainColor=\"black\",\n",
    "                tickColor=\"black\",\n",
    "            ),\n",
    "            scale=alt.Scale(domain=[-1.6,0.1])\n",
    "        ),\n",
    "        tooltip=[\n",
    "            \"site\",\n",
    "            \"wildtype\",\n",
    "            \"mutant\",\n",
    "            \"effect\",\n",
    "            \"escape_\" + antibody_name,\n",
    "            \"single_nucleotide_accessible\"\n",
    "        ],\n",
    "        color=alt.Color(\n",
    "            \"single_nucleotide_accessible:N\", \n",
    "            scale=alt.Scale(\n",
    "                domain=[True, False], \n",
    "                range=[\"#EE7733\", \"#000000\"]\n",
    "            ),\n",
    "            legend=alt.Legend(\n",
    "                title=[\"single nucleotide\", \"accessible mutation\"],\n",
    "            ),\n",
    "        ),\n",
    "    ).properties(\n",
    "        width=150,\n",
    "        height=150,\n",
    "    )\n",
    "    \n",
    "    subplots.append(curr_subplot)\n",
    "\n",
    "# Create row of plots\n",
    "row_2 = alt.hconcat(\n",
    "    subplots[0],\n",
    "    subplots[1],\n",
    "    subplots[2],\n",
    "    subplots[3],\n",
    "    subplots[4],\n",
    "    subplots[5],\n",
    "    spacing=5,\n",
    "    title=\"multi-nucleotide accessible mutations\",\n",
    ")\n",
    "\n",
    "escape_vs_effect = alt.vconcat(\n",
    "    row_1,\n",
    "    row_2,\n",
    "    spacing=10,\n",
    "    title=\"Functional effect vs antibody escape\",\n",
    ").configure_axis(\n",
    "    grid=False,\n",
    "    labelFontSize=16,\n",
    "    titleFontSize=16,\n",
    "    labelFontWeight=\"normal\",\n",
    "    titleFontWeight=\"normal\",\n",
    ").configure_title(\n",
    "    fontSize=24,\n",
    ").configure_view(\n",
    "    stroke=None\n",
    ")\n",
    "\n",
    "# Make output dir if doesn't exist\n",
    "if not os.path.exists(out_dir):\n",
    "    os.mkdir(out_dir)\n",
    "\n",
    "escape_vs_effect.save(func_vs_escape)\n",
    "\n",
    "escape_vs_effect"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bb5e794d-2f2e-4932-a785-5e9577496291",
   "metadata": {},
   "source": [
    "Recreate same plot formatted for paper"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a9b18135-9a56-4948-8dcf-142f50adc1b0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Functions\n",
    "def  plot_func_scores_vs_escape(merged_df, ax, i, antibody_name):\n",
    "    \"\"\"\n",
    "    This function creates a plot of \n",
    "    functional scores vs antibody escape \n",
    "    \"\"\"\n",
    "\n",
    "    # Set lim for each antibody\n",
    "    fixed_min = None\n",
    "    fixed_max = None\n",
    "    if antibody_name == \"2510C\":\n",
    "        fixed_min = -0.6\n",
    "        fixed_max = 7\n",
    "    elif antibody_name == \"121F\":\n",
    "        fixed_min = -0.6\n",
    "        fixed_max = 7\n",
    "    elif antibody_name == \"377H\":\n",
    "        fixed_min = -0.64\n",
    "        fixed_max = 7.5\n",
    "    elif antibody_name == \"256A\":\n",
    "        fixed_min = -0.6\n",
    "        fixed_max = 7\n",
    "    elif antibody_name == \"372D\":\n",
    "        fixed_min = -0.3\n",
    "        fixed_max = 3.5\n",
    "    elif antibody_name == \"89F\":\n",
    "        fixed_min = -0.6\n",
    "        fixed_max = 7\n",
    "    else:\n",
    "        print(\"Error! No ylims set!\")\n",
    "\n",
    "    # Plot escape vs functional score\n",
    "    chart = sns.scatterplot(\n",
    "        data=merged_df,\n",
    "        y=\"effect\",\n",
    "        x=\"escape_\"+antibody_name,\n",
    "        hue=\"single_nucleotide_accessible\",\n",
    "        edgecolor=None,\n",
    "        linewidth=0,\n",
    "        palette={True : \"#EE773366\", False : \"#00000026\"},\n",
    "        s=20,\n",
    "        ax=ax,\n",
    "    )\n",
    "    if antibody_name == \"2510C\":\n",
    "        chart.set_title(\n",
    "            antibody_name[:2] + \".\" + antibody_name[2:], \n",
    "            fontsize=8, \n",
    "            color=\"#44AA99\",\n",
    "        )\n",
    "        xticks = [0, 2, 4, 6]\n",
    "        chart.set_xticks(xticks)\n",
    "        chart.set_xticklabels(labels=map(str, xticks), fontsize=8)\n",
    "    if antibody_name == \"121F\":\n",
    "        chart.set_title(\n",
    "            antibody_name[:2] + \".\" + antibody_name[2:], \n",
    "            fontsize=8, \n",
    "            color=\"#999933\",\n",
    "        )\n",
    "        xticks = [0, 2, 4, 6]\n",
    "        chart.set_xticks(xticks)\n",
    "        chart.set_xticklabels(labels=map(str, xticks), fontsize=8)\n",
    "    if antibody_name == \"377H\" or antibody_name == \"256A\" or antibody_name == \"372D\":\n",
    "        chart.set_title(\n",
    "            antibody_name[:2] + \".\" + antibody_name[2:], \n",
    "            fontsize=8, \n",
    "            color=\"#AA4499\",\n",
    "        )\n",
    "        xticks = None\n",
    "        if antibody_name == \"377H\":\n",
    "            xticks = [0, 2, 4, 6]\n",
    "        elif antibody_name == \"256A\":\n",
    "            xticks = [0, 2, 4, 6]\n",
    "        elif antibody_name == \"372D\":\n",
    "            xticks = [0, 1, 2, 3]\n",
    "        else:\n",
    "            print(\"ERROR\")\n",
    "        chart.set_xticks(xticks)\n",
    "        chart.set_xticklabels(labels=map(str, xticks), fontsize=8)\n",
    "    if antibody_name == \"89F\":\n",
    "        chart.set_title(\n",
    "            antibody_name[:1] + \".\" + antibody_name[1:], \n",
    "            fontsize=8, \n",
    "            color=\"#117733\",\n",
    "        )\n",
    "        xticks = [0, 2, 4, 6]\n",
    "        chart.set_xticks(xticks)\n",
    "        chart.set_xticklabels(labels=map(str, xticks), fontsize=8)\n",
    "        \n",
    "    chart.set_ylabel(\"effect on cell entry\", fontsize=8)\n",
    "    chart.set_xlim(fixed_min, fixed_max)\n",
    "    chart.set_ylim(-1.6,0.1)\n",
    "    yticks = [-1.5, -1, -0.5, 0]\n",
    "    chart.set_yticks(yticks)\n",
    "    chart.set_yticklabels(labels=[\"-1.5\", \"-1.0\", \"-0.5\", \"≥0\"], fontsize=8)\n",
    "    chart.set(xlabel=None)\n",
    "\n",
    "     # Make only one legend\n",
    "    if i == 5:\n",
    "        sns.move_legend(\n",
    "            chart, \n",
    "            \"upper left\", \n",
    "            bbox_to_anchor=(1, 1),\n",
    "            fontsize=8,\n",
    "            markerscale=1,\n",
    "            handletextpad=0.1,\n",
    "            title=\"single\\nnucleotide\\naccessible\\nmutation\",\n",
    "            title_fontproperties = {\n",
    "                \"size\" : 8, \n",
    "                # \"weight\" : \"bold\",\n",
    "            },\n",
    "            frameon=False,\n",
    "            borderaxespad=0.1,\n",
    "            reverse=True,\n",
    "        )\n",
    "        # Add edges to legend markers to match scatter plot\n",
    "        for ha in chart.legend_.legendHandles:\n",
    "            ha.set_edgecolor(None)\n",
    "            ha.set_linewidths(0.5)\n",
    "    else:\n",
    "        ax.get_legend().remove()\n",
    "\n",
    "    \n",
    "    # Change all spines\n",
    "    for axis in [\"top\", \"bottom\", \"left\", \"right\"]:\n",
    "        chart.spines[axis].set_linewidth(1)\n",
    "    chart.tick_params(axis=\"both\", length=3, width=1)\n",
    "\n",
    "    chart.grid(False)\n",
    "    sns.despine()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4292ea7c-22a9-45b0-aae1-21751511f0e4",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, axes = plt.subplots(\n",
    "    2, \n",
    "    6, \n",
    "    figsize=(6.4,3.5),\n",
    "    sharey=True,\n",
    ")\n",
    "for i,antibody_file in enumerate(escape):\n",
    "    \n",
    "    antibody_name = antibody_file.split(\"/\")[-1].split(\"_\")[0]\n",
    "    \n",
    "    plot_func_scores_vs_escape(merged_df.query(\"single_nucleotide_accessible == True\"), axes[0][i], i, antibody_name)\n",
    "    plot_func_scores_vs_escape(merged_df.query(\"single_nucleotide_accessible == False\"), axes[1][i], i, antibody_name)\n",
    "\n",
    "# Common X and Y axis labels\n",
    "fig.text(0.5, 0, \"escape\", ha=\"center\", rotation=\"horizontal\", fontsize=8)\n",
    "\n",
    "fig.tight_layout(w_pad=0.5)\n",
    "\n",
    "# Make output dir if doesn't exist\n",
    "if not os.path.exists(out_dir):\n",
    "    os.mkdir(out_dir)\n",
    "\n",
    "# Save fig\n",
    "plt.savefig(func_vs_escape_svg)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
