{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "4a4d6432-3a14-457e-97d8-2e8d352db94b",
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": ""
    },
    "tags": []
   },
   "source": [
    "# Compare functional effects of mutations measured in DMS to the abundance in natural sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "acc00c87-be55-46af-a737-bc9d09a9b6e8",
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": ""
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Imports\n",
    "import os\n",
    "import warnings\n",
    "import pandas as pd\n",
    "import altair as alt\n",
    "from Bio import SeqIO, AlignIO \n",
    "\n",
    "# Plotting colors\n",
    "# re-arranged for plot\n",
    "tol_muted_adjusted = [\n",
    "    \"#AA4499\",\n",
    "    \"#88CCEE\",\n",
    "    \"#EE7733\",\n",
    "    \"#44AA99\",\n",
    "    \"#1f78b4\",\n",
    "    \"#CC6677\",\n",
    "    \"#117733\",\n",
    "    \"#999933\",\n",
    "    \"#DDCC77\",\n",
    "    \"#CC3311\",\n",
    "    \"#882255\",\n",
    "    \"#000000\",\n",
    "    \"#DDDDDD\",\n",
    "]\n",
    "\n",
    "# Allow more rows for Altair\n",
    "_ = alt.data_transformers.disable_max_rows()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "840242df-5a0f-42f9-89cc-1a4d1abf1fe4",
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": ""
    },
    "tags": [
     "parameters"
    ]
   },
   "outputs": [],
   "source": [
    "# this cell is tagged as `parameters` for papermill parameterization\n",
    "natural_GPC_sequence_alignment = None\n",
    "filtered_func_293T = None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d0f50ff9-8919-4473-b199-20445fbcbb13",
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": ""
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "# # Uncomment for running interactive\n",
    "# natural_GPC_sequence_alignment = \"../non-pipeline_analyses/LASV_phylogeny_analysis/Results/LASV_GPC_protein_alignment.fasta\"\n",
    "# filtered_func_293T = \"../results/filtered_func_effect_CSVs/293T_filtered_func_effects.csv\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "23b6a1d5-0554-4e09-945c-3bbf7c591e64",
   "metadata": {},
   "source": [
    "Process data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aefb98c6-01e5-448a-ae44-6ecdaa56e33b",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_natural_sequence_counts(site, amino_acid, natural_seqs_df):\n",
    "    \"\"\"\n",
    "    Function that counts occurences of an amino acid at a site\n",
    "    across a dataframe of sequences.\n",
    "    \"\"\"\n",
    "    count = 0\n",
    "    for seq in natural_seqs_df[\"sequence\"].tolist():\n",
    "        if seq[site-1] == amino_acid:\n",
    "            count += 1\n",
    "    return count\n",
    "\n",
    "# Load alignment and metadata info\n",
    "natural_seqs_df = pd.DataFrame(columns=[\"strain\", \"sequence\"])\n",
    "\n",
    "# Add alignment sequence to dataframe\n",
    "for curr_fasta in AlignIO.read(natural_GPC_sequence_alignment, \"fasta\"):\n",
    "    natural_seqs_df.loc[len(natural_seqs_df.index)] = [\n",
    "        str(curr_fasta.id),\n",
    "        str(curr_fasta.seq),\n",
    "    ] \n",
    "\n",
    "# Load data as dataframe\n",
    "functional_scores = pd.read_csv(filtered_func_293T)\n",
    "\n",
    "# Filter for no stop codons\n",
    "functional_scores = (\n",
    "    functional_scores.query(\n",
    "        \"mutant != '*'\"\n",
    "    )\n",
    "    .drop(columns=[\"times_seen\", \"effect_std\", \"n_selections\"])\n",
    "    .reset_index(drop=True)\n",
    ")\n",
    "\n",
    "# DMS strain\n",
    "chosen_isolates = [\n",
    "    \"Josiah_NC_004296_reverse_complement_2018-08-13\",\n",
    "]\n",
    "\n",
    "# Create subset of df for chosen isolates\n",
    "validation_isolates = (\n",
    "    natural_seqs_df.loc[natural_seqs_df[\"strain\"].isin(chosen_isolates)]\n",
    "    .reset_index(drop=True)\n",
    ")\n",
    "\n",
    "# Get Josiah sequence for comparison\n",
    "josiah_sequence = validation_isolates.loc[validation_isolates[\"strain\"] == \"Josiah_NC_004296_reverse_complement_2018-08-13\"].at[0,\"sequence\"]\n",
    "\n",
    "# Create a wildtype df to fill in missing Josiah/wildtype measurements as 0\n",
    "josiah_df = pd.DataFrame(\n",
    "    zip(list(range(1,492)), josiah_sequence, josiah_sequence, [0]*491, [True]*491),\n",
    "    columns=[\"site\", \"wildtype\", \"mutant\", \"effect\", \"Josiah reference\"],\n",
    ")\n",
    "\n",
    "# Merge functional effects and josiah reference dataframes\n",
    "AA_level_df = (\n",
    "    pd.concat([functional_scores, josiah_df], ignore_index=True)\n",
    "    .sort_values(by=\"site\")\n",
    "    .reset_index(drop=True)\n",
    "    .fillna(False)\n",
    ")\n",
    "\n",
    "# Get natural sequence counts of each mutant and calculate mutation frequencies\n",
    "# compared to the Josiah reference\n",
    "AA_level_df[\"natural_counts\"] = (\n",
    "    AA_level_df.apply(lambda x: get_natural_sequence_counts(x[\"site\"], x[\"mutant\"], natural_seqs_df), axis=1)\n",
    ")\n",
    "\n",
    "# Mark mutations seen at N times seen in natural sequences\n",
    "AA_level_df[\"natural_count_bins\"] = \"all DMS measured amino-acid mutations\"\n",
    "# At least once\n",
    "at_least_once = (\n",
    "    AA_level_df.loc[AA_level_df[\"natural_counts\"] >= 1].copy()\n",
    ")\n",
    "at_least_once[\"natural_count_bins\"] = \"mutations observed >= 1 in natural GPCs\"\n",
    "AA_level_df = (\n",
    "    pd.concat([\n",
    "        AA_level_df, \n",
    "        at_least_once,\n",
    "    ], ignore_index = True)\n",
    ")\n",
    "# At least X times\n",
    "at_least_x = (\n",
    "    AA_level_df.loc[AA_level_df[\"natural_counts\"] >= 10].copy()\n",
    ")\n",
    "at_least_x[\"natural_count_bins\"] = \"mutations observed >= 10 in natural GPCs\"\n",
    "AA_level_df = (\n",
    "    pd.concat([\n",
    "        AA_level_df, \n",
    "        at_least_x,\n",
    "    ], ignore_index = True)\n",
    ")\n",
    "\n",
    "# Remove any mutations that are the Josiah reference because it skews the median to 0\n",
    "AA_level_df = AA_level_df.loc[(AA_level_df[\"wildtype\"] != AA_level_df[\"mutant\"])].copy()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4cc03a3d-901a-412c-8b97-929239c6ac66",
   "metadata": {},
   "source": [
    "Plot functional score distributions stratified by times observed in natural sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "faabe9c7-f0c1-444f-b553-7a45629b5353",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plot score distrbutions\n",
    "distribution_plot = alt.Chart(\n",
    "        AA_level_df, title=\"Effect on cell entry for mutations observed in natural GPCs\"\n",
    "    ).mark_boxplot(\n",
    "    color=\"black\", \n",
    "    outliers={\"strokeWidth\" : 1, \"opacity\" : 0.25},\n",
    "    median={\"thickness\" : 1.5, \"color\" : \"white\"},\n",
    ").encode(\n",
    "    y=alt.Y(\n",
    "        \"natural_count_bins:N\",\n",
    "        title=None,\n",
    "        sort=[\"mutations observed >= 10 in natural GPCs\", \"mutations observed >= 1 in natural GPCs\", \"all DMS measured amino-acid mutations\"],\n",
    "        axis=alt.Axis(\n",
    "            domainWidth=1,\n",
    "            domainColor=\"black\",\n",
    "            tickColor=\"black\",\n",
    "            labelLimit=400,\n",
    "        ),\n",
    "    ),\n",
    "    x=alt.X(\n",
    "        \"effect:Q\",\n",
    "        title=\"effect on cell entry\",\n",
    "        axis=alt.Axis(\n",
    "            values=[-4,-3,-2,-1,0,1],\n",
    "            domainWidth=1,\n",
    "            domainColor=\"black\",\n",
    "            tickColor=\"black\",\n",
    "        ),\n",
    "        scale=alt.Scale(domain=[-4.5, 1.25])\n",
    "    ),\n",
    ")\n",
    "\n",
    "# Combine plot\n",
    "combined_plot = (\n",
    "    (distribution_plot)\n",
    "    .configure_axis(\n",
    "        grid=False,\n",
    "        labelFontSize=16,\n",
    "        titleFontSize=16,\n",
    "        labelFontWeight=\"normal\",\n",
    "        titleFontWeight=\"normal\",\n",
    "    ).properties(\n",
    "        width=400, \n",
    "        height=100,\n",
    "    ).configure_title(\n",
    "        fontSize=24,\n",
    "    ).configure_view(\n",
    "        stroke=None\n",
    "    )\n",
    ")\n",
    "\n",
    "combined_plot"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "903d1db9-9a35-41b6-8f4f-7d4a131820b5",
   "metadata": {},
   "source": [
    "Recreate **same** plot above but formatted for paper"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5ac24458-e413-49b6-90e0-c5347eceb44f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plot score distrbutions\n",
    "distribution_plot = alt.Chart(\n",
    "        AA_level_df, title=\"Effect on cell entry for mutations observed in natural GPCs\"\n",
    "    ).mark_boxplot(\n",
    "    color=\"black\", \n",
    "    outliers={\"strokeWidth\" : 1, \"opacity\" : 0.25, \"size\" : 10},\n",
    "    median={\"thickness\" : 1.5, \"color\" : \"white\"},\n",
    ").encode(\n",
    "    y=alt.Y(\n",
    "        \"natural_count_bins:N\",\n",
    "        title=None,\n",
    "        sort=[\"mutations observed >= 10 in natural GPCs\", \"mutations observed >= 1 in natural GPCs\", \"all DMS measured amino-acid mutations\"],\n",
    "        axis=alt.Axis(\n",
    "            domainWidth=1,\n",
    "            domainColor=\"black\",\n",
    "            tickColor=\"black\",\n",
    "            labelLimit=400,\n",
    "        ),\n",
    "    ),\n",
    "    x=alt.X(\n",
    "        \"effect:Q\",\n",
    "        title=\"effect on cell entry\",\n",
    "        axis=alt.Axis(\n",
    "            values=[-4,-3,-2,-1,0,1],\n",
    "            domainWidth=1,\n",
    "            domainColor=\"black\",\n",
    "            tickColor=\"black\",\n",
    "        ),\n",
    "        scale=alt.Scale(domain=[-4.5, 1.25])\n",
    "    ),\n",
    ")\n",
    "\n",
    "# Combine plot\n",
    "combined_plot = (\n",
    "    (distribution_plot)\n",
    "    .configure_axis(\n",
    "        grid=False,\n",
    "        labelFontSize=8,\n",
    "        titleFontSize=8,\n",
    "        labelFontWeight=\"normal\",\n",
    "        titleFontWeight=\"normal\",\n",
    "    ).properties(\n",
    "        width=100, \n",
    "        height=75,\n",
    "    ).configure_title(\n",
    "        fontSize=12,\n",
    "    ).configure_view(\n",
    "        stroke=None\n",
    "    )\n",
    ")\n",
    "\n",
    "combined_plot"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
