{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "66273dcc-8100-45d9-b9aa-74e59f843ed2",
   "metadata": {},
   "source": [
    "# Natural sequence and DMS analysis\n",
    "\n",
    "This notebook is preliminary analysis of antigenic selection across Lassa sequences and is not included in the final HTML documents."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a37bc955-f3a8-4870-9148-0dd0c9aea802",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Imports\n",
    "import os\n",
    "import json\n",
    "import warnings\n",
    "import dmslogo\n",
    "import neutcurve\n",
    "import numpy as np\n",
    "import scipy as sp\n",
    "import pandas as pd\n",
    "import altair as alt\n",
    "import seaborn as sns\n",
    "import statsmodels.api\n",
    "import matplotlib.colors\n",
    "from Bio import SeqIO, AlignIO \n",
    "from matplotlib import pyplot as plt\n",
    "from matplotlib import ticker as mticker\n",
    "\n",
    "# Rearranged to make the tree look nicer\n",
    "# re-ordered\n",
    "tol_muted_adjusted = [\n",
    "    \"#000000\",\n",
    "    \"#CC6677\", \n",
    "    \"#1f78b4\", \n",
    "    \"#88CCEE\",\n",
    "    \"#DDDDDD\",\n",
    "    \"#882255\",\n",
    "    \"#117733\", \n",
    "    \"#DDCC77\",\n",
    "    \"#44AA99\", \n",
    "    \"#EE7733\",\n",
    "    \"#AA4499\",\n",
    "    \"#999933\", \n",
    "    \"#CC3311\",\n",
    "]\n",
    "\n",
    "# Seaborn style settings\n",
    "sns.set(rc={\n",
    "    \"figure.dpi\":300, \n",
    "    \"savefig.dpi\":300,\n",
    "    \"svg.fonttype\":\"none\",\n",
    "})\n",
    "sns.set_style(\"ticks\")\n",
    "sns.set_palette(tol_muted_adjusted)\n",
    "\n",
    "# Suppress warnings\n",
    "warnings.simplefilter(\"ignore\")\n",
    "\n",
    "# Allow more rows for Altair\n",
    "_ = alt.data_transformers.disable_max_rows()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8866b67d-a376-4796-b524-fe0ff0735df7",
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": ""
    },
    "tags": [
     "parameters"
    ]
   },
   "outputs": [],
   "source": [
    "# this cell is tagged as `parameters` for papermill parameterization\n",
    "filtered_escape_377H = None\n",
    "filtered_escape_89F = None\n",
    "filtered_escape_2510C = None\n",
    "filtered_escape_121F = None\n",
    "filtered_escape_256A = None\n",
    "filtered_escape_372D = None\n",
    "\n",
    "func_scores = None\n",
    "min_times_seen = None\n",
    "n_selections = None\n",
    "\n",
    "GPC_tree_mutations = None\n",
    "GPC_FEL_results = None\n",
    "GPC_FUBAR_results = None\n",
    "\n",
    "natural_seq_metadata = None\n",
    "natural_seq_alignment = None\n",
    "\n",
    "out_dir_natural = None\n",
    "ols_regression = None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c9357b19-d1fb-4f0b-89d0-257ecc6c0992",
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": ""
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "# # Uncomment for running interactive\n",
    "# filtered_escape_377H = \"../results/filtered_antibody_escape_CSVs/377H_filtered_mut_effect.csv\"\n",
    "# filtered_escape_89F = \"../results/filtered_antibody_escape_CSVs/89F_filtered_mut_effect.csv\"\n",
    "# filtered_escape_2510C = \"../results/filtered_antibody_escape_CSVs/2510C_filtered_mut_effect.csv\"\n",
    "# filtered_escape_121F = \"../results/filtered_antibody_escape_CSVs/121F_filtered_mut_effect.csv\"\n",
    "# filtered_escape_256A = \"../results/filtered_antibody_escape_CSVs/256A_filtered_mut_effect.csv\"\n",
    "# filtered_escape_372D = \"../results/filtered_antibody_escape_CSVs/372D_filtered_mut_effect.csv\"\n",
    "\n",
    "# func_scores = \"../results/func_effects/averages/293T_entry_func_effects.csv\"\n",
    "# min_times_seen = 2\n",
    "# n_selections = 8\n",
    "\n",
    "# GPC_tree_mutations = \"../non-pipeline_analyses/LASV_phylogeny_analysis/Results/GPC_tree_mutations.csv\"\n",
    "# GPC_FEL_results = \"../non-pipeline_analyses/LASV_phylogeny_analysis/Results/GPC_FEL_results.json\"\n",
    "# GPC_FUBAR_results = \"../non-pipeline_analyses/LASV_phylogeny_analysis/Results/GPC_FUBAR_results.json\"\n",
    "\n",
    "# natural_seq_metadata = \"../non-pipeline_analyses/LASV_phylogeny_analysis/Results/LASV_S_segment_metadata.tsv\"\n",
    "# natural_seq_alignment = \"../non-pipeline_analyses/LASV_phylogeny_analysis/Results/LASV_GPC_protein_alignment.fasta\"\n",
    "\n",
    "# out_dir_natural = \"../results/natural_isolate_escape/\"\n",
    "# ols_regression = \"../results/natural_isolate_escape/ols_regression.html\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "165db688-f496-42e0-a066-f49155487833",
   "metadata": {},
   "source": [
    "Process functional scores and antibody escape to create one dataframe with site level information."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3c803710-8b18-46d6-99cf-ac2e38d631af",
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": ""
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Read functional score data\n",
    "functional_scores = pd.read_csv(func_scores)\n",
    "\n",
    "# Filter for minimum selections, times seen and no stop codons\n",
    "functional_scores = (\n",
    "    functional_scores.query(\n",
    "        \"n_selections >= @n_selections and times_seen >= @min_times_seen and mutant != '*'\"\n",
    "    )\n",
    "    .drop(columns=[\"mutant\", \"times_seen\"])\n",
    "    .groupby([\"site\", \"wildtype\"])\n",
    "    .aggregate({\n",
    "        \"effect\" : \"mean\"\n",
    "    })\n",
    "    .reset_index()\n",
    ")\n",
    "\n",
    "# Initialize list of escape files and final df\n",
    "antibody_files = [\n",
    "    filtered_escape_377H,\n",
    "    filtered_escape_89F,\n",
    "    filtered_escape_2510C,\n",
    "    filtered_escape_121F,\n",
    "    filtered_escape_256A,\n",
    "    filtered_escape_372D,\n",
    "]\n",
    "antibody_escape = pd.DataFrame(columns=[\"site\", \"wildtype\"])\n",
    "\n",
    "# Add escape to dataframe for each antibody\n",
    "for index,antibody_file in enumerate(antibody_files):\n",
    "\n",
    "    antibody_name = antibody_file.split(\"/\")[-1].split(\"_\")[0]\n",
    "\n",
    "    # Load data as dataframe\n",
    "    escape_df = pd.read_csv(antibody_file)\n",
    "\n",
    "    # Clip lower scores to 0\n",
    "    escape_df[\"escape_median\"] = escape_df[\"escape_median\"].clip(lower=0)\n",
    "\n",
    "    # # Normalize scores\n",
    "    # escape_df[\"escape_median\"] = (\n",
    "    #     (escape_df[\"escape_median\"] - escape_df[\"escape_median\"].min()) / (escape_df[\"escape_median\"].max() - escape_df[\"escape_median\"].min())  \n",
    "    # )\n",
    "\n",
    "    # Rename escape column to include antibody name\n",
    "    escape_df = (\n",
    "        escape_df\n",
    "        .rename(columns={\"escape_median\" : \"escape_\" + antibody_name})\n",
    "        .groupby([\"site\", \"wildtype\"])\n",
    "        .aggregate({\"escape_\" + antibody_name : \"sum\"})\n",
    "        .reset_index()\n",
    "    )\n",
    "\n",
    "    if index == 0:\n",
    "        antibody_escape = (\n",
    "            pd.concat([\n",
    "                antibody_escape, \n",
    "                escape_df[[\"site\", \"wildtype\", \"escape_\" + antibody_name]],\n",
    "            ])\n",
    "        )\n",
    "    else:\n",
    "        # Merge dataframes\n",
    "        antibody_escape = (\n",
    "            antibody_escape.merge(\n",
    "                escape_df[[\"site\", \"wildtype\", \"escape_\" + antibody_name]],\n",
    "                how=\"left\",\n",
    "                on=[\"site\", \"wildtype\",],\n",
    "                validate=\"one_to_one\",\n",
    "            )\n",
    "        )\n",
    "\n",
    "# Create new column with antibody summed escape\n",
    "antibody_escape[\"total_escape\"] = (\n",
    "    antibody_escape[[\n",
    "        \"escape_377H\",\n",
    "        \"escape_89F\",\n",
    "        \"escape_2510C\",\n",
    "        \"escape_121F\",\n",
    "        \"escape_256A\",\n",
    "        \"escape_372D\",\n",
    "    ]].sum(axis=1)\n",
    ")\n",
    "\n",
    "merged_df = (\n",
    "    functional_scores.merge(\n",
    "        antibody_escape,\n",
    "        how=\"left\",\n",
    "        on=[\"site\", \"wildtype\"],\n",
    "        validate=\"one_to_one\",\n",
    "    )\n",
    ")\n",
    "\n",
    "# Load tree mutation counts\n",
    "tree_mutations = pd.read_csv(GPC_tree_mutations)\n",
    "\n",
    "# Calculate counts of synonymous and nonsynonous mutations and ratio\n",
    "tree_mutations[\"synonymous\"] = (\n",
    "    tree_mutations[\"mut_type\"].apply(lambda x: 1 if x == \"synonymous\" else 0)\n",
    ")\n",
    "tree_mutations[\"nonsynonymous\"] = (\n",
    "    tree_mutations[\"mut_type\"].apply(lambda x: 1 if x == \"nonsynonymous\" else 0)\n",
    ")\n",
    "tree_mutations = (\n",
    "    tree_mutations\n",
    "    .groupby([\"site\"])\n",
    "    .aggregate({\n",
    "        \"mut_type\" : \"count\",\n",
    "        \"synonymous\" : \"sum\",\n",
    "        \"nonsynonymous\" : \"sum\",\n",
    "    })\n",
    "    .reset_index()\n",
    ")\n",
    "tree_mutations[\"non/syn\"] = tree_mutations[\"nonsynonymous\"] / tree_mutations[\"synonymous\"]\n",
    "\n",
    "# Merge tree mutation counts\n",
    "merged_df = (\n",
    "    merged_df.merge(\n",
    "        tree_mutations,\n",
    "        how=\"left\",\n",
    "        on=[\"site\"],\n",
    "        validate=\"one_to_one\",\n",
    "    )\n",
    ")\n",
    "\n",
    "# Load FEL data\n",
    "FEL_results = pd.read_json(GPC_FEL_results)\n",
    "\n",
    "# Process json file and extract site info\n",
    "FEL_results = (\n",
    "    FEL_results\n",
    "    .reset_index()\n",
    ")\n",
    "FEL_results = (\n",
    "    pd.DataFrame(FEL_results.at[0, \"MLE\"])\n",
    "    .rename(columns={\"0\" : \"data\"})\n",
    ")\n",
    "headers = [\n",
    "    \"FEL_alpha\",\n",
    "    \"FEL_beta\",\n",
    "    \"FEL_alpha=beta\",\n",
    "    \"FEL_LRT\",\n",
    "    \"FEL_p-value\",\n",
    "    \"FEL_Total branch length\",\n",
    "    \"FEL_p-asmp\",\n",
    "]\n",
    "FEL_results[headers] = pd.DataFrame(\n",
    "    FEL_results[\"data\"].tolist(), \n",
    "    index=FEL_results.index,\n",
    ")\n",
    "FEL_results = (\n",
    "    FEL_results\n",
    "    .reset_index(drop=False)\n",
    "    .rename(columns={\"index\" : \"site\"})\n",
    "    .drop(columns=[\"data\"])\n",
    ")\n",
    "FEL_results[\"site\"] = FEL_results[\"site\"] + 1\n",
    "\n",
    "# Merge tree mutation counts\n",
    "merged_df = (\n",
    "    merged_df.merge(\n",
    "        FEL_results,\n",
    "        how=\"left\",\n",
    "        on=[\"site\"],\n",
    "        validate=\"one_to_one\",\n",
    "    )\n",
    ")\n",
    "\n",
    "# Load FUBAR data\n",
    "# requires json load because the output file\n",
    "# has mixed dict and series\n",
    "json_data = json.load(open(GPC_FUBAR_results))\n",
    "FUBAR_results = (\n",
    "    pd.DataFrame(json_data[\"MLE\"][\"content\"])\n",
    "    .rename(columns={\"0\" : \"data\"})\n",
    ")\n",
    "\n",
    "# Process json file and extract site info\n",
    "headers = [\n",
    "    \"FUBAR_alpha\",\n",
    "    \"FUBAR_beta\",\n",
    "    \"FUBAR_beta-alpha\",\n",
    "    \"FUBAR_Prob[alpha>beta]\",\n",
    "    \"FUBAR_Prob[alpha<beta]\",\n",
    "    \"FUBAR_BayesFactor[alpha<beta]\",\n",
    "    \"empty_1\", # dummy columns\n",
    "    \"empty_2\", # dummy columns\n",
    "]\n",
    "FUBAR_results[headers] = pd.DataFrame(\n",
    "    FUBAR_results[\"data\"].tolist(), \n",
    "    index=FUBAR_results.index,\n",
    ")\n",
    "FUBAR_results = (\n",
    "    FUBAR_results\n",
    "    .reset_index(drop=False)\n",
    "    .rename(columns={\"index\" : \"site\"})\n",
    "    .drop(columns=[\"data\", \"empty_1\", \"empty_2\"])\n",
    ")\n",
    "FUBAR_results[\"site\"] = FUBAR_results[\"site\"] + 1\n",
    "\n",
    "# Merge tree mutation counts\n",
    "merged_df = (\n",
    "    merged_df.merge(\n",
    "        FUBAR_results,\n",
    "        how=\"left\",\n",
    "        on=[\"site\"],\n",
    "        validate=\"one_to_one\",\n",
    "    )\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "593e895e-91fe-41c6-b6a4-cc1c7ba44022",
   "metadata": {},
   "source": [
    "## Correlation of phylogenetic tree metrics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "289ca1c4-7dae-49d0-80b5-89dbc23f59b2",
   "metadata": {},
   "outputs": [],
   "source": [
    "comparisons = [\n",
    "    (\"FEL_beta\", \"nonsynonymous\"),\n",
    "    (\"FUBAR_beta\", \"nonsynonymous\"),\n",
    "    (\"FUBAR_beta\", \"FEL_beta\"),\n",
    "]\n",
    "subplots = []\n",
    "for pair in comparisons:\n",
    "    stat1 = pair[0]\n",
    "    stat2 = pair[1]\n",
    "\n",
    "    # Calculate correlation \n",
    "    r, p = sp.stats.pearsonr(\n",
    "        merged_df[[stat1, stat2]].dropna()[stat1],\n",
    "        merged_df[[stat1, stat2]].dropna()[stat2],\n",
    "    )\n",
    "    print(f\"r correlation of {stat1} and {stat2}: {r:.2f}\")\n",
    "    \n",
    "    curr_subplot = alt.Chart(\n",
    "        merged_df, \n",
    "    ).mark_point(\n",
    "        filled=True, \n",
    "        color=\"black\", \n",
    "        size=75,\n",
    "        opacity=0.15,\n",
    "    ).encode(\n",
    "        alt.Y(\n",
    "            stat1,\n",
    "            axis=alt.Axis(\n",
    "                title=stat1,\n",
    "                # values=[-4,-3,-2,-1,0,1],\n",
    "                # domainWidth=1,\n",
    "                # domainColor=\"black\",\n",
    "                # tickColor=\"black\",\n",
    "            ),\n",
    "            # scale=alt.Scale(domain=[-4.1,1.1])\n",
    "        ),\n",
    "        alt.X(\n",
    "            stat2,\n",
    "            axis=alt.Axis(\n",
    "                title=stat2,\n",
    "                # values=[-3,-2,-1,0],\n",
    "                # domainWidth=1,\n",
    "                # domainColor=\"black\",\n",
    "                # tickColor=\"black\",\n",
    "            ),\n",
    "            # scale=alt.Scale(type=\"symlog\")\n",
    "        ),\n",
    "        tooltip=[\n",
    "            \"site\",\n",
    "            \"wildtype\",\n",
    "            \"effect\",\n",
    "            \"total_escape\",\n",
    "            \"nonsynonymous\",\n",
    "            \"synonymous\",\n",
    "            \"non/syn\",\n",
    "            \"FEL_beta\",\n",
    "            \"FEL_alpha\",\n",
    "            \"FEL_alpha=beta\",\n",
    "            \"FUBAR_beta\",\n",
    "            \"FUBAR_alpha\",\n",
    "            \"FUBAR_beta-alpha\",\n",
    "        ],\n",
    "    ).properties(\n",
    "        width=300,\n",
    "        height=300,\n",
    "    )\n",
    "\n",
    "    subplots.append(curr_subplot)\n",
    "\n",
    "nonsyn = alt.hconcat(\n",
    "    subplots[0],\n",
    "    subplots[1],\n",
    "    subplots[2],\n",
    "    spacing=5,\n",
    "    title=[\"Correlations of nonsynonymous mutation metrics\"],\n",
    ")\n",
    "\n",
    "\n",
    "comparisons = [\n",
    "    (\"FEL_alpha\", \"synonymous\"),\n",
    "    (\"FUBAR_alpha\", \"synonymous\"),\n",
    "    (\"FUBAR_alpha\", \"FEL_alpha\"),\n",
    "]\n",
    "subplots = []\n",
    "for pair in comparisons:\n",
    "    stat1 = pair[0]\n",
    "    stat2 = pair[1]\n",
    "\n",
    "    # Calculate correlation \n",
    "    r, p = sp.stats.pearsonr(\n",
    "        merged_df[[stat1, stat2]].dropna()[stat1],\n",
    "        merged_df[[stat1, stat2]].dropna()[stat2],\n",
    "    )\n",
    "    print(f\"r correlation of {stat1} and {stat2}: {r:.2f}\")\n",
    "    \n",
    "    curr_subplot = alt.Chart(\n",
    "        merged_df, \n",
    "    ).mark_point(\n",
    "        filled=True, \n",
    "        color=\"black\", \n",
    "        size=75,\n",
    "        opacity=0.15,\n",
    "    ).encode(\n",
    "        alt.Y(\n",
    "            stat1,\n",
    "            axis=alt.Axis(\n",
    "                title=stat1,\n",
    "                # values=[-4,-3,-2,-1,0,1],\n",
    "                # domainWidth=1,\n",
    "                # domainColor=\"black\",\n",
    "                # tickColor=\"black\",\n",
    "            ),\n",
    "            # scale=alt.Scale(domain=[-4.1,1.1])\n",
    "        ),\n",
    "        alt.X(\n",
    "            stat2,\n",
    "            axis=alt.Axis(\n",
    "                title=stat2,\n",
    "                # values=[-3,-2,-1,0],\n",
    "                # domainWidth=1,\n",
    "                # domainColor=\"black\",\n",
    "                # tickColor=\"black\",\n",
    "            ),\n",
    "            # scale=alt.Scale(type=\"symlog\")\n",
    "        ),\n",
    "        tooltip=[\n",
    "            \"site\",\n",
    "            \"wildtype\",\n",
    "            \"effect\",\n",
    "            \"total_escape\",\n",
    "            \"nonsynonymous\",\n",
    "            \"synonymous\",\n",
    "            \"non/syn\",\n",
    "            \"FEL_beta\",\n",
    "            \"FEL_alpha\",\n",
    "            \"FEL_alpha=beta\",\n",
    "            \"FUBAR_beta\",\n",
    "            \"FUBAR_alpha\",\n",
    "            \"FUBAR_beta-alpha\",\n",
    "        ],\n",
    "    ).properties(\n",
    "        width=300,\n",
    "        height=300,\n",
    "    )\n",
    "\n",
    "    subplots.append(curr_subplot)\n",
    "\n",
    "syn = alt.hconcat(\n",
    "    subplots[0],\n",
    "    subplots[1],\n",
    "    subplots[2],\n",
    "    spacing=5,\n",
    "    title=[\"Correlations of synonymous mutation metrics\"],\n",
    ")\n",
    "\n",
    "\n",
    "comparisons = [\n",
    "    (\"FEL_alpha=beta\", \"non/syn\"),\n",
    "    (\"FUBAR_beta-alpha\", \"non/syn\"),\n",
    "    (\"FUBAR_beta-alpha\", \"FEL_alpha=beta\"),\n",
    "]\n",
    "subplots = []\n",
    "for pair in comparisons:\n",
    "    stat1 = pair[0]\n",
    "    stat2 = pair[1]\n",
    "\n",
    "    # Calculate correlation \n",
    "    r, p = sp.stats.pearsonr(\n",
    "        merged_df[[stat1, stat2]].replace([np.inf, -np.inf], np.nan).dropna()[stat1],\n",
    "        merged_df[[stat1, stat2]].replace([np.inf, -np.inf], np.nan).dropna()[stat2],\n",
    "    )\n",
    "    print(f\"r correlation of {stat1} and {stat2}: {r:.2f}\")\n",
    "    \n",
    "    curr_subplot = alt.Chart(\n",
    "        merged_df, \n",
    "    ).mark_point(\n",
    "        filled=True, \n",
    "        color=\"black\", \n",
    "        size=75,\n",
    "        opacity=0.15,\n",
    "    ).encode(\n",
    "        alt.Y(\n",
    "            stat1,\n",
    "            axis=alt.Axis(\n",
    "                title=stat1,\n",
    "                # values=[-4,-3,-2,-1,0,1],\n",
    "                # domainWidth=1,\n",
    "                # domainColor=\"black\",\n",
    "                # tickColor=\"black\",\n",
    "            ),\n",
    "            # scale=alt.Scale(domain=[-4.1,1.1])\n",
    "        ),\n",
    "        alt.X(\n",
    "            stat2,\n",
    "            axis=alt.Axis(\n",
    "                title=stat2,\n",
    "                # values=[-3,-2,-1,0],\n",
    "                # domainWidth=1,\n",
    "                # domainColor=\"black\",\n",
    "                # tickColor=\"black\",\n",
    "            ),\n",
    "            # scale=alt.Scale(type=\"symlog\")\n",
    "        ),\n",
    "        tooltip=[\n",
    "            \"site\",\n",
    "            \"wildtype\",\n",
    "            \"effect\",\n",
    "            \"total_escape\",\n",
    "            \"nonsynonymous\",\n",
    "            \"synonymous\",\n",
    "            \"non/syn\",\n",
    "            \"FEL_beta\",\n",
    "            \"FEL_alpha\",\n",
    "            \"FEL_alpha=beta\",\n",
    "            \"FUBAR_beta\",\n",
    "            \"FUBAR_alpha\",\n",
    "            \"FUBAR_beta-alpha\",\n",
    "        ],\n",
    "    ).properties(\n",
    "        width=300,\n",
    "        height=300,\n",
    "    )\n",
    "\n",
    "    subplots.append(curr_subplot)\n",
    "\n",
    "nonsyn_and_syn = alt.hconcat(\n",
    "    subplots[0],\n",
    "    subplots[1],\n",
    "    subplots[2],\n",
    "    spacing=5,\n",
    "    title=[\"Correlations of combined nonsynonymous and synonymous mutation metrics\"],\n",
    ")\n",
    "\n",
    "combined_plot = alt.vconcat(\n",
    "    nonsyn,\n",
    "    syn,\n",
    "    nonsyn_and_syn,\n",
    "    spacing=5,\n",
    ").configure_axis(\n",
    "    grid=False,\n",
    "    labelFontSize=16,\n",
    "    titleFontSize=16,\n",
    "    labelFontWeight=\"normal\",\n",
    "    titleFontWeight=\"normal\",\n",
    ").configure_axis(\n",
    "    grid=False,\n",
    "    labelFontSize=16,\n",
    "    titleFontSize=16,\n",
    "    labelFontWeight=\"normal\",\n",
    "    titleFontWeight=\"normal\",\n",
    ").configure_title(\n",
    "    fontSize=24,\n",
    ")\n",
    "\n",
    "combined_plot"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d5adffed-4653-43ba-8345-03d85fa31eec",
   "metadata": {},
   "source": [
    "The lower end of nonsynonymous mutations (i.e., 0 to 10) are less correlated even though the correlation looks really good when zoomed out. "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "61de99b1-df08-4071-b966-63a5f6ef4776",
   "metadata": {},
   "source": [
    "## Tree mutations compared to functional effects and antibody escape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7a915adc-67a2-48e7-8d28-48b11da0c002",
   "metadata": {},
   "outputs": [],
   "source": [
    "comparisons = [\n",
    "    (\"effect\", \"nonsynonymous\"),\n",
    "    (\"effect\", \"synonymous\"),\n",
    "    (\"effect\", \"non/syn\"),\n",
    "]\n",
    "subplots = []\n",
    "for pair in comparisons:\n",
    "    stat1 = pair[0]\n",
    "    stat2 = pair[1]\n",
    "\n",
    "    # Calculate correlation \n",
    "    r, p = sp.stats.pearsonr(\n",
    "        merged_df[[stat1, stat2]].replace([np.inf, -np.inf], np.nan).dropna()[stat1],\n",
    "        merged_df[[stat1, stat2]].replace([np.inf, -np.inf], np.nan).dropna()[stat2],\n",
    "    )\n",
    "    print(f\"r correlation of {stat1} and {stat2}: {r:.2f}\")\n",
    "    \n",
    "    curr_subplot = alt.Chart(\n",
    "        merged_df, \n",
    "    ).mark_point(\n",
    "        filled=True, \n",
    "        color=\"black\", \n",
    "        size=75,\n",
    "        opacity=0.15,\n",
    "    ).encode(\n",
    "        alt.Y(\n",
    "            stat1,\n",
    "            axis=alt.Axis(\n",
    "                title=stat1,\n",
    "                # values=[-4,-3,-2,-1,0,1],\n",
    "                # domainWidth=1,\n",
    "                # domainColor=\"black\",\n",
    "                # tickColor=\"black\",\n",
    "            ),\n",
    "            # scale=alt.Scale(domain=[-4.1,1.1])\n",
    "        ),\n",
    "        alt.X(\n",
    "            stat2,\n",
    "            axis=alt.Axis(\n",
    "                title=stat2,\n",
    "                # values=[-3,-2,-1,0],\n",
    "                # domainWidth=1,\n",
    "                # domainColor=\"black\",\n",
    "                # tickColor=\"black\",\n",
    "            ),\n",
    "            # scale=alt.Scale(type=\"symlog\")\n",
    "        ),\n",
    "        tooltip=[\n",
    "            \"site\",\n",
    "            \"wildtype\",\n",
    "            \"effect\",\n",
    "            \"total_escape\",\n",
    "            \"nonsynonymous\",\n",
    "            \"synonymous\",\n",
    "            \"non/syn\",\n",
    "            \"FEL_beta\",\n",
    "            \"FEL_alpha\",\n",
    "            \"FEL_alpha=beta\",\n",
    "            \"FUBAR_beta\",\n",
    "            \"FUBAR_alpha\",\n",
    "            \"FUBAR_beta-alpha\",\n",
    "        ],\n",
    "    ).properties(\n",
    "        width=300,\n",
    "        height=300,\n",
    "    )\n",
    "\n",
    "    subplots.append(curr_subplot)\n",
    "\n",
    "effects = alt.hconcat(\n",
    "    subplots[0],\n",
    "    subplots[1],\n",
    "    subplots[2],\n",
    "    spacing=5,\n",
    "    title=[\"Correlations of functional effects and tree mutation counts\"],\n",
    ")\n",
    "\n",
    "\n",
    "comparisons = [\n",
    "    (\"total_escape\", \"nonsynonymous\"),\n",
    "    (\"total_escape\", \"synonymous\"),\n",
    "    (\"total_escape\", \"non/syn\"),\n",
    "]\n",
    "subplots = []\n",
    "for pair in comparisons:\n",
    "    stat1 = pair[0]\n",
    "    stat2 = pair[1]\n",
    "\n",
    "    # Calculate correlation \n",
    "    r, p = sp.stats.pearsonr(\n",
    "        merged_df[[stat1, stat2]].replace([np.inf, -np.inf], np.nan).dropna()[stat1],\n",
    "        merged_df[[stat1, stat2]].replace([np.inf, -np.inf], np.nan).dropna()[stat2],\n",
    "    )\n",
    "    print(f\"r correlation of {stat1} and {stat2}: {r:.2f}\")\n",
    "    \n",
    "    curr_subplot = alt.Chart(\n",
    "        merged_df, \n",
    "    ).mark_point(\n",
    "        filled=True, \n",
    "        color=\"black\", \n",
    "        size=75,\n",
    "        opacity=0.15,\n",
    "    ).encode(\n",
    "        alt.Y(\n",
    "            stat1,\n",
    "            axis=alt.Axis(\n",
    "                title=stat1,\n",
    "                # values=[-4,-3,-2,-1,0,1],\n",
    "                # domainWidth=1,\n",
    "                # domainColor=\"black\",\n",
    "                # tickColor=\"black\",\n",
    "            ),\n",
    "            # scale=alt.Scale(domain=[-4.1,1.1])\n",
    "        ),\n",
    "        alt.X(\n",
    "            stat2,\n",
    "            axis=alt.Axis(\n",
    "                title=stat2,\n",
    "                # values=[-3,-2,-1,0],\n",
    "                # domainWidth=1,\n",
    "                # domainColor=\"black\",\n",
    "                # tickColor=\"black\",\n",
    "            ),\n",
    "            # scale=alt.Scale(type=\"symlog\")\n",
    "        ),\n",
    "        tooltip=[\n",
    "            \"site\",\n",
    "            \"wildtype\",\n",
    "            \"effect\",\n",
    "            \"total_escape\",\n",
    "            \"nonsynonymous\",\n",
    "            \"synonymous\",\n",
    "            \"non/syn\",\n",
    "            \"FEL_beta\",\n",
    "            \"FEL_alpha\",\n",
    "            \"FEL_alpha=beta\",\n",
    "            \"FUBAR_beta\",\n",
    "            \"FUBAR_alpha\",\n",
    "            \"FUBAR_beta-alpha\",\n",
    "        ],\n",
    "    ).properties(\n",
    "        width=300,\n",
    "        height=300,\n",
    "    )\n",
    "\n",
    "    subplots.append(curr_subplot)\n",
    "\n",
    "escape = alt.hconcat(\n",
    "    subplots[0],\n",
    "    subplots[1],\n",
    "    subplots[2],\n",
    "    spacing=5,\n",
    "    title=[\"Correlations of antibody escape and tree mutation counts\"],\n",
    ")\n",
    "\n",
    "combined_plot = alt.vconcat(\n",
    "    effects,\n",
    "    escape,\n",
    "    spacing=5,\n",
    "    title=\"Tree mutations\",\n",
    ").configure_axis(\n",
    "    grid=False,\n",
    "    labelFontSize=16,\n",
    "    titleFontSize=16,\n",
    "    labelFontWeight=\"normal\",\n",
    "    titleFontWeight=\"normal\",\n",
    ").configure_axis(\n",
    "    grid=False,\n",
    "    labelFontSize=16,\n",
    "    titleFontSize=16,\n",
    "    labelFontWeight=\"normal\",\n",
    "    titleFontWeight=\"normal\",\n",
    ").configure_title(\n",
    "    fontSize=24,\n",
    ")\n",
    "\n",
    "combined_plot"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ef79cc0a-3b24-4909-8a58-506779ea5434",
   "metadata": {},
   "source": [
    "## FEL analysis compared to functional effects and antibody escape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0b96e924-fe20-4145-a94f-717c33d448a2",
   "metadata": {},
   "outputs": [],
   "source": [
    "comparisons = [\n",
    "    (\"effect\", \"FEL_beta\"),\n",
    "    (\"effect\", \"FEL_alpha\"),\n",
    "    (\"effect\", \"FEL_alpha=beta\"),\n",
    "]\n",
    "subplots = []\n",
    "for pair in comparisons:\n",
    "    stat1 = pair[0]\n",
    "    stat2 = pair[1]\n",
    "\n",
    "    # Calculate correlation \n",
    "    r, p = sp.stats.pearsonr(\n",
    "        merged_df[[stat1, stat2]].replace([np.inf, -np.inf], np.nan).dropna()[stat1],\n",
    "        merged_df[[stat1, stat2]].replace([np.inf, -np.inf], np.nan).dropna()[stat2],\n",
    "    )\n",
    "    print(f\"r correlation of {stat1} and {stat2}: {r:.2f}\")\n",
    "    \n",
    "    curr_subplot = alt.Chart(\n",
    "        merged_df, \n",
    "    ).mark_point(\n",
    "        filled=True, \n",
    "        color=\"black\", \n",
    "        size=75,\n",
    "        opacity=0.15,\n",
    "    ).encode(\n",
    "        alt.Y(\n",
    "            stat1,\n",
    "            axis=alt.Axis(\n",
    "                title=stat1,\n",
    "                # values=[-4,-3,-2,-1,0,1],\n",
    "                # domainWidth=1,\n",
    "                # domainColor=\"black\",\n",
    "                # tickColor=\"black\",\n",
    "            ),\n",
    "            # scale=alt.Scale(domain=[-4.1,1.1])\n",
    "        ),\n",
    "        alt.X(\n",
    "            stat2,\n",
    "            axis=alt.Axis(\n",
    "                title=stat2,\n",
    "                # values=[-3,-2,-1,0],\n",
    "                # domainWidth=1,\n",
    "                # domainColor=\"black\",\n",
    "                # tickColor=\"black\",\n",
    "            ),\n",
    "            # scale=alt.Scale(type=\"symlog\")\n",
    "        ),\n",
    "        tooltip=[\n",
    "            \"site\",\n",
    "            \"wildtype\",\n",
    "            \"effect\",\n",
    "            \"total_escape\",\n",
    "            \"nonsynonymous\",\n",
    "            \"synonymous\",\n",
    "            \"non/syn\",\n",
    "            \"FEL_beta\",\n",
    "            \"FEL_alpha\",\n",
    "            \"FEL_alpha=beta\",\n",
    "            \"FUBAR_beta\",\n",
    "            \"FUBAR_alpha\",\n",
    "            \"FUBAR_beta-alpha\",\n",
    "        ],\n",
    "    ).properties(\n",
    "        width=300,\n",
    "        height=300,\n",
    "    )\n",
    "\n",
    "    subplots.append(curr_subplot)\n",
    "\n",
    "effects = alt.hconcat(\n",
    "    subplots[0],\n",
    "    subplots[1],\n",
    "    subplots[2],\n",
    "    spacing=5,\n",
    "    title=[\"Correlations of functional effects and FEL metrics\"],\n",
    ")\n",
    "\n",
    "\n",
    "comparisons = [\n",
    "    (\"total_escape\", \"FEL_beta\"),\n",
    "    (\"total_escape\", \"FEL_alpha\"),\n",
    "    (\"total_escape\", \"FEL_alpha=beta\"),\n",
    "]\n",
    "subplots = []\n",
    "for pair in comparisons:\n",
    "    stat1 = pair[0]\n",
    "    stat2 = pair[1]\n",
    "\n",
    "    # Calculate correlation \n",
    "    r, p = sp.stats.pearsonr(\n",
    "        merged_df[[stat1, stat2]].replace([np.inf, -np.inf], np.nan).dropna()[stat1],\n",
    "        merged_df[[stat1, stat2]].replace([np.inf, -np.inf], np.nan).dropna()[stat2],\n",
    "    )\n",
    "    print(f\"r correlation of {stat1} and {stat2}: {r:.2f}\")\n",
    "    \n",
    "    curr_subplot = alt.Chart(\n",
    "        merged_df, \n",
    "    ).mark_point(\n",
    "        filled=True, \n",
    "        color=\"black\", \n",
    "        size=75,\n",
    "        opacity=0.15,\n",
    "    ).encode(\n",
    "        alt.Y(\n",
    "            stat1,\n",
    "            axis=alt.Axis(\n",
    "                title=stat1,\n",
    "                # values=[-4,-3,-2,-1,0,1],\n",
    "                # domainWidth=1,\n",
    "                # domainColor=\"black\",\n",
    "                # tickColor=\"black\",\n",
    "            ),\n",
    "            # scale=alt.Scale(domain=[-4.1,1.1])\n",
    "        ),\n",
    "        alt.X(\n",
    "            stat2,\n",
    "            axis=alt.Axis(\n",
    "                title=stat2,\n",
    "                # values=[-3,-2,-1,0],\n",
    "                # domainWidth=1,\n",
    "                # domainColor=\"black\",\n",
    "                # tickColor=\"black\",\n",
    "            ),\n",
    "            # scale=alt.Scale(type=\"symlog\")\n",
    "        ),\n",
    "        tooltip=[\n",
    "            \"site\",\n",
    "            \"wildtype\",\n",
    "            \"effect\",\n",
    "            \"total_escape\",\n",
    "            \"nonsynonymous\",\n",
    "            \"synonymous\",\n",
    "            \"non/syn\",\n",
    "            \"FEL_beta\",\n",
    "            \"FEL_alpha\",\n",
    "            \"FEL_alpha=beta\",\n",
    "            \"FUBAR_beta\",\n",
    "            \"FUBAR_alpha\",\n",
    "            \"FUBAR_beta-alpha\",\n",
    "        ],\n",
    "    ).properties(\n",
    "        width=300,\n",
    "        height=300,\n",
    "    )\n",
    "\n",
    "    subplots.append(curr_subplot)\n",
    "\n",
    "escape = alt.hconcat(\n",
    "    subplots[0],\n",
    "    subplots[1],\n",
    "    subplots[2],\n",
    "    spacing=5,\n",
    "    title=[\"Correlations of antibody escape and FEL metrics\"],\n",
    ")\n",
    "\n",
    "combined_plot = alt.vconcat(\n",
    "    effects,\n",
    "    escape,\n",
    "    spacing=5,\n",
    "    title=\"FEL\",\n",
    ").configure_axis(\n",
    "    grid=False,\n",
    "    labelFontSize=16,\n",
    "    titleFontSize=16,\n",
    "    labelFontWeight=\"normal\",\n",
    "    titleFontWeight=\"normal\",\n",
    ").configure_axis(\n",
    "    grid=False,\n",
    "    labelFontSize=16,\n",
    "    titleFontSize=16,\n",
    "    labelFontWeight=\"normal\",\n",
    "    titleFontWeight=\"normal\",\n",
    ").configure_title(\n",
    "    fontSize=24,\n",
    ")\n",
    "\n",
    "combined_plot"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a67a2055-0835-4fa3-83b4-71ca71d8e671",
   "metadata": {},
   "source": [
    "## FUBAR analysis compared to functional effects and antibody escape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e9336b69-71ba-4bb3-9470-29f8a4db5548",
   "metadata": {},
   "outputs": [],
   "source": [
    "comparisons = [\n",
    "    (\"effect\", \"FUBAR_beta\"),\n",
    "    (\"effect\", \"FUBAR_alpha\"),\n",
    "    (\"effect\", \"FUBAR_beta-alpha\"),\n",
    "]\n",
    "subplots = []\n",
    "for pair in comparisons:\n",
    "    stat1 = pair[0]\n",
    "    stat2 = pair[1]\n",
    "\n",
    "    # Calculate correlation \n",
    "    r, p = sp.stats.pearsonr(\n",
    "        merged_df[[stat1, stat2]].replace([np.inf, -np.inf], np.nan).dropna()[stat1],\n",
    "        merged_df[[stat1, stat2]].replace([np.inf, -np.inf], np.nan).dropna()[stat2],\n",
    "    )\n",
    "    print(f\"r correlation of {stat1} and {stat2}: {r:.2f}\")\n",
    "    \n",
    "    curr_subplot = alt.Chart(\n",
    "        merged_df, \n",
    "    ).mark_point(\n",
    "        filled=True, \n",
    "        color=\"black\", \n",
    "        size=75,\n",
    "        opacity=0.15,\n",
    "    ).encode(\n",
    "        alt.Y(\n",
    "            stat1,\n",
    "            axis=alt.Axis(\n",
    "                title=stat1,\n",
    "                # values=[-4,-3,-2,-1,0,1],\n",
    "                # domainWidth=1,\n",
    "                # domainColor=\"black\",\n",
    "                # tickColor=\"black\",\n",
    "            ),\n",
    "            # scale=alt.Scale(domain=[-4.1,1.1])\n",
    "        ),\n",
    "        alt.X(\n",
    "            stat2,\n",
    "            axis=alt.Axis(\n",
    "                title=stat2,\n",
    "                # values=[-3,-2,-1,0],\n",
    "                # domainWidth=1,\n",
    "                # domainColor=\"black\",\n",
    "                # tickColor=\"black\",\n",
    "            ),\n",
    "            # scale=alt.Scale(type=\"symlog\")\n",
    "        ),\n",
    "        tooltip=[\n",
    "            \"site\",\n",
    "            \"wildtype\",\n",
    "            \"effect\",\n",
    "            \"total_escape\",\n",
    "            \"nonsynonymous\",\n",
    "            \"synonymous\",\n",
    "            \"non/syn\",\n",
    "            \"FEL_beta\",\n",
    "            \"FEL_alpha\",\n",
    "            \"FEL_alpha=beta\",\n",
    "            \"FUBAR_beta\",\n",
    "            \"FUBAR_alpha\",\n",
    "            \"FUBAR_beta-alpha\",\n",
    "        ],\n",
    "    ).properties(\n",
    "        width=300,\n",
    "        height=300,\n",
    "    )\n",
    "\n",
    "    subplots.append(curr_subplot)\n",
    "\n",
    "effects = alt.hconcat(\n",
    "    subplots[0],\n",
    "    subplots[1],\n",
    "    subplots[2],\n",
    "    spacing=5,\n",
    "    title=[\"Correlations of functional effects and FUBAR metrics\"],\n",
    ")\n",
    "\n",
    "\n",
    "comparisons = [\n",
    "    (\"total_escape\", \"FUBAR_beta\"),\n",
    "    (\"total_escape\", \"FUBAR_alpha\"),\n",
    "    (\"total_escape\", \"FUBAR_beta-alpha\"),\n",
    "]\n",
    "subplots = []\n",
    "for pair in comparisons:\n",
    "    stat1 = pair[0]\n",
    "    stat2 = pair[1]\n",
    "\n",
    "    # Calculate correlation \n",
    "    r, p = sp.stats.pearsonr(\n",
    "        merged_df[[stat1, stat2]].replace([np.inf, -np.inf], np.nan).dropna()[stat1],\n",
    "        merged_df[[stat1, stat2]].replace([np.inf, -np.inf], np.nan).dropna()[stat2],\n",
    "    )\n",
    "    print(f\"r correlation of {stat1} and {stat2}: {r:.2f}\")\n",
    "    \n",
    "    curr_subplot = alt.Chart(\n",
    "        merged_df, \n",
    "    ).mark_point(\n",
    "        filled=True, \n",
    "        color=\"black\", \n",
    "        size=75,\n",
    "        opacity=0.15,\n",
    "    ).encode(\n",
    "        alt.Y(\n",
    "            stat1,\n",
    "            axis=alt.Axis(\n",
    "                title=stat1,\n",
    "                # values=[-4,-3,-2,-1,0,1],\n",
    "                # domainWidth=1,\n",
    "                # domainColor=\"black\",\n",
    "                # tickColor=\"black\",\n",
    "            ),\n",
    "            # scale=alt.Scale(domain=[-4.1,1.1])\n",
    "        ),\n",
    "        alt.X(\n",
    "            stat2,\n",
    "            axis=alt.Axis(\n",
    "                title=stat2,\n",
    "                # values=[-3,-2,-1,0],\n",
    "                # domainWidth=1,\n",
    "                # domainColor=\"black\",\n",
    "                # tickColor=\"black\",\n",
    "            ),\n",
    "            # scale=alt.Scale(type=\"symlog\")\n",
    "        ),\n",
    "        tooltip=[\n",
    "            \"site\",\n",
    "            \"wildtype\",\n",
    "            \"effect\",\n",
    "            \"total_escape\",\n",
    "            \"nonsynonymous\",\n",
    "            \"synonymous\",\n",
    "            \"non/syn\",\n",
    "            \"FEL_beta\",\n",
    "            \"FEL_alpha\",\n",
    "            \"FEL_alpha=beta\",\n",
    "            \"FUBAR_beta\",\n",
    "            \"FUBAR_alpha\",\n",
    "            \"FUBAR_beta-alpha\",\n",
    "        ],\n",
    "    ).properties(\n",
    "        width=300,\n",
    "        height=300,\n",
    "    )\n",
    "\n",
    "    subplots.append(curr_subplot)\n",
    "\n",
    "escape = alt.hconcat(\n",
    "    subplots[0],\n",
    "    subplots[1],\n",
    "    subplots[2],\n",
    "    spacing=5,\n",
    "    title=[\"Correlations of antibody escape and FUBAR metrics\"],\n",
    ")\n",
    "\n",
    "combined_plot = alt.vconcat(\n",
    "    effects,\n",
    "    escape,\n",
    "    spacing=5,\n",
    "    title=\"FUBAR\",\n",
    ").configure_axis(\n",
    "    grid=False,\n",
    "    labelFontSize=16,\n",
    "    titleFontSize=16,\n",
    "    labelFontWeight=\"normal\",\n",
    "    titleFontWeight=\"normal\",\n",
    ").configure_axis(\n",
    "    grid=False,\n",
    "    labelFontSize=16,\n",
    "    titleFontSize=16,\n",
    "    labelFontWeight=\"normal\",\n",
    "    titleFontWeight=\"normal\",\n",
    ").configure_title(\n",
    "    fontSize=24,\n",
    ")\n",
    "\n",
    "combined_plot"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c7fefc95-0686-43ec-be59-03b69fc031db",
   "metadata": {},
   "source": [
    "## OLS regression for DMS measurements and nonsynonymous substitution metrics\n",
    "\n",
    "Perform multiple linear regression between the DMS measurements (i.e., effect on entry and antibody escape) and the nonsynonymous mutation metrics. We perform OLS regression for both the metric as is and for the log 2 of the metric. Note that when taking the log2 of the metric, only sites with nonzero values remain (i.e., all sites that did not have any mutations observed are not included). We finally plot the predicted metric from the OLS model versus the actual value."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8b20fa39-2fcc-4772-b353-21a988fd4b7b",
   "metadata": {},
   "outputs": [],
   "source": [
    "ols_df = (\n",
    "    merged_df[[\n",
    "        \"site\",\n",
    "        \"wildtype\",\n",
    "        \"FEL_beta\",\n",
    "        \"nonsynonymous\",\n",
    "        \"FUBAR_beta\",\n",
    "        \"effect\",\n",
    "        \"escape_377H\",\n",
    "        \"escape_2510C\",\n",
    "        \"escape_89F\",\n",
    "        \"escape_256A\",\n",
    "        \"escape_372D\",\n",
    "        \"escape_121F\",\n",
    "        \"total_escape\",\n",
    "    ]]\n",
    "    # .replace([np.inf, -np.inf], np.nan)\n",
    "    .rename(columns={\n",
    "        \"effect\" : \"effect on cell entry\",\n",
    "        \"total_escape\" : \"antibody escape\",\n",
    "    })\n",
    "    .dropna() # some sites do not have measurements for antibody escape\n",
    "    .reset_index(drop=True)\n",
    ")\n",
    "\n",
    "\n",
    "X_vars = [\n",
    "    # \"site\",\n",
    "    \"effect on cell entry\",\n",
    "    # \"escape_377H\",\n",
    "    # \"escape_2510C\",\n",
    "    # \"escape_89F\",\n",
    "    # \"escape_256A\",\n",
    "    # \"escape_372D\",\n",
    "    # \"escape_121F\",\n",
    "    \"antibody escape\",\n",
    "]\n",
    "\n",
    "Y_vars = [\n",
    "    \"nonsynonymous\",\n",
    "    \"FEL_beta\",\n",
    "    \"FUBAR_beta\",\n",
    "]\n",
    "\n",
    "subplots = []\n",
    "for Y_var in Y_vars:\n",
    "    \n",
    "    Y_values = ols_df[[Y_var]]\n",
    "    X_values = ols_df[X_vars]\n",
    "\n",
    "    # https://www.statsmodels.org/dev/examples/notebooks/generated/ols.html\n",
    "    ols_model = statsmodels.api.OLS(\n",
    "        endog=Y_values,\n",
    "        exog=statsmodels.api.add_constant(X_values.astype(float)),\n",
    "    )\n",
    "    res_ols = ols_model.fit()\n",
    "    full_r2 = res_ols.rsquared\n",
    "    # print(res_ols.summary())\n",
    "    ols_df[Y_var+\"_predicted\"] = res_ols.predict()\n",
    "\n",
    "    # Calculate correlation \n",
    "    actual_r, p = sp.stats.pearsonr(\n",
    "        ols_df[[Y_var, Y_var+\"_predicted\"]].replace([np.inf, -np.inf], np.nan).dropna()[Y_var+\"_predicted\"],\n",
    "        ols_df[[Y_var, Y_var+\"_predicted\"]].replace([np.inf, -np.inf], np.nan).dropna()[Y_var],\n",
    "    )\n",
    "    print(f\"r correlation of predicted {Y_var} vs actual {Y_var}: {actual_r:.2f}\")\n",
    "    \n",
    "    unique_var = {}\n",
    "    # https://blog.minitab.com/en/adventures-in-statistics-2/how-to-identify-the-most-important-predictor-variables-in-regression-models\n",
    "    for vremove in X_vars:\n",
    "        vremove_ols_model = statsmodels.api.OLS(\n",
    "            endog=Y_values,\n",
    "            exog=statsmodels.api.add_constant(ols_df[[v for v in X_vars if v != vremove]].astype(float)),\n",
    "        )\n",
    "        vremove_res_ols = vremove_ols_model.fit()\n",
    "        unique_var[vremove] = full_r2 - vremove_res_ols.rsquared\n",
    "    \n",
    "    # https://stackoverflow.com/a/53966201\n",
    "    subtitle = [\n",
    "        f\"{var}: {unique_var[var] * 100:.1f}% of variance (coef {res_ols.params[var]:.3f} \\u00B1 {res_ols.bse[var]:.3f})\"\n",
    "        for var in X_vars\n",
    "    ]\n",
    "\n",
    "    chart_title = alt.TitleParams(\n",
    "        Y_var + \" (n = \" + str(Y_values.shape[0]) + \")\",\n",
    "        subtitle=subtitle,\n",
    "        fontSize=16,\n",
    "    )\n",
    "\n",
    "    curr_subplot = alt.Chart(\n",
    "        ols_df,\n",
    "        title=chart_title,\n",
    "    ).mark_point(\n",
    "        filled=True, \n",
    "        color=\"black\", \n",
    "        size=75,\n",
    "        opacity=0.15,\n",
    "    ).encode(\n",
    "        alt.X(\n",
    "            Y_var+\"_predicted\",\n",
    "            axis=alt.Axis(\n",
    "                title=[\"predicted \" + Y_var],\n",
    "                # values=[-4,-3,-2,-1,0,1],\n",
    "                # domainWidth=1,\n",
    "                # domainColor=\"black\",\n",
    "                # tickColor=\"black\",\n",
    "            ),\n",
    "            # scale=alt.Scale(type=\"symlog\")\n",
    "        ),\n",
    "        alt.Y(\n",
    "            Y_var,\n",
    "            axis=alt.Axis(\n",
    "                title=[\"actual \" + Y_var],\n",
    "                # values=[-3,-2,-1,0],\n",
    "                # domainWidth=1,\n",
    "                # domainColor=\"black\",\n",
    "                # tickColor=\"black\",\n",
    "            ),\n",
    "            # scale=alt.Scale(type=\"symlog\")\n",
    "        ),\n",
    "        tooltip=[\n",
    "            \"site\",\n",
    "            \"wildtype\",\n",
    "            Y_var,\n",
    "            Y_var+\"_predicted\",\n",
    "            \"effect on cell entry\",\n",
    "            \"escape_377H\",\n",
    "            \"escape_2510C\",\n",
    "            \"escape_89F\",\n",
    "            \"escape_256A\",\n",
    "            \"escape_372D\",\n",
    "            \"escape_121F\",\n",
    "            \"antibody escape\",\n",
    "            \"nonsynonymous\",\n",
    "            \"FEL_beta\",\n",
    "            \"FUBAR_beta\",\n",
    "            # \"synonymous\",\n",
    "            # \"non/syn\",\n",
    "            # \"FEL_beta\",\n",
    "            # \"FEL_alpha\",\n",
    "            # \"FEL_alpha=beta\",\n",
    "            # \"FUBAR_beta\",\n",
    "            # \"FUBAR_alpha\",\n",
    "            # \"FUBAR_beta-alpha\",\n",
    "        ],\n",
    "    ).properties(\n",
    "        width=400,\n",
    "        height=400,\n",
    "    )\n",
    "\n",
    "    subplots.append(curr_subplot)\n",
    "\n",
    "no_log_plots = alt.hconcat(\n",
    "    subplots[0],\n",
    "    subplots[1],\n",
    "    subplots[2],\n",
    "    spacing=5,\n",
    "    # title=[\"Correlations of predicted vs actual substitution metrics\"],\n",
    ")\n",
    "\n",
    "\n",
    "ols_df[\"log2 nonsynonymous\"] = np.log2(ols_df[\"nonsynonymous\"])\n",
    "ols_df[\"log2 FEL_beta\"] = np.log2(ols_df[\"FEL_beta\"])\n",
    "ols_df[\"log2 FUBAR_beta\"] = np.log2(ols_df[\"FUBAR_beta\"])\n",
    "ols_df = (\n",
    "    ols_df\n",
    "    .replace([np.inf, -np.inf], np.nan) \n",
    "    .dropna()\n",
    "    .reset_index(drop=True)\n",
    ")\n",
    "\n",
    "\n",
    "Y_vars = [\n",
    "    \"log2 nonsynonymous\",\n",
    "    \"log2 FEL_beta\",\n",
    "    \"log2 FUBAR_beta\",\n",
    "]\n",
    "\n",
    "subplots = []\n",
    "for Y_var in Y_vars:\n",
    "    \n",
    "    Y_values = ols_df[[Y_var]]\n",
    "    X_values = ols_df[X_vars]\n",
    "\n",
    "    # https://www.statsmodels.org/dev/examples/notebooks/generated/ols.html\n",
    "    ols_model = statsmodels.api.OLS(\n",
    "        endog=Y_values,\n",
    "        exog=statsmodels.api.add_constant(X_values.astype(float)),\n",
    "    )\n",
    "    res_ols = ols_model.fit()\n",
    "    full_r2 = res_ols.rsquared\n",
    "    # print(res_ols.summary())\n",
    "    ols_df[Y_var+\"_predicted\"] = res_ols.predict()\n",
    "\n",
    "    # Calculate correlation \n",
    "    actual_r, p = sp.stats.pearsonr(\n",
    "        ols_df[[Y_var, Y_var+\"_predicted\"]].replace([np.inf, -np.inf], np.nan).dropna()[Y_var+\"_predicted\"],\n",
    "        ols_df[[Y_var, Y_var+\"_predicted\"]].replace([np.inf, -np.inf], np.nan).dropna()[Y_var],\n",
    "    )\n",
    "    print(f\"r correlation of predicted {Y_var} vs actual {Y_var}: {actual_r:.2f}\")\n",
    "    \n",
    "    unique_var = {}\n",
    "    # https://blog.minitab.com/en/adventures-in-statistics-2/how-to-identify-the-most-important-predictor-variables-in-regression-models\n",
    "    for vremove in X_vars:\n",
    "        vremove_ols_model = statsmodels.api.OLS(\n",
    "            endog=Y_values,\n",
    "            exog=statsmodels.api.add_constant(ols_df[[v for v in X_vars if v != vremove]].astype(float)),\n",
    "        )\n",
    "        vremove_res_ols = vremove_ols_model.fit()\n",
    "        unique_var[vremove] = full_r2 - vremove_res_ols.rsquared\n",
    "    \n",
    "    # https://stackoverflow.com/a/53966201\n",
    "    subtitle = [\n",
    "        f\"{var}: {unique_var[var] * 100:.1f}% of variance (coef {res_ols.params[var]:.3f} \\u00B1 {res_ols.bse[var]:.3f})\"\n",
    "        for var in X_vars\n",
    "    ]\n",
    "\n",
    "    chart_title = alt.TitleParams(\n",
    "        Y_var + \" (n = \" + str(Y_values.shape[0]) + \")\",\n",
    "        subtitle=subtitle,\n",
    "        fontSize=16,\n",
    "    )\n",
    "\n",
    "    curr_subplot = alt.Chart(\n",
    "        ols_df,\n",
    "        title=chart_title,\n",
    "    ).mark_point(\n",
    "        filled=True, \n",
    "        color=\"black\", \n",
    "        size=75,\n",
    "        opacity=0.15,\n",
    "    ).encode(\n",
    "        alt.X(\n",
    "            Y_var+\"_predicted\",\n",
    "            axis=alt.Axis(\n",
    "                title=[\"predicted \" + Y_var],\n",
    "                # values=[-4,-3,-2,-1,0,1],\n",
    "                # domainWidth=1,\n",
    "                # domainColor=\"black\",\n",
    "                # tickColor=\"black\",\n",
    "            ),\n",
    "            # scale=alt.Scale(type=\"symlog\")\n",
    "        ),\n",
    "        alt.Y(\n",
    "            Y_var,\n",
    "            axis=alt.Axis(\n",
    "                title=[\"actual \" + Y_var],\n",
    "                # values=[-3,-2,-1,0],\n",
    "                # domainWidth=1,\n",
    "                # domainColor=\"black\",\n",
    "                # tickColor=\"black\",\n",
    "            ),\n",
    "            # scale=alt.Scale(type=\"symlog\")\n",
    "        ),\n",
    "        tooltip=[\n",
    "            \"site\",\n",
    "            \"wildtype\",\n",
    "            Y_var,\n",
    "            Y_var+\"_predicted\",\n",
    "            \"effect on cell entry\",\n",
    "            \"escape_377H\",\n",
    "            \"escape_2510C\",\n",
    "            \"escape_89F\",\n",
    "            \"escape_256A\",\n",
    "            \"escape_372D\",\n",
    "            \"escape_121F\",\n",
    "            \"antibody escape\",\n",
    "            \"nonsynonymous\",\n",
    "            \"FEL_beta\",\n",
    "            \"FUBAR_beta\",\n",
    "            # \"synonymous\",\n",
    "            # \"non/syn\",\n",
    "            # \"FEL_beta\",\n",
    "            # \"FEL_alpha\",\n",
    "            # \"FEL_alpha=beta\",\n",
    "            # \"FUBAR_beta\",\n",
    "            # \"FUBAR_alpha\",\n",
    "            # \"FUBAR_beta-alpha\",\n",
    "        ],\n",
    "    ).properties(\n",
    "        width=400,\n",
    "        height=400,\n",
    "    )\n",
    "\n",
    "    subplots.append(curr_subplot)\n",
    "\n",
    "log_plots = alt.hconcat(\n",
    "    subplots[0],\n",
    "    subplots[1],\n",
    "    subplots[2],\n",
    "    spacing=5,\n",
    ")\n",
    "\n",
    "combined_plot = alt.vconcat(\n",
    "    no_log_plots,\n",
    "    log_plots,\n",
    "    title=[\"Correlations of predicted vs actual substitution metrics\"],\n",
    ").configure_axis(\n",
    "    grid=False,\n",
    "    labelFontSize=16,\n",
    "    titleFontSize=16,\n",
    "    labelFontWeight=\"normal\",\n",
    "    titleFontWeight=\"normal\",\n",
    ").configure_axis(\n",
    "    grid=False,\n",
    "    labelFontSize=16,\n",
    "    titleFontSize=16,\n",
    "    labelFontWeight=\"normal\",\n",
    "    titleFontWeight=\"normal\",\n",
    ").configure_title(\n",
    "    fontSize=24,\n",
    ")\n",
    "\n",
    "# Make output dir if doesn't exist\n",
    "if not os.path.exists(out_dir_natural):\n",
    "    os.mkdir(out_dir_natural)\n",
    "\n",
    "combined_plot.save(ols_regression)\n",
    "\n",
    "combined_plot"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5e2b7088-8622-46e0-8de0-6fd223686802",
   "metadata": {},
   "source": [
    "## Antibody escape compared to natural sequence time of isolation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9da3a1c5-a376-4c09-86d2-5d6d1e0b2408",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load metadata as dataframe\n",
    "all_metadata = pd.read_csv(natural_seq_metadata, sep=\"\\t\")\n",
    "\n",
    "# Load alignment and metadata info\n",
    "natural_seqs_df = pd.DataFrame(columns=[\"strain\", \"sequence\"])\n",
    "\n",
    "# Load list of sequence names from alignment\n",
    "strains = []\n",
    "for curr_fasta in SeqIO.parse(natural_seq_alignment, \"fasta\"):\n",
    "    strains.append(str(curr_fasta.id))\n",
    "    natural_seqs_df.loc[len(natural_seqs_df.index)] = [\n",
    "        str(curr_fasta.id),\n",
    "        str(curr_fasta.seq),\n",
    "    ]\n",
    "\n",
    "# Filter metadata\n",
    "GPC_metadata = (\n",
    "    all_metadata.loc[all_metadata[\"strain\"].isin(strains)].copy()\n",
    ")\n",
    "# Add Sierra Leone for Josiah strain and fix date for Josiah\n",
    "# using paper Wulff and Johnson, 1979\n",
    "# Note that some dates might be wrong because they were incorrectly\n",
    "# dated on NCBI virus\n",
    "GPC_metadata.at[0, \"country\"] = \"Sierra Leone\"\n",
    "GPC_metadata.at[0, \"date\"] = \"1976-XX-XX\"\n",
    "# Also add host for Pinneo strain\n",
    "GPC_metadata.at[1, \"host\"] = \"Homo sapiens\"\n",
    "\n",
    "# Add sequence data and metadata\n",
    "natural_seqs_df = (\n",
    "    natural_seqs_df.merge(\n",
    "        GPC_metadata,\n",
    "        how=\"left\",\n",
    "        on=\"strain\",\n",
    "        validate=\"one_to_one\",\n",
    "    )\n",
    ")\n",
    "\n",
    "# Create column of collection year\n",
    "natural_seqs_df[\"date_year\"] = (\n",
    "    natural_seqs_df[\"date\"].apply(lambda x: x[0:4])\n",
    ")\n",
    "\n",
    "# Re-group host into simpler categories\n",
    "natural_seqs_df[\"grouped_host\"] = (\n",
    "    natural_seqs_df[\"host\"].apply(lambda x: \"human\" if x == \"Homo sapiens\" else (\"unknown\" if x == \"MISSING\" else \"rodent\"))\n",
    ")\n",
    "\n",
    "# Extract josiah sequence for comparisons\n",
    "josiah_seq = (\n",
    "    natural_seqs_df.loc[\n",
    "        (natural_seqs_df[\"strain\"] == \"Josiah_NC_004296_reverse_complement_2018-08-13\")\n",
    "    ][\"sequence\"][0]\n",
    ")\n",
    "\n",
    "# Read functional score data\n",
    "functional_scores = pd.read_csv(func_scores)\n",
    "\n",
    "# Filter for minimum selections, times seen and no stop codons\n",
    "functional_scores = (\n",
    "    functional_scores.query(\n",
    "        \"n_selections >= @n_selections and times_seen >= @min_times_seen and mutant != '*'\"\n",
    "    )\n",
    "    .reset_index(drop=True)\n",
    ")\n",
    "\n",
    "# Initialize list of escape files and final df\n",
    "antibody_files = [\n",
    "    filtered_escape_377H,\n",
    "    filtered_escape_89F,\n",
    "    filtered_escape_2510C,\n",
    "    filtered_escape_121F,\n",
    "    filtered_escape_256A,\n",
    "    filtered_escape_372D,\n",
    "]\n",
    "antibody_escape = pd.DataFrame(columns=[\"site\", \"wildtype\"])\n",
    "\n",
    "# Add escape to dataframe for each antibody\n",
    "for index,antibody_file in enumerate(antibody_files):\n",
    "\n",
    "    antibody_name = antibody_file.split(\"/\")[-1].split(\"_\")[0]\n",
    "\n",
    "    # Load data as dataframe\n",
    "    escape_df = pd.read_csv(antibody_file)\n",
    "\n",
    "    # Clip lower scores to 0\n",
    "    escape_df[\"escape_median\"] = escape_df[\"escape_median\"].clip(lower=0)\n",
    "\n",
    "    # Rename escape column to include antibody name\n",
    "    escape_df = (\n",
    "        escape_df\n",
    "        .rename(columns={\"escape_median\" : \"escape_\" + antibody_name})\n",
    "    )\n",
    "\n",
    "    if index == 0:\n",
    "        antibody_escape = (\n",
    "            pd.concat([\n",
    "                antibody_escape, \n",
    "                escape_df[[\"site\", \"wildtype\", \"mutant\", \"escape_\" + antibody_name]],\n",
    "            ])\n",
    "        )\n",
    "    else:\n",
    "        # Merge dataframes\n",
    "        antibody_escape = (\n",
    "            antibody_escape.merge(\n",
    "                escape_df[[\"site\", \"wildtype\", \"mutant\", \"escape_\" + antibody_name]],\n",
    "                how=\"left\",\n",
    "                on=[\"site\", \"wildtype\", \"mutant\"],\n",
    "                validate=\"one_to_one\",\n",
    "            )\n",
    "        )\n",
    "\n",
    "merged_df = (\n",
    "    functional_scores.merge(\n",
    "        antibody_escape,\n",
    "        how=\"left\",\n",
    "        on=[\"site\", \"wildtype\", \"mutant\"],\n",
    "        validate=\"one_to_one\",\n",
    "    )\n",
    ")\n",
    "\n",
    "# Add column of mutation\n",
    "merged_df[\"mutation\"] = (\n",
    "    merged_df[\"wildtype\"] + merged_df[\"site\"].astype(str) + merged_df[\"mutant\"]\n",
    ")\n",
    "\n",
    "def calculate_natural_sequence_scores(antibody, sequence, jos_sequence):\n",
    "    \"\"\"\n",
    "    Calculate a cumulative escape or functional \n",
    "    effects score for a natural sequence.\n",
    "    \"\"\"\n",
    "\n",
    "    list_of_muts = []\n",
    "    site = 1\n",
    "    # Get list of mutations\n",
    "    for s1, s2 in zip(jos_sequence, sequence):\n",
    "        if s1 != s2 and s1 != \"-\" and s2 != \"-\":\n",
    "            curr_mut = f\"{s1}{site}{s2}\"\n",
    "            list_of_muts.append(curr_mut)\n",
    "        site += 1\n",
    "\n",
    "    max = merged_df[antibody].max()\n",
    "    min = merged_df[antibody].min()\n",
    "\n",
    "    total_escape = 0\n",
    "    # Iterate through mutations and get escape scores\n",
    "    for mut in list_of_muts:\n",
    "        curr_mut = merged_df.query(\"mutation == @mut\")\n",
    "        if curr_mut.shape[0] == 1:\n",
    "            curr_mut_escape = (\n",
    "                curr_mut\n",
    "                .fillna(0)\n",
    "                .reset_index(drop=True)[antibody][0]\n",
    "            )\n",
    "            # total_escape += (curr_mut_escape - min) / (max-min)\n",
    "            total_escape += curr_mut_escape\n",
    "\n",
    "    # Return total escape\n",
    "    return total_escape \n",
    "\n",
    "def identify_sequences_with_escape_muts(antibody, sequence, jos_sequence, threshold, strain, antibody_name):\n",
    "    \"\"\"\n",
    "    Identify sequences with single escape mutations over a threshold.\n",
    "    \"\"\"\n",
    "\n",
    "    list_of_muts = []\n",
    "    site = 1\n",
    "    # Get list of mutations\n",
    "    for s1, s2 in zip(jos_sequence, sequence):\n",
    "        if s1 != s2 and s1 != \"-\" and s2 != \"-\":\n",
    "            curr_mut = f\"{s1}{site}{s2}\"\n",
    "            list_of_muts.append(curr_mut)\n",
    "        site += 1\n",
    "\n",
    "    # Iterate through mutations and get escape scores\n",
    "    for mut in list_of_muts:\n",
    "        curr_mut = merged_df.query(\"mutation == @mut\")\n",
    "        if curr_mut.shape[0] == 1:\n",
    "            curr_mut_escape = (\n",
    "                curr_mut\n",
    "                .fillna(0)\n",
    "                .reset_index(drop=True)[antibody][0]\n",
    "            )\n",
    "            if curr_mut_escape >= threshold:\n",
    "                # print(f\"{antibody_name} escaped by {curr_mut.fillna(0).reset_index(drop=True)['mutation'][0]} ({curr_mut_escape}) in {strain}\")\n",
    "                return True\n",
    "\n",
    "    # Return false if no muts found\n",
    "    return False \n",
    "\n",
    "# Calculate antibody escape for individual antibodies\n",
    "for antibody in [\"377H\", \"256A\", \"372D\", \"121F\", \"89F\", \"2510C\"]:\n",
    "    natural_seqs_df[\"total_\"+antibody+\"_escape\"] = (\n",
    "        natural_seqs_df[\"sequence\"].apply(lambda x: calculate_natural_sequence_scores(\n",
    "            \"escape_\"+antibody,\n",
    "            x,\n",
    "            josiah_seq\n",
    "        ))\n",
    "    )\n",
    "\n",
    "# Determine if escape muts exist for individual antibodies\n",
    "for antibody in [\"377H\", \"256A\", \"372D\", \"121F\", \"89F\", \"2510C\"]:\n",
    "    natural_seqs_df[\"strong_escape_mut_\"+antibody] = (\n",
    "        natural_seqs_df.apply(lambda x: identify_sequences_with_escape_muts(\n",
    "            \"escape_\"+antibody,\n",
    "            x[\"sequence\"],\n",
    "            josiah_seq,\n",
    "            1,\n",
    "            x[\"strain\"],\n",
    "            antibody,\n",
    "        ), axis=1)\n",
    "    )\n",
    "\n",
    "# Calculate total antibody escape\n",
    "natural_seqs_df[\"total_escape\"] = (\n",
    "    natural_seqs_df[[\n",
    "        \"total_377H_escape\", \n",
    "        \"total_256A_escape\", \n",
    "        \"total_372D_escape\", \n",
    "        \"total_121F_escape\", \n",
    "        \"total_89F_escape\", \n",
    "        \"total_2510C_escape\"\n",
    "    ]].sum(axis=1)\n",
    ")\n",
    "\n",
    "# Calculate total functional scores\n",
    "natural_seqs_df[\"total_func_effect\"] = (\n",
    "    natural_seqs_df[\"sequence\"].apply(lambda x: calculate_natural_sequence_scores(\n",
    "        \"effect\",\n",
    "        x,\n",
    "        josiah_seq\n",
    "    ))\n",
    ")\n",
    "\n",
    "antibody_list = [\n",
    "    \"total_2510C_escape\",\n",
    "    \"total_121F_escape\", \n",
    "    \"total_377H_escape\", \n",
    "    \"total_256A_escape\", \n",
    "    \"total_372D_escape\", \n",
    "    \"total_89F_escape\", \n",
    "    \n",
    "]\n",
    "\n",
    "# Isolates chosen to validate for escape \n",
    "chosen_isolates = [\n",
    "    \"Josiah_NC_004296_reverse_complement_2018-08-13\",\n",
    "    \"GA391_OL774861_reverse_complement_1977-XX-XX\",\n",
    "    \"LASV_H-sapiens-tc_NGA_2016_IRR_007_MK107922_2016-01-18\",\n",
    "    \"Lassa_virus_H-sapiens-wt_NGA_2018_ISTH_1024_MH157037_2018-02-14\",\n",
    "    \"LM395-SLE-2009_KM822115_2009-XX-XX\",\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "16b43783-b1af-46ac-9bbe-6cb8fb765b8e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Print number of human and rodent sequences that have strong escape mutants\n",
    "print(f\"Total number of sequences isolated from humans: {natural_seqs_df.loc[natural_seqs_df['grouped_host'] == 'human'].shape[0]}\")\n",
    "print(f\"Total number of sequences isolated from rodents: {natural_seqs_df.loc[natural_seqs_df['grouped_host'] == 'rodent'].shape[0]}\")\n",
    "print(f\"Total number of sequences isolated from missing: {natural_seqs_df.loc[natural_seqs_df['grouped_host'] == 'unknown'].shape[0]}\")\n",
    "print(f\"Total number of sequences: {natural_seqs_df.shape[0]}\")\n",
    "print(f\"Total number of sequences isolated from humans with strong escape mutations: {natural_seqs_df.loc[natural_seqs_df['grouped_host'] == 'human'].query('strong_escape_mut_2510C == True | strong_escape_mut_121F == True | strong_escape_mut_377H == True | strong_escape_mut_256A == True | strong_escape_mut_372D == True | strong_escape_mut_89F == True').shape[0]}\")\n",
    "print(f\"Total number of sequences isolated from rodents with strong escape mutations: {natural_seqs_df.loc[natural_seqs_df['grouped_host'] == 'rodent'].query('strong_escape_mut_2510C == True | strong_escape_mut_121F == True | strong_escape_mut_377H == True | strong_escape_mut_256A == True | strong_escape_mut_372D == True | strong_escape_mut_89F == True').shape[0]}\")\n",
    "print(f\"Total number of sequences isolated from missing with strong escape mutations: {natural_seqs_df.loc[natural_seqs_df['grouped_host'] == 'unknown'].query('strong_escape_mut_2510C == True | strong_escape_mut_121F == True | strong_escape_mut_377H == True | strong_escape_mut_256A == True | strong_escape_mut_372D == True | strong_escape_mut_89F == True').shape[0]}\")\n",
    "print(f\"Total number of sequences without strong escape mutations: {natural_seqs_df.query('strong_escape_mut_2510C == False & strong_escape_mut_121F == False & strong_escape_mut_377H == False & strong_escape_mut_256A == False & strong_escape_mut_372D == False & strong_escape_mut_89F == False').shape[0]}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0d11ecd4-474a-410b-ade2-225c17236554",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Calculate correlation \n",
    "r, p = sp.stats.pearsonr(\n",
    "    natural_seqs_df[[\"date_year\", \"total_escape\"]].dropna().astype(float)[\"date_year\"],\n",
    "    natural_seqs_df[[\"date_year\", \"total_escape\"]].dropna().astype(float)[\"total_escape\"],\n",
    ")\n",
    "print(f\"r correlation of time and antibody escape: {r:.2f}\")\n",
    "\n",
    "antibody_escape = alt.Chart(\n",
    "    natural_seqs_df, \n",
    ").mark_point(\n",
    "    filled=True, \n",
    "    color=\"black\", \n",
    "    size=75,\n",
    "    opacity=0.5,\n",
    ").encode(\n",
    "    alt.Y(\n",
    "        \"total_escape\",\n",
    "        axis=alt.Axis(\n",
    "            title=\"total antibody escape\",\n",
    "            domainWidth=1,\n",
    "            domainColor=\"black\",\n",
    "            tickColor=\"black\",\n",
    "        ),\n",
    "    ),\n",
    "    alt.X(\n",
    "        \"date_year:T\",\n",
    "        axis=alt.Axis(\n",
    "            title=\"date\",\n",
    "            domainWidth=1,\n",
    "            domainColor=\"black\",\n",
    "            tickColor=\"black\",\n",
    "        ),\n",
    "    ),\n",
    "    color=alt.Color(\n",
    "        \"country:N\", \n",
    "        title=\"country\",\n",
    "        scale=alt.Scale(\n",
    "            domain=natural_seqs_df[\"country\"].unique().tolist(),\n",
    "            range=tol_muted_adjusted[1:]\n",
    "        ),\n",
    "    ), \n",
    "    shape=alt.Shape(\n",
    "        \"grouped_host:N\",\n",
    "        title=\"host\",\n",
    "    ),\n",
    "    tooltip=[\n",
    "        \"strain\",\n",
    "        \"country\",\n",
    "        \"host\",\n",
    "        \"date\",\n",
    "        \"total_escape\",\n",
    "        \"total_377H_escape\", \n",
    "        \"total_256A_escape\", \n",
    "        \"total_372D_escape\", \n",
    "        \"total_121F_escape\", \n",
    "        \"total_89F_escape\", \n",
    "        \"total_2510C_escape\",\n",
    "        \"total_func_effect\",\n",
    "    ],\n",
    ").properties(\n",
    "    width=500,\n",
    "    height=300,\n",
    ")\n",
    "\n",
    "# Calculate correlation \n",
    "r, p = sp.stats.pearsonr(\n",
    "    natural_seqs_df[[\"date_year\", \"total_func_effect\"]].dropna().astype(float)[\"date_year\"],\n",
    "    natural_seqs_df[[\"date_year\", \"total_func_effect\"]].dropna().astype(float)[\"total_func_effect\"],\n",
    ")\n",
    "print(f\"r correlation of time and functional effects: {r:.2f}\")\n",
    "\n",
    "func_effects = alt.Chart(\n",
    "    natural_seqs_df, \n",
    ").mark_point(\n",
    "    filled=True, \n",
    "    color=\"black\", \n",
    "    size=75,\n",
    "    opacity=0.5,\n",
    ").encode(\n",
    "    alt.Y(\n",
    "        \"total_func_effect\",\n",
    "        axis=alt.Axis(\n",
    "            title=\"total functional effect\",\n",
    "            domainWidth=1,\n",
    "            domainColor=\"black\",\n",
    "            tickColor=\"black\",\n",
    "        ),\n",
    "    ),\n",
    "    alt.X(\n",
    "        \"date_year:T\",\n",
    "        axis=alt.Axis(\n",
    "            title=\"date\",\n",
    "            domainWidth=1,\n",
    "            domainColor=\"black\",\n",
    "            tickColor=\"black\",\n",
    "        ),\n",
    "    ),\n",
    "    color=alt.Color(\n",
    "        \"country:N\", \n",
    "        title=\"country\",\n",
    "        scale=alt.Scale(\n",
    "            domain=natural_seqs_df[\"country\"].unique().tolist(),\n",
    "            range=tol_muted_adjusted[1:]\n",
    "        ),\n",
    "    ), \n",
    "    shape=alt.Shape(\n",
    "        \"grouped_host:N\",\n",
    "        title=\"host\",\n",
    "    ),\n",
    "    tooltip=[\n",
    "        \"strain\",\n",
    "        \"country\",\n",
    "        \"host\",\n",
    "        \"date\",\n",
    "        \"total_escape\",\n",
    "        \"total_377H_escape\", \n",
    "        \"total_256A_escape\", \n",
    "        \"total_372D_escape\", \n",
    "        \"total_121F_escape\", \n",
    "        \"total_89F_escape\", \n",
    "        \"total_2510C_escape\",\n",
    "        \"total_func_effect\",\n",
    "    ],\n",
    ").properties(\n",
    "    width=500,\n",
    "    height=300,\n",
    ")\n",
    "\n",
    "combined_plot = alt.hconcat(\n",
    "    antibody_escape,\n",
    "    func_effects,\n",
    "    spacing=5,\n",
    "    title=\"Antibody and functional effects across time\",\n",
    ").configure_axis(\n",
    "    grid=False,\n",
    "    labelFontSize=16,\n",
    "    titleFontSize=16,\n",
    "    labelFontWeight=\"normal\",\n",
    "    titleFontWeight=\"normal\",\n",
    ").configure_axis(\n",
    "    grid=False,\n",
    "    labelFontSize=16,\n",
    "    titleFontSize=16,\n",
    "    labelFontWeight=\"normal\",\n",
    "    titleFontWeight=\"normal\",\n",
    ").configure_title(\n",
    "    fontSize=24,\n",
    ")\n",
    "\n",
    "combined_plot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "71e7f797-efc3-4d68-b7c7-58f7db564185",
   "metadata": {},
   "outputs": [],
   "source": [
    "antibody_escape = alt.Chart(\n",
    "    natural_seqs_df, \n",
    ").mark_point(\n",
    "    filled=True, \n",
    "    color=\"black\", \n",
    "    size=75,\n",
    "    opacity=0.5,\n",
    ").encode(\n",
    "    alt.Y(\n",
    "        \"total_escape\",\n",
    "        axis=alt.Axis(\n",
    "            title=\"total antibody escape\",\n",
    "            domainWidth=1,\n",
    "            domainColor=\"black\",\n",
    "            tickColor=\"black\",\n",
    "        ),\n",
    "    ),\n",
    "    alt.X(\n",
    "        \"grouped_host:N\",\n",
    "        axis=alt.Axis(\n",
    "            title=\"host\",\n",
    "            domainWidth=1,\n",
    "            domainColor=\"black\",\n",
    "            tickColor=\"black\",\n",
    "        ),\n",
    "    ),\n",
    "    xOffset=\"jitter:Q\",\n",
    "    color=alt.Color(\n",
    "        \"country:N\", \n",
    "        title=\"country\",\n",
    "        scale=alt.Scale(\n",
    "            domain=natural_seqs_df[\"country\"].unique().tolist(),\n",
    "            range=tol_muted_adjusted[1:]\n",
    "        ),\n",
    "    ), \n",
    "    tooltip=[\n",
    "        \"strain\",\n",
    "        \"country\",\n",
    "        \"host\",\n",
    "        \"date\",\n",
    "        \"total_escape\",\n",
    "        \"total_377H_escape\", \n",
    "        \"total_256A_escape\", \n",
    "        \"total_372D_escape\", \n",
    "        \"total_121F_escape\", \n",
    "        \"total_89F_escape\", \n",
    "        \"total_2510C_escape\",\n",
    "        \"total_func_effect\",\n",
    "    ],\n",
    ").transform_calculate(\n",
    "    # Generate Gaussian jitter with a Box-Muller transform\n",
    "    jitter=\"sqrt(-2*log(random()))*cos(2*PI*random())\"\n",
    ").properties(\n",
    "    width=500,\n",
    "    height=300,\n",
    ")\n",
    "\n",
    "func_effects = alt.Chart(\n",
    "    natural_seqs_df, \n",
    ").mark_point(\n",
    "    filled=True, \n",
    "    color=\"black\", \n",
    "    size=75,\n",
    "    opacity=0.5,\n",
    ").encode(\n",
    "    alt.Y(\n",
    "        \"total_func_effect\",\n",
    "        axis=alt.Axis(\n",
    "            title=\"total functional effect\",\n",
    "            domainWidth=1,\n",
    "            domainColor=\"black\",\n",
    "            tickColor=\"black\",\n",
    "        ),\n",
    "    ),\n",
    "    alt.X(\n",
    "        \"grouped_host:N\",\n",
    "        axis=alt.Axis(\n",
    "            title=\"host\",\n",
    "            domainWidth=1,\n",
    "            domainColor=\"black\",\n",
    "            tickColor=\"black\",\n",
    "        ),\n",
    "    ),\n",
    "    xOffset=\"jitter:Q\",\n",
    "    color=alt.Color(\n",
    "        \"country:N\", \n",
    "        title=\"country\",\n",
    "        scale=alt.Scale(\n",
    "            domain=natural_seqs_df[\"country\"].unique().tolist(),\n",
    "            range=tol_muted_adjusted[1:]\n",
    "        ),\n",
    "    ), \n",
    "    tooltip=[\n",
    "        \"strain\",\n",
    "        \"country\",\n",
    "        \"host\",\n",
    "        \"date\",\n",
    "        \"total_escape\",\n",
    "        \"total_377H_escape\", \n",
    "        \"total_256A_escape\", \n",
    "        \"total_372D_escape\", \n",
    "        \"total_121F_escape\", \n",
    "        \"total_89F_escape\", \n",
    "        \"total_2510C_escape\",\n",
    "        \"total_func_effect\",\n",
    "    ],\n",
    ").transform_calculate(\n",
    "    # Generate Gaussian jitter with a Box-Muller transform\n",
    "    jitter=\"sqrt(-2*log(random()))*cos(2*PI*random())\"\n",
    ").properties(\n",
    "    width=500,\n",
    "    height=300,\n",
    ")\n",
    "\n",
    "combined_plot = alt.hconcat(\n",
    "    antibody_escape,\n",
    "    func_effects,\n",
    "    spacing=5,\n",
    "    title=\"Antibody and functional effects for different hosts\",\n",
    ").configure_axis(\n",
    "    grid=False,\n",
    "    labelFontSize=16,\n",
    "    titleFontSize=16,\n",
    "    labelFontWeight=\"normal\",\n",
    "    titleFontWeight=\"normal\",\n",
    ").configure_axis(\n",
    "    grid=False,\n",
    "    labelFontSize=16,\n",
    "    titleFontSize=16,\n",
    "    labelFontWeight=\"normal\",\n",
    "    titleFontWeight=\"normal\",\n",
    ").configure_title(\n",
    "    fontSize=24,\n",
    ")\n",
    "\n",
    "combined_plot"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
