{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "3bab6508-4b92-4da1-a9e8-bfef44231f2a",
   "metadata": {},
   "source": [
    "# Convert escape and functional scores to polyclonal b factors for PDB viewing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bbc054a5-bda7-489c-a6d4-0e7d217fcd50",
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": ""
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Imports\n",
    "import os\n",
    "import pandas as pd\n",
    "import polyclonal"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f1c64df0-879a-4e86-a4f7-c02341178e87",
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": ""
    },
    "tags": [
     "parameters"
    ]
   },
   "outputs": [],
   "source": [
    "# this cell is tagged as `parameters` for papermill parameterization\n",
    "func_scores = None\n",
    "pdb_file = None\n",
    "\n",
    "filtered_escape_377H = None\n",
    "filtered_escape_89F = None\n",
    "filtered_escape_2510C = None\n",
    "filtered_escape_121F = None\n",
    "filtered_escape_256A = None\n",
    "filtered_escape_372D = None\n",
    "natural_sequence_variation = None\n",
    "\n",
    "out_dir = None\n",
    "\n",
    "pdb_func_min = None\n",
    "pdb_func_max = None\n",
    "pdb_377H = None\n",
    "pdb_89F = None\n",
    "pdb_2510C = None\n",
    "pdb_121F = None\n",
    "pdb_256A = None\n",
    "pdb_372D = None\n",
    "pdb_natural_variation = None\n",
    "pdb_arevirumab = None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7e9e4c56-ffb6-4e8c-9d85-a2a618beda04",
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": ""
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "# # Uncomment for running interactive\n",
    "# func_scores = \"../results/filtered_func_effect_CSVs/293T_filtered_func_effects.csv\"\n",
    "# pdb_file = \"../data/7puy.pdb\"\n",
    "\n",
    "# filtered_escape_377H = \"../results/filtered_antibody_escape_CSVs/377H_filtered_mut_effect.csv\"\n",
    "# filtered_escape_89F = \"../results/filtered_antibody_escape_CSVs/89F_filtered_mut_effect.csv\"\n",
    "# filtered_escape_2510C = \"../results/filtered_antibody_escape_CSVs/2510C_filtered_mut_effect.csv\"\n",
    "# filtered_escape_121F = \"../results/filtered_antibody_escape_CSVs/121F_filtered_mut_effect.csv\"\n",
    "# filtered_escape_256A = \"../results/filtered_antibody_escape_CSVs/256A_filtered_mut_effect.csv\"\n",
    "# filtered_escape_372D = \"../results/filtered_antibody_escape_CSVs/372D_filtered_mut_effect.csv\"\n",
    "# natural_sequence_variation = \"../non-pipeline_analyses/LASV_phylogeny_analysis/Results/GPC_protein_variation.csv\"\n",
    "\n",
    "# out_dir = \"../results/mapped_scores_onto_pdb/\"\n",
    "\n",
    "# pdb_func_min = \"../results/mapped_scores_onto_pdb/func_scores_min.pdb\"\n",
    "# pdb_func_max = \"../results/mapped_scores_onto_pdb/func_scores_max.pdb\"\n",
    "# pdb_377H = \"../results/mapped_scores_onto_pdb/377H_escape.pdb\"\n",
    "# pdb_89F = \"../results/mapped_scores_onto_pdb/89F_escape.pdb\"\n",
    "# pdb_2510C = \"../results/mapped_scores_onto_pdb/2510C_escape.pdb\"\n",
    "# pdb_121F = \"../results/mapped_scores_onto_pdb/121F_escape.pdb\"\n",
    "# pdb_256A = \"../results/mapped_scores_onto_pdb/256A_escape.pdb\"\n",
    "# pdb_372D = \"../results/mapped_scores_onto_pdb/372D_escape.pdb\"\n",
    "# pdb_natural_variation = \"../results/mapped_scores_onto_pdb/natural_variation.pdb\"\n",
    "# pdb_arevirumab = \"../results/mapped_scores_onto_pdb/arevirumab_escape.pdb\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "73735c1b-6222-4099-9db2-3b1b8746c2c3",
   "metadata": {},
   "outputs": [],
   "source": [
    "def natural_variation_to_b_factors(input_pdb_file, output_pdb_file, natural_variation_file):\n",
    "    \"\"\"\n",
    "    Function to map natural variation to a pdb structure\n",
    "    using mut_escape_pdb_b_factor function from polyclonal.\n",
    "    \"\"\"\n",
    "\n",
    "    natural_variation = pd.read_csv(natural_variation_file)\n",
    "\n",
    "    natural_variation_aA = natural_variation[[\"site\", \"entropy\"]].copy()\n",
    "    natural_variation_bB = natural_variation[[\"site\", \"entropy\"]].copy()\n",
    "    natural_variation_cC = natural_variation[[\"site\", \"entropy\"]].copy()\n",
    "    \n",
    "    natural_variation_aA[\"chain\"] = (\n",
    "        natural_variation_aA.apply(lambda x: \"A\" if x[\"site\"] <= 259 else \"a\", axis=1)\n",
    "    )\n",
    "    \n",
    "    natural_variation_bB[\"chain\"] = (\n",
    "        natural_variation_bB.apply(lambda x: \"B\" if x[\"site\"] <= 259 else \"b\", axis=1)\n",
    "    )\n",
    "    \n",
    "    natural_variation_cC[\"chain\"] = (\n",
    "        natural_variation_cC.apply(lambda x: \"C\" if x[\"site\"] <= 259 else \"c\", axis=1)\n",
    "    )\n",
    "    \n",
    "    natural_variation = (\n",
    "        pd.concat([\n",
    "            natural_variation_aA,\n",
    "            natural_variation_bB,\n",
    "            natural_variation_cC,\n",
    "        ], ignore_index=True)\n",
    "    )\n",
    "\n",
    "    natural_variation = natural_variation.astype({\"site\" : \"int\"}) \n",
    "\n",
    "    print(\"Natural sequence variation:\")\n",
    "    print(f\"50th percentile of entropy: {natural_variation['entropy'].quantile(0.50)}\")\n",
    "    print(f\"75th percentile of entropy: {natural_variation['entropy'].quantile(0.75)}\")\n",
    "    print(f\"90th percentile of entropys: {natural_variation['entropy'].quantile(0.90)}\")\n",
    "    print(f\"Entropy score range: {natural_variation['entropy'].min()} to {natural_variation['entropy'].max()}\")\n",
    "    print()\n",
    "\n",
    "    polyclonal.pdb_utils.reassign_b_factor(\n",
    "        input_pdbfile=input_pdb_file,\n",
    "        output_pdbfile=output_pdb_file,\n",
    "        df=natural_variation,\n",
    "        metric_col=\"entropy\",\n",
    "        site_col=\"site\",\n",
    "        chain_col=\"chain\",\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e622c2ba-e49c-4817-8821-9778b6725775",
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": ""
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "def functional_scores_to_b_factors(input_pdb_file, output_pdb_file, func_scores_file):\n",
    "    \"\"\"\n",
    "    Function to map funcitonal scores to a pdb structure\n",
    "    using mut_escape_pdb_b_factor function from polyclonal.\n",
    "    \"\"\"\n",
    "    \n",
    "    functional_scores = pd.read_csv(func_scores_file)\n",
    "    print(\"Functional DMS scores:\")\n",
    "\n",
    "    # Filter for minimum selections, times seen and no stop codons\n",
    "    functional_scores = (\n",
    "        functional_scores.query(\n",
    "            \"mutant != '*'\"\n",
    "        )\n",
    "        .drop(columns=[\"mutant\", \"times_seen\", \"wildtype\"])\n",
    "        .groupby([\"site\"])\n",
    "        .aggregate({\n",
    "            \"effect\" : \"mean\"\n",
    "        })\n",
    "        .reset_index()\n",
    "    )\n",
    "\n",
    "    # Scores greater than 0\n",
    "    max_scores = functional_scores.loc[functional_scores[\"effect\"] > 0][\"site\"].tolist()\n",
    "    print(f\"Sites greater than 0: {max_scores}\")\n",
    "    \n",
    "    functional_scores_aA = functional_scores.copy()\n",
    "    functional_scores_bB = functional_scores.copy()\n",
    "    functional_scores_cC = functional_scores.copy()\n",
    "    \n",
    "    functional_scores_aA[\"chain\"] = (\n",
    "        functional_scores_aA.apply(lambda x: \"A\" if x[\"site\"] <= 259 else \"a\", axis=1)\n",
    "    )\n",
    "    \n",
    "    functional_scores_bB[\"chain\"] = (\n",
    "        functional_scores_bB.apply(lambda x: \"B\" if x[\"site\"] <= 259 else \"b\", axis=1)\n",
    "    )\n",
    "    \n",
    "    functional_scores_cC[\"chain\"] = (\n",
    "        functional_scores_cC.apply(lambda x: \"C\" if x[\"site\"] <= 259 else \"c\", axis=1)\n",
    "    )\n",
    "    \n",
    "    functional_scores = (\n",
    "        pd.concat([\n",
    "            functional_scores_aA,\n",
    "            functional_scores_bB,\n",
    "            functional_scores_cC,\n",
    "        ], ignore_index=True)\n",
    "    )\n",
    "\n",
    "    functional_scores = functional_scores.astype({\"site\" : \"int\"}) \n",
    "\n",
    "    print(f\"50th percentile of averaged functional scores: {functional_scores['effect'].quantile(0.50)}\")\n",
    "    print(f\"Functional score range: {functional_scores['effect'].min()} to {functional_scores['effect'].max()}\")\n",
    "    print()\n",
    "\n",
    "    polyclonal.pdb_utils.reassign_b_factor(\n",
    "        input_pdbfile=input_pdb_file,\n",
    "        output_pdbfile=output_pdb_file,\n",
    "        df=functional_scores,\n",
    "        metric_col=\"effect\",\n",
    "        site_col=\"site\",\n",
    "        chain_col=\"chain\",\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9c404dff-486e-4f82-b6e3-9ab43cf8d263",
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": ""
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "def escape_scores_to_b_factors(input_pdb_file, output_pdb_file, escape_scores_file):\n",
    "    \"\"\"\n",
    "    Function to map escape scores to a pdb structure\n",
    "    using mut_escape_pdb_b_factor function from polyclonal.\n",
    "    \"\"\"\n",
    "    \n",
    "    escape_scores = pd.read_csv(escape_scores_file)\n",
    "    antibody_name = escape_scores_file.split(\"/\")[-1].split(\"_\")[0]\n",
    "\n",
    "    # Filter escape df for low functional score mutations\n",
    "    escape_scores = (\n",
    "        escape_scores\n",
    "        .query(\"poor_cell_entry == False\")\n",
    "        .copy()\n",
    "        .reset_index(drop=True)\n",
    "    )\n",
    "\n",
    "    # Add dummy phenotype column\n",
    "    escape_scores[\"phenotype\"] = \"escape\"\n",
    "    # Clip lower scores to 0\n",
    "    escape_scores[\"escape_median\"] = escape_scores[\"escape_median\"].clip(lower=0)\n",
    "\n",
    "    # Calculate site sums\n",
    "    escape_scores = (\n",
    "        escape_scores.groupby([\"site\"])\n",
    "        .aggregate({\n",
    "            \"escape_median\" : \"sum\"\n",
    "        })\n",
    "        .reset_index()\n",
    "    )\n",
    "\n",
    "    print(antibody_name)\n",
    "    print(f\"Max summed escape: {escape_scores['escape_median'].max()}\") # Verify max matches altair plots\n",
    "    print(f\"50th percentile of summed escape scores: {escape_scores['escape_median'].quantile(0.50)}\")\n",
    "    print(f\"75th percentile of summed escape scores: {escape_scores['escape_median'].quantile(0.75)}\")\n",
    "    print(f\"90th percentile of summed escape scores: {escape_scores['escape_median'].quantile(0.90)}\")\n",
    "    print(f\"95th percentile of summed escape scores: {escape_scores['escape_median'].quantile(0.95)}\")\n",
    "    print(f\"99th percentile of summed escape scores: {escape_scores['escape_median'].quantile(0.99)}\")\n",
    "    print()\n",
    "    \n",
    "    escape_scores_aA = escape_scores.copy()\n",
    "    escape_scores_bB = escape_scores.copy()\n",
    "    escape_scores_cC = escape_scores.copy()\n",
    "    \n",
    "    escape_scores_aA[\"chain\"] = (\n",
    "        escape_scores_aA.apply(lambda x: \"A\" if x[\"site\"] <= 259 else \"a\", axis=1)\n",
    "    )\n",
    "    \n",
    "    escape_scores_bB[\"chain\"] = (\n",
    "        escape_scores_bB.apply(lambda x: \"B\" if x[\"site\"] <= 259 else \"b\", axis=1)\n",
    "    )\n",
    "    \n",
    "    escape_scores_cC[\"chain\"] = (\n",
    "        escape_scores_cC.apply(lambda x: \"C\" if x[\"site\"] <= 259 else \"c\", axis=1)\n",
    "    )\n",
    "    \n",
    "    escape_scores = (\n",
    "        pd.concat([\n",
    "            escape_scores_aA,\n",
    "            escape_scores_bB,\n",
    "            escape_scores_cC,\n",
    "        ], ignore_index=True)\n",
    "    )\n",
    "\n",
    "    escape_scores = escape_scores.astype({\"site\" : \"int\"}) \n",
    "\n",
    "    polyclonal.pdb_utils.reassign_b_factor(\n",
    "        input_pdbfile=input_pdb_file,\n",
    "        output_pdbfile=output_pdb_file,\n",
    "        df=escape_scores,\n",
    "        metric_col=\"escape_median\",\n",
    "        site_col=\"site\",\n",
    "        chain_col=\"chain\",\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f09a51be-c832-42e5-aa83-63a3003aff32",
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": ""
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Make output dir if doesn't exist\n",
    "if not os.path.exists(out_dir):\n",
    "    os.mkdir(out_dir)\n",
    "\n",
    "# Natural variation (effective amino acids) mapped to pdb structure\n",
    "natural_variation_to_b_factors(pdb_file, pdb_natural_variation, natural_sequence_variation)\n",
    "\n",
    "# Functional scores mapped to pdb structure for scores less than and greater than 0\n",
    "functional_scores_to_b_factors(pdb_file, pdb_func_min, func_scores)\n",
    "functional_scores_to_b_factors(pdb_file, pdb_func_max, func_scores)\n",
    "\n",
    "# Antibody files\n",
    "antibodies = [\n",
    "    filtered_escape_2510C,\n",
    "    filtered_escape_121F,\n",
    "    filtered_escape_377H,\n",
    "    filtered_escape_256A, \n",
    "    filtered_escape_372D,\n",
    "    filtered_escape_89F,\n",
    "]\n",
    "output_files = [\n",
    "    pdb_2510C,\n",
    "    pdb_121F,\n",
    "    pdb_377H,\n",
    "    pdb_256A,\n",
    "    pdb_372D,\n",
    "    pdb_89F,\n",
    "]\n",
    "\n",
    "# Escape scores mapped to pdb structure\n",
    "for index, antibody in enumerate(antibodies):\n",
    "    escape_scores_to_b_factors(pdb_file, output_files[index], antibody)\n",
    "\n",
    "# Load arevirumab antibody files\n",
    "escape_89F = pd.read_csv(filtered_escape_89F)\n",
    "escape_121F = pd.read_csv(filtered_escape_121F)\n",
    "escape_372D = pd.read_csv(filtered_escape_372D)\n",
    "\n",
    "# Filter escape df for low functional score mutations\n",
    "escape_89F = (\n",
    "    escape_89F\n",
    "    .query(\"poor_cell_entry == False\")\n",
    "    .copy()\n",
    "    .reset_index(drop=True)\n",
    ")\n",
    "\n",
    "# Add dummy phenotype column\n",
    "escape_89F[\"phenotype\"] = \"escape\"\n",
    "# Clip lower scores to 0\n",
    "escape_89F[\"escape_median\"] = escape_89F[\"escape_median\"].clip(lower=0)\n",
    "\n",
    "# Calculate site sums\n",
    "escape_89F = (\n",
    "    escape_89F.groupby([\"site\"])\n",
    "    .aggregate({\n",
    "        \"escape_median\" : \"sum\"\n",
    "    })\n",
    "    .rename(columns={\"escape_median\" : \"89F_escape\"})\n",
    "    .reset_index()\n",
    ")\n",
    "\n",
    "# Filter escape df for low functional score mutations\n",
    "escape_121F = (\n",
    "    escape_121F\n",
    "    .query(\"poor_cell_entry == False\")\n",
    "    .copy()\n",
    "    .reset_index(drop=True)\n",
    ")\n",
    "\n",
    "# Add dummy phenotype column\n",
    "escape_121F[\"phenotype\"] = \"escape\"\n",
    "# Clip lower scores to 0\n",
    "escape_121F[\"escape_median\"] = escape_121F[\"escape_median\"].clip(lower=0)\n",
    "\n",
    "# Calculate site sums\n",
    "escape_121F = (\n",
    "    escape_121F.groupby([\"site\"])\n",
    "    .aggregate({\n",
    "        \"escape_median\" : \"sum\"\n",
    "    })\n",
    "    .rename(columns={\"escape_median\" : \"121F_escape\"})\n",
    "    .reset_index()\n",
    ")\n",
    "\n",
    "# Filter escape df for low functional score mutations\n",
    "escape_372D = (\n",
    "    escape_372D\n",
    "    .query(\"poor_cell_entry == False\")\n",
    "    .copy()\n",
    "    .reset_index(drop=True)\n",
    ")\n",
    "\n",
    "# Add dummy phenotype column\n",
    "escape_372D[\"phenotype\"] = \"escape\"\n",
    "# Clip lower scores to 0\n",
    "escape_372D[\"escape_median\"] = escape_372D[\"escape_median\"].clip(lower=0)\n",
    "\n",
    "# Calculate site sums\n",
    "escape_372D = (\n",
    "    escape_372D.groupby([\"site\"])\n",
    "    .aggregate({\n",
    "        \"escape_median\" : \"sum\"\n",
    "    })\n",
    "    .rename(columns={\"escape_median\" : \"372D_escape\"})\n",
    "    .reset_index()\n",
    ")\n",
    "\n",
    "# Merge frames\n",
    "arevirumab_df = (\n",
    "    escape_89F.merge(\n",
    "        escape_121F,\n",
    "        how=\"left\",\n",
    "        on=\"site\",\n",
    "        validate=\"one_to_one\",\n",
    "    )\n",
    ")\n",
    "arevirumab_df = (\n",
    "    arevirumab_df.merge(\n",
    "        escape_372D,\n",
    "        how=\"left\",\n",
    "        on=\"site\",\n",
    "        validate=\"one_to_one\",\n",
    "    )\n",
    ")\n",
    "\n",
    "# calculate average site escape\n",
    "arevirumab_df[\"average_escape\"] = arevirumab_df[[\"89F_escape\", \"121F_escape\", \"372D_escape\",]].mean(axis=1)\n",
    "\n",
    "print(\"Arevirumab-3\")\n",
    "print(f\"Max summed escape: {arevirumab_df['average_escape'].max()}\") # Verify max matches altair plots\n",
    "print(f'Top 25 escape sites: {sorted(arevirumab_df.nlargest(25, \"average_escape\", keep=\"all\")[\"site\"].tolist())}')\n",
    "print(f\"50th percentile of summed escape scores: {arevirumab_df['average_escape'].quantile(0.50)}\")\n",
    "print(f\"75th percentile of summed escape scores: {arevirumab_df['average_escape'].quantile(0.75)}\")\n",
    "print(f\"90th percentile of summed escape scores: {arevirumab_df['average_escape'].quantile(0.90)}\")\n",
    "print(f\"95th percentile of summed escape scores: {arevirumab_df['average_escape'].quantile(0.95)}\")\n",
    "print(f\"99th percentile of summed escape scores: {arevirumab_df['average_escape'].quantile(0.99)}\")\n",
    "print()\n",
    "\n",
    "arevirumab_df_aA = arevirumab_df.copy()\n",
    "arevirumab_df_bB = arevirumab_df.copy()\n",
    "arevirumab_df_cC = arevirumab_df.copy()\n",
    "\n",
    "arevirumab_df_aA[\"chain\"] = (\n",
    "    arevirumab_df_aA.apply(lambda x: \"A\" if x[\"site\"] <= 259 else \"a\", axis=1)\n",
    ")\n",
    "\n",
    "arevirumab_df_bB[\"chain\"] = (\n",
    "    arevirumab_df_bB.apply(lambda x: \"B\" if x[\"site\"] <= 259 else \"b\", axis=1)\n",
    ")\n",
    "\n",
    "arevirumab_df_cC[\"chain\"] = (\n",
    "    arevirumab_df_cC.apply(lambda x: \"C\" if x[\"site\"] <= 259 else \"c\", axis=1)\n",
    ")\n",
    "\n",
    "arevirumab_df = (\n",
    "    pd.concat([\n",
    "        arevirumab_df_aA,\n",
    "        arevirumab_df_bB,\n",
    "        arevirumab_df_cC,\n",
    "    ], ignore_index=True)\n",
    ")\n",
    "\n",
    "arevirumab_df = arevirumab_df.astype({\"site\" : \"int\"}) \n",
    "\n",
    "polyclonal.pdb_utils.reassign_b_factor(\n",
    "    input_pdbfile=pdb_file,\n",
    "    output_pdbfile=pdb_arevirumab,\n",
    "    df=arevirumab_df,\n",
    "    metric_col=\"average_escape\",\n",
    "    site_col=\"site\",\n",
    "    chain_col=\"chain\",\n",
    ")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
