{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "d0000b36-a072-439d-ac9a-ba88b4f5b0de",
   "metadata": {},
   "source": [
    "# Visualize glycoprotein phylogeny\n",
    "\n",
    "This notebook creates a final tree figure using `ete3`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "38d97f47-4aa9-48e9-9c10-d39efb107f03",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Imports\n",
    "import os\n",
    "import yaml\n",
    "import ete3\n",
    "import warnings\n",
    "# Need this line to render ete3 trees\n",
    "# https://github.com/etetoolkit/ete/issues/296\n",
    "os.environ[\"QT_QPA_PLATFORM\"]= \"offscreen\"\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.colors\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "from pymsaviz import MsaViz \n",
    "from Bio import SeqIO, AlignIO\n",
    "\n",
    "# Create color palette\n",
    "def color_gradient_hex(start, end, n):\n",
    "    \"\"\"Color function from polyclonal\"\"\"\n",
    "    cmap = matplotlib.colors.LinearSegmentedColormap.from_list(\n",
    "            name=\"_\", colors=[start, end], N=n\n",
    "        )\n",
    "    return [matplotlib.colors.rgb2hex(tup) for tup in cmap(list(range(0, n)))]\n",
    "\n",
    "# Black to white color scale\n",
    "gray_scale = color_gradient_hex(\"white\", \"#000000\", n=11)\n",
    "\n",
    "# Rearranged to make the tree look nicer\n",
    "tol_muted_adjusted = [\n",
    "    \"#000000\",\n",
    "    \"#CC6677\", \n",
    "    \"#1f78b4\", \n",
    "    \"#88CCEE\",\n",
    "    \"#882255\",\n",
    "    \"#117733\", \n",
    "    \"#DDCC77\",\n",
    "    \"#44AA99\", \n",
    "    \"#999933\", \n",
    "    \"#AA4499\", \n",
    "    \"#EE7733\",\n",
    "    \"#CC3311\",\n",
    "    \"#DDDDDD\",\n",
    "]\n",
    "\n",
    "# Seaborn style settings\n",
    "sns.set(rc={\"figure.dpi\":300, \"savefig.dpi\":300})\n",
    "sns.set_style(\"ticks\")\n",
    "sns.set_palette(tol_muted_adjusted)\n",
    "\n",
    "# Suppress warnings\n",
    "warnings.simplefilter(\"ignore\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b23f43bc-d39e-40c8-a990-7f0a630f643c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Open config file\n",
    "config_path = \"Configure/config.yml\"\n",
    "with open(config_path) as f:\n",
    "    config = yaml.safe_load(f)\n",
    "\n",
    "# Set paths\n",
    "protein_path = config[\"Protein_sequences\"]\n",
    "isolate_sequences = config[\"Protein_validation_sequences\"]\n",
    "isolate_alignment = config[\"GPC_protein_validation_alignment\"]\n",
    "chosen_validations = config[\"Validation_sequence_names\"]\n",
    "alignment_figure = config[\"GPC_protein_validation_alignment_figure\"]\n",
    "codon_tree_file_path = config[\"GPC_codon_reduced_tree_prefix\"] + \".treefile\"\n",
    "protein_tree_file_path = config[\"GPC_protein_reduced_tree_prefix\"] + \".treefile\"\n",
    "alignment_file_path = config[\"GPC_codon_reduced_alignment\"]\n",
    "protein_alignment_file_path = config[\"GPC_protein_reduced_alignment\"]\n",
    "metadata_file_path = config[\"Metadata\"]\n",
    "figure_dir = config[\"Figures_dir\"]\n",
    "codon_tree_figure_file_path = config[\"GPC_codon_reduced_tree_figure\"]\n",
    "percent_identity_to_josiah_figure_file_path = config[\"Amino_acid_identity_to_josiah_figure\"]\n",
    "protein_tree_figure_file_path = config[\"GPC_protein_reduced_tree_figure\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1fc71256-c093-41e9-8a46-5bd96008c5dc",
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Uncomment to run interactively\n",
    "# # Open config file\n",
    "# config_path = \"../Configure/config.yml\"\n",
    "# with open(config_path) as f:\n",
    "#     config = yaml.safe_load(f)\n",
    "\n",
    "# # Set paths\n",
    "# protein_path = \"../\" + config[\"Protein_sequences\"]\n",
    "# isolate_sequences = \"../\" + config[\"Protein_validation_sequences\"]\n",
    "# isolate_alignment = \"../\" + config[\"GPC_protein_validation_alignment\"]\n",
    "# chosen_validations = config[\"Validation_sequence_names\"]\n",
    "# alignment_figure = \"../\" + config[\"GPC_protein_validation_alignment_figure\"]\n",
    "# codon_tree_file_path = \"../\" + config[\"GPC_codon_reduced_tree_prefix\"] + \".treefile\"\n",
    "# protein_tree_file_path = \"../\" + config[\"GPC_protein_reduced_tree_prefix\"] + \".treefile\"\n",
    "# alignment_file_path = \"../\" + config[\"GPC_codon_reduced_alignment\"]\n",
    "# protein_alignment_file_path = \"../\" + config[\"GPC_protein_reduced_alignment\"]\n",
    "# metadata_file_path = \"../\" + config[\"Metadata\"]\n",
    "# figure_dir = \"../\" + config[\"Figures_dir\"]\n",
    "# codon_tree_figure_file_path = \"../\" + config[\"GPC_codon_reduced_tree_figure\"]\n",
    "# percent_identity_to_josiah_figure_file_path = \"../\" + config[\"Amino_acid_identity_to_josiah_figure\"]\n",
    "# protein_tree_figure_file_path = \"../\" + config[\"GPC_protein_reduced_tree_figure\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "82796202-5227-40b8-a78b-d73319ebe521",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load metadata as dataframe\n",
    "all_metadata = pd.read_csv(metadata_file_path, sep=\"\\t\")\n",
    "\n",
    "# Load list of sequence names from alignment\n",
    "strains = []\n",
    "for curr_fasta in SeqIO.parse(alignment_file_path, \"fasta\"):\n",
    "    strains.append(str(curr_fasta.id))\n",
    "\n",
    "# Filter metadata\n",
    "GPC_reduced_metadata = (\n",
    "    all_metadata.loc[all_metadata[\"strain\"].isin(strains)].copy()\n",
    ")\n",
    "# Add Sierra Leone for Josiah strain\n",
    "GPC_reduced_metadata.at[0, \"country\"] = \"Sierra Leone\"\n",
    "\n",
    "# Create more succinct name conversion dict\n",
    "GPC_reduced_metadata[\"succint_name\"] = (\n",
    "    GPC_reduced_metadata.apply(lambda x: str(x[\"strain\"].split(\"_\")[0]) + \"_\" + str(x[\"accession\"]), axis=1)\n",
    ")\n",
    "name_conversion = (\n",
    "    dict(zip(\n",
    "        GPC_reduced_metadata[\"strain\"].tolist(),\n",
    "        GPC_reduced_metadata[\"succint_name\"].tolist()\n",
    "    ))\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "813366bf-dd82-4fd4-9d15-874c8b37fbed",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Map country and strain name to color\n",
    "country_to_color = (\n",
    "    dict(zip(\n",
    "        GPC_reduced_metadata[\"country\"].unique().tolist(),\n",
    "        tol_muted_adjusted[1:]\n",
    "    ))\n",
    ")\n",
    "# Map name to country\n",
    "name_to_country = GPC_reduced_metadata.set_index(\"strain\")[\"country\"].to_dict()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f1680bd6-e133-41d5-836d-d2acdeffc052",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Isolates to highlight in tree\n",
    "to_annotate = config[\"Validation_sequence_names\"]\n",
    "\n",
    "names_to_annotate = (\n",
    "    GPC_reduced_metadata.loc[GPC_reduced_metadata[\"strain\"].isin(to_annotate)][\"strain\"]\n",
    "    .tolist()\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4668ba4b-9ddc-4068-9620-39b3c8e05207",
   "metadata": {},
   "source": [
    "## Create alignmnet of isolates chosen for validation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0432d20b-f253-422a-9d61-c4f6ef159cdf",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Sequences identified in config file\n",
    "isolates = chosen_validations\n",
    "\n",
    "# Extract chosen validation isolates\n",
    "with open(isolate_sequences, \"w\") as isolate_seqs:\n",
    "    for curr_fasta in SeqIO.parse(protein_path, \"fasta\"):\n",
    "        curr_id = str(curr_fasta.id).split(\" \")[0][:-2]\n",
    "        if curr_id in isolates:\n",
    "            curr_fasta.id = name_conversion[str(curr_fasta.id)[:-2]]\n",
    "            curr_fasta.description = \"\"\n",
    "            SeqIO.write(curr_fasta, isolate_seqs, \"fasta\")\n",
    "# Close files\n",
    "isolate_seqs.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d453e1ac-2f59-4f92-8e13-6b173f289c9c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Align chosen validation isolates\n",
    "os.system(f\"mafft --auto --quiet {isolate_sequences} > {isolate_alignment}\")\n",
    "\n",
    "# Visualize alignment\n",
    "mv = MsaViz(\n",
    "    isolate_alignment, \n",
    "    color_scheme=\"Identity\", \n",
    "    consensus_color=\"lightgrey\",\n",
    "    wrap_length=100, \n",
    "    show_grid=True, \n",
    "    show_consensus=True\n",
    ")\n",
    "\n",
    "mv.set_plot_params(\n",
    "    grid_color=\"black\", \n",
    "    identity_color=\"lightgrey\"\n",
    ")\n",
    "\n",
    "# Add text annotations\n",
    "mv.add_text_annotation((1, 58), \"Stable Signal Peptide\", text_color=\"#AA4499\", range_color=\"#AA4499\")\n",
    "mv.add_text_annotation((59, 259), \"Glycoprotein 1\", text_color=\"#88CCEE\", range_color=\"#88CCEE\")\n",
    "mv.add_text_annotation((260, 491), \"Glycoprotein 2\", text_color=\"#EE7733\", range_color=\"#EE7733\")\n",
    "\n",
    "fig = mv.plotfig()\n",
    "\n",
    "# Make output dir if doesn't exist\n",
    "if not os.path.exists(figure_dir):\n",
    "    os.mkdir(figure_dir)\n",
    "\n",
    "fig.savefig(alignment_figure, dpi=300)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a3471f5a-6e91-4261-8fa6-b2a1fce03bb3",
   "metadata": {},
   "source": [
    "## Create a codon tree"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "38dbc80c-98d4-4613-89b1-729393021829",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Function to create and style tree\n",
    "def get_pretty_codon_tree(treefile, country_to_color, name_to_country, names_to_annotate, name_conversion):\n",
    "    \"\"\"\n",
    "    Function that formats a tree using ete3. This \n",
    "    function is modified from Equia et al analysis.\n",
    "    \"\"\"\n",
    "    # Create ete3 tree object\n",
    "    t = ete3.Tree(treefile, format=1)\n",
    "\n",
    "    # Calculate the midpoint node\n",
    "    # and set it as tree outgroup\n",
    "    R = t.get_midpoint_outgroup()\n",
    "    t.set_outgroup(R)\n",
    "    # Ladderize tree\n",
    "    t.ladderize()\n",
    "    \n",
    "    ts = ete3.TreeStyle()\n",
    "    ts.show_leaf_name = False  # add tip names manually\n",
    "    ts.scale = 50\n",
    "    ts.branch_vertical_margin = 2\n",
    "    \n",
    "    name_to_color = {name: country_to_color[country] for name, country in name_to_country.items()}\n",
    "\n",
    "    # Style and annotate each leaf\n",
    "    for n in t.traverse():\n",
    "        nstyle = ete3.NodeStyle()\n",
    "        nstyle[\"hz_line_width\"] = 1\n",
    "        nstyle[\"vt_line_width\"] = 1\n",
    "        nstyle[\"hz_line_color\"] = \"black\"\n",
    "        nstyle[\"vt_line_color\"] = \"black\"\n",
    "        if n.is_leaf():\n",
    "            nstyle[\"fgcolor\"] = name_to_color[n.name]\n",
    "            if n.name in names_to_annotate:\n",
    "                if n.name == \"Josiah_NC_004296_reverse_complement_2018-08-13\":\n",
    "                    nstyle[\"size\"] = 8\n",
    "                    nstyle[\"shape\"] = \"square\"\n",
    "                else:\n",
    "                    nstyle[\"size\"] = 6\n",
    "                    nstyle[\"shape\"] = \"square\"\n",
    "            else:\n",
    "                nstyle[\"size\"] = 4\n",
    "            if n.name in names_to_annotate:\n",
    "                if n.name == \"Josiah_NC_004296_reverse_complement_2018-08-13\":\n",
    "                    n.add_face(ete3.TextFace(f\" {name_conversion[n.name]}\",\n",
    "                                             ftype=\"Arial\",\n",
    "                                             tight_text=True,\n",
    "                                             fsize=8,\n",
    "                                             fgcolor=tol_muted_adjusted[0],\n",
    "                                             ),\n",
    "                               column=0,\n",
    "                               position=\"branch-right\")\n",
    "                else:\n",
    "                    n.add_face(ete3.TextFace(f\" {name_conversion[n.name]}\",\n",
    "                                             ftype=\"Arial\",\n",
    "                                             tight_text=True,\n",
    "                                             fsize=6,\n",
    "                                             fgcolor=tol_muted_adjusted[0],\n",
    "                                             ),\n",
    "                               column=0,\n",
    "                               position=\"branch-right\")\n",
    "        else:\n",
    "            nstyle[\"size\"] = 0\n",
    "        n.set_style(nstyle)\n",
    "\n",
    "    # Add legend\n",
    "    for country,color in country_to_color.items():\n",
    "        # Marker\n",
    "        marker = ete3.CircleFace(5, color=color)\n",
    "        marker.margin_bottom = 1\n",
    "        marker.margin_right = 8\n",
    "        ts.legend.add_face(marker, column=0)\n",
    "        # Text\n",
    "        text = ete3.TextFace(country, ftype=\"Arial\", fsize=8)\n",
    "        text.margin_bottom = 1\n",
    "        ts.legend.add_face(text, column=1)\n",
    "    ts.legend_position = 2\n",
    "        \n",
    "    # Add scale bar but can't edit text\n",
    "    # scale bar: https://github.com/etetoolkit/ete/issues/266\n",
    "    ts.show_scale = True\n",
    "    ts.scale_length = 0.5\n",
    "         \n",
    "    return t, ts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4f71da98-869c-448f-9499-08b181f84e3e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create codon tree\n",
    "t, ts = get_pretty_codon_tree(codon_tree_file_path, country_to_color, name_to_country, names_to_annotate, name_conversion)\n",
    "\n",
    "# Create dictionary to re-order dataframe for percent IDs later\n",
    "tree_order_dict = {}\n",
    "index = len(t.get_leaves())\n",
    "for leaf in t.get_leaves():\n",
    "    tree_order_dict[leaf.name] = index - 1\n",
    "    index -= 1\n",
    "\n",
    "# # Uncomment to display in notebook\n",
    "# display(\n",
    "#     t.render(\n",
    "#         \"%%inline\", \n",
    "#         tree_style=ts,\n",
    "#         # units=\"in\",\n",
    "#         # h=6.5,\n",
    "#         # w=3.5,\n",
    "#         dpi=300,\n",
    "#     )\n",
    "# )\n",
    "\n",
    "# Make output dir if doesn't exist\n",
    "if not os.path.exists(figure_dir):\n",
    "    os.mkdir(figure_dir)\n",
    "\n",
    "# Render image\n",
    "t.render(\n",
    "    codon_tree_figure_file_path, \n",
    "    tree_style=ts,\n",
    "    # units=\"in\",\n",
    "    # h=6.5,\n",
    "    # w=3.5,\n",
    "    dpi=300,\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b73f9be9-ccb5-44ac-a6f7-3b62c09648a4",
   "metadata": {},
   "source": [
    "## Create a percent amino-acid identity bar to accompany codon tree"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5e5aac70-3527-4d2b-908c-98b5cbfbaf59",
   "metadata": {},
   "outputs": [],
   "source": [
    "def percent_ids(seq1, seq2):\n",
    "    \"\"\"\n",
    "    Function to calculate percent similarity between two sequences.\n",
    "    \"\"\"\n",
    "    # Make sure aligned sequences have same length\n",
    "    assert len(seq1) == len(seq2), \"Aligned sequences do not have same length!\"\n",
    "\n",
    "    # Count base similarities \n",
    "    length = len(seq1)\n",
    "    num_gaps = seq1.count('-')\n",
    "    num_bases = length - num_gaps\n",
    "    matching_bases = 0\n",
    "    for i in range(length):\n",
    "        if seq1[i] == seq2[i] and seq1[i] != \"-\":\n",
    "            matching_bases += 1\n",
    "\n",
    "    # Return percent similarity between the two sequences\n",
    "    return ((matching_bases)/num_bases) * 100"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eb5b1714-aff0-423d-8bcb-5afeb014bf19",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load alignment info\n",
    "protein_alignment_df = pd.DataFrame(columns=[\"strain\", \"sequence\"])\n",
    "\n",
    "# Add alignment sequence to dataframe\n",
    "for curr_fasta in AlignIO.read(protein_alignment_file_path, \"fasta\"):\n",
    "    protein_alignment_df.loc[len(protein_alignment_df.index)] = [\n",
    "        str(curr_fasta.id),\n",
    "        str(curr_fasta.seq),\n",
    "    ] \n",
    "\n",
    "josiah_strain = protein_alignment_df.at[0, \"sequence\"]\n",
    "\n",
    "protein_alignment_df[\"percent identity\"] = protein_alignment_df.apply(lambda x: percent_ids(josiah_strain, x[\"sequence\"]), axis=1)\n",
    "protein_alignment_df[\"percent identity rounded\"] = protein_alignment_df[\"percent identity\"].round()\n",
    "protein_alignment_df[\"percent identity rounded\"] = protein_alignment_df[\"percent identity rounded\"].astype(int)\n",
    "percent_to_color_dict = dict(zip(\n",
    "    list(range(90, 101)), \n",
    "    gray_scale\n",
    "))\n",
    "\n",
    "protein_alignment_df[\"color\"] = protein_alignment_df[\"percent identity rounded\"].map(percent_to_color_dict)\n",
    "protein_alignment_df[\"tree order\"] = protein_alignment_df[\"strain\"].map(tree_order_dict)\n",
    "protein_alignment_df = (\n",
    "    protein_alignment_df\n",
    "    .sort_values(by=\"tree order\")\n",
    "    .reset_index(drop=True)\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "973a833e-fbc4-40c0-a269-94501117282c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create amino acid simiarlity bar for phylogentic tree\n",
    "plt.figure(figsize=(6, 1))\n",
    "percent_AA_plot = sns.heatmap(\n",
    "    data=[np.asarray(protein_alignment_df[\"percent identity rounded\"])],\n",
    "    cmap=gray_scale,\n",
    "    vmin=90,\n",
    "    vmax=100,\n",
    "    linewidths=0.5,\n",
    "    linecolor=\"black\",\n",
    "    square=True,\n",
    "    yticklabels=False,\n",
    "    xticklabels=False,\n",
    ")\n",
    "\n",
    "percent_AA_plot.tick_params(axis=\"both\", length=1, width=0.5)\n",
    "cbar = percent_AA_plot.collections[0].colorbar\n",
    "cbar.outline.set_color(\"black\")\n",
    "cbar.outline.set_linewidth(0.5)\n",
    "cbar.set_ticks(ticks=[100, 90], labels=[\"100\", \"90\"])\n",
    "cbar.ax.tick_params(axis=\"both\", length=1, width=0.5)\n",
    "\n",
    "plt.tight_layout()\n",
    "\n",
    "plt.savefig(percent_identity_to_josiah_figure_file_path)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5ba9338e-3688-42f6-9de0-cd6ec5d3fcd2",
   "metadata": {},
   "source": [
    "## Create a protein tree"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0f3ed288-8ff0-4e2f-ba45-487aa2775e64",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Function to create and style tree\n",
    "def get_pretty_protein_tree(treefile, country_to_color, name_to_country, names_to_annotate, name_conversion):\n",
    "    \"\"\"\n",
    "    Function that formats a tree using ete3. This \n",
    "    function is modified from Equia et al analysis.\n",
    "    \"\"\"\n",
    "    # Create ete3 tree object\n",
    "    t = ete3.Tree(treefile, format=1)\n",
    "\n",
    "    # Calculate the midpoint node\n",
    "    # and set it as tree outgroup\n",
    "    R = t.get_midpoint_outgroup()\n",
    "    t.set_outgroup(R)\n",
    "    # Ladderize tree\n",
    "    t.ladderize()\n",
    "    \n",
    "    ts = ete3.TreeStyle()\n",
    "    ts.show_leaf_name = False  # add tip names manually\n",
    "    ts.scale = 250\n",
    "    ts.branch_vertical_margin = 2\n",
    "    \n",
    "    name_to_color = {name: country_to_color[country] for name, country in name_to_country.items()}\n",
    "\n",
    "    # Style and annotate each leaf\n",
    "    for n in t.traverse():\n",
    "        nstyle = ete3.NodeStyle()\n",
    "        nstyle[\"hz_line_width\"] = 1\n",
    "        nstyle[\"vt_line_width\"] = 1\n",
    "        nstyle[\"hz_line_color\"] = \"black\"\n",
    "        nstyle[\"vt_line_color\"] = \"black\"\n",
    "        if n.is_leaf():\n",
    "            nstyle[\"fgcolor\"] = name_to_color[n.name]\n",
    "            if n.name in names_to_annotate:\n",
    "                if n.name == \"Josiah_NC_004296_reverse_complement_2018-08-13\":\n",
    "                    nstyle[\"size\"] = 8\n",
    "                    nstyle[\"shape\"] = \"square\"\n",
    "                else:\n",
    "                    nstyle[\"size\"] = 6\n",
    "                    nstyle[\"shape\"] = \"square\"\n",
    "            else:\n",
    "                nstyle[\"size\"] = 4\n",
    "            if n.name in names_to_annotate:\n",
    "                if n.name == \"Josiah_NC_004296_reverse_complement_2018-08-13\":\n",
    "                    n.add_face(ete3.TextFace(f\" {name_conversion[n.name]}\",\n",
    "                                             ftype=\"Arial\",\n",
    "                                             tight_text=True,\n",
    "                                             fsize=8,\n",
    "                                             fgcolor=tol_muted_adjusted[0],\n",
    "                                             ),\n",
    "                               column=0,\n",
    "                               position=\"branch-right\")\n",
    "                else:\n",
    "                    n.add_face(ete3.TextFace(f\" {name_conversion[n.name]}\",\n",
    "                                             ftype=\"Arial\",\n",
    "                                             tight_text=True,\n",
    "                                             fsize=6,\n",
    "                                             fgcolor=tol_muted_adjusted[0],\n",
    "                                             ),\n",
    "                               column=0,\n",
    "                               position=\"branch-right\")\n",
    "        else:\n",
    "            nstyle[\"size\"] = 0\n",
    "        n.set_style(nstyle)\n",
    "\n",
    "    # Add legend\n",
    "    for country,color in country_to_color.items():\n",
    "        # Marker\n",
    "        marker = ete3.CircleFace(5, color=color)\n",
    "        marker.margin_bottom = 1\n",
    "        marker.margin_right = 8\n",
    "        ts.legend.add_face(marker, column=0)\n",
    "        # Text\n",
    "        text = ete3.TextFace(country, ftype=\"Arial\", fsize=10)\n",
    "        text.margin_bottom = 1\n",
    "        ts.legend.add_face(text, column=1)\n",
    "    ts.legend_position = 2\n",
    "        \n",
    "    # Add scale bar but can't edit text\n",
    "    # scale bar: https://github.com/etetoolkit/ete/issues/266\n",
    "    ts.show_scale = True\n",
    "    ts.scale_length = 0.05\n",
    "         \n",
    "    return t, ts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ce3478e6-2836-42de-aeb5-5d8bbb7a12ee",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create protein tree\n",
    "t, ts = get_pretty_protein_tree(protein_tree_file_path, country_to_color, name_to_country, names_to_annotate, name_conversion)\n",
    "\n",
    "# # Uncomment to display in notebook\n",
    "# display(\n",
    "#     t.render(\n",
    "#         \"%%inline\", \n",
    "#         tree_style=ts,\n",
    "#         # units=\"in\",\n",
    "#         # h=5,\n",
    "#         # w=3.5,\n",
    "#         dpi=300,\n",
    "#     )\n",
    "# )\n",
    "\n",
    "# Make output dir if doesn't exist\n",
    "if not os.path.exists(figure_dir):\n",
    "    os.mkdir(figure_dir)\n",
    "\n",
    "# Render image\n",
    "t.render(\n",
    "    protein_tree_figure_file_path, \n",
    "    tree_style=ts,\n",
    "    # units=\"in\",\n",
    "    # h=5,\n",
    "    # w=3.5,\n",
    "    dpi=300,\n",
    ")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
