{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "be9179b8-2968-4f00-b5e0-a3aa831c685b",
   "metadata": {},
   "source": [
    "# Visualize DAG1 multiple sequence alignments"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dec8deaf-9064-4c79-85ab-5571227d63eb",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Imports\n",
    "import os\n",
    "import yaml\n",
    "import warnings\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "from matplotlib import pyplot as plt\n",
    "from pymsaviz import MsaViz \n",
    "from Bio import SeqIO, AlignIO, Phylo\n",
    "\n",
    "# Plotting colors\n",
    "tol_muted_adjusted = [\n",
    "    \"#000000\",\n",
    "    \"#CC6677\", \n",
    "    \"#1f78b4\", \n",
    "    \"#DDCC77\", \n",
    "    \"#117733\", \n",
    "    \"#882255\", \n",
    "    \"#88CCEE\",\n",
    "    \"#44AA99\", \n",
    "    \"#999933\", \n",
    "    \"#AA4499\", \n",
    "    \"#EE7733\",\n",
    "    \"#CC3311\",\n",
    "    \"#DDDDDD\",\n",
    "]\n",
    "\n",
    "# Seaborn style settings\n",
    "sns.set(rc={\n",
    "    \"figure.dpi\":300, \n",
    "    \"savefig.dpi\":300,\n",
    "    \"svg.fonttype\":\"none\",\n",
    "})\n",
    "sns.set_style(\"ticks\")\n",
    "sns.set_palette(tol_muted_adjusted)\n",
    "\n",
    "# Suppress warnings\n",
    "warnings.simplefilter(\"ignore\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a33f4a43-ccc5-439c-92c7-cce18c99c063",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Open config file\n",
    "config_path = \"Configure/config.yml\"\n",
    "with open(config_path) as f:\n",
    "    config = yaml.safe_load(f)\n",
    "\n",
    "# Set paths\n",
    "reduced_alignment_file_path = config[\"Reduced_protein_alignment\"]\n",
    "alignment_file_path = config[\"Protein_alignment\"]\n",
    "out_dir = \"Results/Figures/\"\n",
    "consensus_glycosylation_site = config[\"Consensus_glycosylation_site\"]\n",
    "human_mastomys_msa = config[\"Human_and_mastomys_msa\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2bd7e861-1507-4a79-8d47-87613513946b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Uncomment to run interactively\n",
    "# # Open config file\n",
    "# config_path = \"../Configure/config.yml\"\n",
    "# with open(config_path) as f:\n",
    "#     config = yaml.safe_load(f)\n",
    "\n",
    "# # Set paths\n",
    "# reduced_alignment_file_path = \"../\" + config[\"Reduced_protein_alignment\"]\n",
    "# alignment_file_path = \"../\" + config[\"Protein_alignment\"]\n",
    "# out_dir = \"../Results/Figures/\"\n",
    "# consensus_glycosylation_site = \"../\" + config[\"Consensus_glycosylation_site\"]\n",
    "# human_mastomys_msa = \"../\" + config[\"Human_and_mastomys_msa\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9c8b7df3-8d07-4ab1-b35c-d609ee71e344",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Count number of sequences with concensus sequence\n",
    "position_counts = [0,0,0,0,0]\n",
    "aa_identities = [\"T\", \"P\", \"T\", \"P\", \"V\"]\n",
    "DAG1_alignment = AlignIO.read(alignment_file_path, \"fasta\")\n",
    "for seq in DAG1_alignment[:, 710:715]:\n",
    "    if seq.seq[0] == \"T\":\n",
    "        position_counts[0] += 1\n",
    "    if seq.seq[1] == \"P\":\n",
    "        position_counts[1] += 1\n",
    "    if seq.seq[2] == \"T\":\n",
    "        position_counts[2] += 1\n",
    "    if seq.seq[3] == \"P\":\n",
    "        position_counts[3] += 1\n",
    "    if seq.seq[4] == \"V\":\n",
    "        position_counts[4] += 1\n",
    "\n",
    "# Create dataframe from data\n",
    "consensus_df = pd.DataFrame({\"amino acid\" : aa_identities, \"counts\" : position_counts})\n",
    "consensus_df[\"percent consensus\"] = consensus_df[\"counts\"] / 527 # number of sequences: 527\n",
    "consensus_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6002a4ed-c0e2-40aa-a5d6-a7f2a3f3ace7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Visualize alignment\n",
    "# number of sequences: 527\n",
    "mv = MsaViz(\n",
    "    alignment_file_path, \n",
    "    color_scheme=\"Identity\", \n",
    "    consensus_color=\"#CC3311\",\n",
    "    wrap_length=5, \n",
    "    show_grid=True, \n",
    "    show_consensus=True,\n",
    "    start=711,\n",
    "    end=715,\n",
    ")\n",
    "\n",
    "mv.set_plot_params(\n",
    "    grid_color=\"black\", \n",
    "    identity_color=\"lightgrey\"\n",
    ")\n",
    "\n",
    "fig = mv.plotfig()\n",
    "\n",
    "# Make output dir if doesn't exist\n",
    "if not os.path.exists(out_dir):\n",
    "    os.mkdir(out_dir)\n",
    "\n",
    "# Save fig\n",
    "plt.savefig(consensus_glycosylation_site)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d3470af3-b00c-4d06-a3dc-87b346c885ce",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Visualize alignment\n",
    "mv = MsaViz(\n",
    "    reduced_alignment_file_path, \n",
    "    color_scheme=\"Identity\", \n",
    "    consensus_color=\"lightgrey\",\n",
    "    wrap_length=75, \n",
    "    show_grid=True, \n",
    "    show_consensus=False,\n",
    ")\n",
    "\n",
    "mv.set_plot_params(\n",
    "    grid_color=\"black\", \n",
    "    identity_color=\"lightgrey\",\n",
    ")\n",
    "\n",
    "# Add text annotations\n",
    "mv.add_text_annotation((1, 653), \"alpha-dystroglycan\", text_color=\"#AA4499\", range_color=\"#AA4499\")\n",
    "mv.add_text_annotation((654, 896), \"beta-dystrolgycan\", text_color=\"#88CCEE\", range_color=\"#88CCEE\")\n",
    "\n",
    "fig = mv.plotfig()\n",
    "\n",
    "# Make output dir if doesn't exist\n",
    "if not os.path.exists(out_dir):\n",
    "    os.mkdir(out_dir)\n",
    "\n",
    "# Save fig\n",
    "plt.savefig(human_mastomys_msa)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
